{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "bff64795-ac92-4c1d-9ca3-00a7885ff712",
   "metadata": {},
   "source": [
    "## Annotation of cells from LN samples"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "40506a85-8ec8-4660-be16-2fcafb64d0f3",
   "metadata": {},
   "outputs": [],
   "source": [
    "import numpy as np\n",
    "import seaborn as sns\n",
    "\n",
    "import matplotlib\n",
    "import matplotlib.pyplot as plt\n",
    "from matplotlib.legend import Legend\n",
    "import matplotlib.colors as colors\n",
    "from matplotlib import cm\n",
    "from mpl_toolkits.mplot3d import axes3d\n",
    "import pandas as pd\n",
    "import scipy\n",
    "import scanpy as sc\n",
    "import anndata as ad\n",
    "\n",
    "from sklearn import datasets\n",
    "from sklearn.decomposition import PCA\n",
    "\n",
    "from numba import jit\n",
    "\n",
    "import celltypist\n",
    "from celltypist import models\n",
    "\n",
    "from matplotlib.cm import ScalarMappable"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "858b2b14-4ab3-4dff-b882-212090c7ce45",
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "#Custom colormap\n",
    "\n",
    "from matplotlib.cm import register_cmap\n",
    "from matplotlib.colors import ListedColormap\n",
    "\n",
    "tab20b = matplotlib.colormaps['tab20b']\n",
    "tab20c = matplotlib.colormaps['tab20c']\n",
    "colors1 = tab20b(np.linspace(3.001/5., 1, 9))\n",
    "colors2 = tab20c(np.linspace(0, 3.999/5., 16))\n",
    "\n",
    "colors = np.concatenate([colors1, colors2])\n",
    "\n",
    "map_name = 'op_tab25'\n",
    "op_cmap = ListedColormap(colors, name=map_name )\n",
    "matplotlib.colormaps.register(name=map_name, cmap=op_cmap)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "067ef764-d890-4f3e-baae-32d3d66ef8ac",
   "metadata": {},
   "outputs": [],
   "source": [
    "sc.set_figure_params(scanpy=True, dpi=600, dpi_save=600, frameon=True, vector_friendly=True, fontsize=14,\n",
    "                         figsize=(9,8),  format='pdf', facecolor=None, transparent=False, ipython_format='png2x')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a8198d1a-4772-4f3d-ac54-2cbf76c78b45",
   "metadata": {},
   "outputs": [],
   "source": [
    "#Load lymph node data\n",
    "adata = sc.read(\"maranou_032024_LN_integrated.h5ad\")\n",
    "tissue = 'LN'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ad0d8dcc-4cb1-48ef-b16e-d1e6fe509621",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Create a new index by combining 'sample' column with the current index\n",
    "adata.obs['unique_cell_name'] = adata.obs['sample'].astype(str) + '_' + adata.obs.index.astype(str)\n",
    "\n",
    "# Set this new column as the index\n",
    "adata.obs.index = adata.obs['unique_cell_name']\n",
    "\n",
    "# Optionally, you might want to drop the 'unique_cell_name' column afterward if it's no longer needed\n",
    "adata.obs.drop('unique_cell_name', axis=1, inplace=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2d16729c-b532-44e9-b468-5e79234d42fb",
   "metadata": {},
   "outputs": [],
   "source": [
    "# 1. Perform initial clustering at lower resolution\n",
    "sc.tl.leiden(adata, resolution=0.95, key_added='leiden_initial')\n",
    "\n",
    "# 2. Identify the cluster(s) containing DC2, monocytes, and macrophages\n",
    "target_clusters = ['16','18'] \n",
    "\n",
    "# 3. Subset the data\n",
    "adata_subset = adata[adata.obs['leiden_initial'].isin(target_clusters)].copy()\n",
    "\n",
    "# 4. Recompute the neighborhood graph on the subset\n",
    "sc.pp.neighbors(adata_subset)\n",
    "\n",
    "# 5. Perform Leiden clustering at higher resolution on the subset\n",
    "sc.tl.leiden(adata_subset, resolution=0.7, key_added='leiden_refined')\n",
    "\n",
    "# 6. Prepare categories for the combined clustering\n",
    "initial_categories = list(adata.obs['leiden_initial'].cat.categories)\n",
    "refined_categories = list(adata_subset.obs['leiden_refined'].cat.categories)\n",
    "\n",
    "# Remove target clusters from initial categories\n",
    "initial_categories_filtered = [cat for cat in initial_categories if cat not in target_clusters]\n",
    "\n",
    "# Create new category names for refined clusters\n",
    "refined_categories_renamed = [f'r{cat}' for cat in refined_categories]\n",
    "\n",
    "# Combine filtered initial categories with renamed refined categories\n",
    "combined_categories = initial_categories_filtered + refined_categories_renamed\n",
    "\n",
    "# 7. Create new column for combined clustering\n",
    "adata.obs['leiden'] = pd.Categorical(\n",
    "    adata.obs['leiden_initial'],\n",
    "    categories=combined_categories\n",
    ")\n",
    "\n",
    "# 8. Update the combined clustering for the refined subset\n",
    "for idx in adata_subset.obs.index:\n",
    "    refined_value = adata_subset.obs.loc[idx, 'leiden_refined']\n",
    "    adata.obs.at[idx, 'leiden'] = f'r{refined_value}'\n",
    "\n",
    "# 9. Optionally, sort the categories for better readability\n",
    "adata.obs['leiden'] = adata.obs['leiden'].cat.reorder_categories(sorted(adata.obs['leiden'].cat.categories))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5d796d48-38c9-43b6-ad7f-9f56609bd3ca",
   "metadata": {},
   "outputs": [],
   "source": [
    "sc.pl.umap(adata, color=['leiden'], title=[str(tissue) +' Leiden'], palette='op_tab25', legend_loc='on data')\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f84d712e-15e2-4d94-b851-e1e159381d39",
   "metadata": {},
   "outputs": [],
   "source": [
    "sc.pl.tsne(adata, color=['leiden'], title=[str(tissue) +' Leiden','Naive '+str(tissue)+' WT/KO'], palette='op_tab25', legend_loc='on data')\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "30537785-d314-4b77-9195-ee467eff936c",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Define matplotlib Axes\n",
    "# Number of Axes & plot size\n",
    "ncols = 2\n",
    "nrows = 2\n",
    "figsize = 8\n",
    "wspace = 0.1\n",
    "fig, axs = plt.subplots(\n",
    "    nrows=nrows,\n",
    "    ncols=ncols,\n",
    "    figsize=(ncols * figsize + figsize * wspace * (ncols - 1), nrows * figsize),\n",
    ")\n",
    "plt.subplots_adjust(wspace=wspace)\n",
    "\n",
    "sc.pl.umap(adata[adata.obs['sample']=='wt_naive'], ax=axs[0,0], show=False, color=['leiden'], title=['Naive WT '+ str(tissue) +' Leiden'], palette='op_tab25', legend_loc='on data')\n",
    "sc.pl.umap(adata[adata.obs['sample']=='wt_pathogenic'],  ax=axs[0,1], show=False, color=['leiden'], title=['Pathogenic WT '+ str(tissue) +' Leiden'], palette='op_tab25', legend_loc='on data')\n",
    "sc.pl.umap(adata[adata.obs['sample']=='ko_naive'],  ax=axs[1,0],show=False, color=['leiden'], title=['Naive Cd74 KO '+ str(tissue) +' Leiden'], palette='op_tab25', legend_loc='on data')\n",
    "sc.pl.umap(adata[adata.obs['sample']=='ko_pathogenic'],  ax=axs[1,1],color=['leiden'], title=['Pathogenic Cd74 KO '+ str(tissue) +' Leiden'], palette='op_tab25', legend_loc='on data')\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5be12125-ec88-454f-9f9b-56a7695735f6",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Define matplotlib Axes\n",
    "# Number of Axes & plot size\n",
    "ncols = 2\n",
    "nrows = 2\n",
    "figsize = 8\n",
    "wspace = 0.1\n",
    "fig, axs = plt.subplots(\n",
    "    nrows=nrows,\n",
    "    ncols=ncols,\n",
    "    figsize=(ncols * figsize + figsize * wspace * (ncols - 1), nrows * figsize),\n",
    ")\n",
    "plt.subplots_adjust(wspace=wspace)\n",
    "\n",
    "sc.pl.tsne(adata[adata.obs['sample']=='wt_naive'], ax=axs[0,0], show=False, color=['leiden'], title=['Naive WT '+ str(tissue) +' Leiden'], palette='op_tab25', legend_loc='on data')\n",
    "sc.pl.tsne(adata[adata.obs['sample']=='wt_pathogenic'],  ax=axs[0,1], show=False, color=['leiden'], title=['Pathogenic WT '+ str(tissue) +' Leiden'], palette='op_tab25', legend_loc='on data')\n",
    "sc.pl.tsne(adata[adata.obs['sample']=='ko_naive'],  ax=axs[1,0],show=False, color=['leiden'], title=['Naive Cd74 KO '+ str(tissue) +' Leiden'], palette='op_tab25', legend_loc='on data')\n",
    "sc.pl.tsne(adata[adata.obs['sample']=='ko_pathogenic'],  ax=axs[1,1],color=['leiden'], title=['Pathogenic Cd74 KO '+ str(tissue) +' Leiden'], palette='op_tab25', legend_loc='on data')\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "2efd7843-e89b-4319-ad0d-0ad20d8b5f4a",
   "metadata": {},
   "source": [
    "### Celltypist for making initial guesses for cell types"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2c3b7466-4815-4813-9f6b-51349b525d9e",
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "import celltypist\n",
    "from celltypist import models\n",
    "\n",
    "adata_human = adata.copy()\n",
    "\n",
    "#Mouse-human gene conversion\n",
    "mouse_human_genes = pd.read_csv(\"http://www.informatics.jax.org/downloads/reports/HOM_MouseHumanSequence.rpt\",sep=\"\\t\", index_col=False)\n",
    "mouse = mouse_human_genes[mouse_human_genes['Common Organism Name']=='mouse, laboratory']\n",
    "human = mouse_human_genes[mouse_human_genes['Common Organism Name']=='human']\n",
    "\n",
    "# Collect dataframes for conversion\n",
    "mouse = mouse[['DB Class Key', 'Symbol']]\n",
    "mouse.index = np.arange(mouse.shape[0])\n",
    "\n",
    "human = human[['DB Class Key', 'Symbol']]\n",
    "human.index = np.arange(human.shape[0])\n",
    "\n",
    "gene_list = np.array([])\n",
    "\n",
    "# Convert from human to mouse\n",
    "for sym_idx in np.arange(adata.var_names.shape[0]):\n",
    "\n",
    "    row_idx  = np.where(mouse['Symbol']== adata_human.var_names[sym_idx])[0]\n",
    "\n",
    "    if len(row_idx)>0:\n",
    "        \n",
    "        dbck = mouse['DB Class Key'][row_idx].values[0]\n",
    "        human_gene = human[human['DB Class Key']==dbck]['Symbol'].values\n",
    "\n",
    "        if len(human_gene)>0:\n",
    "            gene_list =  np.append(gene_list,human_gene[0])\n",
    "\n",
    "        else:\n",
    "\n",
    "            gene_list =  np.append(gene_list,adata_human.var_names[sym_idx])\n",
    "\n",
    "    else:\n",
    "\n",
    "        gene_list =  np.append(gene_list,adata_human.var_names[sym_idx])\n",
    "\n",
    "sc.pp.normalize_per_cell(adata_human, counts_per_cell_after=1e4)\n",
    "sc.pp.log1p(adata_human)\n",
    "np.expm1(adata_human.X).sum(1)\n",
    "\n",
    "adata_human.var_names = gene_list\n",
    "predictions = celltypist.annotate(adata_human, model = 'Immune_All_Low.pkl', majority_voting = True)\n",
    "print(predictions.predicted_labels)\n",
    "adata_human = predictions.to_adata()\n",
    "\n",
    "sc.pl.umap(adata_human, color = 'majority_voting', palette='tab20',  legend_loc = 'on data', legend_fontsize=12, legend_fontweight='medium')\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "34ca2270-31ec-448a-8b0d-62710c7ba665",
   "metadata": {},
   "source": [
    "### DEGs in Leiden clusters"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "462da0da-30a3-4693-ad26-5d0f739c1154",
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "sc.tl.rank_genes_groups(adata,'leiden', method='wilcoxon')\n",
    "# sc.pl.rank_genes_groups(adata, n_genes=20, sharey=False)\n",
    "\n",
    "de_markers = sc.get.rank_genes_groups_df(adata, None)\n",
    "de_markers = de_markers[(de_markers.pvals_adj < 0.03) & (de_markers.logfoldchanges > 1.0)]\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d4af223b-8ee7-428f-a255-f07769f2867a",
   "metadata": {},
   "outputs": [],
   "source": [
    "de_markers"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "565ca299-efcb-436f-bbf1-19b0a42632bf",
   "metadata": {},
   "outputs": [],
   "source": [
    "sc.set_figure_params(scanpy=True, dpi=300, dpi_save=300, frameon=True, vector_friendly=True, fontsize=14,\n",
    "                         figsize=(9,8),  format='pdf', facecolor=None, transparent=False, ipython_format='png2x')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "190e5e48-ee86-4921-ae64-7bc45754d91d",
   "metadata": {},
   "outputs": [],
   "source": [
    "sc.pl.rank_genes_groups(adata, n_genes=20, sharey=False)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8760a0bd-00ab-48eb-b2f5-f2e72054cbc9",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Give cluster name to see DEGs compared to other clusters\n",
    "cluster ='10'\n",
    "de_markers[de_markers['group']==cluster].iloc[np.argsort(de_markers[de_markers['group']==cluster]['scores'])[::-1],:][0:50]\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "091c5c40-fb24-47a1-a8a3-d053aaf27127",
   "metadata": {},
   "outputs": [],
   "source": [
    "sc.pl.umap(adata, color=['leiden'], title=[str(tissue) +' Leiden'], palette='op_tab25', legend_loc='on data')\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a8575c40-9b9c-42c5-b76e-044447dd0339",
   "metadata": {},
   "outputs": [],
   "source": [
    "#Leiden refined annotations\n",
    "\n",
    "annotation_dict = {\"0\":\"Antigen-presenting B cells\",\n",
    "                       \"1\":\"Naive B cells\", #high expression of MHC Class II genes (H2-Aa, H2-Ab1, H2-Eb1) and Cd74\n",
    "                       \"2\":\"CD8+\",\n",
    "                       \"3\":\"Activated B cells/Early plasmablasts\",\n",
    "                       \"4\":\"Developing B cells\",\n",
    "                       \"5\":\"Activated CD8+\", # These are Ctla4-\n",
    "                       \"6\":\"CD8+\",\n",
    "                       \"7\":\"CTLA4+ CD8+ T cells\",\n",
    "                       \"8\":\"CD4+\",\n",
    "                       \"9\":\"Treg\",\n",
    "                       #\"10\":\"CD8-CD4-CD3E- T cell-like\", #Have T cell lineage but some are CD8- CD4- CD3E-\n",
    "                       \"10\":\"Heterogeneous T cells\", \n",
    "                       \"11\":\"NK cells\",\n",
    "                       \"12\":\"Th cells\",\n",
    "                       \"13\":\"MZB and B-1 cells\",\n",
    "                       \"14\":\"Interferon-activated B cells\",\n",
    "                       \"15\":\"Activated CD4+\",\n",
    "                       \"17\":\"Activated CD4+\",\n",
    "                       \"19\":\"$\\gamma \\delta$T cells\",\n",
    "                       \"20\":\"CCR7hi CD40+ CD80+ CD86+ DC\",\n",
    "                       \"21\":\"pDC\",\n",
    "                       \"22\":\"Plasma cells\",\n",
    "                       \"r0\":\"Monocytes and macrophages\", #Probably Classical monocytes\n",
    "                       \"r1\":\"cDC2\", # \n",
    "                       \"r2\":\"Lymphoid-resident cDC1\", #XCR1+ [Gurka et al]\n",
    "                       \"r3\":\"Undefined DC\", #[Gurka et al]\n",
    "                       \"r4\":\"CCR7hi CD40+ CD80- CD86- DC\", # Properties of both cDC and migratory \n",
    "                       \"r5\":\"cDC2\",\n",
    "                       \"r6\":\"Plasma cells\", # Previously \"High proliferation/synthesis\". Marked by Rpl/Rps hi\n",
    "                       \"r7\":\"Mast cells\",\n",
    "                       \"r8\":\"Neutrophils\",\n",
    "                       \"r9\":\"Monocytes and macrophages\", #Probably Non-classical monocytes\n",
    "                       }\n",
    "\n",
    "ann_colors = plt.colormaps['tab20'].colors\n",
    "ann_palette={\"Antigen-presenting B cells\": ann_colors[18],\n",
    "                       \"Naive B cells\": ann_colors[0],\n",
    "                       \"CD8+\":ann_colors[4],\n",
    "                       \"Activated B cells/Early plasmablasts\": ann_colors[13],\n",
    "                       \"Developing B cells\":plt.matplotlib.colors.to_rgb('dodgerblue'),\n",
    "                       \"Activated CD8+\":ann_colors[2],\n",
    "                       \"CD4+\":plt.matplotlib.colors.to_rgb('greenyellow'),\n",
    "                       \"Treg\":ann_colors[16],\n",
    "                        \"CTLA4+ CD8+ T cells\":ann_colors[3],\n",
    "                        #\"CD8-CD4-CD3E- T cell-like\":ann_colors[15],\n",
    "                        \"Heterogeneous T cells\":ann_colors[15],\n",
    "                       \"NK cells\":ann_colors[17],\n",
    "                       \"Th cells\":plt.matplotlib.colors.to_rgb('mediumseagreen'),\n",
    "                       \"MZB and B-1 cells\":ann_colors[1],\n",
    "                       \"Interferon-activated B cells\":ann_colors[9],\n",
    "                       \"Activated CD4+\":plt.matplotlib.colors.to_rgb('greenyellow'),\n",
    "                       \"$\\gamma \\delta$T cells\":ann_colors[5],\n",
    "                       \"CCR7hi CD40+ CD80+ CD86+ DC\":ann_colors[12],\n",
    "                       \"pDC\":ann_colors[6],\n",
    "                       \"Plasma cells\":ann_colors[19],\n",
    "                       \"Monocytes and macrophages\":ann_colors[7], #Probably Classical monocytes\n",
    "                       \"Lymphoid-resident cDC1\":plt.matplotlib.colors.to_rgb('palevioletred'), \n",
    "                       \"Undefined DC\":plt.matplotlib.colors.to_rgb('mediumvioletred'), #\n",
    "                       \"CyC(low) DC1\":ann_colors[0], \n",
    "                       \"cDC2\":plt.matplotlib.colors.to_rgb('darkorchid'),\n",
    "                       \"CCR7hi CD40+ CD80- CD86- DC\":plt.matplotlib.colors.to_rgb('deeppink'), \n",
    "                       \"High proliferation/synthesis\":plt.matplotlib.colors.to_rgb('b'),\n",
    "                       \"Mast cells\":plt.matplotlib.colors.to_rgb('cornflowerblue'),\n",
    "                       \"Neutrophils\":plt.matplotlib.colors.to_rgb('coral'),}\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "503ef355-4778-4fd7-ae5e-216642211f01",
   "metadata": {},
   "outputs": [],
   "source": [
    "adata.obs['cell_type'] = adata.obs.leiden.map(annotation_dict)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "32672a63-824c-49b6-b77c-0c4f49547117",
   "metadata": {},
   "outputs": [],
   "source": [
    "sc.pl.umap(adata, color=['cell_type'],add_outline=True, outline_width = (0.2,0.5), palette=ann_palette, alpha=0.7, s=10, title='Lymph node',legend_loc='on data',legend_fontsize=8, legend_fontweight='heavy', save='_LN_annotations.pdf')\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "46ea8f46-c40b-4be3-be72-6c2dad1e054e",
   "metadata": {},
   "source": [
    "### UMAPS of Naive/Pathogenic WT/KO samples"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "80d37f42-df02-4dd0-ba1f-2e8b03084df6",
   "metadata": {},
   "outputs": [],
   "source": [
    "file_path = '/Users/oipulk/Documents/scRNASeq/data/Eleftheria_Maranou_Mar2024/analysis/figures/SI_Fig7/LN_annotations_naive_wt.pdf'\n",
    "with plt.rc_context():  # Use this to set figure params like size and dpi\n",
    "    sc.pl.umap(adata[adata.obs['sample']=='wt_naive'], color=['cell_type'],add_outline=True, outline_width = (0.2,0.3), palette=ann_palette, alpha=0.75, s=12, title='Lymph node, Naive WT',legend_loc='on data',legend_fontsize=8, legend_fontweight='heavy', show=False)\n",
    "    plt.savefig(file_path, bbox_inches=\"tight\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6176b5c4-a579-4819-8fce-f88fc6a271b0",
   "metadata": {},
   "outputs": [],
   "source": [
    "file_path = '/Users/oipulk/Documents/scRNASeq/data/Eleftheria_Maranou_Mar2024/analysis/figures/SI_Fig7/LN_annotations_naive_ko.pdf'\n",
    "with plt.rc_context():  # Use this to set figure params like size and dpi\n",
    "    sc.pl.umap(adata[adata.obs['sample']=='ko_naive'], color=['cell_type'],add_outline=True, outline_width = (0.2,0.3), palette=ann_palette, alpha=0.75, s=12, title='Lymph node, Naive Cd74 KO',legend_loc='on data',legend_fontsize=8, legend_fontweight='heavy', show=False)\n",
    "    plt.savefig(file_path, bbox_inches=\"tight\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f90b8128-cd68-449e-8065-08e1a9d56c0b",
   "metadata": {},
   "outputs": [],
   "source": [
    "file_path = '/Users/oipulk/Documents/scRNASeq/data/Eleftheria_Maranou_Mar2024/analysis/figures/SI_Fig7/LN_annotations_pathogenic_wt.pdf'\n",
    "with plt.rc_context():  # Use this to set figure params like size and dpi\n",
    "    sc.pl.umap(adata[adata.obs['sample']=='wt_pathogenic'], color=['cell_type'],add_outline=True, outline_width = (0.2,0.3), palette=ann_palette, alpha=0.75, s=12, title='Lymph node, Pathogenic WT',legend_loc='on data',legend_fontsize=8, legend_fontweight='heavy', show=False)\n",
    "    plt.savefig(file_path, bbox_inches=\"tight\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "87b78a91-b38c-432f-a4a8-c16db8f88dc1",
   "metadata": {},
   "outputs": [],
   "source": [
    "file_path = '/Users/oipulk/Documents/scRNASeq/data/Eleftheria_Maranou_Mar2024/analysis/figures/SI_Fig7/LN_annotations_pathogenic_ko.pdf'\n",
    "with plt.rc_context():  # Use this to set figure params like size and dpi\n",
    "    sc.pl.umap(adata[adata.obs['sample']=='ko_pathogenic'], color=['cell_type'],add_outline=True, outline_width = (0.2,0.3), palette=ann_palette, alpha=0.75, s=12, title='Lymph node, Pathogenic Cd74 KO',legend_loc='on data',legend_fontsize=8, legend_fontweight='heavy', show=False)\n",
    "    plt.savefig(file_path, bbox_inches=\"tight\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8a284fa6-652d-4d6f-8d72-932ef36a1e2f",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Define matplotlib Axes\n",
    "# Number of Axes & plot size\n",
    "ncols = 2\n",
    "nrows = 1\n",
    "figsize = 8\n",
    "wspace = 0.1\n",
    "fig, axs = plt.subplots(\n",
    "    nrows=nrows,\n",
    "    ncols=ncols,\n",
    "    figsize=(ncols * figsize + figsize * wspace * (ncols - 1), nrows * figsize),\n",
    ")\n",
    "plt.subplots_adjust(wspace=wspace)\n",
    "\n",
    "sc.pl.umap(adata[adata.obs['sample']=='wt_pathogenic'],  ax=axs[0], show=False, color=['cell_type'], title=['Lymph node, Pathogenic WT'], add_outline=True, outline_width = (0.2,0.5), palette=ann_palette, alpha=0.7, s=10,legend_loc='on data',legend_fontsize=12, legend_fontweight='medium')\n",
    "sc.pl.umap(adata[adata.obs['sample']=='ko_pathogenic'], ax=axs[1],show=False, color=['cell_type'], title=['Lymph node, Pathogenic Cd74 KO'], add_outline=True, outline_width = (0.2,0.5), palette=ann_palette, alpha=0.7, s=10,legend_loc='on data',legend_fontsize=12, legend_fontweight='medium')\n",
    "\n",
    "plt.tight_layout()\n",
    "plt.savefig('/Users/oipulk/Documents/scRNASeq/data/Eleftheria_Maranou_Mar2024/analysis/figures/annotations/LN/umap_LN_annotations_pathogenic_wt_ko.pdf',dpi=600)\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "23e18910-c317-47db-ad08-f54d02de0882",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Define matplotlib Axes\n",
    "# Number of Axes & plot size\n",
    "ncols = 2\n",
    "nrows = 2\n",
    "figsize = 8\n",
    "wspace = 0.1\n",
    "fig, axs = plt.subplots(\n",
    "    nrows=nrows,\n",
    "   ncols=ncols,\n",
    "    figsize=(ncols * figsize + figsize * wspace * (ncols - 1), nrows * figsize),\n",
    ")\n",
    "plt.subplots_adjust(wspace=wspace)\n",
    "\n",
    "sc.pl.umap(adata[adata.obs['sample']=='wt_naive'], ax=axs[0,0], show=False, color=['cell_type'], title=['Lymph node, Naive WT'], add_outline=True, outline_width = (0.2,0.5), palette=ann_palette, alpha=0.7, s=10,legend_loc='on data',legend_fontsize=12, legend_fontweight='medium')\n",
    "sc.pl.umap(adata[adata.obs['sample']=='wt_pathogenic'],  ax=axs[0,1], show=False, color=['cell_type'], title=['Lymph node, Pathogenic WT'],add_outline=True, outline_width = (0.2,0.5), palette=ann_palette, alpha=0.7, s=10,legend_loc='on data',legend_fontsize=12, legend_fontweight='medium')\n",
    "sc.pl.umap(adata[adata.obs['sample']=='ko_naive'],  ax=axs[1,0],show=False, color=['cell_type'], title=['Lymph node, Naive Cd74 KO'], add_outline=True, outline_width = (0.2,0.5), palette=ann_palette, alpha=0.7, s=10,legend_loc='on data',legend_fontsize=12, legend_fontweight='medium')\n",
    "sc.pl.umap(adata[adata.obs['sample']=='ko_pathogenic'], show=False, ax=axs[1,1],color=['cell_type'], title=['Lymph node, Pathogenic Cd74 KO'], add_outline=True, outline_width = (0.2,0.5), palette=ann_palette, alpha=0.7, s=10,legend_loc='on data',legend_fontsize=12, legend_fontweight='medium')\n",
    "\n",
    "plt.tight_layout()\n",
    "plt.savefig('/Users/oipulk/Documents/scRNASeq/data/Eleftheria_Maranou_Mar2024/analysis/figures/annotations/LN/umap_LN_annotations_samples.pdf',dpi=600)\n",
    "plt.show()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "437a4344-247c-4f12-8efd-f5e9c78793aa",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Save data\n",
    "adata.write_h5ad(\"maranou_032024_LN_annotated.h5ad\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "1907835b-7799-4c08-9804-7c813c103451",
   "metadata": {},
   "source": [
    "## DCs/Monocytes/Macrophages"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "71a6febd-67f9-4325-a4aa-9abd3a31254a",
   "metadata": {},
   "outputs": [],
   "source": [
    "#Load annotated lymph node data\n",
    "adata = sc.read(\"maranou_032024_LN_annotated.h5ad\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "88352f37-2881-419d-8f3a-0313313ace05",
   "metadata": {},
   "outputs": [],
   "source": [
    "DC_subset = ['CCR7hi CD40+ CD80- CD86- DC','CCR7hi CD40+ CD80+ CD86+ DC','Lymphoid-resident cDC1','cDC2','pDC','Undefined DC']\n",
    "dc_mono_macro = DC_subset+['Monocytes and macrophages']+['Neutrophils']\n",
    "adata_dc_mono_macro = adata[adata.obs['cell_type'].isin(dc_mono_macro)].copy()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9f3c4a86-ef40-47db-98af-83bbbed7053f",
   "metadata": {},
   "outputs": [],
   "source": [
    "file_path = '/Users/oipulk/Documents/scRNASeq/data/Eleftheria_Maranou_Mar2024/analysis/figures/SI_Fig7/LN_DCMonoMacro_annotations.pdf'\n",
    "with plt.rc_context():  # Use this to set figure params like size and dpi\n",
    "    sc.pl.umap(adata_dc_mono_macro, color=['cell_type'],add_outline=True, outline_width = (0.2,0.3), palette=ann_palette, alpha=0.75, s=30, title='Lymph node myeloid cells',legend_loc='on data',legend_fontsize=12, legend_fontweight='heavy', show=False)\n",
    "    plt.savefig(file_path, bbox_inches=\"tight\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "17ed8b08-7201-408a-80be-fa9604e552fa",
   "metadata": {},
   "source": [
    "## Cell proportions in samples"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e8d8df7e-0e41-4aef-af90-01c812fd4bcf",
   "metadata": {},
   "outputs": [],
   "source": [
    "sns.set_style(\"ticks\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3e9c4e54-5a0f-4459-9c75-35c3050fdf46",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Get cell type counts\n",
    "cell_counts = adata[adata.obs['sample']=='wt_naive'].obs['cell_type'].value_counts()\n",
    "\n",
    "# Calculate proportions\n",
    "total_cells = cell_counts.sum()\n",
    "cell_proportions = cell_counts / total_cells\n",
    "\n",
    "# Sort cell types by proportion (descending)\n",
    "cell_proportions_sorted = cell_proportions.sort_values(ascending=True)\n",
    "\n",
    "# Create the plot\n",
    "fig, ax = plt.subplots(figsize=(6, 6))\n",
    "\n",
    "# Plot the stacked bar\n",
    "bottom = 0\n",
    "top = 1\n",
    "for cell_type, proportion in cell_proportions_sorted.items():\n",
    "    ax.bar(0, proportion, bottom=bottom, width=0.5, alpha=0.6,\n",
    "           color=ann_palette[cell_type], label=cell_type)\n",
    "    bottom += proportion\n",
    "    top -= proportion\n",
    "\n",
    "# Customize the plot\n",
    "ax.set_ylabel('Proportion of Cells')\n",
    "ax.set_title('Cell Type Proportions', fontsize=16)\n",
    "ax.set_xlim(-0.25, 0.25)\n",
    "ax.set_xticks([])  # Remove x-axis ticks\n",
    "\n",
    "bottom = 0\n",
    "top = 1\n",
    "# Add cell count and proportion labels\n",
    "for i, (cell_type, proportion) in enumerate(cell_proportions_sorted.items()):\n",
    "    count = cell_counts[cell_type]\n",
    "    y_position = bottom + proportion/2\n",
    "\n",
    "    if proportion>0.015:\n",
    "        ax.text(0, y_position, f'{cell_type} {count} ({proportion:.1%})', \n",
    "                ha='center', va='center', fontsize=12)\n",
    "    bottom += proportion\n",
    "    top -= proportion\n",
    "\n",
    "# Add a legend\n",
    "# plt.legend(bbox_to_anchor=(1.05, 1), loc='upper left', borderaxespad=0.)\n",
    "\n",
    "plt.tight_layout()\n",
    "plt.show()\n",
    "\n",
    "# Optionally, save the figure\n",
    "# plt.savefig('cell_type_proportions_stacked.png', dpi=300, bbox_inches='tight')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e3311887-262d-4a27-91b3-1542f2334b37",
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_cell_proportions(adata):\n",
    "    \n",
    "    cell_counts = adata.obs['cell_type'].value_counts()\n",
    "    total_cells = cell_counts.sum()\n",
    "    return cell_counts / total_cells\n",
    "\n",
    "# Get cell proportions for both AnnData objects\n",
    "adata1 = adata[adata.obs['sample']=='wt_naive']\n",
    "adata2 = adata[adata.obs['sample']=='ko_naive']\n",
    "adata3 = adata[adata.obs['sample']=='wt_pathogenic']\n",
    "adata4 = adata[adata.obs['sample']=='ko_pathogenic']\n",
    "\n",
    "cell_proportions1 = get_cell_proportions(adata1)\n",
    "cell_proportions2 = get_cell_proportions(adata2)\n",
    "cell_proportions3 = get_cell_proportions(adata3)\n",
    "cell_proportions4 = get_cell_proportions(adata4)\n",
    "\n",
    "# Combine all cell types from both datasets\n",
    "all_cell_types = sorted(set(cell_proportions1.index) | set(cell_proportions2.index) | \n",
    "                        set(cell_proportions3.index) | set(cell_proportions4.index))\n",
    "# Create a DataFrame with proportions from all datasets\n",
    "df = pd.DataFrame({\n",
    "    'WT Naive': cell_proportions1.reindex(all_cell_types).fillna(0),\n",
    "    'KO Naive': cell_proportions2.reindex(all_cell_types).fillna(0),\n",
    "    'WT Pathogenic': cell_proportions3.reindex(all_cell_types).fillna(0),\n",
    "    'KO Pathogenic': cell_proportions4.reindex(all_cell_types).fillna(0)\n",
    "})\n",
    "\n",
    "# Sort by the maximum proportion across all datasets\n",
    "df = df.sort_values(by=df.columns.tolist(), ascending=False)\n",
    "\n",
    "# Create the plot\n",
    "fig, ax = plt.subplots(figsize=(15, 8))\n",
    "\n",
    "ax.set_xlim(0., 4.0)  # Adjust x-axis limits\n",
    "\n",
    "# Plot the stacked bars\n",
    "x = [0.1, 0.9, 1.9, 2.7]  # x-coordinates for the four bars\n",
    "width = 0.7  # width of the bars\n",
    "\n",
    "for cell_type in df.index:\n",
    "    bottoms = [df.loc[:cell_type, col].sum() - df.loc[cell_type, col] for col in df.columns]\n",
    "    \n",
    "    ax.bar(x, df.loc[cell_type], bottom=bottoms, width=width, alpha=0.6,\n",
    "           color=ann_palette[cell_type], label=cell_type if x[0] == 0 else \"\")\n",
    "\n",
    "# Customize the plot\n",
    "ax.set_ylabel('Proportion of Cells', fontsize=14)\n",
    "# ax.set_title('Cell Type Proportions Comparison', fontsize=18)\n",
    "ax.set_xticks(x)\n",
    "ax.set_xticklabels(['WT', 'Cd74 KO','WT', 'Cd74 KO'], fontsize=12, rotation=0, ha='center')\n",
    "ax.set_xlim(-0.5, 3.5)\n",
    "\n",
    "# Add cell type labels\n",
    "for i, dataset in enumerate(df.columns):\n",
    "    bottom = 0\n",
    "    for cell_type in df.index:\n",
    "        if df.loc[cell_type, dataset] > 0.015:  # Only label if proportion > 1%\n",
    "            ax.text(x[i], bottom + df.loc[cell_type, dataset]/2, \n",
    "                    f'{cell_type} {df.loc[cell_type, dataset]:.1%}', \n",
    "                    ha='center', va='center', fontsize=12)\n",
    "        bottom += df.loc[cell_type, dataset]\n",
    "\n",
    "# Add a legend\n",
    "# plt.legend(bbox_to_anchor=(1.05, 1), loc='upper left', borderaxespad=0., fontsize=10)\n",
    "\n",
    "# Remove top and right spines\n",
    "ax.spines['top'].set_visible(False)\n",
    "ax.spines['right'].set_visible(False)\n",
    "\n",
    "ax.set_xlim(-0.4, 3.1)  # Adjust x-axis limits\n",
    "ax.text(0.25, 1.02, 'Naive', transform=ax.transAxes, ha='center', va='bottom', fontsize=16, fontweight='bold')\n",
    "ax.text(0.77, 1.02, 'Pathogenic', transform=ax.transAxes, ha='center', va='bottom', fontsize=16, fontweight='bold')\n",
    "\n",
    "# Optionally, save the figure\n",
    "plt.savefig('/Users/oipulk/Documents/scRNASeq/data/Eleftheria_Maranou_Mar2024/analysis/figures/annotations/LN/LN_cell_type_proportions_naive_vs_pathogenic.pdf', dpi=600, bbox_inches='tight')\n",
    "\n",
    "plt.show()\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8517c30d-61d1-4c66-bc8f-65e718e25cf0",
   "metadata": {},
   "outputs": [],
   "source": [
    "list(adata.obs['cell_type'].unique())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2119f2d1-a8b3-4e64-b46f-9a2c85fe04b2",
   "metadata": {},
   "outputs": [],
   "source": [
    "T_cell_subset = [\"CD8+\",'Treg','CD4+','Th cells','Heterogeneous T cells','CTLA4+ CD8+ T cells','$\\\\gamma \\\\delta$T cells','Activated CD4+','Activated CD8+']\n",
    "DC_subset = ['CCR7hi CD40+ CD80- CD86- DC','CCR7hi CD40+ CD80+ CD86+ DC','Lymphoid-resident cDC1','cDC2','pDC','Undefined DC']\n",
    "B_cell_subset = [\"Anti?gen-presenting B cells\",\"Naive B cells\",\"Activated B cells/Early plasmablasts\",\"Developing B cells\",\"MZB and B-1 cells\",\"Interferon-activated B cells\"]\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5985abfb-35dd-4285-9b41-c08006b0519d",
   "metadata": {},
   "outputs": [],
   "source": [
    "## Proportions of B cells\n",
    "\n",
    "adata_subset = adata[adata.obs['cell_type'].isin(B_cell_subset)].copy()\n",
    "\n",
    "# Get cell proportions for both AnnData objects\n",
    "adata1 = adata_subset[adata_subset.obs['sample']=='wt_naive']\n",
    "adata2 = adata_subset[adata_subset.obs['sample']=='ko_naive']\n",
    "adata3 = adata_subset[adata_subset.obs['sample']=='wt_pathogenic']\n",
    "adata4 = adata_subset[adata_subset.obs['sample']=='ko_pathogenic']\n",
    "\n",
    "cell_proportions1 = get_cell_proportions(adata1)\n",
    "cell_proportions2 = get_cell_proportions(adata2)\n",
    "cell_proportions3 = get_cell_proportions(adata3)\n",
    "cell_proportions4 = get_cell_proportions(adata4)\n",
    "\n",
    "# Combine all cell types from both datasets\n",
    "all_cell_types = sorted(set(cell_proportions1.index) | set(cell_proportions2.index) | \n",
    "                        set(cell_proportions3.index) | set(cell_proportions4.index))\n",
    "# Create a DataFrame with proportions from all datasets\n",
    "df = pd.DataFrame({\n",
    "    'WT Naive': cell_proportions1.reindex(all_cell_types).fillna(0),\n",
    "    'KO Naive': cell_proportions2.reindex(all_cell_types).fillna(0),\n",
    "    'WT Pathogenic': cell_proportions3.reindex(all_cell_types).fillna(0),\n",
    "    'KO Pathogenic': cell_proportions4.reindex(all_cell_types).fillna(0)\n",
    "})\n",
    "\n",
    "# Sort by the maximum proportion across all datasets\n",
    "df = df.sort_values(by=df.columns.tolist(), ascending=False)\n",
    "\n",
    "# Create the plot\n",
    "fig, ax = plt.subplots(figsize=(15, 8))\n",
    "\n",
    "ax.set_xlim(0., 4.0)  # Adjust x-axis limits\n",
    "\n",
    "# Plot the stacked bars\n",
    "x = [0.1, 0.9, 1.9, 2.7]  # x-coordinates for the four bars\n",
    "width = 0.7  # width of the bars\n",
    "\n",
    "for cell_type in df.index:\n",
    "    bottoms = [df.loc[:cell_type, col].sum() - df.loc[cell_type, col] for col in df.columns]\n",
    "    \n",
    "    ax.bar(x, df.loc[cell_type], bottom=bottoms, width=width, alpha=0.5,\n",
    "           color=ann_palette[cell_type], label=cell_type if x[0] == 0 else \"\")\n",
    "\n",
    "# Customize the plot\n",
    "ax.set_ylabel('Proportion of B Cells', fontsize=14)\n",
    "# ax.set_title('Cell Type Proportions Comparison', fontsize=18)\n",
    "ax.set_xticks(x)\n",
    "ax.set_xticklabels(['WT', 'Cd74 KO','WT', 'Cd74 KO'], fontsize=12, rotation=0, ha='center')\n",
    "ax.set_xlim(-0.5, 3.5)\n",
    "\n",
    "# Add cell type labels\n",
    "for i, dataset in enumerate(df.columns):\n",
    "    bottom = 0\n",
    "    for cell_type in df.index:\n",
    "        if df.loc[cell_type, dataset] > 0.01:  # Only label if proportion > 1%\n",
    "            ax.text(x[i], bottom + df.loc[cell_type, dataset]/2, \n",
    "                    f'{cell_type} {df.loc[cell_type, dataset]:.1%}', \n",
    "                    ha='center', va='center', fontsize=12)\n",
    "        bottom += df.loc[cell_type, dataset]\n",
    "\n",
    "# Add a legend\n",
    "# plt.legend(bbox_to_anchor=(1.05, 1), loc='upper left', borderaxespad=0., fontsize=10)\n",
    "\n",
    "# Remove top and right spines\n",
    "ax.spines['top'].set_visible(False)\n",
    "ax.spines['right'].set_visible(False)\n",
    "\n",
    "ax.set_xlim(-0.4, 3.1)  # Adjust x-axis limits\n",
    "ax.text(0.25, 1.02, 'Naive', transform=ax.transAxes, ha='center', va='bottom', fontsize=16, fontweight='bold')\n",
    "ax.text(0.77, 1.02, 'Pathogenic', transform=ax.transAxes, ha='center', va='bottom', fontsize=16, fontweight='bold')\n",
    "\n",
    "# Optionally, save the figure\n",
    "plt.savefig('/Users/oipulk/Documents/scRNASeq/data/Eleftheria_Maranou_Mar2024/analysis/figures/annotations/LN/LN_B_cell_type_proportions_naive_vs_pathogenic.pdf', dpi=600, bbox_inches='tight')\n",
    "\n",
    "plt.show()\n",
    "\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "81834e28-b71e-46e8-8a12-5ec74149b044",
   "metadata": {},
   "outputs": [],
   "source": [
    "## Proprotions of T cells\n",
    "\n",
    "adata_subset = adata[adata.obs['cell_type'].isin(T_cell_subset)].copy()\n",
    "\n",
    "# Get cell proportions for both AnnData objects\n",
    "adata1 = adata_subset[adata_subset.obs['sample']=='wt_naive']\n",
    "adata2 = adata_subset[adata_subset.obs['sample']=='ko_naive']\n",
    "adata3 = adata_subset[adata_subset.obs['sample']=='wt_pathogenic']\n",
    "adata4 = adata_subset[adata_subset.obs['sample']=='ko_pathogenic']\n",
    "\n",
    "cell_proportions1 = get_cell_proportions(adata1)\n",
    "cell_proportions2 = get_cell_proportions(adata2)\n",
    "cell_proportions3 = get_cell_proportions(adata3)\n",
    "cell_proportions4 = get_cell_proportions(adata4)\n",
    "\n",
    "# Combine all cell types from both datasets\n",
    "all_cell_types = sorted(set(cell_proportions1.index) | set(cell_proportions2.index) | \n",
    "                        set(cell_proportions3.index) | set(cell_proportions4.index))\n",
    "# Create a DataFrame with proportions from all datasets\n",
    "df = pd.DataFrame({\n",
    "    'WT Naive': cell_proportions1.reindex(all_cell_types).fillna(0),\n",
    "    'KO Naive': cell_proportions2.reindex(all_cell_types).fillna(0),\n",
    "    'WT Pathogenic': cell_proportions3.reindex(all_cell_types).fillna(0),\n",
    "    'KO Pathogenic': cell_proportions4.reindex(all_cell_types).fillna(0)\n",
    "})\n",
    "\n",
    "# Sort by the maximum proportion across all datasets\n",
    "df = df.sort_values(by=df.columns.tolist(), ascending=False)\n",
    "\n",
    "# Create the plot\n",
    "fig, ax = plt.subplots(figsize=(15, 8))\n",
    "\n",
    "ax.set_xlim(0., 4.0)  # Adjust x-axis limits\n",
    "\n",
    "# Plot the stacked bars\n",
    "x = [0.1, 0.9, 1.9, 2.7]  # x-coordinates for the four bars\n",
    "width = 0.7  # width of the bars\n",
    "\n",
    "for cell_type in df.index:\n",
    "    bottoms = [df.loc[:cell_type, col].sum() - df.loc[cell_type, col] for col in df.columns]\n",
    "    \n",
    "    ax.bar(x, df.loc[cell_type], bottom=bottoms, width=width, alpha=0.5,\n",
    "           color=ann_palette[cell_type], label=cell_type if x[0] == 0 else \"\")\n",
    "\n",
    "# Customize the plot\n",
    "ax.set_ylabel('Proportion of T Cells', fontsize=14)\n",
    "# ax.set_title('Cell Type Proportions Comparison', fontsize=18)\n",
    "ax.set_xticks(x)\n",
    "ax.set_xticklabels(['WT', 'Cd74 KO','WT', 'Cd74 KO'], fontsize=12, rotation=0, ha='center')\n",
    "ax.set_xlim(-0.5, 3.5)\n",
    "\n",
    "# Add cell type labels\n",
    "for i, dataset in enumerate(df.columns):\n",
    "    bottom = 0\n",
    "    for cell_type in df.index:\n",
    "        if df.loc[cell_type, dataset] > 0.01:  # Only label if proportion > 1.5%\n",
    "            ax.text(x[i], bottom + df.loc[cell_type, dataset]/2, \n",
    "                    f'{cell_type} {df.loc[cell_type, dataset]:.1%}', \n",
    "                    ha='center', va='center', fontsize=12)\n",
    "        bottom += df.loc[cell_type, dataset]\n",
    "\n",
    "# Add a legend\n",
    "# plt.legend(bbox_to_anchor=(1.05, 1), loc='upper left', borderaxespad=0., fontsize=10)\n",
    "\n",
    "# Remove top and right spines\n",
    "ax.spines['top'].set_visible(False)\n",
    "ax.spines['right'].set_visible(False)\n",
    "\n",
    "ax.set_xlim(-0.4, 3.1)  # Adjust x-axis limits\n",
    "ax.text(0.25, 1.02, 'Naive', transform=ax.transAxes, ha='center', va='bottom', fontsize=16, fontweight='bold')\n",
    "ax.text(0.77, 1.02, 'Pathogenic', transform=ax.transAxes, ha='center', va='bottom', fontsize=16, fontweight='bold')\n",
    "\n",
    "# Optionally, save the figure\n",
    "plt.savefig('/Users/oipulk/Documents/scRNASeq/data/Eleftheria_Maranou_Mar2024/analysis/figures/annotations/LN/LN_T_cell_type_proportions_naive_vs_pathogenic.pdf', dpi=600, bbox_inches='tight')\n",
    "\n",
    "plt.show()\n",
    "\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "898fab94-0bcc-4a63-9824-4f4468703eba",
   "metadata": {},
   "outputs": [],
   "source": [
    "## Proportions of DCs\n",
    "\n",
    "adata_subset = adata[adata.obs['cell_type'].isin(DC_subset)].copy()\n",
    "\n",
    "# Get cell proportions for both AnnData objects\n",
    "adata1 = adata_subset[adata_subset.obs['sample']=='wt_naive']\n",
    "adata2 = adata_subset[adata_subset.obs['sample']=='ko_naive']\n",
    "adata3 = adata_subset[adata_subset.obs['sample']=='wt_pathogenic']\n",
    "adata4 = adata_subset[adata_subset.obs['sample']=='ko_pathogenic']\n",
    "\n",
    "cell_proportions1 = get_cell_proportions(adata1)\n",
    "cell_proportions2 = get_cell_proportions(adata2)\n",
    "cell_proportions3 = get_cell_proportions(adata3)\n",
    "cell_proportions4 = get_cell_proportions(adata4)\n",
    "\n",
    "# Combine all cell types from both datasets\n",
    "all_cell_types = sorted(set(cell_proportions1.index) | set(cell_proportions2.index) | \n",
    "                        set(cell_proportions3.index) | set(cell_proportions4.index))\n",
    "# Create a DataFrame with proportions from all datasets\n",
    "df = pd.DataFrame({\n",
    "    'WT Naive': cell_proportions1.reindex(all_cell_types).fillna(0),\n",
    "    'KO Naive': cell_proportions2.reindex(all_cell_types).fillna(0),\n",
    "    'WT Pathogenic': cell_proportions3.reindex(all_cell_types).fillna(0),\n",
    "    'KO Pathogenic': cell_proportions4.reindex(all_cell_types).fillna(0)\n",
    "})\n",
    "\n",
    "# Sort by the maximum proportion across all datasets\n",
    "df = df.sort_values(by=df.columns.tolist(), ascending=False)\n",
    "\n",
    "# Create the plot\n",
    "fig, ax = plt.subplots(figsize=(15, 8))\n",
    "\n",
    "ax.set_xlim(0., 4.0)  # Adjust x-axis limits\n",
    "\n",
    "# Plot the stacked bars\n",
    "x = [0.1, 0.9, 1.9, 2.7]  # x-coordinates for the four bars\n",
    "width = 0.7  # width of the bars\n",
    "\n",
    "for cell_type in df.index:\n",
    "    bottoms = [df.loc[:cell_type, col].sum() - df.loc[cell_type, col] for col in df.columns]\n",
    "    \n",
    "    ax.bar(x, df.loc[cell_type], bottom=bottoms, width=width, alpha=0.5,\n",
    "           color=ann_palette[cell_type], label=cell_type if x[0] == 0 else \"\")\n",
    "\n",
    "# Customize the plot\n",
    "ax.set_ylabel('Proportion of DCs', fontsize=14)\n",
    "# ax.set_title('Cell Type Proportions Comparison', fontsize=18)\n",
    "ax.set_xticks(x)\n",
    "ax.set_xticklabels(['WT', 'Cd74 KO','WT', 'Cd74 KO'], fontsize=12, rotation=0, ha='center')\n",
    "ax.set_xlim(-0.5, 3.5)\n",
    "\n",
    "# Add cell type labels\n",
    "for i, dataset in enumerate(df.columns):\n",
    "    bottom = 0\n",
    "    for cell_type in df.index:\n",
    "\n",
    "        if df.loc[cell_type, dataset] < 0.015: \n",
    "            if df.loc[cell_type, dataset] < 0.01: \n",
    "                print(cell_type)            \n",
    "                ax.text(x[i], bottom + 0.02, \n",
    "                    f'{cell_type} {df.loc[cell_type, dataset]:.1%}', \n",
    "                    ha='center', va='center', fontsize=12)\n",
    "                \n",
    "            else:\n",
    "                \n",
    "                print(cell_type)            \n",
    "                ax.text(x[i], bottom + 0.015, \n",
    "                    f'{cell_type} {df.loc[cell_type, dataset]:.1%}', \n",
    "                    ha='center', va='center', fontsize=12)\n",
    "                \n",
    "        else:\n",
    "\n",
    "            ax.text(x[i], bottom + df.loc[cell_type, dataset]/2, \n",
    "                f'{cell_type} {df.loc[cell_type, dataset]:.1%}', \n",
    "                ha='center', va='center', fontsize=12)\n",
    "        \n",
    "        bottom += df.loc[cell_type, dataset]\n",
    "\n",
    "# Add a legend\n",
    "# plt.legend(bbox_to_anchor=(1.05, 1), loc='upper left', borderaxespad=0., fontsize=10)\n",
    "\n",
    "# Remove top and right spines\n",
    "ax.spines['top'].set_visible(False)\n",
    "ax.spines['right'].set_visible(False)\n",
    "\n",
    "ax.set_xlim(-0.4, 3.1)  # Adjust x-axis limits\n",
    "ax.text(0.25, 1.02, 'Naive', transform=ax.transAxes, ha='center', va='bottom', fontsize=16, fontweight='bold')\n",
    "ax.text(0.77, 1.02, 'Pathogenic', transform=ax.transAxes, ha='center', va='bottom', fontsize=16, fontweight='bold')\n",
    "\n",
    "# Optionally, save the figure\n",
    "plt.savefig('/Users/oipulk/Documents/scRNASeq/data/Eleftheria_Maranou_Mar2024/analysis/figures/annotations/LN/LN_DC_type_proportions_naive_vs_pathogenic.pdf', dpi=600, bbox_inches='tight')\n",
    "\n",
    "plt.show()\n",
    "\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f175aa1e-bc53-498a-851d-cf6056f0167d",
   "metadata": {},
   "outputs": [],
   "source": [
    "## WT vs Cd74 KO\n",
    "\n",
    "def get_cell_proportions(adata):\n",
    "    \n",
    "    cell_counts = adata.obs['cell_type'].value_counts()\n",
    "    total_cells = cell_counts.sum()\n",
    "    return cell_counts / total_cells\n",
    "\n",
    "# Get cell proportions for both AnnData objects\n",
    "adata1 = adata[adata.obs['sample']=='wt_naive']\n",
    "adata2 = adata[adata.obs['sample']=='wt_pathogenic']\n",
    "adata3 = adata[adata.obs['sample']=='ko_naive']\n",
    "adata4 = adata[adata.obs['sample']=='ko_pathogenic']\n",
    "\n",
    "cell_proportions1 = get_cell_proportions(adata1)\n",
    "cell_proportions2 = get_cell_proportions(adata2)\n",
    "cell_proportions3 = get_cell_proportions(adata3)\n",
    "cell_proportions4 = get_cell_proportions(adata4)\n",
    "\n",
    "# Combine all cell types from both datasets\n",
    "all_cell_types = sorted(set(cell_proportions1.index) | set(cell_proportions2.index) | \n",
    "                        set(cell_proportions3.index) | set(cell_proportions4.index))\n",
    "# Create a DataFrame with proportions from all datasets\n",
    "df = pd.DataFrame({\n",
    "    'WT Naive': cell_proportions1.reindex(all_cell_types).fillna(0),\n",
    "    'WT Pathogenic': cell_proportions2.reindex(all_cell_types).fillna(0),\n",
    "    'KO Naive': cell_proportions3.reindex(all_cell_types).fillna(0),\n",
    "    'KO Pathogenic': cell_proportions4.reindex(all_cell_types).fillna(0)\n",
    "})\n",
    "\n",
    "# Sort by the maximum proportion across all datasets\n",
    "df = df.sort_values(by=df.columns.tolist(), ascending=False)\n",
    "\n",
    "# Create the plot\n",
    "fig, ax = plt.subplots(figsize=(15, 8))\n",
    "\n",
    "\n",
    "\n",
    "# Plot the stacked bars\n",
    "x = [0.1, 0.9, 1.9, 2.7]  # x-coordinates for the four bars\n",
    "width = 0.7  # width of the bars\n",
    "\n",
    "for cell_type in df.index:\n",
    "    bottoms = [df.loc[:cell_type, col].sum() - df.loc[cell_type, col] for col in df.columns]\n",
    "    \n",
    "    ax.bar(x, df.loc[cell_type], bottom=bottoms, width=width, alpha=0.7,\n",
    "           color=ann_palette[cell_type], label=cell_type if x[0] == 0 else \"\")\n",
    "\n",
    "# Customize the plot\n",
    "ax.set_ylabel('Proportion of Cells', fontsize=14)\n",
    "# ax.set_title('Cell Type Proportions Comparison', fontsize=18)\n",
    "ax.set_xticks(x)\n",
    "ax.set_xticklabels(['Naive', 'Pathogenic','Naive', 'Pathogenic'], fontsize=12, rotation=0, ha='center')\n",
    "ax.set_xlim(-0.5, 3.5)\n",
    "\n",
    "# Add cell type labels\n",
    "for i, dataset in enumerate(df.columns):\n",
    "    bottom = 0\n",
    "    for cell_type in df.index:\n",
    "        if df.loc[cell_type, dataset] > 0.015:  # Only label if proportion > 1%\n",
    "            ax.text(x[i], bottom + df.loc[cell_type, dataset]/2, \n",
    "                    f'{cell_type} {df.loc[cell_type, dataset]:.1%}', \n",
    "                    ha='center', va='center', fontsize=12)\n",
    "        bottom += df.loc[cell_type, dataset]\n",
    "\n",
    "# Add a legend\n",
    "# plt.legend(bbox_to_anchor=(1.05, 1), loc='upper left', borderaxespad=0., fontsize=10)\n",
    "\n",
    "# Remove top and right spines\n",
    "ax.spines['top'].set_visible(False)\n",
    "ax.spines['right'].set_visible(False)\n",
    "\n",
    "\n",
    "ax.text(0.25, 1.02, 'WT', transform=ax.transAxes, ha='center', va='bottom', fontsize=14, fontweight='bold')\n",
    "ax.text(0.77, 1.02, 'Cd74 KO', transform=ax.transAxes, ha='center', va='bottom', fontsize=14, fontweight='bold')\n",
    "ax.set_xlim(-0.4, 3.1)  # Adjust x-axis limits\n",
    "\n",
    "plt.savefig('/Users/oipulk/Documents/scRNASeq/data/Eleftheria_Maranou_Mar2024/analysis/figures/annotations/LN/LN_cell_type_proportions_wt_vs_ko.pdf', dpi=600, bbox_inches='tight')\n",
    "\n",
    "plt.show()\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "676a1412-934e-4b05-95aa-7ac94d11272e",
   "metadata": {},
   "source": [
    "## DC marker expression in DC clusters"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c1d0ae49-7472-4ce4-a419-92f67492dd16",
   "metadata": {},
   "outputs": [],
   "source": [
    "### Dotplot of key markers in DC clusters \n",
    "common_cdc_markers = ['Flt3', 'Zbtb46', 'Kit']\n",
    "cdc1_markers = ['Irf8', 'Batf3', 'Nfil3', 'Id2', 'Bcl6', 'Xcr1', 'Rab43', 'Itgax', 'Itgae', 'Cd24a', 'Cd8a']\n",
    "cdc2_markers = ['Relb', 'Esam', 'Itgam', 'Irf4', 'Sirpa']\n",
    "additional_cdc_markers = ['Clec9a','Ccr2', 'Ccr7']\n",
    "dc_activation_markers = ['Cd40', 'Cd80', 'Cd86'] \n",
    "pDC_markers = ['Ptprc','Bst2','Siglech','Tcf4','Zeb2','Irf7','Pacsin1','Spib','Bcl11b']\n",
    "\n",
    "all_dc_markers = np.sort(common_cdc_markers+cdc1_markers+cdc2_markers+additional_cdc_markers+dc_activation_markers+pDC_markers)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1ce283df-5656-41ff-8c36-b9aec8e21fcf",
   "metadata": {},
   "outputs": [],
   "source": [
    "adata_dc_hr = adata[adata.obs['cell_type'].isin(DC_subset)].copy()\n",
    "\n",
    "sc.tl.dendrogram(adata_dc_hr, groupby='cell_type')\n",
    "# sc.pl.dendrogram(adata_dc_lr, groupby='cell_type_low_res')\n",
    "sc.pl.dotplot(adata_dc_hr, var_names=all_dc_markers, groupby='cell_type', cmap='Blues', save='LN_dc_markers_annotations.pdf')\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a294f1b6-b7b3-4a38-8f49-8a01ff9d3b31",
   "metadata": {},
   "outputs": [],
   "source": [
    "sc.pl.umap(adata_dc_hr, color=['cell_type'],add_outline=True, outline_width = (0.2,0.5), palette=ann_palette, alpha=0.7, s=10, title='Lymph node',legend_loc='on data',legend_fontsize=8, legend_fontweight='heavy', save='_LN_DC_annotations.pdf')\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "5d86fc70-a594-49da-a210-465d69cea528",
   "metadata": {},
   "source": [
    "## Markers used in annotation"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "cf94970d-b03f-4396-aa2b-f9fe6b92b915",
   "metadata": {},
   "source": [
    "## T cell markers"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f992fd7c-c164-4e6d-9135-de9956a21f80",
   "metadata": {},
   "outputs": [],
   "source": [
    "# cd8_genes = ['Cd3e', 'Cd160', 'Cd8a', 'Pdcd1']\n",
    "\n",
    "#  TM tissue specific markers \n",
    "# cd8_genes = ['Cd3e',  'Cd8a','Gzma', 'Prf1']\n",
    "# adata.obs['cd8_score'] = adata[:,cd8_genes].X.sum(1)\n",
    "\n",
    "#Cd8+ T cells\n",
    "cd8_genes = ['Cd3e',  'Cd8a']\n",
    "adata.obs['cd8_score'] = adata[:,cd8_genes].X.sum(1)\n",
    "\n",
    "#Cd4+ T cells\n",
    "cd4_genes = ['Cd3e','Cd4']\n",
    "adata.obs['cd4_score'] = adata[:,cd4_genes].X.sum(1)\n",
    "\n",
    "Treg_genes =['Cd4','Foxp3', 'Ctla4']\n",
    "adata.obs['Treg_score'] = adata[:,Treg_genes].X.sum(1)\n",
    "\n",
    "Th_genes = ['Cxcr6','Bcl6','Prdm1','Tbx21']\n",
    "adata.obs['Th_score'] = adata[:,Th_genes].X.sum(1)\n",
    "\n",
    "cd8effector_genes = ['Cd8a','Cx3cr1','Gzmb','Lgals1','S1pr5']\n",
    "adata.obs['cd8effector_score'] = adata[:,cd8effector_genes].X.sum(1)\n",
    "TRM_genes = ['Itgae','Cd69']\n",
    "adata.obs['TRM_score'] = adata[:,TRM_genes].X.sum(1)\n",
    "\n",
    "\n",
    "sc.pl.umap(adata, color=['cd8_score', 'cd4_score'],palette='tab20',cmap='coolwarm',vmax=4)\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "433c1eb3-c086-40f5-8e27-3c6519c8eafd",
   "metadata": {},
   "outputs": [],
   "source": [
    "sc.pl.umap(adata, color=['Treg_score','Th_score'],palette='tab20',cmap='coolwarm', vmax=4)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "692cef8c-6d14-4505-b4eb-2e2cefb22a24",
   "metadata": {},
   "outputs": [],
   "source": [
    "ncols = 2\n",
    "nrows = 2\n",
    "figsize = 8\n",
    "wspace = 0.1\n",
    "fig, axs = plt.subplots(\n",
    "    nrows=nrows,\n",
    "    ncols=ncols,\n",
    "    figsize=(ncols * figsize + figsize * wspace * (ncols - 1), nrows * figsize),\n",
    ")\n",
    "plt.subplots_adjust(wspace=wspace)\n",
    "sc.pl.umap(adata[adata.obs['sample']=='wt_naive'], ax=axs[0,0], show=False, color=['Cd3e'],palette='tab20',cmap='coolwarm',vmax=4)\n",
    "sc.pl.umap(adata[adata.obs['sample']=='wt_pathogenic'], ax=axs[0,1], show=False,color=['Cd3e'],palette='tab20',cmap='coolwarm',vmax=4)\n",
    "sc.pl.umap(adata[adata.obs['sample']=='ko_naive'], ax=axs[1,0], show=False, color=['Cd3e'],palette='tab20',cmap='coolwarm',vmax=4)\n",
    "sc.pl.umap(adata[adata.obs['sample']=='ko_pathogenic'], ax=axs[1,1], color=['Cd3e'],palette='tab20',cmap='coolwarm',vmax=4)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "26abf1d5-9fc8-48de-926d-aceec74d41d7",
   "metadata": {},
   "outputs": [],
   "source": [
    "ncols = 2\n",
    "nrows = 2\n",
    "figsize = 8\n",
    "wspace = 0.1\n",
    "fig, axs = plt.subplots(\n",
    "    nrows=nrows,\n",
    "    ncols=ncols,\n",
    "    figsize=(ncols * figsize + figsize * wspace * (ncols - 1), nrows * figsize),\n",
    ")\n",
    "plt.subplots_adjust(wspace=wspace)\n",
    "sc.pl.umap(adata[adata.obs['sample']=='wt_naive'], ax=axs[0,0], show=False, color=['Cd8a'],palette='tab20',cmap='coolwarm',vmax=3)\n",
    "sc.pl.umap(adata[adata.obs['sample']=='wt_pathogenic'], ax=axs[0,1], show=False,color=['Cd8a'],palette='tab20',cmap='coolwarm',vmax=3)\n",
    "sc.pl.umap(adata[adata.obs['sample']=='ko_naive'], ax=axs[1,0], show=False, color=['Cd8a'],palette='tab20',cmap='coolwarm',vmax=3)\n",
    "sc.pl.umap(adata[adata.obs['sample']=='ko_pathogenic'], ax=axs[1,1], color=['Cd8a'],palette='tab20',cmap='coolwarm',vmax=3)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9129cb0a-8aef-4620-b629-9784da1d1ce8",
   "metadata": {},
   "outputs": [],
   "source": [
    "ncols = 2\n",
    "nrows = 2\n",
    "figsize = 8\n",
    "wspace = 0.1\n",
    "fig, axs = plt.subplots(\n",
    "    nrows=nrows,\n",
    "    ncols=ncols,\n",
    "    figsize=(ncols * figsize + figsize * wspace * (ncols - 1), nrows * figsize),\n",
    ")\n",
    "plt.subplots_adjust(wspace=wspace)\n",
    "sc.pl.umap(adata[adata.obs['sample']=='wt_naive'], ax=axs[0,0], show=False, color=['Cd4'],palette='tab20',cmap='coolwarm',vmax=3)\n",
    "sc.pl.umap(adata[adata.obs['sample']=='wt_pathogenic'], ax=axs[0,1], show=False,color=['Cd4'],palette='tab20',cmap='coolwarm',vmax=3)\n",
    "sc.pl.umap(adata[adata.obs['sample']=='ko_naive'], ax=axs[1,0], show=False, color=['Cd4'],palette='tab20',cmap='coolwarm',vmax=3)\n",
    "sc.pl.umap(adata[adata.obs['sample']=='ko_pathogenic'], ax=axs[1,1], color=['Cd4'],palette='tab20',cmap='coolwarm',vmax=3)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3a00a463-79f6-40b9-8329-448488a7007f",
   "metadata": {},
   "outputs": [],
   "source": [
    "# This shows CTLA4+ CD8+ T cells \n",
    "sc.pl.umap(adata, color=['Cd8a','Ctla4'],palette='tab20',cmap='coolwarm',vmax=2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d1d49264-8d06-4a3e-853f-34f792034e2e",
   "metadata": {},
   "outputs": [],
   "source": [
    "#CD3E+ CD8A- CD4-\n",
    "adata.obs['DNTscore'] = (np.asarray(1-adata[:,'Cd8a'].X.sum(1)/np.max(adata[:,'Cd8a'].X.sum(1)))*np.asarray(1-adata[:,'Cd4'].X.sum(1)/np.max(adata[:,'Cd4'].X.sum(1))))*np.asarray(adata[:,'Cd3e'].X.sum(1)/np.max(adata[:,'Cd3e'].X.sum(1)))\n",
    "\n",
    "# CD3E+ CD8A+ CD4+\n",
    "adata.obs['DPTscore'] = (np.asarray(adata[:,'Cd8a'].X.sum(1)/np.max(adata[:,'Cd8a'].X.sum(1)))*np.asarray(adata[:,'Cd4'].X.sum(1)/np.max(adata[:,'Cd4'].X.sum(1))))\n",
    "\n",
    "#CD3E- CD8A- CD4-\n",
    "adata.obs['TNTscore'] = (np.asarray(1 - adata[:,'Cd8a'].X.sum(1)/np.max(adata[:,'Cd8a'].X.sum(1)))*np.asarray(1 - adata[:,'Cd4'].X.sum(1)/np.max(adata[:,'Cd4'].X.sum(1))))*np.asarray(1- adata[:,'Cd3e'].X.sum(1)/np.max(adata[:,'Cd3e'].X.sum(1)))\n",
    "ncols = 2\n",
    "nrows = 1\n",
    "figsize = 8\n",
    "wspace = 0.1\n",
    "fig, axs = plt.subplots(\n",
    "    nrows=nrows,\n",
    "    ncols=ncols,\n",
    "    figsize=(ncols * figsize + figsize * wspace * (ncols - 1), nrows * figsize),\n",
    ")\n",
    "sc.pl.umap(adata, color=['DNTscore'],palette='tab20',cmap='coolwarm', ax=axs[0],vmax=1, show=False)\n",
    "sc.pl.umap(adata, color=['TNTscore'],palette='tab20',cmap='coolwarm', ax=axs[1],vmax=1)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "fe499a6e-f971-47ac-af5e-75edaae8cc0d",
   "metadata": {},
   "outputs": [],
   "source": [
    "sc.pl.umap(adata, color=['Pdcd1'],palette='tab20',cmap='coolwarm',vmax=0.8)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9e649b16-183a-499a-84e8-2cb91900eabf",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Fcgr3 is the gene name for CD16\n",
    "sc.pl.umap(adata, color=['Ncam1', 'Fcgr3'],palette='tab20',cmap='coolwarm', vmax=1)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ae14f69a-84f7-42d7-b8ca-4c3bbb3a6504",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Sell is alias for CD62L (naive/early memory T cell marker for central memory T cells)\n",
    "memory_genes = ['Cd44','Sell','Il7r']\n",
    "Trm_genes = ['Cd44','Sell','Il7r', 'Cd69','Itgae' ]\n",
    "\n",
    "adata.obs['memory_score'] = adata[:,memory_genes].X.sum(1)\n",
    "adata.obs['Trm_score'] = adata[:,Trm_genes].X.sum(1)\n",
    "sc.pl.umap(adata, color=['memory_score','Trm_score'],palette='tab20',cmap='coolwarm')\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "eb57d10a-5033-4ceb-88a1-fb5042333987",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Tcm markers\n",
    "sc.pl.umap(adata, color=['Sell', 'Ccr7'],palette='tab20',cmap='coolwarm', vmax=3)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9240b933-cc73-4e59-bddc-835022b1917b",
   "metadata": {},
   "source": [
    "### NK cells"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "46888644-88f2-4a19-8571-cc1172330acb",
   "metadata": {},
   "outputs": [],
   "source": [
    "nk_genes = ['Klrb1c', 'Ncr1','Fcgr3','Itgal']\n",
    "adata.obs['nk_score'] = adata[:,nk_genes].X.sum(1)\n",
    "sc.pl.umap(adata, color=['nk_score'],palette='tab20',cmap='coolwarm')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "409b973f-16ea-47c3-9b3c-1efe298f831a",
   "metadata": {},
   "source": [
    "### Naive B"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2552314a-bb09-4d0f-8dd8-1659a31f04f3",
   "metadata": {},
   "outputs": [],
   "source": [
    "adata.var_names[adata.var_names.str.startswith('Pdcd1')]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0927670b-d34d-4af7-bf24-a1a41e6ee52c",
   "metadata": {},
   "outputs": [],
   "source": [
    "naiveB_genes = ['Ighd', 'Ighm']\n",
    "memoryB_genes = ['Itgam','Cd80','Cxcr3','Nt5e','Pdcd1lg2']\n",
    "adata.obs['naiveB_score'] = adata[:,naiveB_genes].X.sum(1)\n",
    "adata.obs['memoryB_score'] = adata[:,memoryB_genes].X.sum(1)\n",
    "sc.pl.umap(adata, color=['naiveB_score','memoryB_score'],palette='tab20',cmap='coolwarm')\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d1d92615-f222-408b-ab9c-70f849c0833f",
   "metadata": {},
   "source": [
    "### Age-associated B"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "252bbccc-6acc-4be7-87ac-6a6a00f33187",
   "metadata": {},
   "outputs": [],
   "source": [
    "ageB_genes = ['Ighm','Il10', 'Ifng', 'Itgax', 'Tbx2', 'Itgam', 'Fas']\n",
    "\n",
    "adata.obs['ageB_score'] = adata[:,ageB_genes].X.sum(1)\n",
    "sc.pl.umap(adata, color=['ageB_score'],palette='tab20',cmap='coolwarm',vmax=5)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f42969d4-0bd1-43f9-a6dd-0f1f5f62c469",
   "metadata": {},
   "source": [
    "### marginal Zone B cells"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8b73c01c-60ff-42ef-92c6-666ac69ea84a",
   "metadata": {},
   "outputs": [],
   "source": [
    "mzB_genes = ['Cd9','Cr2','Spib']\n",
    "\n",
    "adata.obs['mzB_score'] = adata[:,mzB_genes].X.sum(1)\n",
    "sc.pl.umap(adata, color=mzB_genes,palette='tab20',cmap='coolwarm',vmax=2)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "1c6c7df6-c9e0-4bff-abed-55a39cd90a6b",
   "metadata": {},
   "source": [
    "### B-1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9f1c1001-f27a-4b6a-bde9-0451fc32e819",
   "metadata": {},
   "outputs": [],
   "source": [
    "B1_genes = ['Cd5','Spn','Ptpn22']\n",
    "sc.pl.umap(adata, color=B1_genes,palette='tab20',cmap='coolwarm',vmax=1)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "bd4b5836-ecce-4518-aebf-51800307e9bc",
   "metadata": {},
   "source": [
    "### Common innate-like B cell markers"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "bc37507c-0b19-4793-81f0-b2e9fe97c386",
   "metadata": {},
   "outputs": [],
   "source": [
    "innate_genes = ['Cd19','Cd24a','Cd38']\n",
    "sc.pl.umap(adata, color=innate_genes,palette='tab20',cmap='coolwarm',vmax=2)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "3ac36284-f37a-4e8b-b299-9ccbe27b7a96",
   "metadata": {},
   "source": [
    "### DC"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a5ed7161-afad-4b78-bf13-1e2fb33547ca",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Gene ref: Saito 2022, The role of type-2 conventional dendritic cells in the regulation of tumor immunity\n",
    "dc1_saito = ['Xcr1','Itgax','Clec9a','Cd8a','Ly75']\n",
    "dc1_act = ['Clec9a','Itgax']\n",
    "dc1_shi_etal = ['Xcr1','Tlr3','Clec9a','Cadm1']\n",
    "\n",
    "adata_subset = adata[adata.obs['leiden'].isin(['r0','r1','r2','r3','r4','r5','r6','r7','r8','r9','20','21'])].copy()\n",
    "\n",
    "sc.tl.score_genes(adata, dc1_saito, score_name='dc1_saito_score')\n",
    "sc.tl.score_genes(adata, dc1_act, score_name='dc1_act_score')\n",
    "sc.tl.score_genes(adata, dc1_shi_etal, score_name='dc1_shi_etal_score')\n",
    "\n",
    "sc.pl.umap(adata_subset, color=['dc1_shi_etal_score','dc1_act_score'],palette='tab20',cmap='coolwarm',vmax=1.5)\n",
    "\n",
    "# Mouse pDC's express CD11c (Itgax) at low level. cDC2 express Itgax, Itgam and Sirpa\n",
    "# Shi et al state that Sirpa is expressed in humans, and Cd11b (Itgam) in mice. We do observe Sirpa\n",
    "\n",
    "dc2_saito = ['Itgax','Itgam','Sirpa']\n",
    "dc2_act = ['Fcgr1','Itgam','Sirpa']\n",
    "dc2_combined = list(np.append(dc2_saito,dc2_act))\n",
    "\n",
    "sc.tl.score_genes(adata, dc2_saito, score_name='dc2_saito_score')\n",
    "sc.tl.score_genes(adata, dc2_act, score_name='dc2_act_score')\n",
    "sc.tl.score_genes(adata, dc2_combined, score_name='dc2_combined_score')\n",
    "\n",
    "\n",
    "sc.pl.umap(adata_subset, color=['dc2_saito_score','dc2_act_score'],palette='tab20',cmap='coolwarm',vmax=1.0)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a594f3c4-b37d-4dcc-b50c-db6cdde1c975",
   "metadata": {},
   "outputs": [],
   "source": [
    "sc.pl.umap(adata_DC1_subset, color=['Irf8','Irf4'],palette='tab20',cmap='coolwarm',vmin=0)\n",
    "sc.pl.umap(adata_DC1_subset, color=['Xcr1','Itgax'],palette='tab20',cmap='coolwarm',vmin=0)\n",
    "sc.pl.umap(adata_DC1_subset, color=['Clec9a','Cd24a'],palette='tab20',cmap='coolwarm',vmin=0)\n",
    "sc.pl.umap(adata_DC1_subset, color=['Klf4','Tbx21'],palette='tab20',cmap='coolwarm',vmin=0)\n",
    "sc.pl.umap(adata_DC1_subset, color=['Relb','Tbx21'],palette='tab20',cmap='coolwarm',vmin=0)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4bec2289-92fd-42db-884f-8d4e85e9e54b",
   "metadata": {},
   "outputs": [],
   "source": [
    "sc.pl.umap(adata_DC1_subset, color=['Cd8a','Sirpa','Xcr1'],palette='tab20',cmap='coolwarm',vmin=0)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c1215ac8-2f52-4c4c-9b0f-51d2c9378ae6",
   "metadata": {},
   "outputs": [],
   "source": [
    "sc.pl.umap(adata_DC1_subset, color=['Cd80', 'Cd86'],palette='tab20',cmap='coolwarm',vmin=0)\n",
    "sc.pl.umap(adata_DC1_subset, color=['Cd40','Ccr7'],palette='tab20',cmap='coolwarm',vmin=0)\n",
    "\n",
    "# ['Ccr7','Itgae', 'Cd207', 'Cx3cr1', 'Itgax', 'H2-Ab1', 'Cd40', 'Cd80', 'Cd86', 'Relb', 'Zbtb46']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f44f8cfd-38d0-4da0-9b13-a3f26794617c",
   "metadata": {},
   "outputs": [],
   "source": [
    "# More detailed comparison between DC1 subsets\n",
    "sc.tl.rank_genes_groups(adata, 'leiden', groups = ['r1','r2','r3','r4'], reference='rest', method='wilcoxon')\n",
    "sc.pl.rank_genes_groups(adata, n_genes=20, sharey=False)\n",
    "de_markers = sc.get.rank_genes_groups_df(adata, None)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "74bc4370-eedf-4078-a63d-6e05c81e15a2",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Check markers in DC1: The DEGs can be used to separate subgroups. We will use Cst3 (Cystatin C, CyC)\n",
    "\n",
    "#Ref: The protease inhibitor cystatin C is differentially expressed among dendritic cell populations, but does not control antigen presentation\n",
    "# El-Sukkari et al., J Immunol. 2003 Nov 15;171(10):5003-11.  doi: 10.4049/jimmunol.171.10.5003.\n",
    "\n",
    "adata_DC1_subset = adata[adata.obs['leiden'].isin(['r1','r2','r3','r4','r5','r6','r7','r8'])].copy()\n",
    "sc.pl.umap(adata_DC1_subset, color=['Cd8a','Cd4','Itgax','Wdfy4'],palette='tab20',cmap='coolwarm',vmin=0)\n",
    "sc.pl.umap(adata_DC1_subset, color=['Ccr2','Cst3','Xcr1','Sirpa'],palette='tab20',cmap='coolwarm',vmin=0)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "260c3c3e-4118-4e8f-bead-a0ea8f634cc0",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Check markers in DC1: The DEGs can be used to separate subgroups. We will use Cst3 (Cystatin C, CyC)\n",
    "\n",
    "#Ref: The protease inhibitor cystatin C is differentially expressed among dendritic cell populations, but does not control antigen presentation\n",
    "# El-Sukkari et al., J Immunol. 2003 Nov 15;171(10):5003-11.  doi: 10.4049/jimmunol.171.10.5003.\n",
    "\n",
    "adata_DC1_dsubset = adata[adata.obs['leiden'].isin(['r1','r2','r3','r4','r5','r6','r7','r8','20'])].copy()\n",
    "sc.pl.umap(adata_DC1_dsubset[adata_DC1_dsubset.obs['WT/KO']=='wt'], color=['Cd8a','Cd4','Sirpa','Itgax'],palette='tab20',cmap='coolwarm',vmin=0)\n",
    "sc.pl.umap(adata_DC1_dsubset[adata_DC1_dsubset.obs['WT/KO']=='ko'], color=['Cd8a','Cd4','Sirpa','Itgax'],palette='tab20',cmap='coolwarm',vmin=0)\n",
    "\n",
    "sc.pl.umap(adata_DC1_dsubset[adata_DC1_dsubset.obs['WT/KO']=='wt'], color=['Ccr2','Cst3','Xcr1','Cd74'],palette='tab20',cmap='coolwarm',vmin=0)\n",
    "sc.pl.umap(adata_DC1_dsubset[adata_DC1_dsubset.obs['WT/KO']=='ko'], color=['Ccr2','Cst3','Xcr1','Cd74'],palette='tab20',cmap='coolwarm',vmin=0)\n",
    "'Cd40','Ccr7'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a60a91a9-e2c9-4866-9ccc-71baab3557ed",
   "metadata": {},
   "outputs": [],
   "source": [
    "adata_DC1_dsubset = adata[adata.obs['leiden'].isin(['r1','r2','r3','r4','r5','r6','r7','r8','20'])].copy()\n",
    "sc.pl.umap(adata_DC1_dsubset[adata_DC1_dsubset.obs['pathogenicity']=='naive'], color=['Ccr2','Ccr7'],palette='tab20',cmap='coolwarm',vmin=0)\n",
    "sc.pl.umap(adata_DC1_dsubset[adata_DC1_dsubset.obs['pathogenicity']=='pathogenic'], color=['Ccr2','Ccr7'],palette='tab20',cmap='coolwarm',vmin=0)\n",
    "\n",
    "sc.pl.umap(adata_DC1_dsubset[adata_DC1_dsubset.obs['pathogenicity']=='naive'], color=['Cd80','Cd86','Cd40'],palette='tab20',cmap='coolwarm',vmin=0)\n",
    "sc.pl.umap(adata_DC1_dsubset[adata_DC1_dsubset.obs['pathogenicity']=='pathogenic'], color=['Cd80','Cd86','Cd40'],palette='tab20',cmap='coolwarm',vmin=0)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3dd44f58-5b79-4db8-8ac8-c564a30f66bf",
   "metadata": {},
   "outputs": [],
   "source": [
    "### Beutler lab study Ccr2+ cDC2 cells. \n",
    "sc.pl.umap(adata_subset, color=['Ccr2','Wdfy4'],palette='tab20',cmap='coolwarm',vmax=3)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4ef8c2dd-fce4-401e-8cac-8afe9222267a",
   "metadata": {},
   "outputs": [],
   "source": [
    "#Let's check monocyte-derived DC markers (Itgam=Cd11b, Fcgr1=Cd64, Csf1r = Cd115, Adgre = F4/80)\n",
    "# Elodie Segura, Alice Coillard. Antigen presentation by mouse monocyte-derived cells: Re-evaluating\n",
    "# the concept of monocyte-derived dendritic cells. Molecular Immunology, 2021, 135, pp.165-169.\n",
    "# 10.1016/j.molimm.2021.04.012. inserm-03381917\n",
    "\n",
    "modc = ['Ccr2','Itgam','Fcgr1','Csf1r','Adgre1']\n",
    "\n",
    "sc.tl.score_genes(adata, modc , score_name='modc_score')\n",
    "\n",
    "sc.pl.umap(adata_subset, color=['modc_score'],palette='tab20',cmap='coolwarm',vmax=2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "18f650ab-bc5d-47ef-8d91-21fe43e0201a",
   "metadata": {},
   "outputs": [],
   "source": [
    "sc.pl.umap(adata_subset, color=['Bst2','Tcf4'],palette='tab20',cmap='coolwarm')\n",
    "sc.pl.umap(adata_subset, color=['Zeb2','Irf7'],palette='tab20',cmap='coolwarm')\n",
    "sc.pl.umap(adata_subset, color=['Pacsin1','Spib'],palette='tab20',cmap='coolwarm')\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5b3f8147-fcec-465a-b54a-32614f47ec27",
   "metadata": {},
   "outputs": [],
   "source": [
    "pdc_saito = ['Ptprc','Bst2','Siglech']\n",
    "\n",
    "adata.obs['pdc_score_saito'] = adata[:,pdc_saito].X.sum(1)\n",
    "sc.pl.umap(adata, color=['pdc_score_saito'],palette='tab20',cmap='coolwarm')\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f0bc13bf-e51e-4fb0-8253-dc6aa5bb9b5b",
   "metadata": {},
   "outputs": [],
   "source": [
    "migratory_dc_act = ['Arc','Ccr7','Irf4'] \n",
    "sc.tl.score_genes(adata, migratory_dc_act , score_name='migratory_dc_act_score')\n",
    "adata_subset = adata[adata.obs['leiden'].isin(['r0','r1','r2','r3','r4','r5','r6','r7','r8','r9','20','21'])].copy()\n",
    "sc.pl.umap(adata_subset, color=['migratory_dc_act_score'],palette='tab20',cmap='coolwarm',vmax=1.2)\n",
    "sc.pl.umap(adata_subset, color= ['Arc','Ccr7','Irf4'] ,palette='tab20',cmap='coolwarm',vmax=1.2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "76553830-7f25-4281-a6a6-533ee458cf8b",
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "sc.pl.umap(adata, color=['Bst2','Klk1'],palette='tab20',cmap='coolwarm')\n",
    "sc.pl.umap(adata, color=['Cmah', 'Bcl119'],palette='tab20',cmap='coolwarm')\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0465fc5e-8fcf-45d0-bddc-942733d6381c",
   "metadata": {},
   "outputs": [],
   "source": [
    "# dc1_genes = ['Cd8a','Irf8']\n",
    "dc1_genes = ['Cd8a','Irf8', 'Batf3', 'Nfil3','Id2', 'Bcl6', 'Xcr1','Rab43', 'Itgax', 'Itgae', 'Cd24a']\n",
    "adata.obs['dc1_score'] = adata[:,dc1_genes].X.sum(1)\n",
    "# dc2_genes = ['Relb', 'Esam', 'Itgam','Irf4', 'Sirpa']\n",
    "dc2_genes = ['H2-Ab1', 'Sirpa', 'Cd4', 'Notch2', 'Clec4a2','Esam', 'Irf4', 'Relb', 'Zeb2', 'Klf4']\n",
    "adata.obs['dc2_score'] = np.asarray(adata[:,dc2_genes].X.sum(1)/np.max(adata[:,dc2_genes].X.sum(1)))*np.asarray(1-adata[:,'Cd8a'].X.sum(1)/np.max(adata[:,'Cd8a'].X.sum(1)))\n",
    "pdc_genes =['Cd209a','Lifr','Tcf4', 'Zeb2', 'Pacsin1', 'Spib']\n",
    "adata.obs['pdc_score'] = adata[:,pdc_genes].X.sum(1)\n",
    "\n",
    "migratorydc_genes = ['Ccr7','Itgae', 'Cd207', 'Cx3cr1', 'Itgax', 'H2-Ab1', 'Cd40', 'Cd80', 'Cd86', 'Relb', 'Zbtb46']\n",
    "adata.obs['migratorydc_score'] = adata[:,migratorydc_genes].X.sum(1)\n",
    "\n",
    "\n",
    "sc.pl.umap(adata, color=['dc1_score'],palette='tab20',cmap='coolwarm')\n",
    "sc.pl.umap(adata, color=['dc2_score'],palette='tab20',cmap='coolwarm',vmax=0.9)\n",
    "sc.pl.umap(adata, color=['pdc_score'],palette='tab20',cmap='coolwarm',vmax=10)\n",
    "sc.pl.umap(adata, color=['migratorydc_score'],palette='tab20',cmap='coolwarm')\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9b387e49-f97e-4c88-a9d4-6651eff7e0f5",
   "metadata": {},
   "outputs": [],
   "source": [
    "## Division to Cross-presenting and Sirpa+ DC\n",
    "## See Gurka et al, Front. Immunol., 04 February 2015 Sec. Antigen Presenting Cell Biology\n",
    "## Volume 6 - 2015 | https://doi.org/10.3389/fimmu.2015.00035\n",
    "\n",
    "cdc1_genes1 = ['Cd8a','Clec9a','Batf3']\n",
    "cdc1_genes2 = ['Xcr1','Sirpa']\n",
    "sc.pl.umap(adata, color=cdc1_genes1,palette='tab20',cmap='coolwarm')\n",
    "sc.pl.umap(adata, color=cdc1_genes2,palette='tab20',cmap='coolwarm')\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "778309e1-57c5-4342-a8ee-fd84674a96ee",
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "### Activated cDCs\n",
    "acdc_genes1 = ['Itgam' ,'Itgax']\n",
    "acdc_genes2 = ['Cd80' ,'Cd86']\n",
    "sc.pl.umap(adata, color=acdc_genes1,palette='tab20',cmap='coolwarm')\n",
    "sc.pl.umap(adata, color=acdc_genes2,palette='tab20',cmap='coolwarm')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "594184fc-71dc-43dd-ab6d-0806994f0b53",
   "metadata": {},
   "outputs": [],
   "source": [
    "### Monocyte-derived DC\n",
    "mddc_genes1 = ['Itgam' ,'Itgax']\n",
    "mddc_genes2 = ['Il12a','Ly6g']\n",
    "sc.pl.umap(adata, color=mddc_genes1,palette='tab20',cmap='coolwarm')\n",
    "sc.pl.umap(adata, color=mddc_genes2,palette='tab20',cmap='coolwarm')\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "59a8e913-b2a8-4e51-b74a-7e39dbc4aede",
   "metadata": {},
   "outputs": [],
   "source": [
    "#pDC markers separately\n",
    "sc.pl.umap(adata, color=['Irf8','Tcf4'],palette='tab20',cmap='coolwarm')\n",
    "sc.pl.umap(adata, color=['Zbtb46','Siglech'],palette='tab20',cmap='coolwarm')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "67aee9b2-fcda-4a0e-9cdb-9bc04d27cd7c",
   "metadata": {},
   "outputs": [],
   "source": [
    "sc.pl.umap(adata, color=['Itgae','Xcr1'],palette='tab20',cmap='coolwarm')\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b231e6d5-e0e7-411f-a10e-192e710c7f30",
   "metadata": {},
   "source": [
    "### Monocytes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "937d0f49-afeb-4851-a1b8-5885ae483d4c",
   "metadata": {},
   "outputs": [],
   "source": [
    "clmono_genes=['Ccl9', 'Ccr2','Cd68','Ly6c2']\n",
    "nclmono_genes = ['Csf1r','Cx3cr1','Fabp4']\n",
    "adata.obs['clmono_score'] = adata[:,clmono_genes].X.sum(1)\n",
    "adata.obs['nclmono_score'] = adata[:,nclmono_genes].X.sum(1)\n",
    "sc.pl.umap(adata, color=['clmono_score', 'nclmono_score'],palette='tab20',cmap='coolwarm')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9ecb9c3d-3344-40cc-b38b-7581142d2a1d",
   "metadata": {},
   "source": [
    "### Macrophages"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "69c82d2f-f1c4-4388-a04d-cf685a99cbe1",
   "metadata": {},
   "outputs": [],
   "source": [
    "# macroph_genes = ['Adgre1','Itgam','Ly6g']\n",
    "\n",
    "#PanglaoDB\n",
    "macroph_genes = ['Cd68','Fcgr1','Naaa','Lyz2','Ccl12']\n",
    "#['Adgre1','Itgam','Apoe','C1qa','Cx3cr1']\n",
    "adata.obs['macroph_score'] = adata[:,macroph_genes].X.sum(1)\n",
    "\n",
    "sc.pl.umap(adata, color=['macroph_score'],palette='tab20',cmap='coolwarm')\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d4d0cb01-1db8-4b95-8090-7105fc7f6941",
   "metadata": {},
   "source": [
    "### Plasma cells"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2df43b1a-220d-47b4-a52a-1a3c60fb43b3",
   "metadata": {},
   "outputs": [],
   "source": [
    "sc.pl.umap(adata, color=['Mzb1','Ighg1'], palette='tab20',cmap='coolwarm')\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "882a7fc1-2fb2-4462-a2e3-3ff094526a7a",
   "metadata": {},
   "outputs": [],
   "source": [
    "plasma_genes = ['Jchain','Sdc1']\n",
    "adata.obs['plasma_score'] = adata[:,plasma_genes].X.sum(1)\n",
    "sc.pl.umap(adata, color=['plasma_score'], palette='tab20',cmap='coolwarm', vmax=1.)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "94022699-bca6-43a7-98ee-75a3452e1491",
   "metadata": {},
   "outputs": [],
   "source": [
    "sc.pl.umap(adata, color=['Ltc4s','Batf3'], palette='tab20',cmap='coolwarm', vmax=4)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "0db8472e-1c02-496b-926b-36bf45320dc5",
   "metadata": {},
   "source": [
    "### Mast cells"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0a11552e-65a6-497e-97c5-19996617bc91",
   "metadata": {},
   "outputs": [],
   "source": [
    "mast_genes = ['Kit','Ltc4s','Il1rl1','Hdc']\n",
    "adata.obs['mast_score'] = adata[:,mast_genes].X.sum(1)\n",
    "sc.pl.umap(adata, color=['mast_score'], palette='tab20',cmap='coolwarm', vmax=4)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "173a8b7b-138a-4813-8315-534c4f71cf3f",
   "metadata": {},
   "outputs": [],
   "source": [
    "adata.obs['hi_mast_score'] =  (np.asarray(adata[:,mast_genes].X.sum(1).flatten())[0] > np.percentile(np.asarray(adata[:,mast_genes].X.sum(1).flatten())[0], 99.9)).astype(str)\n",
    "adata.obs['hi_mast_score'] = adata.obs['hi_mast_score'].astype('category')\n",
    "sc.pl.umap(adata, color=['hi_mast_score'])"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b98f87df-aab1-4e52-9503-e09ac62f3d0d",
   "metadata": {},
   "source": [
    "### Neutrophils"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "72aaec6e-be82-4063-8bc3-24350e014851",
   "metadata": {},
   "outputs": [],
   "source": [
    "neutrophil_genes1 = ['Csf3r','Ly6g']\n",
    "neutrophil_genes2 = ['S100a8','Il1r2']\n",
    "sc.pl.umap(adata, color=neutrophil_genes1, palette='tab20',cmap='coolwarm')\n",
    "sc.pl.umap(adata, color=neutrophil_genes2, palette='tab20',cmap='coolwarm')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "7db030bd-1863-401f-89ba-83a55d1b6bc7",
   "metadata": {},
   "source": [
    "### Basophils"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "99b19943-356b-4186-bb8b-82ef30399597",
   "metadata": {},
   "outputs": [],
   "source": [
    "basophil_genes = ['Fcer1a', 'Cpa3','Ms4a2','Gata2','Il3ra','Ccr3','Hdc']\n",
    "adata.obs['basophil_score'] = adata[:,basophil_genes].X.sum(1)\n",
    "sc.pl.umap(adata, color=['basophil_score'], palette='tab20',cmap='coolwarm',vmax=4)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b2f241e2-dab9-4dbd-b635-34a513d16d9c",
   "metadata": {},
   "source": [
    "### Progenitor cells"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c2074839-686f-4ad1-b202-fbe07cf1234f",
   "metadata": {},
   "outputs": [],
   "source": [
    "progenitor_genes = ['Cxcr5', 'Id3','Slamf6','Tcf7']\n",
    "adata.obs['progenitor_score'] = adata[:,progenitor_genes].X.sum(1)\n",
    "sc.pl.umap(adata, color=['progenitor_score'], palette='tab20',cmap='coolwarm', vmax=5)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8b087c12-0266-48ae-8226-e49d7bcdabf2",
   "metadata": {},
   "outputs": [],
   "source": [
    "ILC2_genes = ['Il1rl1', 'Gata3', 'Icos']\n",
    "adata.obs['ILC2_score'] = adata[:,ILC2_genes].X.sum(1)\n",
    "sc.pl.umap(adata, color=['ILC2_score'], palette='tab20',cmap='coolwarm', vmax=5)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "947fedf2-4a28-41f6-a270-d98ded845cd6",
   "metadata": {},
   "source": [
    "### NKT cells"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5643b876-fce4-4a03-8a85-398a1ef96bbe",
   "metadata": {},
   "outputs": [],
   "source": [
    "sc.pl.umap(adata, color=['Zbtb16','Klrb1c'], palette='tab20',cmap='coolwarm', vmax=2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2850993d-519c-4ef7-8a37-7aa0e9693dc6",
   "metadata": {},
   "outputs": [],
   "source": [
    "NKT_genes = ['Ncam1', 'Gata3', 'Il2rb']\n",
    "adata.obs['NKT_score'] = adata[:,NKT_genes].X.sum(1)\n",
    "sc.pl.umap(adata, color=NKT_genes, palette='tab20',cmap='coolwarm', vmax=1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ed2ce68b-4d8c-40d5-9bf2-69d633139ce5",
   "metadata": {},
   "outputs": [],
   "source": [
    "#Invariant alpha chain genes for NKT cells\n",
    "sc.pl.umap(adata, color=['Trav11','Traj18'], palette='tab20',cmap='coolwarm',vmax=0.5)\n",
    "sc.pl.umap(adata, color=['Zbtb16','Tbx21'], palette='tab20',cmap='coolwarm', vmax=2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6e5d7b55-ed99-4faf-86b0-3d12b7536d0b",
   "metadata": {},
   "outputs": [],
   "source": [
    "# common NKT  chain genes\n",
    "sc.pl.umap(adata, color=['Trbv1','Trbv13-2','Trbv29'], palette='tab20',cmap='coolwarm', vmax=1)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "fe21a80b-7453-4f01-865b-b3c0f521b486",
   "metadata": {},
   "source": [
    "### gamma delta T"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "244c4d3f-47ed-4f42-bbf2-7a63bba6204d",
   "metadata": {},
   "outputs": [],
   "source": [
    "gamma_delta_genes = ['Tcrg-V4', 'Tcrg-V6', 'Tcrg-V1','Tcrg', 'Tcrd', 'Trgv1', 'Trgv2', 'Trgv3', 'Trgv4', 'Trgv5', 'Trgv6', 'Trgv7','Trdv1', 'Trdv3', 'Trdv4', 'Trdv5','Trg', 'Trd', 'Sox13', 'Id3', 'Blk', 'Il17a', 'Ifng']\n",
    "present_gamma_delta_genes = [gene for gene in gamma_delta_genes if gene in adata.var_names]\n",
    "print(\"Present  T cell genes:\", present_gamma_delta_genes)\n",
    "adata.obs['gamma_delta_score'] = adata[:,present_gamma_delta_genes].X.sum(1)\n",
    "sc.pl.umap(adata, color=['gamma_delta_score'], palette='tab20',cmap='coolwarm', vmax=5)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e4ba1428-019b-4641-8ae1-cceb49916d2c",
   "metadata": {},
   "outputs": [],
   "source": [
    "[i for i in adata.var_names.tolist() if i.startswith('Tcr')]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "98f55174-eb94-4ea7-aab5-f0a2342889f0",
   "metadata": {},
   "outputs": [],
   "source": [
    "# gamma delta T markers\n",
    "['Tcrg-V4', 'Tcrg-V6', 'Tcrg-V1']\n",
    "sc.pl.umap(adata, color=['Tcrg-V4', 'Tcrg-V6', 'Tcrg-V1'], palette='tab20',cmap='coolwarm', vmax=0.5)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ae935d63-0599-473c-afee-401089378de6",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.12"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
