{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "5b569639-6aa8-46c3-b153-a61cfde20a1f",
   "metadata": {},
   "source": [
    "# Gene ontology analysis of spleen samples"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "df1112b8-d475-4fca-902f-0942fc20dbe4",
   "metadata": {},
   "outputs": [],
   "source": [
    "import numpy as np\n",
    "import seaborn as sns\n",
    "\n",
    "import matplotlib\n",
    "import matplotlib.pyplot as plt\n",
    "from matplotlib.legend import Legend\n",
    "import matplotlib.colors as colors\n",
    "from matplotlib import cm\n",
    "from mpl_toolkits.mplot3d import axes3d\n",
    "import matplotlib.patches as patches\n",
    "import pandas as pd\n",
    "import scipy\n",
    "import scanpy as sc\n",
    "import anndata as ad\n",
    "\n",
    "import random\n",
    "\n",
    "import pydeseq2\n",
    "from pydeseq2.dds import DeseqDataSet\n",
    "from pydeseq2.ds import DeseqStats\n",
    "\n",
    "import gseapy as gp\n",
    "from gseapy.plot import gseaplot\n",
    "\n",
    "eps= 10.**(-300.)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "39a14a86-a984-4e12-bc0a-486b85e71752",
   "metadata": {},
   "outputs": [],
   "source": [
    "sc.set_figure_params(scanpy=True, dpi=600, dpi_save=1200, frameon=True, vector_friendly=False, fontsize=14,\n",
    "                         figsize=(9,8),  format='pdf', facecolor=None, transparent=False, ipython_format='png2x')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "3291a12f-d320-4f71-9a84-b82b674d8de0",
   "metadata": {},
   "source": [
    "#### GOA based on Sanbomics tutorial https://www.youtube.com/watch?v=ONiWugVEf2s"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "249adda1-77a6-4974-a234-1dc464bf4767",
   "metadata": {},
   "source": [
    "#!pip install goatools"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "4cacd977-f83c-4647-bd37-e49ba0c3307f",
   "metadata": {},
   "source": [
    "### biomaRt not used in this notebook.\n",
    "###!pip install biomart"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "6d764d02-d438-4eab-a372-e523a9b74faa",
   "metadata": {},
   "source": [
    "###!pip install adjustText"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "3df7a636-6816-4e8e-bb1d-cd61cc5cf363",
   "metadata": {},
   "source": [
    "###!pip install gprofiler"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c81effff-145a-4190-a71c-2f46b92473f0",
   "metadata": {},
   "outputs": [],
   "source": [
    "#!mv /Users/oipulk/Downloads/gene_result.txt .\n",
    "#!python /Users/oipulk/miniconda3/envs/py3/bin/ncbi_gene_results_to_python.py -o genes_ncbi_mus_musculus_proteincoding.py gene_result.txt"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a8ab4e53-0fd9-4f00-bc71-425825e47806",
   "metadata": {},
   "outputs": [],
   "source": [
    "adata = sc.read(\"maranou_032024_spleen_annotated.h5ad\")\n",
    "dc_types = [\"Lymphoid-resident cDC1\",\"CD8(low) cDC1\",\"CD8- CCR2+ cDC1\",\"CD8- CCR2- cDC1\", \"CCR7+ DC1\", \n",
    "                        \"Relb(int.) cDC2\",\"Migratory cDC2\",\"WDFY4+ cDC2\",\"Relb(low) cDC2\",\"pDC\",\"Mixed DC\",\"Undefined DC\"]\n",
    "adata_subset = adata[adata.obs['cell_type_high_res'].isin(dc_types)].copy()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "16b044cd-7067-432d-a16b-f540a71c3a98",
   "metadata": {},
   "outputs": [],
   "source": [
    "adata_subset.var_names[adata_subset.var['highly_variable']]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ebfe92f3-48f7-4747-9758-62b7e91c4ad4",
   "metadata": {},
   "outputs": [],
   "source": [
    "hvgs = set(adata_subset.var_names[adata_subset.var['highly_variable']])\n",
    "len(hvgs)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "140e140c-b585-4498-971e-4d08ae03ae47",
   "metadata": {},
   "outputs": [],
   "source": [
    "min_expr_threshold = 0.1  # adjust based on your normalization\n",
    "dc_expressed = set(adata_subset.var_names[np.asarray(adata_subset.X.mean(axis=0))[0] >= min_expr_threshold])\n",
    "len(dc_expressed)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d8ca4b8a-9bff-49ea-8d68-1369bafe3a77",
   "metadata": {},
   "outputs": [],
   "source": [
    "len(hvgs.intersection(dc_expressed))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "48979b7b-139d-4bcf-bd5f-a3ddbf1026da",
   "metadata": {},
   "outputs": [],
   "source": [
    "from genes_ncbi_mus_musculus_proteincoding import GENEID2NT as GeneID2nt_mus\n",
    "from goatools.base import download_go_basic_obo, download_ncbi_associations\n",
    "from goatools.obo_parser import GODag\n",
    "from goatools.anno.genetogo_reader import Gene2GoReader\n",
    "from goatools.goea.go_enrichment_ns import GOEnrichmentStudyNS\n",
    "import pandas as pd\n",
    "\n",
    "from genes_ncbi_mus_musculus_proteincoding import GENEID2NT as GeneID2nt_mus\n",
    "\n",
    "obo_fname = download_go_basic_obo()\n",
    "fin_gene2go = download_ncbi_associations()\n",
    "obodag = GODag(\"go-basic.obo\")\n",
    "\n",
    "# Define map and inverse map between gene symbol and ID\n",
    "mapper = {}\n",
    "\n",
    "for key in GeneID2nt_mus:\n",
    "    mapper[GeneID2nt_mus[key].Symbol] = GeneID2nt_mus[key].GeneID\n",
    "    \n",
    "inv_map = {v: k for k, v in mapper.items()}\n",
    "\n",
    "\n",
    "# Read NCBI's gene2go. Store annotations in a list of namedtuples\n",
    "objanno = Gene2GoReader(fin_gene2go, taxids=[10090])\n",
    "# Get namespace2association where:\n",
    "#    namespace is:\n",
    "#        BP: biological_process               \n",
    "#        MF: molecular_function\n",
    "#        CC: cellular_component\n",
    "#    assocation is a dict:\n",
    "#        key: NCBI GeneID\n",
    "#        value: A set of GO IDs associated with that gene\n",
    "ns2assoc = objanno.get_ns2assc()\n",
    "\n",
    "### To narrow down the set of background genes, we can take only the highly variable ones in our data\n",
    "# expressed_genes = set(adata.var_names[adata.var['highly_variable']])\n",
    "# # Convert gene symbols to NCBI IDs\n",
    "# expressed_gene_ids = {mapper[gene] for gene in expressed_genes \n",
    "#                      if gene in mapper}\n",
    "\n",
    "# Or get genes with meaningful expression\n",
    "min_expr_threshold = 0.1  # adjust based on your normalization\n",
    "dc_genes =set(adata_subset.var_names[np.asarray(adata_subset.X.mean(axis=0))[0] >= min_expr_threshold])\n",
    "\n",
    "# Convert to NCBI IDs\n",
    "dc_gene_ids = {mapper[gene] for gene in dc_genes \n",
    "               if gene in mapper}\n",
    "\n",
    "goeaobj = GOEnrichmentStudyNS(\n",
    "        # dc_gene_ids,\n",
    "        GeneID2nt_mus.keys(), # List of mouse protein-coding genes\n",
    "        ns2assoc, # geneid/GO associations\n",
    "        obodag, # Ontologies\n",
    "        propagate_counts = True,\n",
    "        alpha = 0.05, # default significance cut-off\n",
    "        methods = ['fdr_bh']) # default multipletest correction method\n",
    "\n",
    "GO_items = []\n",
    "\n",
    "temp = goeaobj.ns2objgoea['BP'].assoc\n",
    "for item in temp:\n",
    "    GO_items += temp[item]\n",
    "    \n",
    "temp = goeaobj.ns2objgoea['CC'].assoc\n",
    "for item in temp:\n",
    "    GO_items += temp[item]\n",
    "    \n",
    "temp = goeaobj.ns2objgoea['MF'].assoc\n",
    "for item in temp:\n",
    "    GO_items += temp[item]\n",
    "\n",
    "\n",
    "# # Define immune-relevant root terms and keywords\n",
    "IMMUNE_ROOT_TERMS = {\n",
    "\n",
    "    # #All GO terms \n",
    "    # 'GO:0008150', # BP\n",
    "    # 'GO:0003674', # MF\n",
    "    # 'GO:0005575', # CC\n",
    " \n",
    "    # Core immune processes\n",
    "    'GO:0002376',  # immune system process\n",
    "    'GO:0050896',  # response to stimulus\n",
    "    'GO:0006955',  # response to stimulus - immune response\n",
    "    'GO:0009987',  # Cellular process\n",
    "    'GO:0050789', # Regulation of biological process\n",
    "    'GO:0008219', # Cell death\n",
    "    'GO:0045321',  # leukocyte activation\n",
    "    'GO:0005515', # cytokine binding\n",
    "    \n",
    "    'GO:0002520',  # immune system development\n",
    "    'GO:0002682',  # regulation of immune system process\n",
    "    'GO:0006954',  # inflammatory response\n",
    "    'GO:0002250',  # adaptive immune response\n",
    "    'GO:0045087',  # innate immune response\n",
    "    'GO:0030333',  # antigen processing and presentation\n",
    "    \n",
    "    # # DC-specific processes\n",
    "    'GO:0002396',  # MHC protein complex assembly\n",
    "    \n",
    "    'GO:0071404',  # cellular response to low-density lipoprotein particle stimulus\n",
    "    'GO:0071396',  # cellular response to lipid\n",
    "    'GO:0019886',  # antigen processing and presentation via MHC class II\n",
    "    'GO:0002606',  # positive regulation of dendritic cell differentiation\n",
    "    'GO:0002819',  # regulation of adaptive immune response\n",
    "    \n",
    "    # # Tumor microenvironment relevant\n",
    "    'GO:0002837',  # regulation of immune response to tumor cell\n",
    "    'GO:0097530',  # granulocyte migration\n",
    "    'GO:0002548',  # monocyte chemotaxis\n",
    "    'GO:0042102',  # positive regulation of T cell proliferation\n",
    "    'GO:0042110',  # T cell activation\n",
    "    'GO:0050863',  # regulation of T cell activation\n",
    "    \n",
    "    # # Cellular processes relevant to DCs\n",
    "    'GO:0016192',  # vesicle-mediated transport\n",
    "    'GO:0006897',  # endocytosis\n",
    "    'GO:0034341',  # response to interferon-gamma\n",
    "    'GO:0071347',  # cellular response to interleukin-1\n",
    "    'GO:0035872',  # nucleotide-binding domain, leucine rich repeat containing receptor signaling pathway\n",
    "    \n",
    "    # # Additional immune processes\n",
    "    'GO:0002764',  # immune response-regulating signaling pathway\n",
    "    'GO:0050776',  # regulation of immune response\n",
    "    'GO:0002684',  # positive regulation of immune system process\n",
    "    'GO:0002683',  # negative regulation of immune system process\n",
    "    'GO:0002521',  # leukocyte differentiation\n",
    "    'GO:0002366',  # leukocyte activation involved in immune response\n",
    "    'GO:0002460',  # adaptive immune response based on somatic recombination\n",
    "    'GO:0002757',  # immune response-activating signal transduction\n",
    "}\n",
    "\n",
    "IMMUNE_KEYWORDS = {\n",
    "    # Core immune terms\n",
    "    'immune', 'lymphocyte', 'leukocyte', 'cytokine',\n",
    "    'chemokine', 'inflammation', 'inflammatory', 'antigen', 'antibody',\n",
    "    'tcr', 'bcr', 'mhc', 'nfkb', 'toll', 'interferon',\n",
    "    'interleukin', 'tnf', 'immunoglobulin',\n",
    "    \n",
    "    # DC-specific terms\n",
    "    'dendritic', 'cd11c', 'cd83', 'cd86', 'cd40',\n",
    "    'batf3', 'zbtb46', 'xcr1', 'clec9a', 'cd103',\n",
    "    'sirpa', 'cd207', 'cd1c', 'cd141',\n",
    "    \n",
    "    # Antigen presentation and processing\n",
    "    'tap', 'cathepsin', 'lamp', 'rab', 'snap',\n",
    "    'vamp', 'h2-dm', 'h2-do', 'clip', 'ciita',\n",
    "    \n",
    "    # Signal transduction\n",
    "    'stat', 'jak', 'traf', 'irf', 'syk',\n",
    "    'mapk', 'pi3k', 'nf-kb','nf-kappab', 'tgf', 'notch',\n",
    "    \n",
    "    # Migration and adhesion\n",
    "    'ccr', 'cxcr', 'integrin', 'selectin', 'vcam',\n",
    "    'icam', 'mmp', 'itga', 'itgb', 'migratory','migration',\n",
    "    \n",
    "    # Tumor microenvironment\n",
    "    'pd-l1', 'cd80', 'ido', 'arg1', 'vegf',\n",
    "    'tgfb', 'il10', 'stat3', 'hippo', 'ros',\n",
    "    \n",
    "    # Metabolism and stress\n",
    "    'mtor', 'ampk', 'hif', 'ros', 'autophagy',\n",
    "    'glycolysis', 'oxidative', 'mitochondrial',\n",
    "    \n",
    "    # Trafficking and processing\n",
    "    'endosome', 'lysosome', 'golgi', 'vesicle',\n",
    "    'phagosome', 'autophagy', 'endocytic',\n",
    "    \n",
    "    # Transcriptional regulation\n",
    "    'transcription', 'chromatin', 'epigenetic',\n",
    "    'methylation', 'acetylation', 'histone',\n",
    "    \n",
    "    # Additional immune terms\n",
    "    'granzyme', 'perforin', 'complement',\n",
    "    'inflammasome', 'tlr', 'nod', 'rig-i',\n",
    "    'mda5', 'nlrp3', 'asc', 'caspase',\n",
    "    \n",
    "    # Cytokine signaling\n",
    "    'il1', 'il2', 'il4', 'il6', 'il10',\n",
    "    'il12', 'il17', 'il23', 'ifng', 'tnfa',\n",
    "    \n",
    "    # T cell related\n",
    "    't cell', 'cd3', 'cd4', 'cd8', 'foxp3', 'rorc',\n",
    "    'tbet', 'gata3', 'cd25', 'cd69',\n",
    "    \n",
    "    # B cell related\n",
    "    'b cell', 'cd19', 'cd20', 'cd27', 'cd38', 'cd138',\n",
    "    'baff', 'april', 'plasma cell',\n",
    "    \n",
    "    # Myeloid cell related\n",
    "    'myeloid', 'cd14', 'cd16', 'cd64', 'cd68', 'cd163',\n",
    "    'marco', 'mertk', 'trem2',\n",
    "    \n",
    "    # Additional trafficking terms\n",
    "    'cxcl', 'ccl', 'sphingosine', 'ceramide',\n",
    "    's1p', 'lpa', 'prostaglandin',\n",
    "\n",
    "    # Additional Immune System Terms:\n",
    "    'innate', 'adaptive', 'spleen', \n",
    "    'nk cell', 'natural killer',\n",
    "    'myeloid', 'neutrophil', 'eosinophil', 'basophil',\n",
    "    'ifna', 'ifnb',\n",
    "    \n",
    "    #Cancer/Oncology Related Terms:\n",
    "    'neoplasm', 'oncogene', 'tumor suppressor',\n",
    "    'metastasis', 'angiogenesis', 'invasion',\n",
    "    'warburg', 'emt', 'mesenchymal',\n",
    "    'p53', 'myc', 'ras', 'brca', \n",
    "    \n",
    "    #Cancer Hallmark Terms:\n",
    "    'telomerase', 'senescence', 'apoptosis'\n",
    "    'stemness', 'cancer stem cell'\n",
    "    'dna repair', 'genome instability'\n",
    "    'immortalization', 'proliferation', \n",
    "    \n",
    "    #Tumor Microenvironment Additions:\n",
    "    'mdsc','cam','caf', 'suppressor', \n",
    "    'cancer', 'macrophage', 'fibroblast',\n",
    "    'hypoxia', 'angiogenesis', \n",
    "    \n",
    "    #Immune Checkpoint Terms:\n",
    "    'checkpoint', 'ctla4', 'lag3', 'tim3', 'vista',\n",
    "    'pd1', 'pdcd1', 'btla',\n",
    "    'icos', 'gitr', 'ox40', \n",
    "    \n",
    "    #Signaling Pathway Completeness:\n",
    "    'wnt', 'hedgehog', 'hippo',\n",
    "    'akt', 'pten', 'pi3k',\n",
    "    'smad', 'bmp', 'jun', 'fos' \n",
    "\n",
    "}\n",
    "\n",
    "\n",
    "def get_immune_relevant_children(obodag, root_terms=IMMUNE_ROOT_TERMS):\n",
    "    \"\"\"Get all children of immune-relevant root terms\"\"\"\n",
    "    relevant_terms = set()\n",
    "    for term in root_terms:\n",
    "        if term in obodag:\n",
    "            relevant_terms.add(term)\n",
    "            relevant_terms.update(obodag[term].get_all_children())\n",
    "    return relevant_terms\n",
    "\n",
    "def is_immune_relevant(go_term, immune_terms, immune_keywords=IMMUNE_KEYWORDS):\n",
    "    \"\"\"Check if a GO term is immune-relevant\"\"\"\n",
    "    if go_term.id in immune_terms:\n",
    "        return True\n",
    "    return any(keyword in go_term.name.lower() for keyword in immune_keywords)\n",
    "\n",
    "# Your existing setup code remains the same until the go_it function\n",
    "def go_it(test_genes, immune_filter=True):\n",
    "    print(f'input genes: {len(test_genes)}')\n",
    "    \n",
    "    mapped_genes = []\n",
    "    for gene in test_genes:\n",
    "        try:\n",
    "            mapped_genes.append(mapper[gene])\n",
    "        except:\n",
    "            pass\n",
    "    print(f'mapped genes: {len(mapped_genes)}')\n",
    "    \n",
    "    # Get immune-relevant terms\n",
    "    immune_terms = get_immune_relevant_children(obodag)\n",
    "    \n",
    "    goea_results_all = goeaobj.run_study(mapped_genes)\n",
    "    goea_results_sig = [r for r in goea_results_all if r.p_fdr_bh < 0.05]\n",
    "    \n",
    "    # Filter for immune-relevant terms if requested\n",
    "    if immune_filter:\n",
    "        goea_results_sig = [r for r in goea_results_sig \n",
    "                           if is_immune_relevant(r.goterm, immune_terms)]\n",
    "    \n",
    "    GO = pd.DataFrame(\n",
    "        list(map(\n",
    "            lambda x: [\n",
    "                x.GO, \n",
    "                x.goterm.name, \n",
    "                x.goterm.namespace,\n",
    "                x.p_uncorrected,\n",
    "                x.p_fdr_bh,\n",
    "                x.ratio_in_study[0],\n",
    "                x.ratio_in_study[1],\n",
    "                GO_items.count(x.GO),\n",
    "                list(map(lambda y: inv_map[y], x.study_items)),\n",
    "            ],\n",
    "            goea_results_sig\n",
    "        )),\n",
    "        columns=['GO', 'term', 'class', 'p', 'p_corr', 'n_genes',\n",
    "                'n_study', 'n_go', 'study_genes']\n",
    "    )\n",
    "    \n",
    "    # Split by class and handle each separately\n",
    "    go_classes = []\n",
    "    for class_name in GO['class'].unique():\n",
    "        class_df = GO[GO['class'] == class_name].copy()\n",
    "        if len(class_df) > 0:  # Only process if class has any terms\n",
    "            class_df['immune_relevance'] = class_df.apply(\n",
    "                lambda row: sum(1 for kw in IMMUNE_KEYWORDS \n",
    "                if kw in row['term'].lower()) / len(IMMUNE_KEYWORDS),\n",
    "                axis=1\n",
    "            )\n",
    "            # Filter and sort within each class\n",
    "            class_df = class_df[class_df.n_genes > 1].sort_values(\n",
    "                by=['immune_relevance', 'p_corr'],\n",
    "                ascending=[False, True]\n",
    "            )\n",
    "            go_classes.append(class_df)\n",
    "\n",
    "    # Combine results if any exist\n",
    "    if go_classes:\n",
    "        GO = pd.concat(go_classes, axis=0)\n",
    "    else:\n",
    "        # Create empty DataFrame with same structure if no results at all\n",
    "        GO = pd.DataFrame(columns=GO.columns)\n",
    "        GO['immune_relevance'] = pd.Series(dtype=float)\n",
    "    \n",
    "    return GO\n",
    "\n",
    "\n",
    "  \n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9499139e-ab4e-4947-afb6-33f400456059",
   "metadata": {},
   "outputs": [],
   "source": [
    "immune_terms = get_immune_relevant_children(obodag)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "08ff7c10-148e-45b1-b24c-83cd00a5b923",
   "metadata": {},
   "outputs": [],
   "source": [
    "len(immune_terms)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "82291d57-bf98-4b82-8140-170e9b39dbd1",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Whether to use interaction DEA or not\n",
    "\n",
    "interaction = True\n",
    "# interaction = False\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a1e8c3d0-7503-441c-8161-ef748bfd4db2",
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "if interaction:\n",
    "\n",
    "    ## Aggregated DC with interaction DEA\n",
    "    dc_markers = pd.read_csv('/Users/oipulk/Documents/scRNASeq/data/Eleftheria_Maranou_Mar2024/analysis/figures/pseudobulk/spleen/deseq2_interaction_results/Spleen_kopat_vs_rest_interaction_degs_aggregatedDC.csv')\n",
    "    dc_markers = dc_markers.rename(columns={\"gene\": \"Gene\"})\n",
    "\n",
    "##################################################\n",
    "else:\n",
    "    \n",
    "    ## Aggregated DC:\n",
    "    dc_markers = pd.read_csv('/Users/oipulk/Documents/scRNASeq/data/Eleftheria_Maranou_Mar2024/analysis/figures/pseudobulk/spleen/deseq2_pathogenic_results/Spleen_pat_kowt_degs_aggregatedDC.csv')\n",
    "    dc_markers = dc_markers.rename(columns={\"Unnamed: 0\": \"Gene\"})\n",
    "\n",
    "# Restrict the set of genes to high LFC\n",
    "# dc_markers_up = dc_markers[dc_markers['log2FoldChange']>1.0]\n",
    "# dc_markers_down = dc_markers[dc_markers['log2FoldChange']<-1.0]\n",
    "\n",
    "dc_markers_up = dc_markers[dc_markers['log2FoldChange']>0.5]\n",
    "dc_markers_down = dc_markers[dc_markers['log2FoldChange']<-0.5]\n",
    "\n",
    "# Drop the KO gene if it is still in the set (not necessarily in interaction analysis)\n",
    "cd74_up_idx = dc_markers_up[dc_markers_up['Gene']=='Cd74'].index\n",
    "if len(cd74_up_idx)>0:\n",
    "    dc_markers_up = dc_markers_up.drop(cd74_up_idx[0], axis=0)\n",
    "\n",
    "cd74_down_idx = dc_markers_down[dc_markers_down['Gene']=='Cd74'].index\n",
    "if len(cd74_down_idx)>0:\n",
    "    dc_markers_down = dc_markers_down.drop(cd74_down_idx[0], axis=0)\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f5b9f66f-5dff-47f7-86f2-8617c60ef98f",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Get all GO terms\n",
    "\n",
    "immune_filter = True\n",
    "\n",
    "# df_all = go_it(dc_markers.Gene.values, immune_filter=False)\n",
    "\n",
    "# Get only immune-relevant terms\n",
    "df_immune_up = go_it(dc_markers_up.Gene.values, immune_filter=immune_filter)\n",
    "df_immune_down = go_it(dc_markers_down.Gene.values, immune_filter=immune_filter)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "48b90e0e-934f-422a-972e-9e7a7e3c6156",
   "metadata": {},
   "outputs": [],
   "source": [
    "df_immune_up['up_down'] = 'up'\n",
    "df_immune_down['up_down'] = 'down'\n",
    "\n",
    "df_immune_up['-logp_corr']= -np.log(eps+df_immune_up['p_corr'])\n",
    "df_immune_up['per'] = (df_immune_up.n_genes/df_immune_up.n_go).astype(float)\n",
    "\n",
    "df_immune_up['score'] =  np.log(df_immune_up['-logp_corr']) + np.log(df_immune_up['per'])\n",
    "df_immune_up = df_immune_up.sort_values('-logp_corr')[::-1]\n",
    "\n",
    "\n",
    "df_immune_down['-logp_corr']= -np.log(eps+df_immune_down['p_corr'])\n",
    "df_immune_down['per'] = (df_immune_down.n_genes/df_immune_down.n_go).astype(float)\n",
    "\n",
    "df_immune_down['score'] =  np.log(df_immune_down['-logp_corr']) + np.log(df_immune_down['per'])\n",
    "df_immune_down = df_immune_down.sort_values('-logp_corr')[::-1]\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c641767c-fe50-42dc-820e-be0326a4aef5",
   "metadata": {},
   "outputs": [],
   "source": [
    "## Here you can choose between up and down regulated DEGs, or all of them\n",
    "## For interaction DEA, use up-regulated only (only a few DR, inparticular Hba, Hbb)\n",
    "upregulated_degs_only = False\n",
    "downregulated_degs_only = False\n",
    "\n",
    "if downregulated_degs_only|upregulated_degs_only:\n",
    "\n",
    "    if downregulated_degs_only:\n",
    "        df = df_immune_down.copy()\n",
    "    \n",
    "    if upregulated_degs_only:\n",
    "        df = df_immune_up.copy()\n",
    "\n",
    "else:\n",
    "\n",
    "    df = pd.concat([df_immune_up, df_immune_down])\n",
    "\n",
    "# Sort by p_corrected\n",
    "df= df.sort_values('-logp_corr')[::-1]\n",
    "# or by score\n",
    "# df= df.sort_values('score')[::-1]\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f2738784-7548-4435-8810-52df8dc1389f",
   "metadata": {},
   "outputs": [],
   "source": [
    "### Save GO data\n",
    "############\n",
    "\n",
    "if interaction:\n",
    "    \n",
    "    if upregulated_degs_only:\n",
    "        \n",
    "        df.to_csv('/Users/oipulk/Documents/scRNASeq/data/Eleftheria_Maranou_Mar2024/analysis/figures/GO/spleen/interactionDEA/aggregatedDC/spleen_GO_aggreagatedDC_interaction_patKO_upregulated_degs.csv')\n",
    "\n",
    "    else:\n",
    "            \n",
    "        df.to_csv('/Users/oipulk/Documents/scRNASeq/data/Eleftheria_Maranou_Mar2024/analysis/figures/GO/spleen/interactionDEA/aggregatedDC/spleen_GO_aggreagatedDC_interaction_patKO_all_degs.csv')\n",
    "\n",
    "\n",
    "\n",
    "else:\n",
    "    \n",
    "    if upregulated_degs_only:\n",
    "\n",
    "        df.to_csv('/Users/oipulk/Documents/scRNASeq/data/Eleftheria_Maranou_Mar2024/analysis/figures/GO/spleen/KO_vs_WT_in_pathogenic/aggregatedDC/spleen_GO_aggreagatedDC_KO_vs_WT_upregulated_degs.csv')\n",
    "\n",
    "    else:\n",
    "\n",
    "        df.to_csv('/Users/oipulk/Documents/scRNASeq/data/Eleftheria_Maranou_Mar2024/analysis/figures/GO/spleen/KO_vs_WT_in_pathogenic/aggregatedDC/spleen_GO_aggreagatedDC_KO_vs_WT_all_degs.csv')\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6955e992-b9d9-4a7f-bee2-4910c975e2da",
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "immune_relevance_filter = True\n",
    "\n",
    "if immune_relevance_filter:\n",
    "    print(df[df['immune_relevance']>0])\n",
    "\n",
    "else: \n",
    "    print(df)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "27f96267-d435-42e2-9ba6-a265ae8e55e0",
   "metadata": {},
   "outputs": [],
   "source": [
    "list(df[df['immune_relevance']>0]['term'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "eabcfbf6-c478-4ffc-a69b-07c23e62db98",
   "metadata": {},
   "outputs": [],
   "source": [
    "df[df['term']=='dendritic cell migration']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "bdffe625-af62-45d7-8e51-34afbc715e1c",
   "metadata": {},
   "outputs": [],
   "source": [
    "df[(df['per']>0.05)&(df['p_corr']<0.02)]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b7f6de47-3918-482b-b010-e4698f4b0afb",
   "metadata": {},
   "outputs": [],
   "source": [
    "plt.scatter(np.log(df['per']),-np.log(df['p_corr']))\n",
    "plt.scatter(np.log(df[df['term']=='dendritic cell migration']['per']),-np.log(df[df['term']=='dendritic cell migration']['p_corr']))\n",
    "plt.axvline(np.log(0.05),color='mediumseagreen')\n",
    "plt.axhline(-np.log(0.02),color='mediumseagreen')\n",
    "plt.xlabel('genes per pathway')\n",
    "plt.ylabel('-log(p_corrected)')\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "451cb403-82d9-499d-aabd-bd320698bc73",
   "metadata": {},
   "outputs": [],
   "source": [
    "df[df['class']=='cellular_component']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6e10cccb-6b3c-4193-aa11-c7fd7828646c",
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "\n",
    "#Take only low p-value enrichments and the ones with a decent coverage\n",
    "if interaction:\n",
    "    \n",
    "    # For interaction DEA, we need a less strict threshold\n",
    "    df = df[(df['p_corr']<0.02)&(df['per']>=0.05)] ## Defaults: p_corr<0.01, per>0.05\n",
    "\n",
    "else:\n",
    "\n",
    "    df = df[(df['p_corr']<0.02)&(df['per']>=0.05)]\n",
    "\n",
    "# Colorbar from all p_corr (BP, MF, CC)\n",
    "\n",
    "import matplotlib as mpl\n",
    "import textwrap\n",
    "\n",
    "fig, ax = plt.subplots(figsize = (0.5, 2.75), dpi=600)\n",
    "\n",
    "cmap = mpl.cm.coolwarm_r\n",
    "norm = mpl.colors.Normalize(vmin = df.p_corr.min(), vmax =df.p_corr.max())\n",
    "\n",
    "mapper = cm.ScalarMappable(norm = norm, cmap = cmap)\n",
    "\n",
    "cbl = mpl.colorbar.ColorbarBase(ax, cmap = cmap, norm = norm, orientation = 'vertical')\n",
    "plt.savefig('colorbar.pdf', bbox_inches='tight')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "538502d0-fd69-4ff7-9db4-9c585db2ec6d",
   "metadata": {},
   "outputs": [],
   "source": [
    "immune_relevance_filter = True"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a7e41d36-1008-4480-b57a-fb08bdfa6d39",
   "metadata": {},
   "outputs": [],
   "source": [
    "plot_fraction=True\n",
    "\n",
    "if upregulated_degs_only==True:\n",
    "\n",
    "    data_df = df[df['up_down']=='up'].copy()\n",
    "\n",
    "if downregulated_degs_only==True:\n",
    "\n",
    "    data_df = df[df['up_down']=='down'].copy()\n",
    "\n",
    "if (upregulated_degs_only==False)&(downregulated_degs_only==False):\n",
    "    \n",
    "    data_df = df.copy()\n",
    "\n",
    "if immune_relevance_filter:\n",
    "\n",
    "    data_df = data_df[data_df['immune_relevance']>0]\n",
    "\n",
    "\n",
    "data_df = data_df[(data_df['class']=='biological_process')|(data_df['class']=='cellular_component')].copy()\n",
    "\n",
    "# data_df['score'] =  data_df['-logp_corr']/np.max(data_df['-logp_corr']) + data_df['per']/np.max(data_df['per'])\n",
    "\n",
    "# data_df = data_df[(data_df['class']=='biological_process')].sort_values('p_corr')\n",
    "# data_df = data_df.sort_values('score')[::-1]\n",
    "\n",
    "\n",
    "plt.figure(figsize = (4,28))\n",
    "\n",
    "if plot_fraction:\n",
    "    \n",
    "    ax = sns.barplot(data = data_df, x = 'per', y = 'term', width=0.7, palette = mapper.to_rgba(data_df.p_corr.values))\n",
    "\n",
    "else:\n",
    "    \n",
    "    ax = sns.barplot(data = data_df, x = 'n_genes', y = 'term', width=0.7, palette = mapper.to_rgba(data_df.p_corr.values))\n",
    "\n",
    "\n",
    "\n",
    "ax.set_yticklabels([textwrap.fill(e, 32) for e in data_df['term']],fontsize=10)\n",
    "ax.set_ylabel('Biological process', fontsize=12)\n",
    "\n",
    "if plot_fraction:\n",
    "\n",
    "    ax.set_xlabel('GO term gene fraction', fontsize=12)\n",
    "\n",
    "else:\n",
    "\n",
    "    ax.set_xlabel('GO term genes', fontsize=12)\n",
    "\n",
    "\n",
    "\n",
    "##### TITLE #####\n",
    "\n",
    "### Interaction DEA\n",
    "if interaction:\n",
    "\n",
    "    if upregulated_degs_only:\n",
    "    \n",
    "        ax.set_title('GO: Spleen aggregated DC, interaction DEA', fontsize=12)\n",
    "\n",
    "    else:\n",
    "\n",
    "        ax.set_title('GO: Spleen aggregated DC, interaction DEA', fontsize=12)\n",
    "\n",
    "        \n",
    "### KO vs WT in pathogenic mice\n",
    "\n",
    "else:\n",
    "\n",
    "    if upregulated_degs_only:\n",
    "\n",
    "        ax.set_title('GO: Spleen aggregated DC, Pathogenic KO vs WT', fontsize=12)\n",
    "\n",
    "    else:\n",
    "        \n",
    "        ax.set_title('GO: Spleen aggregated DC, Pathogenic KO vs WT', fontsize=12)\n",
    "\n",
    "\n",
    "\n",
    "#### SAVE FIG #####\n",
    "\n",
    "### Interaction DEA\n",
    "if interaction:\n",
    "\n",
    "    if upregulated_degs_only:\n",
    "\n",
    "        if plot_fraction:\n",
    "\n",
    "            plt.savefig('/Users/oipulk/Documents/scRNASeq/data/Eleftheria_Maranou_Mar2024/analysis/figures/GO/spleen/interactionDEA/aggregatedDC/spleen_GO_BP_aggreagatedDC_interaction_patKO_upregulated_degs_per.pdf', dpi=600,bbox_inches = \"tight\")\n",
    "\n",
    "        else:\n",
    "\n",
    "            plt.savefig('/Users/oipulk/Documents/scRNASeq/data/Eleftheria_Maranou_Mar2024/analysis/figures/GO/spleen/interactionDEA/aggregatedDC/spleen_GO_BP_aggreagatedDC_interaction_patKO_upregulated_degs_ngenes.pdf', dpi=600,bbox_inches = \"tight\")\n",
    "\n",
    "    else:\n",
    "\n",
    "        if plot_fraction:\n",
    "            \n",
    "            plt.savefig('/Users/oipulk/Documents/scRNASeq/data/Eleftheria_Maranou_Mar2024/analysis/figures/GO/spleen/interactionDEA/aggregatedDC/spleen_GO_BP_aggreagatedDC_interaction_patKO_all_degs_per.pdf', dpi=600,bbox_inches = \"tight\")\n",
    "       \n",
    "        else:\n",
    "\n",
    "            plt.savefig('/Users/oipulk/Documents/scRNASeq/data/Eleftheria_Maranou_Mar2024/analysis/figures/GO/spleen/interactionDEA/aggregatedDC/spleen_GO_BP_aggreagatedDC_interaction_patKO_all_degs_ngenes.pdf', dpi=600,bbox_inches = \"tight\")\n",
    "\n",
    "\n",
    "### KO vs WT in pathogenic mice\n",
    "else:\n",
    "\n",
    "    if upregulated_degs_only:\n",
    "\n",
    "        if plot_fraction:\n",
    "\n",
    "            plt.savefig('/Users/oipulk/Documents/scRNASeq/data/Eleftheria_Maranou_Mar2024/analysis/figures/GO/spleen/KO_vs_WT_in_pathogenic/aggregatedDC/spleen_GO_BP_aggreagatedDC_KO_vs_WT_upregulated_degs_per.pdf', dpi=600,bbox_inches = \"tight\")\n",
    "\n",
    "        else:\n",
    "\n",
    "            plt.savefig('/Users/oipulk/Documents/scRNASeq/data/Eleftheria_Maranou_Mar2024/analysis/figures/GO/spleen/KO_vs_WT_in_pathogenic/aggregatedDC/spleen_GO_BP_aggreagatedDC_KO_vs_WT_upregulated_degs_ngenes.pdf', dpi=600,bbox_inches = \"tight\")\n",
    "\n",
    "    else:\n",
    "\n",
    "        if plot_fraction:\n",
    "            \n",
    "           plt.savefig('/Users/oipulk/Documents/scRNASeq/data/Eleftheria_Maranou_Mar2024/analysis/figures/GO/spleen/KO_vs_WT_in_pathogenic/aggregatedDC/spleen_GO_BP_aggreagatedDC_KO_vs_WT_all_degs_per.pdf', dpi=600,bbox_inches = \"tight\")\n",
    "\n",
    "        else:\n",
    "\n",
    "            plt.savefig('/Users/oipulk/Documents/scRNASeq/data/Eleftheria_Maranou_Mar2024/analysis/figures/GO/spleen/KO_vs_WT_in_pathogenic/aggregatedDC/spleen_GO_BP_aggreagatedDC_KO_vs_WT_all_degs_ngenes.pdf', dpi=600,bbox_inches = \"tight\")\n",
    "\n",
    "\n",
    "\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "86596ff6-2afe-4b21-87ea-8361ba61d95f",
   "metadata": {},
   "outputs": [],
   "source": [
    "data_df[data_df['class']=='cellular_component']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "911db017-9a2f-42f5-87b0-1e4d5a04f1c2",
   "metadata": {},
   "outputs": [],
   "source": [
    "from itertools import combinations\n",
    "import networkx as nx\n",
    "\n",
    "# Calculate Jaccard similarity between gene sets\n",
    "def jaccard_similarity(set1, set2):\n",
    "    intersection = len(set(set1) & set(set2))\n",
    "    union = len(set(set1) | set(set2))\n",
    "    return intersection / union if union > 0 else 0\n",
    "\n",
    "# Create a similarity matrix between terms\n",
    "def create_similarity_matrix(df):\n",
    "    n = len(df)\n",
    "    sim_matrix = np.zeros((n, n))\n",
    "    \n",
    "    for i, j in combinations(range(n), 2):\n",
    "        # Convert string representation of gene lists to actual sets\n",
    "        genes1 = set(df.iloc[i]['study_genes'])\n",
    "        genes2 = set(df.iloc[j]['study_genes'])\n",
    "        \n",
    "        sim = jaccard_similarity(genes1, genes2)\n",
    "        sim_matrix[i, j] = sim\n",
    "        sim_matrix[j, i] = sim\n",
    "    \n",
    "    return sim_matrix"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1e608f8c-838a-4e07-ba21-3c82b3780d50",
   "metadata": {},
   "outputs": [],
   "source": [
    "sim_matrix = create_similarity_matrix(data_df)\n",
    "\n",
    "# Create heatmap\n",
    "plt.figure(figsize=(12, 10))\n",
    "sns.heatmap(sim_matrix, \n",
    "            xticklabels=data_df['term'].str[:30],\n",
    "            yticklabels=data_df['term'].str[:30],\n",
    "            cmap='YlOrRd')\n",
    "plt.xticks(rotation=45, ha='right')\n",
    "plt.yticks(rotation=0)\n",
    "plt.title('Gene Set Overlap (Jaccard Similarity)')\n",
    "plt.tight_layout()\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "28ac734c-27f1-4c04-884e-02c88f5dd6bb",
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "from scipy.cluster.hierarchy import linkage, fcluster\n",
    "\n",
    "# Create a network graph based on similarity threshold\n",
    "G = nx.Graph()\n",
    "n = len(data_df)\n",
    "G.add_nodes_from(range(n))  \n",
    "\n",
    "for i in range(n):\n",
    "    for j in range(i+1, n):\n",
    "        if sim_matrix[i,j] >= 0.4:\n",
    "            G.add_edge(i, j, weight=sim_matrix[i,j])\n",
    "\n",
    "# Find connected components (clusters)\n",
    "clusters = list(nx.connected_components(G))\n",
    "\n",
    "# Create visualization\n",
    "plt.figure(figsize=(15, 10))\n",
    "\n",
    "# Use spring layout for node positioning\n",
    "pos = nx.spring_layout(G, k=1, iterations=50)\n",
    "\n",
    "# Draw the network\n",
    "nx.draw_networkx_nodes(G, pos, node_size=200, node_color='lightblue')\n",
    "nx.draw_networkx_edges(G, pos, width=2, alpha=0.5)\n",
    "\n",
    "# # Add labels\n",
    "# labels = {i: f\"{data_df['term'].iloc[i]}...\\n(-logp={data_df['-logp_corr'].iloc[i]:.2f})\" for i in range(len(data_df))}\n",
    "# nx.draw_networkx_labels(G, pos, labels, font_size=8)\n",
    "\n",
    "plt.title(\"GO Term Clusters (Similarity Threshold ≥ 0.5)\")\n",
    "plt.axis('off')\n",
    "plt.tight_layout()\n",
    "plt.show()\n",
    "\n",
    "# Print clusters with details\n",
    "print(\"\\nDetailed Clusters (Similarity ≥ 0.75):\")\n",
    "print(\"=\" * 80)\n",
    "for i, cluster in enumerate(clusters):\n",
    "    print(f\"\\nCluster {i+1}:\")\n",
    "    print(\"-\" * 40)\n",
    "    cluster_list = list(cluster)\n",
    "    # Sort cluster members by -logp_corr\n",
    "    cluster_list.sort(key=lambda x: data_df['score'].iloc[x], reverse=True)\n",
    "    \n",
    "    for idx in cluster_list:\n",
    "        print(f\"GO: {data_df['GO'].iloc[idx]}\")\n",
    "        print(f\"Term: {data_df['term'].iloc[idx]}\")\n",
    "        print(f\"-logp_corr: {data_df['-logp_corr'].iloc[idx]:.3f}\")\n",
    "        print(f\" p_corr: {data_df['p_corr'].iloc[idx]:.3f}\")\n",
    "        print(f\"score: {data_df['score'].iloc[idx]:.3f}\")\n",
    "        print(f\"Genes: {data_df['study_genes'].iloc[idx]}\")\n",
    "        print(\"-\" * 40)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "353bb04b-d354-4ac0-bae8-e3a039e1acaf",
   "metadata": {},
   "source": [
    "# Filter out more highest p-value pathways if necessary\n",
    "\n",
    "sig_cluster_ids = []\n",
    "\n",
    "for i, cluster in enumerate(clusters):\n",
    "    \n",
    "    cluster_list = list(cluster)\n",
    "    # Sort cluster members by -logp_corr\n",
    "    cluster_list.sort(key=lambda x: data_df['-logp_corr'].iloc[x], reverse=True)\n",
    "        \n",
    "    if np.exp(-data_df['-logp_corr'].iloc[cluster_list].mean())<0.01:\n",
    "\n",
    "        sig_cluster_ids.append(i)\n",
    "\n",
    "sig_clusters = list(np.array(clusters)[sig_cluster_ids])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6061514a-410a-4053-b586-9536301c3d96",
   "metadata": {},
   "outputs": [],
   "source": [
    "for k, cluster in enumerate(clusters):\n",
    "    print(k, cluster)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "873db327-87d7-4460-ba01-94abf1891917",
   "metadata": {},
   "outputs": [],
   "source": [
    "data_df['term']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "39118f92-a97f-475b-8ff2-e0c4e217c71d",
   "metadata": {},
   "outputs": [],
   "source": [
    "data_df['term']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5f30eacb-ddeb-4b80-8b25-e2d2ee68c393",
   "metadata": {},
   "outputs": [],
   "source": [
    "cluster_name_dict = {}\n",
    "cluster_gene_dict = {}\n",
    "cluster_p_dict = {}\n",
    "cluster_GO_dict = {}\n",
    "\n",
    "for k, cluster in enumerate(clusters):\n",
    "\n",
    "    # Let's remove Fat cell differentiation and Fibroblast GF pathways because pathways because these are likely \n",
    "    # secondary annotations due to genes' pleiotropic effects\n",
    "    if ((data_df['term'].iloc[list(clusters[k])[0]] != 'fat cell differentiation')&(data_df['term'].iloc[list(clusters[k])[0]] != 'response to fibroblast growth factor')):\n",
    "    \n",
    "        cluster_name_dict[k]= data_df['term'].iloc[list(clusters[k])[0]]\n",
    "        \n",
    "        cluster_GO_dict[k] = set(data_df['GO'].iloc[list(clusters[k])])\n",
    "        cluster_p_vals = data_df['p_corr'].iloc[list(clusters[k])[0]]\n",
    "        cluster_genes = set(data_df['study_genes'].iloc[list(clusters[k])[0]])\n",
    "        \n",
    "        for j in np.arange(1,len(list(clusters[k]))):\n",
    "        \n",
    "            cluster_genes = cluster_genes.union(set(data_df['study_genes'].iloc[list(clusters[k])[j]]))\n",
    "            cluster_p_vals = np.append(cluster_p_vals,data_df['p_corr'].iloc[list(clusters[k])[j]])\n",
    "            # cluster_terms = cluster_terms.union(set(data_df['term'].iloc[list(clusters[k])[j]]))\n",
    "        \n",
    "        cluster_gene_dict[k]=cluster_genes\n",
    "        cluster_p_dict[k]=cluster_p_vals\n",
    "        "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3d3919d2-37d0-4efc-84c3-94e08f9bac83",
   "metadata": {},
   "outputs": [],
   "source": [
    "cluster_name_dict"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f722b036-5993-4b3f-964d-6ad72435cf03",
   "metadata": {},
   "outputs": [],
   "source": [
    "#Simplify names\n",
    "cluster_name_dict = {0: 'TF AP-1 Complex',\n",
    " 1: 'p38MAPK Cascade',\n",
    " 2: 'T Cell Differentiation and Activation',\n",
    " 3: 'Interleukin-2 Production',\n",
    " 5: 'Leukocyte Homeostasis',\n",
    " 6: 'Non-canonical NF-kappaB Signaling',\n",
    " 7: 'MHC class II Antigen Processing and Presentation ',\n",
    " 8: 'Canonical NF-kappaB Signaling',\n",
    " 9: 'cAMP Response',\n",
    " 10: 'p53-mediated Signaling',\n",
    " 11: 'Apoptotic Signaling',\n",
    " 12: 'Cytoskeletal Signaling',\n",
    " 13: 'Mature B Cell Differentiation',\n",
    " 15: 'Interleukin-1 Production'}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3a652c73-1c92-4f78-856f-b1bc0f5488f4",
   "metadata": {},
   "outputs": [],
   "source": [
    "cluster_GO_dict"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "73a2d52f-2a2b-4d24-8721-b83f67c41df2",
   "metadata": {},
   "outputs": [],
   "source": [
    "cluster_p_dict"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5329ea1c-929b-425f-87c0-96b5b5db9daa",
   "metadata": {},
   "outputs": [],
   "source": [
    "cluster_gene_dict"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9a31f965-1f98-47c9-9bb2-1b2c2d91f19a",
   "metadata": {},
   "outputs": [],
   "source": [
    "from statsmodels.stats.multitest import multipletests\n",
    "\n",
    "## Read DESeq2 results \n",
    "def read_results(filename):\n",
    "    res = pd.read_csv(filename)\n",
    "    res.set_index('gene', inplace=True)\n",
    "    return res"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3a2cb3fa-f94f-4562-aba9-711fc86abe49",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Initialize dictionary to store results across repeats\n",
    "all_results = {}\n",
    "eps = 1e-300  # small constant to avoid log(0)\n",
    "\n",
    "for n_repeat in np.arange(1, 6):\n",
    "    # Read results for this repeat\n",
    "    results_df = read_results('/Users/oipulk/Documents/scRNASeq/data/Eleftheria_Maranou_Mar2024/analysis/figures/pseudobulk/spleen/deseq2_interaction_results/spleen_aggregatedDC_deseq2_interaction_results_'+str(n_repeat)+'.csv')\n",
    "\n",
    "    # Filter by base mean\n",
    "    res = results_df[results_df.baseMean > 0.25]\n",
    "    \n",
    "    # Store log of raw p-values (and other statistics)\n",
    "    res['log_pvalue'] = -np.log(eps + res['pvalue'])\n",
    "    \n",
    "    if n_repeat == 1:\n",
    "        # Initialize storage for all genes we'll see\n",
    "        de_res = res.copy()\n",
    "        # Initialize storage for running average of log p-values\n",
    "        all_genes = set(res.index)\n",
    "    else:\n",
    "        # Update set of all genes we've seen\n",
    "        all_genes.update(res.index)\n",
    "    \n",
    "    # Store results for each repeat\n",
    "    for idx in res.index:\n",
    "        if idx not in all_results:\n",
    "            all_results[idx] = {\n",
    "                'log_pvalue': [],\n",
    "                'baseMean': [],\n",
    "                'log2FoldChange': [],\n",
    "                'lfcSE': [],\n",
    "                'stat': []\n",
    "            }\n",
    "        \n",
    "        all_results[idx]['log_pvalue'].append(res.loc[idx, 'log_pvalue'])\n",
    "        all_results[idx]['baseMean'].append(res.loc[idx, 'baseMean'])\n",
    "        all_results[idx]['log2FoldChange'].append(res.loc[idx, 'log2FoldChange'])\n",
    "        all_results[idx]['lfcSE'].append(res.loc[idx, 'lfcSE'])\n",
    "        all_results[idx]['stat'].append(res.loc[idx, 'stat'])\n",
    "\n",
    "# Create final results dataframe\n",
    "final_results = []\n",
    "for gene in all_genes:\n",
    "    if gene in all_results:\n",
    "        # Calculate mean of log p-values\n",
    "        mean_log_pvalue = np.mean(all_results[gene]['log_pvalue'])\n",
    "        # Convert back to p-value\n",
    "        pvalue = np.exp(-mean_log_pvalue)\n",
    "        \n",
    "        # Calculate means of other statistics\n",
    "        row = {\n",
    "            'gene': gene,\n",
    "            'pvalue': pvalue,\n",
    "            'baseMean': np.mean(all_results[gene]['baseMean']),\n",
    "            'log2FoldChange': np.mean(all_results[gene]['log2FoldChange']),\n",
    "            'lfcSE': np.mean(all_results[gene]['lfcSE']),\n",
    "            'stat': np.mean(all_results[gene]['stat'])\n",
    "        }\n",
    "        final_results.append(row)\n",
    "\n",
    "# Convert to dataframe\n",
    "final_df = pd.DataFrame(final_results)\n",
    "final_df.set_index('gene', inplace=True)\n",
    "\n",
    "# Remove NaN values before running multipletests\n",
    "valid_mask = ~np.isnan(final_df['pvalue'])\n",
    "valid_pvals = final_df.loc[valid_mask, 'pvalue']\n",
    "\n",
    "# Run multipletests only on valid p-values\n",
    "adj_pvals = np.full(len(final_df), np.nan)\n",
    "adj_pvals[valid_mask] = multipletests(valid_pvals.values, method='fdr_bh')[1]\n",
    "final_df['padj'] = adj_pvals\n",
    "\n",
    "# # Store log-transformed versions\n",
    "final_df['log_pvalue'] = -np.log(eps + final_df['pvalue'])\n",
    "final_df['log_padj'] = -np.log(eps + final_df['padj'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d12280de-ebee-4a29-a32e-e771f23701b3",
   "metadata": {},
   "outputs": [],
   "source": [
    "de_res = final_df\n",
    "\n",
    "sns.set_style(\"white\")\n",
    "plt.figure(figsize=(9,6))\n",
    "\n",
    "\n",
    "plt.scatter(de_res['log2FoldChange'], np.log(de_res['log_padj']) , facecolor='dodgerblue', edgecolor='blue',linewidths=0.5, zorder=10,alpha=0.7, s=14)\n",
    "\n",
    "plt.axhline( np.log(-np.log(0.05)) , linestyle='--', linewidth=0.8, alpha=0.7)\n",
    "\n",
    "degs = de_res[(de_res['log_padj'] > -np.log(eps+0.05))&(np.abs(de_res['log2FoldChange'])>0.5)]\n",
    "\n",
    "sig_logfoldchanges = degs['log2FoldChange']\n",
    "sig_log_padj = degs['log_padj']\n",
    "sig_var_names = list(degs.index)\n",
    "\n",
    "top_ind = np.argsort(np.abs(sig_logfoldchanges)+3*np.log(sig_log_padj))[::-1]\n",
    "top_names = np.asarray(sig_var_names)[top_ind]\n",
    "\n",
    "for i, gene_name in enumerate(top_names[0:100]):\n",
    "\n",
    "    # if gene_name=='Cd86':\n",
    "    # print(i, gene_name)\n",
    "    plt.text(sig_logfoldchanges[top_ind[i]]+0.03, np.log(sig_log_padj[top_ind[i]])+0.05, str(gene_name), fontsize=10, zorder=15)\n",
    "\n",
    "select_names = ['Fos','Jun','Junb']\n",
    "select_ind = []\n",
    "[select_ind.append(np.where(np.array(sig_var_names, dtype='str')==select_names[k])[0][0]) for k in np.arange(0,len(select_names))]\n",
    "\n",
    "for i, gene_name in enumerate(select_names):\n",
    "    \n",
    "    plt.text(sig_logfoldchanges[select_ind[i]]+0.03, np.log(sig_log_padj[select_ind[i]])+0.05, str(gene_name), fontsize=10, zorder=15)\n",
    "\n",
    "\n",
    "plt.xlim([-8,9])\n",
    "plt.ylim([0.75,5])\n",
    "\n",
    "plt.ylabel('log(-log p)', fontsize=13)\n",
    "plt.xlabel('LFC', fontsize=13)\n",
    "\n",
    "# plt.ylim([-0.2,np.max(sig_log_padj_kowt[np.asarray(sig_var_names_kowt)!='Cd74'])+0.5])\n",
    "# plt.title('Difference in response to pathogenic condition in spleen: KO vs WT')\n",
    "plt.title('Diff. response to pathogenicity in spleen aggregated DCs: KO vs WT', fontsize='small')\n",
    "\n",
    "# plt.savefig('/Users/oipulk/Documents/scRNASeq/data/Eleftheria_Maranou_Mar2024/analysis/figures/pseudobulk/spleen/deseq2_interaction_results/volcano_kopat_vs_rest_pathogenic_spleen_DC2_interaction.pdf', dpi=600,bbox_inches = \"tight\")\n",
    "# plt.savefig('/Users/oipulk/Documents/scRNASeq/data/Eleftheria_Maranou_Mar2024/analysis/figures/pseudobulk/spleen/deseq2_interaction_results/volcano_kopat_vs_rest_pathogenic_spleen_CD8minusDC1_interaction.pdf', dpi=600,bbox_inches = \"tight\")\n",
    "plt.savefig('/Users/oipulk/Documents/scRNASeq/data/Eleftheria_Maranou_Mar2024/analysis/figures/pseudobulk/spleen/deseq2_interaction_results/volcano_kopat_vs_rest_pathogenic_spleen_aggregatedDC_interaction.pdf', dpi=600,bbox_inches = \"tight\")\n",
    "\n",
    "plt.show()\n",
    "\n",
    "top_genes=np.asarray(sig_var_names)[np.argsort(np.abs(sig_logfoldchanges)+sig_log_padj)[::-1]]\n",
    "\n",
    "print('top_genes:',top_genes )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "93dd9eb3-4f22-4cf1-b363-f60cea5a2a02",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Read the files made with DESeq2 in R\n",
    "raw_counts = pd.read_csv(\"/Users/oipulk/Documents/scRNASeq/data/Eleftheria_Maranou_Mar2024/analysis/figures/pseudobulk/spleen/deseq2_interaction_results/res_raw_counts.csv\", index_col=0)\n",
    "norm_counts = pd.read_csv(\"/Users/oipulk/Documents/scRNASeq/data/Eleftheria_Maranou_Mar2024/analysis/figures/pseudobulk/spleen/deseq2_interaction_results/res_normalized_counts.csv\", index_col=0)\n",
    "gene_info = pd.read_csv(\"/Users/oipulk/Documents/scRNASeq/data/Eleftheria_Maranou_Mar2024/analysis/figures/pseudobulk/spleen/deseq2_interaction_results/res_gene_info.csv\", index_col=0)\n",
    "sample_info = pd.read_csv(\"/Users/oipulk/Documents/scRNASeq/data/Eleftheria_Maranou_Mar2024/analysis/figures/pseudobulk/spleen/deseq2_interaction_results/res_sample_info.csv\", index_col=0)\n",
    "\n",
    "# Create AnnData object with raw counts as main matrix\n",
    "dds_adata = ad.AnnData(\n",
    "    X=raw_counts.T.values,\n",
    "    var=gene_info,\n",
    "    obs=sample_info\n",
    ")\n",
    "\n",
    "# Set the index names\n",
    "dds_adata.var_names = raw_counts.index\n",
    "dds_adata.obs_names = raw_counts.columns\n",
    "\n",
    "# Add normalized counts in layers\n",
    "dds_adata.layers['normalized'] = norm_counts.T.values\n",
    "\n",
    "# Add size factors to obs\n",
    "dds_adata.obs['size_factors'] = sample_info['sizeFactor']\n",
    "\n",
    "# Store information about what's in the layers\n",
    "dds_adata.uns['layer_description'] = {\n",
    "    'X': 'raw_counts',\n",
    "    'normalized': 'deseq2_normalized_counts'\n",
    "}\n",
    "\n",
    "# Verify the structure\n",
    "print(dds_adata)\n",
    "print(\"\\nAvailable layers:\", list(dds_adata.layers.keys()))\n",
    "print(\"\\nShape of raw counts:\", dds_adata.X.shape)\n",
    "print(\"Shape of normalized counts:\", dds_adata.layers['normalized'].shape)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ba00fe82-9822-429e-977c-4962cc5aa390",
   "metadata": {},
   "outputs": [],
   "source": [
    "import matplotlib.patches as patches\n",
    "import matplotlib.path as mpath\n",
    "from matplotlib.collections import PatchCollection"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "29a7515c-1e1d-429e-bf9b-8a0a2ff07080",
   "metadata": {},
   "outputs": [],
   "source": [
    "len(clusters)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "640b4188-f73d-4179-8163-5459c93402d5",
   "metadata": {},
   "outputs": [],
   "source": [
    "#Define colormap for GO terms\n",
    "\n",
    "# Get colors from tab20\n",
    "tab20_colors = plt.cm.tab20.colors\n",
    "tab20b_colors = plt.cm.tab20b.colors\n",
    "\n",
    "# Combine colors from tab20 and tab20b\n",
    "# combined_colors = [plt.cm.tab20b.colors[0]] + list(tab20_colors)[0:10]+ [plt.cm.tab20c.colors[6]]+ [plt.cm.tab20b.colors[11]] + list(tab20_colors)[16:20]  + [plt.cm.tab20b.colors[2]]\n",
    "combined_colors = list(tab20_colors)[0:10]+  list(tab20_colors)[16:17] + list(tab20_colors)[18:19]  + [plt.cm.tab20b.colors[2]] + [plt.cm.tab20b.colors[0]] \n",
    "\n",
    "# Create a new ListedColormap\n",
    "from matplotlib.colors import ListedColormap\n",
    "custom_cmap = ListedColormap(combined_colors)\n",
    "\n",
    "# Verify the number of colors\n",
    "print(f\"Number of colors in new colormap: {len(custom_cmap.colors)}\")\n",
    "\n",
    "# Optional: Visualize the colormap\n",
    "fig, ax = plt.subplots(figsize=(10, 1))\n",
    "gradient = np.linspace(0, 1, len(custom_cmap.colors))\n",
    "gradient = np.vstack((gradient, gradient))\n",
    "ax.imshow(gradient, aspect='auto', cmap=custom_cmap)\n",
    "ax.set_axis_off()\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "538126ec-20f5-400e-a924-2048f6ea91f5",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Z-score normalization over rows\n",
    "\n",
    "dds =dds_adata.copy()\n",
    "\n",
    "# degs = de_res[(de_res['log_padj'] > -np.log(eps+0.05))&(np.abs(de_res['log2FoldChange'])>0.5)]\n",
    "\n",
    "# Only upregulated degs\n",
    "degs = de_res[(de_res['log_padj'] > -np.log(eps+0.05))&(de_res['log2FoldChange']>0.5)]\n",
    "\n",
    "\n",
    "sig_logfoldchanges = degs['log2FoldChange']\n",
    "sig_log_padj = degs['log_padj']\n",
    "sig_var_names = list(degs.index)\n",
    "\n",
    "top_genes=np.asarray(sig_var_names)[np.argsort(np.abs(sig_logfoldchanges)+sig_log_padj)[::-1]]\n",
    "\n",
    "dds.layers['log1p'] = np.log1p(dds.layers['normalized'])\n",
    "dds_sig = dds[:, sig_var_names]\n",
    "\n",
    "log1p_counts = pd.DataFrame(dds_sig.layers['log1p'], index=dds_sig.obs.index, columns=dds_sig.var_names)\n",
    "\n",
    "# conditions = ['wt_naive','wt_pathogenic','ko_naive','ko_pathogenic']  \n",
    "conditions = ['ko_pathogenic','ko_naive','wt_pathogenic','wt_naive']  \n",
    "\n",
    "#Construct a dataframe that will hold the mean expression levels of the filtered genes in various conditions\n",
    "mean_log1p_counts = pd.DataFrame(index=conditions, columns=dds_sig.var_names, dtype='float')\n",
    "\n",
    "for condition in conditions:\n",
    "    # Get all rows that start with the current condition\n",
    "    rows = [row for row in log1p_counts.index if row.startswith(condition)]\n",
    "    \n",
    "    # Calculate the mean of these rows\n",
    "    condition_mean = np.nanmean(log1p_counts.loc[rows],axis=0)\n",
    "    \n",
    "    mean_log1p_counts.loc[condition] =  condition_mean.astype(float)\n",
    "\n",
    "\n",
    "diff_ko = mean_log1p_counts.loc['ko_pathogenic'] - mean_log1p_counts.loc['ko_naive']\n",
    "diff_wt = mean_log1p_counts.loc['wt_pathogenic'] - mean_log1p_counts.loc['wt_naive']\n",
    "ddiff_ko_wt = diff_ko - diff_wt\n",
    "\n",
    "plot_genes = top_genes\n",
    "\n",
    "# Create a list of '+' and '-' signs based on the difference\n",
    "signs_ko = {gene: '+' if d > 0 else '-' for gene, d in diff_ko[plot_genes].items()}\n",
    "signs_wt = {gene: '+' if d > 0 else '-' for gene, d in diff_wt[plot_genes].items()}\n",
    "signs_ko_wt = {gene: '+' if d > 0 else '-' for gene, d in ddiff_ko_wt[plot_genes].items()}\n",
    "\n",
    "\n",
    "sns.set_style(\"white\")\n",
    "plt.figure(dpi=1200)\n",
    "\n",
    "g=sns.clustermap(mean_log1p_counts[plot_genes].T,\n",
    "               z_score=0,\n",
    "               # standard_scale=0,  \n",
    "               cmap='RdYlBu_r',\n",
    "               yticklabels=True,\n",
    "               xticklabels=['KO Pathogenic', 'KO Naive' ,'WT Pathogenic','WT Naive'],\n",
    "               col_cluster=False,\n",
    "               row_cluster=True,\n",
    "               cbar_pos=None,\n",
    "               dendrogram_ratio=(.3, 0.),\n",
    "               tree_kws=dict(linewidths=1.0, colors=(0.2, 0.2, 0.4)),\n",
    "               figsize = (8,24),\n",
    "               # cbar_pos=(0.85, 0.8, 0.1, 0.05),\n",
    "              )\n",
    "\n",
    "# 2. Label modifications\n",
    "g.ax_heatmap.set_yticklabels(g.ax_heatmap.get_ymajorticklabels(), fontsize=9, rotation=180)\n",
    "g.ax_heatmap.set_xticklabels(g.ax_heatmap.get_xmajorticklabels(), fontsize=9, rotation=180)\n",
    "\n",
    "# 3. Get the figure and axis references\n",
    "fig = plt.gcf()\n",
    "ax = g.ax_heatmap\n",
    "\n",
    "# 4. Add the signs (your existing code)\n",
    "reordered_genes = g.dendrogram_row.reordered_ind\n",
    "for idx, gene_idx in enumerate(reordered_genes):\n",
    "    gene = plot_genes[gene_idx]\n",
    "    sign_ko = signs_ko[gene]\n",
    "    sign_wt = signs_wt[gene]\n",
    "    sign_ko_wt = signs_ko_wt[gene]\n",
    "\n",
    "    #Sign of ko response\n",
    "    rect = patches.Rectangle((0.925, idx), 0.15, 1, fill=True, facecolor='white', edgecolor='none', alpha=0.8)\n",
    "    ax.add_patch(rect)\n",
    "    if sign_ko=='+':\n",
    "        ax.text(1.01, idx+0.5, sign_ko, ha='center', va='center', fontweight='medium', fontsize=10, rotation=90)\n",
    "    else:\n",
    "        ax.text(1.0, idx+0.5, sign_ko, ha='center', va='center', fontweight='medium', fontsize=10, rotation=90)\n",
    "\n",
    "    \n",
    "    #Sign of wt response\n",
    "    rect = patches.Rectangle((2.925, idx), 0.15, 1, fill=True, facecolor='white', edgecolor='none', alpha=0.8)\n",
    "    ax.add_patch(rect)\n",
    "    if sign_wt=='+':\n",
    "        ax.text(3.01, idx+0.5, sign_wt, ha='center', va='center', fontweight='medium', fontsize=10, rotation=90)\n",
    "    else:\n",
    "        ax.text(3.0, idx+0.5, sign_wt, ha='center', va='center', fontweight='medium', fontsize=10, rotation=90)\n",
    "\n",
    "    #Sign of ko-wt response difference\n",
    "    rect = patches.Rectangle((1.9, idx), 0.2, 1, fill=True, facecolor='white', edgecolor='none', alpha=0.9)\n",
    "    ax.add_patch(rect)\n",
    "\n",
    "    if sign_ko_wt=='+':\n",
    "        ax.text(2.01, idx+0.5, sign_ko_wt, ha='center', va='center', fontweight='bold', fontsize=12, rotation=90)\n",
    "    else:\n",
    "        ax.text(2.0, idx+0.5, sign_ko_wt, ha='center', va='center', fontweight='bold', fontsize=12, rotation=90)\n",
    "\n",
    "\n",
    "# Define colors for 12 clusters\n",
    "# cluster_colors = plt.cm.tab20(np.reshape(np.linspace(0, 1, len(custom_cmap.colors)),(int(len(custom_cmap.colors)/2),2)).flatten(order='F'))\n",
    "cluster_colors = custom_cmap(np.linspace(0, 1, len(custom_cmap.colors)))\n",
    "# cluster_colors = plt.cm.tab20(np.linspace(0, 1, len(custom_cmap.colors)))\n",
    "\n",
    "\n",
    "# Create a new axes for the cluster labels\n",
    "ax_labels = fig.add_axes([\n",
    "    ax.get_position().x1 + 0.05,\n",
    "    ax.get_position().y0,\n",
    "    0.5,\n",
    "    ax.get_position().height\n",
    "])\n",
    "\n",
    "ax_labels.set_frame_on(False)\n",
    "ax_labels.set_xticks([])\n",
    "ax_labels.set_yticks([])\n",
    "\n",
    "# Get the actual y-coordinates from the heatmap\n",
    "heatmap_height = ax.get_position().height\n",
    "y_scale = heatmap_height / len(plot_genes)\n",
    "y_offset = ax.get_position().y0\n",
    "\n",
    "# Create correct gene position mapping using scaled coordinates\n",
    "gene_positions = {}\n",
    "for idx, gene_idx in enumerate(reordered_genes):\n",
    "    gene = plot_genes[gene_idx]\n",
    "    # Scale the position to match the heatmap coordinates. Use the small deviation after -idx-1 to adjust the centre)\n",
    "    scaled_pos = y_offset + (len(plot_genes) - idx - 1 + 0.5) * y_scale\n",
    "    gene_positions[gene] = scaled_pos\n",
    "\n",
    "# # Print some debugging info\n",
    "# print(\"Y-scale:\", y_scale)\n",
    "# print(\"Y-offset:\", y_offset)\n",
    "# print(\"Sample of gene positions:\", dict(list(gene_positions.items())[:5]))\n",
    "\n",
    "def curved_path(start_pos, end_pos):\n",
    "    return mpath.Path([\n",
    "        (0, start_pos),\n",
    "        (0.5, start_pos),\n",
    "        (1.0, end_pos),\n",
    "        (1.5, end_pos)\n",
    "    ], [\n",
    "        mpath.Path.MOVETO,\n",
    "        mpath.Path.CURVE4,\n",
    "        mpath.Path.CURVE4,\n",
    "        mpath.Path.CURVE4\n",
    "    ])\n",
    "\n",
    "# Calculate mean positions for each cluster\n",
    "cluster_mean_positions = {}\n",
    "for cluster_idx, genes in cluster_gene_dict.items():\n",
    "    valid_positions = [gene_positions[gene] for gene in genes if gene in gene_positions]\n",
    "    if valid_positions:\n",
    "        cluster_mean_positions[cluster_idx] = np.mean(valid_positions)\n",
    "\n",
    "# Sort clusters by their mean position\n",
    "sorted_clusters = sorted(cluster_mean_positions.items(), key=lambda x: x[1])\n",
    "\n",
    "# Calculate evenly spaced positions within the same coordinate system\n",
    "y_min = ax.get_position().y0\n",
    "y_max = ax.get_position().y1\n",
    "total_height = y_max - y_min\n",
    "n_clusters = len(sorted_clusters)\n",
    "spacing = total_height / (n_clusters + 1)\n",
    "\n",
    "evenly_spaced_positions = {\n",
    "    cluster_idx: y_min + spacing * (i + 1)\n",
    "    for i, (cluster_idx, _) in enumerate(sorted_clusters)\n",
    "}\n",
    "\n",
    "# Draw connections for each cluster\n",
    "for cluster_idx, mean_pos in sorted_clusters:\n",
    "    genes = cluster_gene_dict[cluster_idx]\n",
    "    cluster_genes = [gene for gene in genes if gene in gene_positions]\n",
    "    cluster_GO_terms = list(cluster_GO_dict[cluster_idx])\n",
    "    go_terms_formatted = '\\n'.join(cluster_GO_terms)\n",
    "    p_values = cluster_p_dict[cluster_idx]\n",
    "    cluster_min_p = np.min(cluster_p_dict[cluster_idx])\n",
    "\n",
    "    # print(cluster_idx)\n",
    "    \n",
    "    if not cluster_genes:\n",
    "        continue\n",
    "    \n",
    "    # Use evenly spaced position for label\n",
    "    label_pos = evenly_spaced_positions[cluster_idx]\n",
    "\n",
    "    box_dist = 1.6\n",
    "    # Draw curved lines to each gene\n",
    "    for gene in cluster_genes:\n",
    "        gene_pos = gene_positions[gene]\n",
    "        path = curved_path(gene_pos, label_pos)\n",
    "        patch = patches.PathPatch(\n",
    "            path,\n",
    "            facecolor='none',\n",
    "            edgecolor=cluster_colors[np.where(np.array(sorted_clusters, dtype=int)[:,0]==cluster_idx)],\n",
    "            alpha=0.8,\n",
    "            linewidth=1.75\n",
    "        )\n",
    "        ax_labels.add_patch(patch)\n",
    "    \n",
    "    # Add cluster label\n",
    "    bbox_props = dict(\n",
    "        boxstyle=\"round,pad=0.3\",\n",
    "        fc=\"white\",\n",
    "        ec=cluster_colors[np.where(np.array(sorted_clusters, dtype=int)[:,0]==cluster_idx)],\n",
    "        alpha=1\n",
    "    )\n",
    "\n",
    "    \n",
    "    if (len(cluster_GO_terms)>1)&(len(cluster_GO_terms)<4):\n",
    "            ax_labels.text(\n",
    "                box_dist + 0.1 * cluster_idx, \n",
    "                label_pos,\n",
    "                f'{cluster_name_dict[cluster_idx]}\\n'\n",
    "                f'{len(cluster_genes)} genes\\n'\n",
    "                f'min. p={cluster_min_p:.3g}\\n'\n",
    "                f'{go_terms_formatted}',  # Now each GO term will be on its own line\n",
    "                color='black',\n",
    "                va='center',\n",
    "                fontsize=10,\n",
    "                bbox=bbox_props,\n",
    "                rotation=90\n",
    "            )\n",
    "\n",
    "    if len(cluster_GO_terms)==1:\n",
    "        ax_labels.text(\n",
    "            box_dist + 0.1 * cluster_idx, \n",
    "            label_pos,\n",
    "            f'{cluster_name_dict[cluster_idx]}\\n'\n",
    "            f'{len(cluster_genes)} genes\\n'\n",
    "            f'p={cluster_min_p:.3g}\\n'\n",
    "            f'{go_terms_formatted}',  # Now each GO term will be on its own line\n",
    "            color='black',\n",
    "            va='center',\n",
    "            fontsize=10,\n",
    "            bbox=bbox_props,\n",
    "            rotation=90\n",
    "        )\n",
    "        \n",
    "\n",
    "    if (len(cluster_GO_terms)>3):\n",
    "\n",
    "        ax_labels.text(\n",
    "                box_dist + 0.1 * cluster_idx, \n",
    "                label_pos,\n",
    "                f'{cluster_name_dict[cluster_idx]}\\n'\n",
    "                f'{len(cluster_genes)} genes\\n'\n",
    "                f'min. p={cluster_min_p:.3g}\\n'\n",
    "                f'{len(cluster_GO_terms)} Go terms',  # Now each GO term will be on its own line\n",
    "                color='black',\n",
    "                va='center',\n",
    "                fontsize=10,\n",
    "                bbox=bbox_props,\n",
    "                rotation=90\n",
    "            )\n",
    "# Set proper limits\n",
    "ax_labels.set_xlim(-0.1, 2.7)\n",
    "ax_labels.set_ylim(y_min, y_max)\n",
    "\n",
    "# plt.savefig('/Users/oipulk/Documents/scRNASeq/data/Eleftheria_Maranou_Mar2024/analysis/figures/Fig5/interactionDEA_heatmap_with_GO.pdf', dpi=600,bbox_inches = \"tight\")\n",
    "\n",
    "plt.show()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6d41dbcb-bffe-406a-8df1-86d362ea785d",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Without Z-score normalization\n",
    "\n",
    "dds =dds_adata.copy()\n",
    "\n",
    "# degs = de_res[(de_res['log_padj'] > -np.log(eps+0.05))&(np.abs(de_res['log2FoldChange'])>0.5)]\n",
    "\n",
    "# Only upregulated degs\n",
    "degs = de_res[(de_res['log_padj'] > -np.log(eps+0.05))&(de_res['log2FoldChange']>0.5)]\n",
    "\n",
    "\n",
    "sig_logfoldchanges = degs['log2FoldChange']\n",
    "sig_log_padj = degs['log_padj']\n",
    "sig_var_names = list(degs.index)\n",
    "\n",
    "top_genes=np.asarray(sig_var_names)[np.argsort(np.abs(sig_logfoldchanges)+sig_log_padj)[::-1]]\n",
    "\n",
    "dds.layers['log1p'] = np.log1p(dds.layers['normalized'])\n",
    "dds_sig = dds[:, sig_var_names]\n",
    "\n",
    "log1p_counts = pd.DataFrame(dds_sig.layers['log1p'], index=dds_sig.obs.index, columns=dds_sig.var_names)\n",
    "\n",
    "# conditions = ['wt_naive','wt_pathogenic','ko_naive','ko_pathogenic']  \n",
    "conditions = ['ko_pathogenic','ko_naive','wt_pathogenic','wt_naive']  \n",
    "\n",
    "#Construct a dataframe that will hold the mean expression levels of the filtered genes in various conditions\n",
    "mean_log1p_counts = pd.DataFrame(index=conditions, columns=dds_sig.var_names, dtype='float')\n",
    "\n",
    "for condition in conditions:\n",
    "    # Get all rows that start with the current condition\n",
    "    rows = [row for row in log1p_counts.index if row.startswith(condition)]\n",
    "    \n",
    "    # Calculate the mean of these rows\n",
    "    condition_mean = np.nanmean(log1p_counts.loc[rows],axis=0)\n",
    "    \n",
    "    mean_log1p_counts.loc[condition] =  condition_mean.astype(float)\n",
    "\n",
    "\n",
    "diff_ko = mean_log1p_counts.loc['ko_pathogenic'] - mean_log1p_counts.loc['ko_naive']\n",
    "diff_wt = mean_log1p_counts.loc['wt_pathogenic'] - mean_log1p_counts.loc['wt_naive']\n",
    "ddiff_ko_wt = diff_ko - diff_wt\n",
    "\n",
    "plot_genes = top_genes\n",
    "\n",
    "# Create a list of '+' and '-' signs based on the difference\n",
    "signs_ko = {gene: '+' if d > 0 else '-' for gene, d in diff_ko[plot_genes].items()}\n",
    "signs_wt = {gene: '+' if d > 0 else '-' for gene, d in diff_wt[plot_genes].items()}\n",
    "signs_ko_wt = {gene: '+' if d > 0 else '-' for gene, d in ddiff_ko_wt[plot_genes].items()}\n",
    "\n",
    "\n",
    "sns.set_style(\"white\")\n",
    "plt.figure(dpi=1200)\n",
    "\n",
    "g=sns.clustermap(mean_log1p_counts[plot_genes].T,\n",
    "               z_score=None,\n",
    "               # standard_scale=0,  \n",
    "               cmap='RdYlBu_r',\n",
    "               yticklabels=True,\n",
    "               xticklabels=['KO Pathogenic', 'KO Naive' ,'WT Pathogenic','WT Naive'],\n",
    "               col_cluster=False,\n",
    "               row_cluster=True,\n",
    "               cbar_pos=None,\n",
    "               dendrogram_ratio=(.3, 0.),\n",
    "               tree_kws=dict(linewidths=1.0, colors=(0.2, 0.2, 0.4)),\n",
    "               figsize = (8,24),\n",
    "               # cbar_pos=(0.85, 0.8, 0.1, 0.05),\n",
    "              )\n",
    "\n",
    "# 2. Label modifications\n",
    "g.ax_heatmap.set_yticklabels(g.ax_heatmap.get_ymajorticklabels(), fontsize=9, rotation=180)\n",
    "g.ax_heatmap.set_xticklabels(g.ax_heatmap.get_xmajorticklabels(), fontsize=9, rotation=180)\n",
    "\n",
    "# 3. Get the figure and axis references\n",
    "fig = plt.gcf()\n",
    "ax = g.ax_heatmap\n",
    "\n",
    "# 4. Add the signs (your existing code)\n",
    "reordered_genes = g.dendrogram_row.reordered_ind\n",
    "for idx, gene_idx in enumerate(reordered_genes):\n",
    "    gene = plot_genes[gene_idx]\n",
    "    sign_ko = signs_ko[gene]\n",
    "    sign_wt = signs_wt[gene]\n",
    "    sign_ko_wt = signs_ko_wt[gene]\n",
    "\n",
    "    #Sign of ko response\n",
    "    rect = patches.Rectangle((0.925, idx), 0.15, 1, fill=True, facecolor='white', edgecolor='none', alpha=0.8)\n",
    "    ax.add_patch(rect)\n",
    "    if sign_ko=='+':\n",
    "        ax.text(1.01, idx+0.5, sign_ko, ha='center', va='center', fontweight='medium', fontsize=10, rotation=90)\n",
    "    else:\n",
    "        ax.text(1.0, idx+0.5, sign_ko, ha='center', va='center', fontweight='medium', fontsize=10, rotation=90)\n",
    "\n",
    "    \n",
    "    #Sign of wt response\n",
    "    rect = patches.Rectangle((2.925, idx), 0.15, 1, fill=True, facecolor='white', edgecolor='none', alpha=0.8)\n",
    "    ax.add_patch(rect)\n",
    "    if sign_wt=='+':\n",
    "        ax.text(3.01, idx+0.5, sign_wt, ha='center', va='center', fontweight='medium', fontsize=10, rotation=90)\n",
    "    else:\n",
    "        ax.text(3.0, idx+0.5, sign_wt, ha='center', va='center', fontweight='medium', fontsize=10, rotation=90)\n",
    "\n",
    "    #Sign of ko-wt response difference\n",
    "    rect = patches.Rectangle((1.9, idx), 0.2, 1, fill=True, facecolor='white', edgecolor='none', alpha=0.9)\n",
    "    ax.add_patch(rect)\n",
    "\n",
    "    if sign_ko_wt=='+':\n",
    "        ax.text(2.01, idx+0.5, sign_ko_wt, ha='center', va='center', fontweight='bold', fontsize=12, rotation=90)\n",
    "    else:\n",
    "        ax.text(2.0, idx+0.5, sign_ko_wt, ha='center', va='center', fontweight='bold', fontsize=12, rotation=90)\n",
    "\n",
    "\n",
    "# Define colors for 12 clusters\n",
    "# cluster_colors = plt.cm.tab20(np.reshape(np.linspace(0, 1, len(custom_cmap.colors)),(int(len(custom_cmap.colors)/2),2)).flatten(order='F'))\n",
    "cluster_colors = custom_cmap(np.linspace(0, 1, len(custom_cmap.colors)))\n",
    "# cluster_colors = plt.cm.tab20(np.linspace(0, 1, len(custom_cmap.colors)))\n",
    "\n",
    "\n",
    "# Create a new axes for the cluster labels\n",
    "ax_labels = fig.add_axes([\n",
    "    ax.get_position().x1 + 0.05,\n",
    "    ax.get_position().y0,\n",
    "    0.5,\n",
    "    ax.get_position().height\n",
    "])\n",
    "\n",
    "ax_labels.set_frame_on(False)\n",
    "ax_labels.set_xticks([])\n",
    "ax_labels.set_yticks([])\n",
    "\n",
    "# Get the actual y-coordinates from the heatmap\n",
    "heatmap_height = ax.get_position().height\n",
    "y_scale = heatmap_height / len(plot_genes)\n",
    "y_offset = ax.get_position().y0\n",
    "\n",
    "# Create correct gene position mapping using scaled coordinates\n",
    "gene_positions = {}\n",
    "for idx, gene_idx in enumerate(reordered_genes):\n",
    "    gene = plot_genes[gene_idx]\n",
    "    # Scale the position to match the heatmap coordinates. Use the small deviation after -idx-1 to adjust the centre)\n",
    "    scaled_pos = y_offset + (len(plot_genes) - idx - 1 + 0.5) * y_scale\n",
    "    gene_positions[gene] = scaled_pos\n",
    "\n",
    "# # Print some debugging info\n",
    "# print(\"Y-scale:\", y_scale)\n",
    "# print(\"Y-offset:\", y_offset)\n",
    "# print(\"Sample of gene positions:\", dict(list(gene_positions.items())[:5]))\n",
    "\n",
    "def curved_path(start_pos, end_pos):\n",
    "    return mpath.Path([\n",
    "        (0, start_pos),\n",
    "        (0.5, start_pos),\n",
    "        (1.0, end_pos),\n",
    "        (1.5, end_pos)\n",
    "    ], [\n",
    "        mpath.Path.MOVETO,\n",
    "        mpath.Path.CURVE4,\n",
    "        mpath.Path.CURVE4,\n",
    "        mpath.Path.CURVE4\n",
    "    ])\n",
    "\n",
    "# Calculate mean positions for each cluster\n",
    "cluster_mean_positions = {}\n",
    "for cluster_idx, genes in cluster_gene_dict.items():\n",
    "    valid_positions = [gene_positions[gene] for gene in genes if gene in gene_positions]\n",
    "    if valid_positions:\n",
    "        cluster_mean_positions[cluster_idx] = np.mean(valid_positions)\n",
    "\n",
    "# Sort clusters by their mean position\n",
    "sorted_clusters = sorted(cluster_mean_positions.items(), key=lambda x: x[1])\n",
    "\n",
    "# Calculate evenly spaced positions within the same coordinate system\n",
    "y_min = ax.get_position().y0\n",
    "y_max = ax.get_position().y1\n",
    "total_height = y_max - y_min\n",
    "n_clusters = len(sorted_clusters)\n",
    "spacing = total_height / (n_clusters + 1)\n",
    "\n",
    "evenly_spaced_positions = {\n",
    "    cluster_idx: y_min + spacing * (i + 1)\n",
    "    for i, (cluster_idx, _) in enumerate(sorted_clusters)\n",
    "}\n",
    "\n",
    "# Draw connections for each cluster\n",
    "for cluster_idx, mean_pos in sorted_clusters:\n",
    "    genes = cluster_gene_dict[cluster_idx]\n",
    "    cluster_genes = [gene for gene in genes if gene in gene_positions]\n",
    "    cluster_GO_terms = list(cluster_GO_dict[cluster_idx])\n",
    "    go_terms_formatted = '\\n'.join(cluster_GO_terms)\n",
    "    p_values = cluster_p_dict[cluster_idx]\n",
    "    cluster_min_p = np.min(cluster_p_dict[cluster_idx])\n",
    "\n",
    "    # print(cluster_idx)\n",
    "    \n",
    "    if not cluster_genes:\n",
    "        continue\n",
    "    \n",
    "    # Use evenly spaced position for label\n",
    "    label_pos = evenly_spaced_positions[cluster_idx]\n",
    "\n",
    "    box_dist = 1.6\n",
    "    # Draw curved lines to each gene\n",
    "    for gene in cluster_genes:\n",
    "        gene_pos = gene_positions[gene]\n",
    "        path = curved_path(gene_pos, label_pos)\n",
    "        patch = patches.PathPatch(\n",
    "            path,\n",
    "            facecolor='none',\n",
    "            edgecolor=cluster_colors[np.where(np.array(sorted_clusters, dtype=int)[:,0]==cluster_idx)],\n",
    "            alpha=0.8,\n",
    "            linewidth=1.75\n",
    "        )\n",
    "        ax_labels.add_patch(patch)\n",
    "    \n",
    "    # Add cluster label\n",
    "    bbox_props = dict(\n",
    "        boxstyle=\"round,pad=0.3\",\n",
    "        fc=\"white\",\n",
    "        ec=cluster_colors[np.where(np.array(sorted_clusters, dtype=int)[:,0]==cluster_idx)],\n",
    "        alpha=1\n",
    "    )\n",
    "\n",
    "    \n",
    "    if (len(cluster_GO_terms)>1)&(len(cluster_GO_terms)<4):\n",
    "            ax_labels.text(\n",
    "                box_dist + 0.1 * cluster_idx, \n",
    "                label_pos,\n",
    "                f'{cluster_name_dict[cluster_idx]}\\n'\n",
    "                f'{len(cluster_genes)} genes\\n'\n",
    "                f'min. p={cluster_min_p:.3g}\\n'\n",
    "                f'{go_terms_formatted}',  # Now each GO term will be on its own line\n",
    "                color='black',\n",
    "                va='center',\n",
    "                fontsize=10,\n",
    "                bbox=bbox_props,\n",
    "                rotation=90\n",
    "            )\n",
    "\n",
    "    if len(cluster_GO_terms)==1:\n",
    "        ax_labels.text(\n",
    "            box_dist + 0.1 * cluster_idx, \n",
    "            label_pos,\n",
    "            f'{cluster_name_dict[cluster_idx]}\\n'\n",
    "            f'{len(cluster_genes)} genes\\n'\n",
    "            f'p={cluster_min_p:.3g}\\n'\n",
    "            f'{go_terms_formatted}',  # Now each GO term will be on its own line\n",
    "            color='black',\n",
    "            va='center',\n",
    "            fontsize=10,\n",
    "            bbox=bbox_props,\n",
    "            rotation=90\n",
    "        )\n",
    "        \n",
    "\n",
    "    if (len(cluster_GO_terms)>3):\n",
    "\n",
    "        ax_labels.text(\n",
    "                box_dist + 0.1 * cluster_idx, \n",
    "                label_pos,\n",
    "                f'{cluster_name_dict[cluster_idx]}\\n'\n",
    "                f'{len(cluster_genes)} genes\\n'\n",
    "                f'min. p={cluster_min_p:.3g}\\n'\n",
    "                f'{len(cluster_GO_terms)} Go terms',  # Now each GO term will be on its own line\n",
    "                color='black',\n",
    "                va='center',\n",
    "                fontsize=10,\n",
    "                bbox=bbox_props,\n",
    "                rotation=90\n",
    "            )\n",
    "# Set proper limits\n",
    "ax_labels.set_xlim(-0.1, 2.7)\n",
    "ax_labels.set_ylim(y_min, y_max)\n",
    "\n",
    "plt.savefig('/Users/oipulk/Documents/scRNASeq/data/Eleftheria_Maranou_Mar2024/analysis/figures/Fig5/interactionDEA_heatmap_with_GO_noZ_score.pdf', dpi=600,bbox_inches = \"tight\")\n",
    "\n",
    "plt.show()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c498387c-be8f-4156-a206-2bd30ffaaf21",
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "\n",
    "# Create a separate figure for the colorbar\n",
    "fig_cbar, ax_cbar = plt.subplots(figsize=(4, 2))\n",
    "\n",
    "# Get the normalization from the main heatmap\n",
    "norm = g.ax_heatmap.collections[0].norm\n",
    "\n",
    "# Create a colorbar\n",
    "sm = plt.cm.ScalarMappable(cmap='RdYlBu_r', norm=norm)\n",
    "sm.set_array([])\n",
    "\n",
    "# Add colorbar to the new figure\n",
    "cbar = plt.colorbar(sm, ax=ax_cbar, orientation='horizontal', \n",
    "                    aspect=20,  # Decrease this number to make the colorbar thicker\n",
    "                    shrink=1.5)  # Adjust the length of the colorbar\n",
    "ax_cbar.remove()  # Remove the axes, leaving only the colorbar\n",
    "\n",
    "# cbar.set_ticks([-1.5, 0,  1.5])\n",
    "# cbar.set_ticklabels(['%.1f' % i for i in [-1.5,  0, 1.5]])\n",
    "\n",
    "# Adjust the figure layout\n",
    "plt.tight_layout()\n",
    "\n",
    "plt.savefig('/Users/oipulk/Documents/scRNASeq/data/Eleftheria_Maranou_Mar2024/analysis/figures/Fig5/interactionDEA_heatmap_with_GO_cbar_horizontal_noZ.pdf', dpi=600,bbox_inches = \"tight\")\n",
    "\n",
    "\n",
    "# Show both figures\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2dd5b574-7f32-4205-9f50-c0c6ee0aed08",
   "metadata": {},
   "outputs": [],
   "source": [
    "# antigen_terms = np.array(list([s for s in list(reactome_results['description']) if \"Antigen\" in s]))\n",
    "# data_df[data_df['np.asarray(antigen_terms[0:5],dtype=str)]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "11d902a3-37ef-424a-aaa7-77aa9df7f245",
   "metadata": {},
   "outputs": [],
   "source": [
    "df[(df['class']=='cellular_component')]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "74b66542-6498-457a-a853-19f5a155f4b5",
   "metadata": {},
   "outputs": [],
   "source": [
    "immune_relevance_filter=True\n",
    "\n",
    "if upregulated_degs_only==True:\n",
    "\n",
    "    data_df = df[df['up_down']=='up'].copy()\n",
    "\n",
    "if downregulated_degs_only==True:\n",
    "\n",
    "    data_df = df[df['up_down']=='down'].copy()\n",
    "\n",
    "if (upregulated_degs_only==False)&(downregulated_degs_only==False):\n",
    "    \n",
    "    data_df = df.copy()\n",
    "\n",
    "if immune_relevance_filter:\n",
    "    data_df = data_df[data_df['immune_relevance']>0]\n",
    "\n",
    "data_df = data_df[(data_df['class']=='cellular_component')]\n",
    "data_df = data_df.sort_values('score')[::-1]\n",
    "\n",
    "plt.figure(figsize = (4,6))\n",
    "\n",
    "if plot_fraction:\n",
    "    ax = sns.barplot(data = data_df, x = 'per', y = 'term', palette = mapper.to_rgba(data_df.p_corr.values))\n",
    "    ax.set_xlabel('GO term gene fraction', fontsize=12)\n",
    "    \n",
    "else:\n",
    "    ax = sns.barplot(data = data_df, x = 'n_genes', y = 'term', palette = mapper.to_rgba(data_df.p_corr.values))\n",
    "    ax.set_xlabel('GO term genes', fontsize=12)\n",
    "\n",
    "\n",
    "ax.set_yticklabels([textwrap.fill(e, 22) for e in data_df['term']],fontsize=10)\n",
    "ax.set_ylabel('Cellular component', fontsize=12)\n",
    "    \n",
    "\n",
    "### Interaction DEA\n",
    "if interaction:\n",
    "\n",
    "    if upregulated_degs_only:\n",
    "    \n",
    "        ax.set_title('GO: Spleen aggregated DC, interaction DEA', fontsize=12)\n",
    "\n",
    "    else:\n",
    "\n",
    "        ax.set_title('GO: Spleen aggregated DC, interaction DEA', fontsize=12)\n",
    "\n",
    "        \n",
    "### KO vs WT in pathogenic mice\n",
    "\n",
    "else:\n",
    "\n",
    "    if upregulated_degs_only:\n",
    "\n",
    "        ax.set_title('GO: Spleen aggregated DC, Pathogenic KO vs WT', fontsize=12)\n",
    "\n",
    "    else:\n",
    "        \n",
    "        ax.set_title('GO: Spleen aggregated DC, Pathogenic KO vs WT', fontsize=12)\n",
    "\n",
    "\n",
    "#### SAVE FIG #####\n",
    "\n",
    "### Interaction DEA\n",
    "if interaction:\n",
    "\n",
    "    if upregulated_degs_only:\n",
    "\n",
    "        if plot_fraction:\n",
    "\n",
    "            plt.savefig('/Users/oipulk/Documents/scRNASeq/data/Eleftheria_Maranou_Mar2024/analysis/figures/GO/spleen/interactionDEA/aggregatedDC/spleen_GO_CC_aggreagatedDC_interaction_patKO_upregulated_degs_per.pdf', dpi=600,bbox_inches = \"tight\")\n",
    "\n",
    "        else:\n",
    "           \n",
    "            plt.savefig('/Users/oipulk/Documents/scRNASeq/data/Eleftheria_Maranou_Mar2024/analysis/figures/GO/spleen/interactionDEA/aggregatedDC/spleen_GO_CC_aggreagatedDC_interaction_patKO_upregulated_degs_ngenes.pdf', dpi=600,bbox_inches = \"tight\")\n",
    "\n",
    "    else:\n",
    "\n",
    "        if plot_fraction:\n",
    "\n",
    "            plt.savefig('/Users/oipulk/Documents/scRNASeq/data/Eleftheria_Maranou_Mar2024/analysis/figures/GO/spleen/interactionDEA/aggregatedDC/spleen_GO_CC_aggreagatedDC_interaction_patKO_all_degs_per.pdf', dpi=600,bbox_inches = \"tight\")\n",
    "\n",
    "        else:\n",
    "    \n",
    "            plt.savefig('/Users/oipulk/Documents/scRNASeq/data/Eleftheria_Maranou_Mar2024/analysis/figures/GO/spleen/interactionDEA/aggregatedDC/spleen_GO_CC_aggreagatedDC_interaction_patKO_all_deg_ngeness.pdf', dpi=600,bbox_inches = \"tight\")\n",
    "\n",
    "\n",
    "\n",
    "### KO vs WT in pathogenic mice\n",
    "else:\n",
    "\n",
    "    if upregulated_degs_only:\n",
    "\n",
    "        if plot_fraction:\n",
    "\n",
    "            plt.savefig('/Users/oipulk/Documents/scRNASeq/data/Eleftheria_Maranou_Mar2024/analysis/figures/GO/spleen/KO_vs_WT_in_pathogenic/aggregatedDC/spleen_GO_CC_aggreagatedDC_KO_vs_WT_upregulated_degs_per.pdf', dpi=600,bbox_inches = \"tight\")\n",
    "\n",
    "        else: \n",
    "\n",
    "            plt.savefig('/Users/oipulk/Documents/scRNASeq/data/Eleftheria_Maranou_Mar2024/analysis/figures/GO/spleen/KO_vs_WT_in_pathogenic/aggregatedDC/spleen_GO_CC_aggreagatedDC_KO_vs_WT_upregulated_degs_ngenes.pdf', dpi=600,bbox_inches = \"tight\")\n",
    "    \n",
    "    else:\n",
    "\n",
    "        if plot_fraction:\n",
    "\n",
    "            plt.savefig('/Users/oipulk/Documents/scRNASeq/data/Eleftheria_Maranou_Mar2024/analysis/figures/GO/spleen/KO_vs_WT_in_pathogenic/aggregatedDC/spleen_GO_CC_aggreagatedDC_KO_vs_WT_all_degs_per.pdf', dpi=600,bbox_inches = \"tight\")\n",
    "\n",
    "        else:\n",
    "            \n",
    "            plt.savefig('/Users/oipulk/Documents/scRNASeq/data/Eleftheria_Maranou_Mar2024/analysis/figures/GO/spleen/KO_vs_WT_in_pathogenic/aggregatedDC/spleen_GO_CC_aggreagatedDC_KO_vs_WT_all_degs_ngenes.pdf', dpi=600,bbox_inches = \"tight\")\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1f0a550a-81b9-4965-bae3-f72117b02ec3",
   "metadata": {},
   "outputs": [],
   "source": [
    "data_df.iloc[0:10,:][['term','study_genes']]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d3236492-5a28-4e19-a33a-3d2a8b1e3060",
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "plt.figure(figsize = (4,8))\n",
    "if upregulated_degs_only==True:\n",
    "\n",
    "    data_df = df[df['up_down']=='up'].copy()\n",
    "\n",
    "if downregulated_degs_only==True:\n",
    "\n",
    "    data_df = df[df['up_down']=='down'].copy()\n",
    "\n",
    "if (upregulated_degs_only==False)&(downregulated_degs_only==False):\n",
    "    \n",
    "    data_df = df.copy()\n",
    "\n",
    "if immune_filter:\n",
    "    data_df = data_df[data_df['immune_relevance']>0]\n",
    "\n",
    "\n",
    "data_df = data_df[(data_df['class']=='molecular_function')]\n",
    "data_df = data_df.sort_values('score')[::-1]\n",
    "\n",
    "\n",
    "if plot_fraction:\n",
    "\n",
    "    ax = sns.barplot(data = data_df, x = 'per', y = 'term', palette = mapper.to_rgba(data_df.p_corr.values))\n",
    "    ax.set_xlabel('GO term gene fraction', fontsize=12)\n",
    "\n",
    "else:\n",
    "    \n",
    "    ax = sns.barplot(data = data_df, x = 'n_genes', y = 'term', palette = mapper.to_rgba(data_df.p_corr.values))\n",
    "    ax.set_xlabel('GO term genes', fontsize=12)\n",
    "\n",
    "ax.set_yticklabels([textwrap.fill(e, 32) for e in data_df['term']], fontsize=10)\n",
    "ax.set_ylabel('Molecular function', fontsize=12)\n",
    "\n",
    "\n",
    "### Interaction DEA\n",
    "if interaction:\n",
    "\n",
    "    if upregulated_degs_only:\n",
    "    \n",
    "        ax.set_title('GO: Spleen aggregated DC, interaction DEA', fontsize=12)\n",
    "\n",
    "    else:\n",
    "\n",
    "        ax.set_title('GO: Spleen aggregated DC, interaction DEA', fontsize=12)\n",
    "\n",
    "        \n",
    "### KO vs WT in pathogenic mice\n",
    "\n",
    "else:\n",
    "\n",
    "    if upregulated_degs_only:\n",
    "\n",
    "        ax.set_title('GO: Spleen aggregated DC, Pathogenic KO vs WT', fontsize=12)\n",
    "\n",
    "    else:\n",
    "        \n",
    "        ax.set_title('GO: Spleen aggregated DC, Pathogenic KO vs WT', fontsize=12)\n",
    "\n",
    "#### SAVE FIG #####\n",
    "\n",
    "### Interaction DEA\n",
    "\n",
    "if interaction:\n",
    "\n",
    "    if upregulated_degs_only:\n",
    "\n",
    "        if plot_fraction:\n",
    "\n",
    "            plt.savefig('/Users/oipulk/Documents/scRNASeq/data/Eleftheria_Maranou_Mar2024/analysis/figures/GO/spleen/interactionDEA/aggregatedDC/spleen_GO_MF_aggreagatedDC_interaction_patKO_upregulated_degs_per.pdf', dpi=600,bbox_inches = \"tight\")\n",
    "        \n",
    "        else:\n",
    "\n",
    "            plt.savefig('/Users/oipulk/Documents/scRNASeq/data/Eleftheria_Maranou_Mar2024/analysis/figures/GO/spleen/interactionDEA/aggregatedDC/spleen_GO_MF_aggreagatedDC_interaction_patKO_upregulated_degs_ngenes.pdf', dpi=600,bbox_inches = \"tight\")\n",
    "\n",
    "    else:\n",
    "\n",
    "        if plot_fraction:\n",
    "\n",
    "            plt.savefig('/Users/oipulk/Documents/scRNASeq/data/Eleftheria_Maranou_Mar2024/analysis/figures/GO/spleen/interactionDEA/aggregatedDC/spleen_GO_MF_aggreagatedDC_interaction_patKO_all_degs_per.pdf', dpi=600,bbox_inches = \"tight\")\n",
    "        \n",
    "        else:\n",
    "            \n",
    "            plt.savefig('/Users/oipulk/Documents/scRNASeq/data/Eleftheria_Maranou_Mar2024/analysis/figures/GO/spleen/interactionDEA/aggregatedDC/spleen_GO_MF_aggreagatedDC_interaction_patKO_all_degs_ngenes.pdf', dpi=600,bbox_inches = \"tight\")\n",
    "\n",
    "\n",
    "### KO vs WT in pathogenic mice\n",
    "else:\n",
    "\n",
    "    if upregulated_degs_only:\n",
    "\n",
    "        if plot_fraction:\n",
    "\n",
    "            plt.savefig('/Users/oipulk/Documents/scRNASeq/data/Eleftheria_Maranou_Mar2024/analysis/figures/GO/spleen/KO_vs_WT_in_pathogenic/aggregatedDC/spleen_GO_MF_aggreagatedDC_KO_vs_WT_upregulated_degs_per.pdf', dpi=600,bbox_inches = \"tight\")\n",
    "\n",
    "        else:\n",
    "            \n",
    "            plt.savefig('/Users/oipulk/Documents/scRNASeq/data/Eleftheria_Maranou_Mar2024/analysis/figures/GO/spleen/KO_vs_WT_in_pathogenic/aggregatedDC/spleen_GO_MF_aggreagatedDC_KO_vs_WT_upregulated_degs_ngenes.pdf', dpi=600,bbox_inches = \"tight\")\n",
    "\n",
    "    else:\n",
    "\n",
    "        if plot_fraction:\n",
    "\n",
    "            plt.savefig('/Users/oipulk/Documents/scRNASeq/data/Eleftheria_Maranou_Mar2024/analysis/figures/GO/spleen/KO_vs_WT_in_pathogenic/aggregatedDC/spleen_GO_MF_aggreagatedDC_KO_vs_WT_all_degs_per.pdf', dpi=600,bbox_inches = \"tight\")\n",
    "        \n",
    "        else:    \n",
    "        \n",
    "            plt.savefig('/Users/oipulk/Documents/scRNASeq/data/Eleftheria_Maranou_Mar2024/analysis/figures/GO/spleen/KO_vs_WT_in_pathogenic/aggregatedDC/spleen_GO_MF_aggreagatedDC_KO_vs_WT_all_degs_ngenes.pdf', dpi=600,bbox_inches = \"tight\")\n",
    "\n",
    "\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "85eb1722-f32f-4945-af5f-379d8a3eab0b",
   "metadata": {},
   "outputs": [],
   "source": [
    "df[df['term']=='transcription factor AP-1 complex']['study_genes']"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f42109c8-1f38-4cf2-bba1-b8614c965fee",
   "metadata": {},
   "source": [
    "### Enrichr"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ee9d7568-0a2f-4dd9-a52d-125dc0a4e3b8",
   "metadata": {},
   "outputs": [],
   "source": [
    "import json\n",
    "import requests\n",
    "\n",
    "def enrichr_analysis(genes, database='KEGG_2021_Mouse'):\n",
    "    # Get the enrichr link\n",
    "    ENRICHR_URL = 'https://maayanlab.cloud/Enrichr/addList'\n",
    "    genes_str = '\\n'.join(genes)\n",
    "    payload = {\n",
    "        'list': (None, genes_str),\n",
    "    }\n",
    "    \n",
    "    response = requests.post(ENRICHR_URL, files=payload)\n",
    "    if not response.ok:\n",
    "        raise Exception('Error analyzing gene list')\n",
    "    \n",
    "    data = json.loads(response.text)\n",
    "    user_list_id = data['userListId']\n",
    "    \n",
    "    # Get the enrichment results\n",
    "    ENRICHR_URL = f'https://maayanlab.cloud/Enrichr/enrich?userListId={user_list_id}&backgroundType={database}'\n",
    "    response = requests.get(ENRICHR_URL)\n",
    "    if not response.ok:\n",
    "        raise Exception('Error getting enrichment results')\n",
    "    \n",
    "    data = json.loads(response.text)\n",
    "    \n",
    "    # Create a pandas DataFrame\n",
    "    results = pd.DataFrame(data[database], \n",
    "                         columns=[\"rank\", \"description\", \"p_value\", \"z_score\", \"combined_score\", \"genes\", \"adjusted_p_value\",'q1','q2'])\n",
    "\n",
    "    results = results.drop(['q1','q2'], axis=1)\n",
    "    \n",
    "    return results.sort_values('combined_score')\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d63a6cea-6b5c-4edb-a0fc-0b735356f5b8",
   "metadata": {},
   "outputs": [],
   "source": [
    "for k in dc_markers_up.index:\n",
    " print(dc_markers_up['Gene'][k].upper())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "687ef0b4-3904-43d4-a550-4944fe934f5b",
   "metadata": {},
   "outputs": [],
   "source": [
    "if interaction:\n",
    "\n",
    "    ## Aggregated DC with interaction DEA\n",
    "    dc_markers = pd.read_csv('/Users/oipulk/Documents/scRNASeq/data/Eleftheria_Maranou_Mar2024/analysis/figures/pseudobulk/spleen/deseq2_interaction_results/Spleen_kopat_vs_rest_interaction_degs_aggregatedDC.csv')\n",
    "    dc_markers = dc_markers.rename(columns={\"gene\": \"Gene\"})\n",
    "\n",
    "##################################################\n",
    "else:\n",
    "    \n",
    "    ## Aggregated DC:\n",
    "    dc_markers = pd.read_csv('/Users/oipulk/Documents/scRNASeq/data/Eleftheria_Maranou_Mar2024/analysis/figures/pseudobulk/spleen/deseq2_pathogenic_results/Spleen_pat_kowt_degs_aggregatedDC.csv')\n",
    "    dc_markers = dc_markers.rename(columns={\"Unnamed: 0\": \"Gene\"})\n",
    "\n",
    "dc_markers_up = dc_markers[dc_markers['log2FoldChange']>0.5]\n",
    "dc_markers_down = dc_markers[dc_markers['log2FoldChange']<-0.5]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6580f6de-19de-4bcf-a210-43e191708711",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Run the analysis\n",
    "genes = list(dc_markers_up.Gene.values)\n",
    "# Analyze same gene list with different databases\n",
    "kegg_results = enrichr_analysis(genes, database='KEGG_2021_Mouse')\n",
    "wiki_results = enrichr_analysis(genes, database='WikiPathways_2024_Mouse')\n",
    "reactome_results = enrichr_analysis(genes, database='Reactome_Pathways_2024')\n",
    "go_bp_results = enrichr_analysis(genes, database='GO_Biological_Process_2023')\n",
    "go_cc_results = enrichr_analysis(genes, database='GO_Cellular_Component_2023')\n",
    "go_mf_results = enrichr_analysis(genes, database='GO_Molecular_Function_2023')\n",
    "\n",
    "\n",
    "# Find overlapping biological themes\n",
    "kegg_pathways = list(kegg_results[kegg_results['adjusted_p_value'] < 0.05].sort_values('combined_score')[::-1]['description'])\n",
    "wiki_pathways = list(wiki_results[wiki_results['adjusted_p_value'] < 0.05].sort_values('combined_score')[::-1]['description'])\n",
    "reactome_pathways = list(reactome_results[reactome_results['adjusted_p_value'] < 0.05].sort_values('combined_score')[::-1]['description'])\n",
    "go_bp_terms = list(go_bp_results[go_bp_results['adjusted_p_value'] < 0.05].sort_values('combined_score')[::-1]['description'])\n",
    "go_cc_terms = list(go_cc_results[go_cc_results['adjusted_p_value'] < 0.05].sort_values('combined_score')[::-1]['description'])\n",
    "go_mf_terms = list(go_mf_results[go_mf_results['adjusted_p_value'] < 0.05].sort_values('combined_score')[::-1]['description'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "dfea7786-fff3-4272-b9f4-cb26f0d6db5e",
   "metadata": {},
   "outputs": [],
   "source": [
    "n_genes = list([])\n",
    "for row in np.arange(reactome_results.shape[0]):\n",
    "    \n",
    "    n_genes_row = len(reactome_results.iloc[row,:]['genes'])\n",
    "    n_genes = np.append(n_genes,n_genes_row)\n",
    "\n",
    "reactome_results['n_genes'] = n_genes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "69f66a6b-db85-4597-bf4b-4772e4671c41",
   "metadata": {},
   "outputs": [],
   "source": [
    "n_genes = list([])\n",
    "for row in np.arange(kegg_results.shape[0]):\n",
    "    \n",
    "    n_genes_row = len(kegg_results.iloc[row,:]['genes'])\n",
    "    n_genes = np.append(n_genes,n_genes_row)\n",
    "\n",
    "kegg_results['n_genes'] = n_genes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "df5dc502-d8a1-456c-abf7-61908bde2214",
   "metadata": {},
   "outputs": [],
   "source": [
    "n_genes = list([])\n",
    "for row in np.arange(go_bp_results.shape[0]):\n",
    "    \n",
    "    n_genes_row = len(go_bp_results.iloc[row,:]['genes'])\n",
    "    n_genes = np.append(n_genes,n_genes_row)\n",
    "\n",
    "go_bp_results['n_genes'] = n_genes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d77815d6-7a56-46c0-a454-b9f781c0ad67",
   "metadata": {},
   "outputs": [],
   "source": [
    "n_genes = list([])\n",
    "for row in np.arange(go_cc_results.shape[0]):\n",
    "    \n",
    "    n_genes_row = len(go_cc_results.iloc[row,:]['genes'])\n",
    "    n_genes = np.append(n_genes,n_genes_row)\n",
    "\n",
    "go_cc_results['n_genes'] = n_genes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c9346ecf-0b1d-4e34-9d09-e4cdd6a82237",
   "metadata": {},
   "outputs": [],
   "source": [
    "reactome_sigs = reactome_results[reactome_results['adjusted_p_value'] < 0.01].sort_values('adjusted_p_value')\n",
    "reactome_sigs.to_csv('reactome_sigs.csv')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0b9e666f-dda6-4105-b0e9-62da65bb99dc",
   "metadata": {},
   "outputs": [],
   "source": [
    "go_bp_sigs = go_bp_results[go_bp_results['adjusted_p_value'] < 0.01].sort_values('adjusted_p_value')[::-1]\n",
    "go_bp_sigs.to_csv('go_bp_sigs.csv')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1a50ec9c-e2d2-4d5f-b2f3-7c332110a0d8",
   "metadata": {},
   "outputs": [],
   "source": [
    "go_cc_sigs = go_cc_results[go_cc_results['adjusted_p_value'] < 0.05].sort_values('adjusted_p_value')[::-1]\n",
    "go_cc_sigs.to_csv('go_cc_sigs.csv')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9db90e84-aade-43aa-890f-cbeb6c2be4c3",
   "metadata": {},
   "outputs": [],
   "source": [
    "kegg_sigs = kegg_results[kegg_results['adjusted_p_value'] < 0.05].sort_values('adjusted_p_value')[::-1]\n",
    "kegg_sigs.to_csv('kegg_sigs.csv')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "bdf837ca-7ff1-4ffc-b279-30647a00676e",
   "metadata": {},
   "outputs": [],
   "source": [
    "wiki_sigs = wiki_results[wiki_results['adjusted_p_value'] < 0.05].sort_values('adjusted_p_value')[::-1]\n",
    "wiki_sigs.to_csv('wiki_sigs.csv')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "af6cf312-0127-42a5-a471-425b3d52de92",
   "metadata": {},
   "outputs": [],
   "source": [
    "plot_fraction = False"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5f87e828-adf8-4f6b-a424-103b882a1b9c",
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "import matplotlib as mpl\n",
    "import textwrap\n",
    "\n",
    "fig, ax = plt.subplots(figsize = (0.5, 2.75), dpi=600)\n",
    "\n",
    "cmap = mpl.cm.coolwarm_r\n",
    "norm = mpl.colors.Normalize(vmin = 0, vmax =0.01)\n",
    "\n",
    "mapper = cm.ScalarMappable(norm = norm, cmap = cmap)\n",
    "\n",
    "cbl = mpl.colorbar.ColorbarBase(ax, cmap = cmap, norm = norm, orientation = 'vertical')\n",
    "plt.savefig('colorbar.pdf', bbox_inches='tight')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a7484340-ae8f-45ef-847e-518a80c70382",
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "plt.figure(figsize = (4,16))\n",
    "\n",
    "data_df = reactome_sigs\n",
    "\n",
    "if plot_fraction:\n",
    "\n",
    "    ax = sns.barplot(data = data_df, x = 'n_genes', y = 'description', palette = mapper.to_rgba(data_df.adjusted_p_value.values))\n",
    "    ax.set_xlabel('GO term gene fraction', fontsize=12)\n",
    "\n",
    "else:\n",
    "    \n",
    "    ax = sns.barplot(data = data_df, x = 'n_genes', y = 'description', palette = mapper.to_rgba(data_df.adjusted_p_value.values))\n",
    "    ax.set_xlabel('N genes', fontsize=12)\n",
    "\n",
    "ax.set_yticklabels([textwrap.fill(e, 32) for e in data_df['description']], fontsize=8)\n",
    "ax.set_ylabel('Reactome Pathways', fontsize=12)\n",
    "\n",
    "\n",
    "### Interaction DEA\n",
    "if interaction:\n",
    "\n",
    "    if upregulated_degs_only:\n",
    "    \n",
    "        ax.set_title('Reactome: Spleen aggregated DC, interaction DEA', fontsize=12)\n",
    "\n",
    "    else:\n",
    "\n",
    "        ax.set_title('Reactome: Spleen aggregated DC, interaction DEA', fontsize=12)\n",
    "\n",
    "        \n",
    "### KO vs WT in pathogenic mice\n",
    "\n",
    "else:\n",
    "\n",
    "    if upregulated_degs_only:\n",
    "\n",
    "        ax.set_title('Reactome: Spleen aggregated DC, Pathogenic KO vs WT', fontsize=12)\n",
    "\n",
    "    else:\n",
    "        \n",
    "        ax.set_title('Reactome: Spleen aggregated DC, Pathogenic KO vs WT', fontsize=12)\n",
    "\n",
    "#### SAVE FIG #####\n",
    "\n",
    "### Interaction DEA\n",
    "\n",
    "if interaction:\n",
    "\n",
    "    if upregulated_degs_only:\n",
    "\n",
    "        if plot_fraction:\n",
    "\n",
    "            plt.savefig('/Users/oipulk/Documents/scRNASeq/data/Eleftheria_Maranou_Mar2024/analysis/figures/GO/spleen/interactionDEA/aggregatedDC/spleen_GO_MF_aggreagatedDC_interaction_patKO_upregulated_degs_per.pdf', dpi=600,bbox_inches = \"tight\")\n",
    "        \n",
    "        else:\n",
    "\n",
    "            plt.savefig('/Users/oipulk/Documents/scRNASeq/data/Eleftheria_Maranou_Mar2024/analysis/figures/GO/spleen/interactionDEA/aggregatedDC/spleen_GO_MF_aggreagatedDC_interaction_patKO_upregulated_degs_ngenes.pdf', dpi=600,bbox_inches = \"tight\")\n",
    "\n",
    "    else:\n",
    "\n",
    "        if plot_fraction:\n",
    "\n",
    "            plt.savefig('/Users/oipulk/Documents/scRNASeq/data/Eleftheria_Maranou_Mar2024/analysis/figures/GO/spleen/interactionDEA/aggregatedDC/spleen_GO_MF_aggreagatedDC_interaction_patKO_all_degs_per.pdf', dpi=600,bbox_inches = \"tight\")\n",
    "        \n",
    "        else:\n",
    "            \n",
    "            plt.savefig('/Users/oipulk/Documents/scRNASeq/data/Eleftheria_Maranou_Mar2024/analysis/figures/GO/spleen/interactionDEA/aggregatedDC/spleen_GO_MF_aggreagatedDC_interaction_patKO_all_degs_ngenes.pdf', dpi=600,bbox_inches = \"tight\")\n",
    "\n",
    "\n",
    "### KO vs WT in pathogenic mice\n",
    "else:\n",
    "\n",
    "    if upregulated_degs_only:\n",
    "\n",
    "        if plot_fraction:\n",
    "\n",
    "            plt.savefig('/Users/oipulk/Documents/scRNASeq/data/Eleftheria_Maranou_Mar2024/analysis/figures/GO/spleen/KO_vs_WT_in_pathogenic/aggregatedDC/spleen_GO_MF_aggreagatedDC_KO_vs_WT_upregulated_degs_per.pdf', dpi=600,bbox_inches = \"tight\")\n",
    "\n",
    "        else:\n",
    "            \n",
    "            plt.savefig('/Users/oipulk/Documents/scRNASeq/data/Eleftheria_Maranou_Mar2024/analysis/figures/GO/spleen/KO_vs_WT_in_pathogenic/aggregatedDC/spleen_GO_MF_aggreagatedDC_KO_vs_WT_upregulated_degs_ngenes.pdf', dpi=600,bbox_inches = \"tight\")\n",
    "\n",
    "    else:\n",
    "\n",
    "        if plot_fraction:\n",
    "\n",
    "            plt.savefig('/Users/oipulk/Documents/scRNASeq/data/Eleftheria_Maranou_Mar2024/analysis/figures/GO/spleen/KO_vs_WT_in_pathogenic/aggregatedDC/spleen_GO_MF_aggreagatedDC_KO_vs_WT_all_degs_per.pdf', dpi=600,bbox_inches = \"tight\")\n",
    "        \n",
    "        else:    \n",
    "        \n",
    "            plt.savefig('/Users/oipulk/Documents/scRNASeq/data/Eleftheria_Maranou_Mar2024/analysis/figures/GO/spleen/KO_vs_WT_in_pathogenic/aggregatedDC/spleen_GO_MF_aggreagatedDC_KO_vs_WT_all_degs_ngenes.pdf', dpi=600,bbox_inches = \"tight\")\n",
    "\n",
    "\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "21de97f7-b667-4583-a8fe-4c6be502f6ba",
   "metadata": {},
   "outputs": [],
   "source": [
    "data_df.iloc[0:10][['description','genes']]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6c97c2f4-dbb8-42ea-b612-5065ab2ee705",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Create a similarity matrix between terms\n",
    "def create_similarity_matrix_enrichr(df):\n",
    "    n = len(df)\n",
    "    sim_matrix = np.zeros((n, n))\n",
    "    \n",
    "    for i, j in combinations(range(n), 2):\n",
    "        # Convert string representation of gene lists to actual sets\n",
    "        genes1 = set(df.iloc[i]['genes'])\n",
    "        genes2 = set(df.iloc[j]['genes'])\n",
    "        \n",
    "        sim = jaccard_similarity(genes1, genes2)\n",
    "        sim_matrix[i, j] = sim\n",
    "        sim_matrix[j, i] = sim\n",
    "    \n",
    "    return sim_matrix"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a73da77c-173f-4d60-baa0-2eeea9e1c32f",
   "metadata": {},
   "outputs": [],
   "source": [
    "sim_matrix = create_similarity_matrix_enrichr(data_df)\n",
    "\n",
    "# Create heatmap\n",
    "plt.figure(figsize=(12, 10))\n",
    "sns.heatmap(sim_matrix, \n",
    "            xticklabels=data_df['description'].str[:30],\n",
    "            yticklabels=data_df['description'].str[:30],\n",
    "            cmap='YlOrRd')\n",
    "plt.xticks(rotation=45, ha='right')\n",
    "plt.yticks(rotation=0)\n",
    "plt.title('Gene Set Overlap (Jaccard Similarity)')\n",
    "plt.tight_layout()\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c59d9f0e-0de7-4bd6-9116-65d5dd9f59c4",
   "metadata": {},
   "outputs": [],
   "source": [
    "data_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0650a13a-febb-465c-913a-44f40190fc94",
   "metadata": {},
   "outputs": [],
   "source": [
    "from scipy.cluster.hierarchy import linkage, fcluster\n",
    "\n",
    "\n",
    "# Create a network graph based on similarity threshold\n",
    "G = nx.Graph()\n",
    "n = len(data_df) \n",
    "for i in range(n):\n",
    "    for j in range(i+1, n):\n",
    "        if sim_matrix[i,j] >= 0.75:\n",
    "            G.add_edge(i, j, weight=sim_matrix[i,j])\n",
    "\n",
    "# Find connected components (clusters)\n",
    "clusters = list(nx.connected_components(G))\n",
    "\n",
    "# Create visualization\n",
    "plt.figure(figsize=(15, 10))\n",
    "\n",
    "# Use spring layout for node positioning\n",
    "pos = nx.spring_layout(G, k=1, iterations=50)\n",
    "\n",
    "# Draw the network\n",
    "nx.draw_networkx_nodes(G, pos, node_size=500, node_color='lightblue')\n",
    "nx.draw_networkx_edges(G, pos, width=2, alpha=0.5)\n",
    "\n",
    "# # Add labels\n",
    "# labels = {i: f\"{data_df['term'].iloc[i]}...\\n(-logp={data_df['-logp_corr'].iloc[i]:.2f})\" for i in range(len(data_df))}\n",
    "# nx.draw_networkx_labels(G, pos, labels, font_size=8)\n",
    "\n",
    "plt.title(\"GO Term Clusters (Similarity Threshold ≥ 0.75)\")\n",
    "plt.axis('off')\n",
    "plt.tight_layout()\n",
    "plt.show()\n",
    "\n",
    "# Print clusters with details\n",
    "print(\"\\nDetailed Clusters (Similarity ≥ 0.75):\")\n",
    "print(\"=\" * 80)\n",
    "for i, cluster in enumerate(clusters):\n",
    "    print(f\"\\nCluster {i+1}:\")\n",
    "    print(\"-\" * 40)\n",
    "    cluster_list = list(cluster)\n",
    "    # Sort cluster members by -logp_corr\n",
    "    cluster_list.sort(key=lambda x: data_df['adjusted_p_value'].iloc[x], reverse=False)\n",
    "    \n",
    "    for idx in cluster_list:\n",
    "        print(f\"GO: {data_df['rank'].iloc[idx]}\")\n",
    "        print(f\"Term: {data_df['description'].iloc[idx]}\")\n",
    "        print(f\"p_corr: {data_df['adjusted_p_value'].iloc[idx]:.3f}\")\n",
    "        print(f\"Genes: {data_df['genes'].iloc[idx]}\")\n",
    "        print(\"-\" * 40)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "bbd16d76-fce0-4d8b-8cc1-1b9aa2e04466",
   "metadata": {},
   "outputs": [],
   "source": [
    "plt.figure(figsize = (4,16))\n",
    "\n",
    "data_df = go_bp_results[go_bp_results['adjusted_p_value'] < 0.01].sort_values('adjusted_p_value')\n",
    "\n",
    "if plot_fraction:\n",
    "\n",
    "    ax = sns.barplot(data = data_df, x = 'n_genes', y = 'description', palette = mapper.to_rgba(data_df.adjusted_p_value.values))\n",
    "    ax.set_xlabel('GO term gene fraction', fontsize=12)\n",
    "\n",
    "else:\n",
    "    \n",
    "    ax = sns.barplot(data = data_df, x = 'n_genes', y = 'description', palette = mapper.to_rgba(data_df.adjusted_p_value.values))\n",
    "    ax.set_xlabel('GO term genes', fontsize=12)\n",
    "\n",
    "ax.set_yticklabels([textwrap.fill(e, 32) for e in data_df['description']], fontsize=8)\n",
    "ax.set_ylabel('GO BP', fontsize=12)\n",
    "\n",
    "\n",
    "### Interaction DEA\n",
    "if interaction:\n",
    "\n",
    "    if upregulated_degs_only:\n",
    "    \n",
    "        ax.set_title('GO: Spleen aggregated DC, interaction DEA', fontsize=12)\n",
    "\n",
    "    else:\n",
    "\n",
    "        ax.set_title('GO: Spleen aggregated DC, interaction DEA', fontsize=12)\n",
    "\n",
    "        \n",
    "### KO vs WT in pathogenic mice\n",
    "\n",
    "else:\n",
    "\n",
    "    if upregulated_degs_only:\n",
    "\n",
    "        ax.set_title('GO: Spleen aggregated DC, Pathogenic KO vs WT', fontsize=12)\n",
    "\n",
    "    else:\n",
    "        \n",
    "        ax.set_title('GO: Spleen aggregated DC, Pathogenic KO vs WT', fontsize=12)\n",
    "\n",
    "#### SAVE FIG #####\n",
    "\n",
    "### Interaction DEA\n",
    "\n",
    "if interaction:\n",
    "\n",
    "    if upregulated_degs_only:\n",
    "\n",
    "        if plot_fraction:\n",
    "\n",
    "            plt.savefig('/Users/oipulk/Documents/scRNASeq/data/Eleftheria_Maranou_Mar2024/analysis/figures/GO/spleen/interactionDEA/aggregatedDC/spleen_GO_MF_aggreagatedDC_interaction_patKO_upregulated_degs_per.pdf', dpi=600,bbox_inches = \"tight\")\n",
    "        \n",
    "        else:\n",
    "\n",
    "            plt.savefig('/Users/oipulk/Documents/scRNASeq/data/Eleftheria_Maranou_Mar2024/analysis/figures/GO/spleen/interactionDEA/aggregatedDC/spleen_GO_MF_aggreagatedDC_interaction_patKO_upregulated_degs_ngenes.pdf', dpi=600,bbox_inches = \"tight\")\n",
    "\n",
    "    else:\n",
    "\n",
    "        if plot_fraction:\n",
    "\n",
    "            plt.savefig('/Users/oipulk/Documents/scRNASeq/data/Eleftheria_Maranou_Mar2024/analysis/figures/GO/spleen/interactionDEA/aggregatedDC/spleen_GO_MF_aggreagatedDC_interaction_patKO_all_degs_per.pdf', dpi=600,bbox_inches = \"tight\")\n",
    "        \n",
    "        else:\n",
    "            \n",
    "            plt.savefig('/Users/oipulk/Documents/scRNASeq/data/Eleftheria_Maranou_Mar2024/analysis/figures/GO/spleen/interactionDEA/aggregatedDC/spleen_GO_MF_aggreagatedDC_interaction_patKO_all_degs_ngenes.pdf', dpi=600,bbox_inches = \"tight\")\n",
    "\n",
    "\n",
    "### KO vs WT in pathogenic mice\n",
    "else:\n",
    "\n",
    "    if upregulated_degs_only:\n",
    "\n",
    "        if plot_fraction:\n",
    "\n",
    "            plt.savefig('/Users/oipulk/Documents/scRNASeq/data/Eleftheria_Maranou_Mar2024/analysis/figures/GO/spleen/KO_vs_WT_in_pathogenic/aggregatedDC/spleen_GO_MF_aggreagatedDC_KO_vs_WT_upregulated_degs_per.pdf', dpi=600,bbox_inches = \"tight\")\n",
    "\n",
    "        else:\n",
    "            \n",
    "            plt.savefig('/Users/oipulk/Documents/scRNASeq/data/Eleftheria_Maranou_Mar2024/analysis/figures/GO/spleen/KO_vs_WT_in_pathogenic/aggregatedDC/spleen_GO_MF_aggreagatedDC_KO_vs_WT_upregulated_degs_ngenes.pdf', dpi=600,bbox_inches = \"tight\")\n",
    "\n",
    "    else:\n",
    "\n",
    "        if plot_fraction:\n",
    "\n",
    "            plt.savefig('/Users/oipulk/Documents/scRNASeq/data/Eleftheria_Maranou_Mar2024/analysis/figures/GO/spleen/KO_vs_WT_in_pathogenic/aggregatedDC/spleen_GO_MF_aggreagatedDC_KO_vs_WT_all_degs_per.pdf', dpi=600,bbox_inches = \"tight\")\n",
    "        \n",
    "        else:    \n",
    "        \n",
    "            plt.savefig('/Users/oipulk/Documents/scRNASeq/data/Eleftheria_Maranou_Mar2024/analysis/figures/GO/spleen/KO_vs_WT_in_pathogenic/aggregatedDC/spleen_GO_MF_aggreagatedDC_KO_vs_WT_all_degs_ngenes.pdf', dpi=600,bbox_inches = \"tight\")\n",
    "\n",
    "\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4046bf58-af56-4739-a88a-118c8a3f6570",
   "metadata": {},
   "outputs": [],
   "source": [
    "list(go_cc_results['description'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f2c80724-ce9d-4406-acb8-ba7774d3d738",
   "metadata": {},
   "outputs": [],
   "source": [
    "plt.figure(figsize = (4,16))\n",
    "\n",
    "data_df = go_cc_results[go_cc_results['adjusted_p_value'] < 0.05].sort_values('adjusted_p_value')\n",
    "\n",
    "if plot_fraction:\n",
    "\n",
    "    ax = sns.barplot(data = data_df, x = 'n_genes', y = 'description', palette = mapper.to_rgba(data_df.adjusted_p_value.values))\n",
    "    ax.set_xlabel('GO term gene fraction', fontsize=12)\n",
    "\n",
    "else:\n",
    "    \n",
    "    ax = sns.barplot(data = data_df, x = 'n_genes', y = 'description', palette = mapper.to_rgba(data_df.adjusted_p_value.values))\n",
    "    ax.set_xlabel('GO term genes', fontsize=12)\n",
    "\n",
    "ax.set_yticklabels([textwrap.fill(e, 32) for e in data_df['description']], fontsize=8)\n",
    "ax.set_ylabel('GO CC', fontsize=12)\n",
    "\n",
    "\n",
    "### Interaction DEA\n",
    "if interaction:\n",
    "\n",
    "    if upregulated_degs_only:\n",
    "    \n",
    "        ax.set_title('GO: Spleen aggregated DC, interaction DEA', fontsize=12)\n",
    "\n",
    "    else:\n",
    "\n",
    "        ax.set_title('GO: Spleen aggregated DC, interaction DEA', fontsize=12)\n",
    "\n",
    "        \n",
    "### KO vs WT in pathogenic mice\n",
    "\n",
    "else:\n",
    "\n",
    "    if upregulated_degs_only:\n",
    "\n",
    "        ax.set_title('GO: Spleen aggregated DC, Pathogenic KO vs WT', fontsize=12)\n",
    "\n",
    "    else:\n",
    "        \n",
    "        ax.set_title('GO: Spleen aggregated DC, Pathogenic KO vs WT', fontsize=12)\n",
    "\n",
    "#### SAVE FIG #####\n",
    "\n",
    "### Interaction DEA\n",
    "\n",
    "if interaction:\n",
    "\n",
    "    if upregulated_degs_only:\n",
    "\n",
    "        if plot_fraction:\n",
    "\n",
    "            plt.savefig('/Users/oipulk/Documents/scRNASeq/data/Eleftheria_Maranou_Mar2024/analysis/figures/GO/spleen/interactionDEA/aggregatedDC/spleen_GO_MF_aggreagatedDC_interaction_patKO_upregulated_degs_per.pdf', dpi=600,bbox_inches = \"tight\")\n",
    "        \n",
    "        else:\n",
    "\n",
    "            plt.savefig('/Users/oipulk/Documents/scRNASeq/data/Eleftheria_Maranou_Mar2024/analysis/figures/GO/spleen/interactionDEA/aggregatedDC/spleen_GO_MF_aggreagatedDC_interaction_patKO_upregulated_degs_ngenes.pdf', dpi=600,bbox_inches = \"tight\")\n",
    "\n",
    "    else:\n",
    "\n",
    "        if plot_fraction:\n",
    "\n",
    "            plt.savefig('/Users/oipulk/Documents/scRNASeq/data/Eleftheria_Maranou_Mar2024/analysis/figures/GO/spleen/interactionDEA/aggregatedDC/spleen_GO_MF_aggreagatedDC_interaction_patKO_all_degs_per.pdf', dpi=600,bbox_inches = \"tight\")\n",
    "        \n",
    "        else:\n",
    "            \n",
    "            plt.savefig('/Users/oipulk/Documents/scRNASeq/data/Eleftheria_Maranou_Mar2024/analysis/figures/GO/spleen/interactionDEA/aggregatedDC/spleen_GO_MF_aggreagatedDC_interaction_patKO_all_degs_ngenes.pdf', dpi=600,bbox_inches = \"tight\")\n",
    "\n",
    "\n",
    "### KO vs WT in pathogenic mice\n",
    "else:\n",
    "\n",
    "    if upregulated_degs_only:\n",
    "\n",
    "        if plot_fraction:\n",
    "\n",
    "            plt.savefig('/Users/oipulk/Documents/scRNASeq/data/Eleftheria_Maranou_Mar2024/analysis/figures/GO/spleen/KO_vs_WT_in_pathogenic/aggregatedDC/spleen_GO_MF_aggreagatedDC_KO_vs_WT_upregulated_degs_per.pdf', dpi=600,bbox_inches = \"tight\")\n",
    "\n",
    "        else:\n",
    "            \n",
    "            plt.savefig('/Users/oipulk/Documents/scRNASeq/data/Eleftheria_Maranou_Mar2024/analysis/figures/GO/spleen/KO_vs_WT_in_pathogenic/aggregatedDC/spleen_GO_MF_aggreagatedDC_KO_vs_WT_upregulated_degs_ngenes.pdf', dpi=600,bbox_inches = \"tight\")\n",
    "\n",
    "    else:\n",
    "\n",
    "        if plot_fraction:\n",
    "\n",
    "            plt.savefig('/Users/oipulk/Documents/scRNASeq/data/Eleftheria_Maranou_Mar2024/analysis/figures/GO/spleen/KO_vs_WT_in_pathogenic/aggregatedDC/spleen_GO_MF_aggreagatedDC_KO_vs_WT_all_degs_per.pdf', dpi=600,bbox_inches = \"tight\")\n",
    "        \n",
    "        else:    \n",
    "        \n",
    "            plt.savefig('/Users/oipulk/Documents/scRNASeq/data/Eleftheria_Maranou_Mar2024/analysis/figures/GO/spleen/KO_vs_WT_in_pathogenic/aggregatedDC/spleen_GO_MF_aggreagatedDC_KO_vs_WT_all_degs_ngenes.pdf', dpi=600,bbox_inches = \"tight\")\n",
    "\n",
    "\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e75bb547-6d62-4f39-817c-42123396042e",
   "metadata": {},
   "outputs": [],
   "source": [
    "plt.figure(figsize = (4,16))\n",
    "\n",
    "data_df = kegg_results[kegg_results['adjusted_p_value'] < 0.05].sort_values('combined_score')[::-1][0:50]\n",
    "\n",
    "if plot_fraction:\n",
    "\n",
    "    ax = sns.barplot(data = data_df, x = 'n_genes', y = 'description', palette = mapper.to_rgba(data_df.adjusted_p_value.values))\n",
    "    ax.set_xlabel('gene fraction', fontsize=12)\n",
    "\n",
    "else:\n",
    "    \n",
    "    ax = sns.barplot(data = data_df, x = 'n_genes', y = 'description', palette = mapper.to_rgba(data_df.adjusted_p_value.values))\n",
    "    ax.set_xlabel('N genes', fontsize=12)\n",
    "\n",
    "ax.set_yticklabels([textwrap.fill(e, 32) for e in data_df['description']], fontsize=8)\n",
    "ax.set_ylabel('KEGG Pathways', fontsize=12)\n",
    "\n",
    "\n",
    "### Interaction DEA\n",
    "if interaction:\n",
    "\n",
    "    if upregulated_degs_only:\n",
    "    \n",
    "        ax.set_title('KEGG: Spleen aggregated DC, interaction DEA', fontsize=12)\n",
    "\n",
    "    else:\n",
    "\n",
    "        ax.set_title('KEGG: Spleen aggregated DC, interaction DEA', fontsize=12)\n",
    "\n",
    "        \n",
    "### KO vs WT in pathogenic mice\n",
    "\n",
    "else:\n",
    "\n",
    "    if upregulated_degs_only:\n",
    "\n",
    "        ax.set_title('KEGG: Spleen aggregated DC, Pathogenic KO vs WT', fontsize=12)\n",
    "\n",
    "    else:\n",
    "        \n",
    "        ax.set_title('KEGG: Spleen aggregated DC, Pathogenic KO vs WT', fontsize=12)\n",
    "\n",
    "#### SAVE FIG #####\n",
    "\n",
    "### Interaction DEA\n",
    "\n",
    "if interaction:\n",
    "\n",
    "    if upregulated_degs_only:\n",
    "\n",
    "        if plot_fraction:\n",
    "\n",
    "            plt.savefig('/Users/oipulk/Documents/scRNASeq/data/Eleftheria_Maranou_Mar2024/analysis/figures/GO/spleen/interactionDEA/aggregatedDC/spleen_GO_MF_aggreagatedDC_interaction_patKO_upregulated_degs_per.pdf', dpi=600,bbox_inches = \"tight\")\n",
    "        \n",
    "        else:\n",
    "\n",
    "            plt.savefig('/Users/oipulk/Documents/scRNASeq/data/Eleftheria_Maranou_Mar2024/analysis/figures/GO/spleen/interactionDEA/aggregatedDC/spleen_GO_MF_aggreagatedDC_interaction_patKO_upregulated_degs_ngenes.pdf', dpi=600,bbox_inches = \"tight\")\n",
    "\n",
    "    else:\n",
    "\n",
    "        if plot_fraction:\n",
    "\n",
    "            plt.savefig('/Users/oipulk/Documents/scRNASeq/data/Eleftheria_Maranou_Mar2024/analysis/figures/GO/spleen/interactionDEA/aggregatedDC/spleen_GO_MF_aggreagatedDC_interaction_patKO_all_degs_per.pdf', dpi=600,bbox_inches = \"tight\")\n",
    "        \n",
    "        else:\n",
    "            \n",
    "            plt.savefig('/Users/oipulk/Documents/scRNASeq/data/Eleftheria_Maranou_Mar2024/analysis/figures/GO/spleen/interactionDEA/aggregatedDC/spleen_GO_MF_aggreagatedDC_interaction_patKO_all_degs_ngenes.pdf', dpi=600,bbox_inches = \"tight\")\n",
    "\n",
    "\n",
    "### KO vs WT in pathogenic mice\n",
    "else:\n",
    "\n",
    "    if upregulated_degs_only:\n",
    "\n",
    "        if plot_fraction:\n",
    "\n",
    "            plt.savefig('/Users/oipulk/Documents/scRNASeq/data/Eleftheria_Maranou_Mar2024/analysis/figures/GO/spleen/KO_vs_WT_in_pathogenic/aggregatedDC/spleen_GO_MF_aggreagatedDC_KO_vs_WT_upregulated_degs_per.pdf', dpi=600,bbox_inches = \"tight\")\n",
    "\n",
    "        else:\n",
    "            \n",
    "            plt.savefig('/Users/oipulk/Documents/scRNASeq/data/Eleftheria_Maranou_Mar2024/analysis/figures/GO/spleen/KO_vs_WT_in_pathogenic/aggregatedDC/spleen_GO_MF_aggreagatedDC_KO_vs_WT_upregulated_degs_ngenes.pdf', dpi=600,bbox_inches = \"tight\")\n",
    "\n",
    "    else:\n",
    "\n",
    "        if plot_fraction:\n",
    "\n",
    "            plt.savefig('/Users/oipulk/Documents/scRNASeq/data/Eleftheria_Maranou_Mar2024/analysis/figures/GO/spleen/KO_vs_WT_in_pathogenic/aggregatedDC/spleen_GO_MF_aggreagatedDC_KO_vs_WT_all_degs_per.pdf', dpi=600,bbox_inches = \"tight\")\n",
    "        \n",
    "        else:    \n",
    "        \n",
    "            plt.savefig('/Users/oipulk/Documents/scRNASeq/data/Eleftheria_Maranou_Mar2024/analysis/figures/GO/spleen/KO_vs_WT_in_pathogenic/aggregatedDC/spleen_GO_MF_aggreagatedDC_KO_vs_WT_all_degs_ngenes.pdf', dpi=600,bbox_inches = \"tight\")\n",
    "\n",
    "\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6c82cd0c-fc6b-4e13-a52c-31e502c9b324",
   "metadata": {},
   "outputs": [],
   "source": [
    "kegg_results[0:20]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "60b762a7-699f-479f-b739-7596ae49715f",
   "metadata": {},
   "outputs": [],
   "source": [
    "enrichr_results= kegg_results\n",
    "\n",
    "if enrichr_results is not None:\n",
    "    # Convert to DataFrame\n",
    "    enrichr_df = pd.DataFrame(enrichr_results)\n",
    "    \n",
    "    if not enrichr_df.empty:\n",
    "        # Sort by p-value\n",
    "        enrichr_df = enrichr_df.sort_values('combined_score')[::-1]\n",
    "        \n",
    "        # Create visualization\n",
    "        plt.figure(figsize=(4, 8))\n",
    "        \n",
    "        # Take top 10 pathways\n",
    "        top = enrichr_df.head(30)\n",
    "        \n",
    "        # Create bar plot\n",
    "        plt.barh(range(len(top)), -np.log10(top['adjusted_p_value']))\n",
    "        \n",
    "        # Customize the plot\n",
    "        plt.yticks(range(len(top)), top['description'], fontsize=10)\n",
    "        plt.xlabel('-log10(adj. p-value)', fontsize=12)\n",
    "        plt.title('KEGG Pathway Enrichment Analysis by Enrichr', fontsize=14)\n",
    "        \n",
    "        # Add value labels\n",
    "        for i, v in enumerate(-np.log10(top['adjusted_p_value'])):\n",
    "            plt.text(v, i, f'{v:.2f}', va='center')\n",
    "        \n",
    "        plt.tight_layout()\n",
    "        plt.show()\n",
    "        \n",
    "        # # Print detailed results\n",
    "        # print(\"\\nDetailed KEGG Pathway Results:\")\n",
    "        # print(enrichr_df[['name', 'p_value', 'term_size', 'query_size', \n",
    "        #               'intersection_size', 'effective_domain_size']].to_string())\n",
    "else:\n",
    "    print(\"No significant KEGG pathways found\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "445384d7-454b-46e5-aed4-4ff70a9e435d",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Run the analysis for downregulated genes\n",
    "\n",
    "degs = list(dc_markers_down.Gene.values)\n",
    "\n",
    "enrichr_results = enrichr_analysis(degs)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "78b919d5-5c00-4cd0-a02f-0bcab0586be5",
   "metadata": {},
   "outputs": [],
   "source": [
    "if enrichr_results is not None:\n",
    "    # Convert to DataFrame\n",
    "    enrichr_df = pd.DataFrame(enrichr_results)\n",
    "    \n",
    "    if not enrichr_df.empty:\n",
    "        # Sort by p-value\n",
    "        enrichr_df = enrichr_df.sort_values('adjusted_p_value')\n",
    "        \n",
    "        # Create visualization\n",
    "        plt.figure(figsize=(6, 10))\n",
    "        \n",
    "        # Take top 10 pathways\n",
    "        top = enrichr_df.head(30)\n",
    "        \n",
    "        # Create bar plot\n",
    "        plt.barh(range(len(top)), -np.log10(top['adjusted_p_value']))\n",
    "        \n",
    "        # Customize the plot\n",
    "        plt.yticks(range(len(top)), top['description'], fontsize=10)\n",
    "        plt.xlabel('-log10(adj. p-value)', fontsize=12)\n",
    "        plt.title('KEGG Pathway Enrichment Analysis by Enrichr', fontsize=14)\n",
    "        \n",
    "        # Add value labels\n",
    "        for i, v in enumerate(-np.log10(top['adjusted_p_value'])):\n",
    "            plt.text(v, i, f'{v:.2f}', va='center')\n",
    "        \n",
    "        plt.tight_layout()\n",
    "        plt.show()\n",
    "        \n",
    "        # # Print detailed results\n",
    "        # print(\"\\nDetailed KEGG Pathway Results:\")\n",
    "        # print(enrichr_df[['name', 'p_value', 'term_size', 'query_size', \n",
    "        #               'intersection_size', 'effective_domain_size']].to_string())\n",
    "else:\n",
    "    print(\"No significant KEGG pathways found\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5b040c13-c187-4560-963e-8bff392e83e9",
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "enrichr_results[enrichr_results['description']=='Th17 cell differentiation']"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a4a12c69-4087-4c39-bc77-359551944c75",
   "metadata": {
    "jp-MarkdownHeadingCollapsed": true
   },
   "source": [
    "### gprofiler"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a0d0020e-5537-4a23-ada2-d38fe6e311c0",
   "metadata": {},
   "outputs": [],
   "source": [
    "import gprofiler\n",
    "from gprofiler import GProfiler\n",
    "# Create a list of gene symbols\n",
    "degs = list(dc_markers_up.Gene.values)\n",
    "\n",
    "# Run KEGG pathway analysis\n",
    "gp = GProfiler(return_dataframe=True)\n",
    "kegg_results = gp.profile(organism='mmusculus',query=degs)\n",
    "\n",
    "kegg_df = kegg_results[kegg_results['source']=='KEGG']\n",
    "\n",
    "if kegg_df is not None:\n",
    "    \n",
    "    if not kegg_df.empty:\n",
    "        # Sort by p-value\n",
    "        kegg_df = kegg_df.sort_values('p_value')\n",
    "        \n",
    "        # Create visualization\n",
    "        plt.figure(figsize=(8, 8))\n",
    "        \n",
    "        # Take top 10 pathways\n",
    "        top_10 = kegg_df.head(30)\n",
    "        \n",
    "        # Create bar plot\n",
    "        plt.barh(range(len(top_10)), -np.log10(top_10['p_value']))\n",
    "        \n",
    "        # Customize the plot\n",
    "        plt.yticks(range(len(top_10)), top_10['name'], fontsize=10)\n",
    "        plt.xlabel('-log10(p-value)', fontsize=12)\n",
    "        plt.title('KEGG Pathways', fontsize=14)\n",
    "        \n",
    "        # Add value labels\n",
    "        for i, v in enumerate(-np.log10(top_10['p_value'])):\n",
    "            plt.text(v, i, f'{v:.2f}', va='center')\n",
    "        \n",
    "        plt.tight_layout()\n",
    "        plt.show()\n",
    "        \n",
    "        # Print detailed results\n",
    "        # print(\"\\nDetailed KEGG Pathway Results:\")\n",
    "        # print(kegg_df[['name', 'p_value', 'term_size', 'query_size', \n",
    "        #               'intersection_size', 'effective_domain_size']].to_string())\n",
    "else:\n",
    "    print(\"No significant KEGG pathways found\")\n",
    "\n",
    "\n",
    "kegg_df = kegg_results[kegg_results['source']=='GO:BP']\n",
    "\n",
    "if kegg_results is not None:\n",
    "    \n",
    "    if not kegg_df.empty:\n",
    "        # Sort by p-value\n",
    "        kegg_df = kegg_df.sort_values('p_value')\n",
    "        \n",
    "        # Create visualization\n",
    "        plt.figure(figsize=(8, 8))\n",
    "        \n",
    "        # Take top 10 pathways\n",
    "        top_10 = kegg_df.head(30)\n",
    "        \n",
    "        # Create bar plot\n",
    "        plt.barh(range(len(top_10)), -np.log10(top_10['p_value']))\n",
    "        \n",
    "        # Customize the plot\n",
    "        plt.yticks(range(len(top_10)), top_10['name'], fontsize=10)\n",
    "        plt.xlabel('-log10(p-value)', fontsize=12)\n",
    "        plt.title('GO: BP', fontsize=14)\n",
    "        \n",
    "        # Add value labels\n",
    "        for i, v in enumerate(-np.log10(top_10['p_value'])):\n",
    "            plt.text(v, i, f'{v:.2f}', va='center')\n",
    "        \n",
    "        plt.tight_layout()\n",
    "        plt.show()\n",
    "        \n",
    "        # Print detailed results\n",
    "        # print(\"\\nDetailed KEGG Pathway Results:\")\n",
    "        # print(kegg_df[['name', 'p_value', 'term_size', 'query_size', \n",
    "        #               'intersection_size', 'effective_domain_size']].to_string())\n",
    "else:\n",
    "    print(\"No significant KEGG pathways found\")\n",
    "\n",
    "kegg_df = kegg_results[kegg_results['source']=='GO:CC']\n",
    "\n",
    "if kegg_results is not None:\n",
    "    \n",
    "    if not kegg_df.empty:\n",
    "        # Sort by p-value\n",
    "        kegg_df = kegg_df.sort_values('p_value')\n",
    "        \n",
    "        # Create visualization\n",
    "        plt.figure(figsize=(8, 8))\n",
    "        \n",
    "        # Take top 10 pathways\n",
    "        top_10 = kegg_df.head(30)\n",
    "        \n",
    "        # Create bar plot\n",
    "        plt.barh(range(len(top_10)), -np.log10(top_10['p_value']))\n",
    "        \n",
    "        # Customize the plot\n",
    "        plt.yticks(range(len(top_10)), top_10['name'], fontsize=10)\n",
    "        plt.xlabel('-log10(p-value)', fontsize=12)\n",
    "        plt.title('GO: CC', fontsize=14)\n",
    "        \n",
    "        # Add value labels\n",
    "        for i, v in enumerate(-np.log10(top_10['p_value'])):\n",
    "            plt.text(v, i, f'{v:.2f}', va='center')\n",
    "        \n",
    "        plt.tight_layout()\n",
    "        plt.show()\n",
    "        \n",
    "        # Print detailed results\n",
    "        # print(\"\\nDetailed KEGG Pathway Results:\")\n",
    "        # print(kegg_df[['name', 'p_value', 'term_size', 'query_size', \n",
    "        #               'intersection_size', 'effective_domain_size']].to_string())\n",
    "else:\n",
    "    print(\"No significant KEGG pathways found\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "bb29e80b-e98b-4e5f-b3cc-f15f8814f624",
   "metadata": {},
   "outputs": [],
   "source": [
    "def create_enrichment_dotplot(df, n_terms=20):\n",
    "    \"\"\"\n",
    "    Create a dot plot of GO enrichment results with different metrics\n",
    "    \n",
    "    Parameters:\n",
    "    -----------\n",
    "    df : pd.DataFrame\n",
    "        Enrichment results\n",
    "    \"\"\"\n",
    "    # Get top terms\n",
    "    plot_df = df.nsmallest(n_terms, 'p_corr').copy()\n",
    "\n",
    "    plot_df = plot_df.sort_values('p_corr', ascending=False)\n",
    "    \n",
    "    # Create figure\n",
    "    fig, ax = plt.subplots(figsize=(12, 8))\n",
    "    \n",
    "    # Plot scatter using percentage for color\n",
    "    scatter = ax.scatter(x=-np.log10(plot_df['p_corr']),\n",
    "                        y=plot_df['term'],\n",
    "                        s=plot_df['n_go'].astype('float'),  # Size represents number of genes\n",
    "                        c=plot_df['per'],  # Color represents percentage\n",
    "                        cmap='winter_r')\n",
    "    \n",
    "    # Customize plot\n",
    "    ax.set_xlabel('-log10(Adjusted p-value)')\n",
    "    ax.set_ylabel('GO Term')\n",
    "    \n",
    "    # Add colorbar\n",
    "    plt.colorbar(scatter, label='Percentage of genes')\n",
    "    \n",
    "    return fig\n",
    "\n",
    "\n",
    "def create_go_network(df, go_relationships, n_terms=15):\n",
    "    \"\"\"\n",
    "    Create a network visualization of GO terms\n",
    "    \"\"\"\n",
    "    import networkx as nx\n",
    "    from adjustText import adjust_text\n",
    "    \n",
    "    # Create network\n",
    "    G = nx.Graph()\n",
    "    \n",
    "    # Get top terms\n",
    "    top_terms = df.nsmallest(n_terms, 'p_corr')\n",
    "    \n",
    "    # Add nodes\n",
    "    for _, row in top_terms.iterrows():\n",
    "        G.add_node(row['term'], pvalue=row['p_corr'], size=row['n_genes'])\n",
    "    \n",
    "    # Add edges based on GO relationships\n",
    "    for term1 in G.nodes():\n",
    "        for term2 in G.nodes():\n",
    "            if term1 != term2 and (term1, term2) in go_relationships:\n",
    "                G.add_edge(term1, term2)\n",
    "    \n",
    "    # Create plot\n",
    "    fig, ax = plt.subplots(figsize=(12, 12))\n",
    "    \n",
    "    # Calculate node colors and sizes\n",
    "    node_colors = [-np.log10(G.nodes[node]['pvalue']) for node in G.nodes()]\n",
    "    node_sizes = [G.nodes[node]['size'] for node in G.nodes()]\n",
    "    \n",
    "    # Draw network\n",
    "    pos = nx.spring_layout(G, k=1.5, iterations=50)\n",
    "    nx.draw_networkx_nodes(G, pos, \n",
    "                          node_color=node_colors,\n",
    "                          node_size=50*np.sqrt(node_sizes),\n",
    "                          cmap='YlOrRd',\n",
    "                          alpha=0.7)\n",
    "    nx.draw_networkx_edges(G, pos, alpha=0.2)\n",
    "    \n",
    "    # Add labels with adjustText\n",
    "    texts = []\n",
    "    for node, (x, y) in pos.items():\n",
    "        texts.append(plt.text(x, y, node, fontsize=8))\n",
    "    \n",
    "    adjust_text(texts, arrowprops=dict(arrowstyle='-', color='gray', alpha=0.5))\n",
    "    \n",
    "    plt.title('GO Term Network')\n",
    "    plt.axis('off')\n",
    "    \n",
    "    return fig\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "cf8bb3c3-6a86-484a-947c-170877bc81f8",
   "metadata": {},
   "outputs": [],
   "source": [
    "data_df = df[df['up_down']=='up']\n",
    "data_df = data_df[(data_df['class']=='biological_process')].sort_values('p_corr')\n",
    "# data_df = data_df[(data_df['class']=='molecular_function')].sort_values('p_corr')\n",
    "\n",
    "\n",
    "# Now let's create the visualizations\n",
    "# First, the dot plot\n",
    "fig1 = create_enrichment_dotplot(data_df, n_terms=30)\n",
    "plt.title('GO Term Enrichment Analysis')\n",
    "plt.tight_layout()\n",
    "plt.show()\n",
    "\n",
    "# Create simple GO relationships based on p-values\n",
    "go_relationships = {}\n",
    "sorted_terms = data_df.sort_values('p_corr').head(20)\n",
    "for i, row1 in sorted_terms.iterrows():\n",
    "    for j, row2 in sorted_terms.iterrows():\n",
    "        if i != j:\n",
    "            # Create relationship if terms are significantly enriched\n",
    "            if row1['p_corr'] < 0.01 and row2['p_corr'] < 0.0001:\n",
    "                go_relationships[(row1['term'], row2['term'])] = 1\n",
    "\n",
    "# Create network visualization\n",
    "fig2 = create_go_network(data_df, go_relationships, n_terms=15)\n",
    "plt.show()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f476f025-192a-4d89-a800-b6620865f4d1",
   "metadata": {},
   "outputs": [],
   "source": [
    "def create_enrichment_dotplot(df, n_terms=20):\n",
    "    \"\"\"\n",
    "    Create side-by-side dot plots for up and down regulated genes\n",
    "    \n",
    "    Parameters:\n",
    "    -----------\n",
    "    df : pd.DataFrame\n",
    "        Enrichment results with 'up_down' column indicating 'up' or 'down'\n",
    "    n_terms : int\n",
    "        Number of terms to show for each regulation direction\n",
    "    \"\"\"\n",
    "    # Create figure with two subplots side by side\n",
    "    fig, (ax1, ax2) = plt.subplots(1, 2, figsize=(24, 8))\n",
    "    \n",
    "    # Process upregulated genes\n",
    "    df_up = df[df['up_down'] == 'up']\n",
    "    plot_df_up = df_up.nsmallest(n_terms, 'p_corr').copy()\n",
    "    plot_df_up = plot_df_up.sort_values('p_corr', ascending=False)\n",
    "    \n",
    "    # Process downregulated genes\n",
    "    df_down = df[df['up_down'] == 'down']\n",
    "    plot_df_down = df_down.nsmallest(n_terms, 'p_corr').copy()\n",
    "    plot_df_down = plot_df_down.sort_values('p_corr', ascending=False)\n",
    "    \n",
    "    # Plot upregulated genes\n",
    "    plt.sca(ax1)\n",
    "    scatter1 = ax1.scatter(x=-np.log10(plot_df_up['p_corr']),\n",
    "                          y=plot_df_up['term'],\n",
    "                          s=1000 * plot_df_up['per'] / (plot_df_up['per'].max()),\n",
    "                          c=plot_df_up['n_go'],\n",
    "                          cmap='winter')\n",
    "    ax1.set_xlabel('-log10(Adjusted p-value)')\n",
    "    ax1.set_ylabel('GO Term')\n",
    "    ax1.set_title('Upregulated DEGs')\n",
    "    plt.colorbar(scatter1, ax=ax1, label='Number of genes in GO term')\n",
    "    \n",
    "    # Plot downregulated genes\n",
    "    plt.sca(ax2)\n",
    "    scatter2 = ax2.scatter(x=-np.log10(plot_df_down['p_corr']),\n",
    "                          y=plot_df_down['term'],\n",
    "                          s=1000 * plot_df_down['per'] / (plot_df_down['per'].max()),\n",
    "                          c=plot_df_down['n_go'],\n",
    "                          cmap='winter')\n",
    "    ax2.set_xlabel('-log10(Adjusted p-value)')\n",
    "    ax2.set_ylabel('GO Term')\n",
    "    ax2.set_title('Downregulated DEGs')\n",
    "    plt.colorbar(scatter2, ax=ax2, label='Number of genes in GO term')\n",
    "    \n",
    "    # Adjust layout\n",
    "    plt.tight_layout()\n",
    "    \n",
    "    return fig\n",
    "\n",
    "# Use the function with your concatenated DataFrame\n",
    "fig = create_enrichment_dotplot(df, n_terms=20)\n",
    "plt.show()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a20278c1-5b1f-4ba3-9139-2a6fcae6d4d8",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ad145636-d36d-4915-a95f-baf27202e30b",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.12"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
