{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "5e136caf-7d6f-4fc9-a923-1de591283d2c",
   "metadata": {},
   "source": [
    "# Pseudobulk DEA of spleen samples"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d0579a61-b218-4cc1-9bfe-10e417fbeaf8",
   "metadata": {},
   "outputs": [],
   "source": [
    "import numpy as np\n",
    "import seaborn as sns\n",
    "\n",
    "import matplotlib\n",
    "import matplotlib.pyplot as plt\n",
    "from matplotlib.legend import Legend\n",
    "import matplotlib.colors as colors\n",
    "from matplotlib import cm\n",
    "from mpl_toolkits.mplot3d import axes3d\n",
    "import matplotlib.patches as patches\n",
    "import pandas as pd\n",
    "import scipy\n",
    "import scanpy as sc\n",
    "import anndata as ad\n",
    "\n",
    "import random\n",
    "\n",
    "from pydeseq2.dds import DeseqDataSet\n",
    "from pydeseq2.ds import DeseqStats\n",
    "\n",
    "eps = 1e-300"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a3f64541-8532-4486-8547-006a174a7363",
   "metadata": {},
   "outputs": [],
   "source": [
    "sc.set_figure_params(scanpy=True, dpi=300, dpi_save=1200, frameon=True, vector_friendly=False, fontsize=14,\n",
    "                         figsize=(9,8),  format='pdf', facecolor=None, transparent=False, ipython_format='png2x')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8bb38df4-a867-4d00-acff-c0dd5ce4c389",
   "metadata": {},
   "outputs": [],
   "source": [
    "# adata = sc.read_h5ad(\"maranou_032024_spleen_annotated.h5ad\")\n",
    "adata = sc.read_h5ad(\"maranou_032024_spleen_annotated.h5ad\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "60cbbed9-e1fe-4826-9f2e-16ee595a71bf",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Based on DEA and marker genes form PanglaoDB, Tabula Muris and Cell Marker 2.0 (currently dysfunctional), Annotation of Cell Types: ACT\n",
    "annotation_dict = {\"0\":\"Naive B cells\",#\n",
    "                   \"1\":\"Naive B cells\",#\n",
    "                   \"2\":\"MZB and B-1 cells\",#\n",
    "                   \"3\":\"Naive B cells\",#\n",
    "                   \"4\":\"Activated B cells\",#\n",
    "                   \"5\":\"Immature B cells\",#\n",
    "                   \"6\":\"NK cells\",#\n",
    "                   \"7\":\"CD8+\",#\n",
    "                   \"8\":\"Treg\",#\n",
    "                   \"9\":\"Naive B cells\",#\n",
    "                   \"10\":\"CD4+\",#\n",
    "                   \"11\":\"Monocytes and macrophages\", #Classical monocytes here\n",
    "                   \"12\":\"MZB and B-1 cells\",#  \n",
    "                   # \"13\":\"DC2\",#\n",
    "                    \"dc2r0\":\"WDFY4+ cDC2\",#\n",
    "                   \"dc2r1\":\"Relb(low) cDC2\", #\n",
    "                   \"dc2r2\":\"Migratory cDC2\",#\n",
    "                   \n",
    "                   \n",
    "                   \"14\":\"Tcm\",#\n",
    "                   \"15\":\"Mature follicular B cells\", #\n",
    "                   \"17\":\"Th\",#\n",
    "                   \"18\":\"Monocytes and macrophages\", #Nonclassical monocytes\n",
    "                   \"19\":\"Proliferative B cells\",# Activated follicular B cells?\n",
    "                   \"20\":\"Lymphoid-resident cDC1\",#\n",
    "                   \"21\":\"Heterogenous T cells\",#\n",
    "                   \"22\":\"Neutrophils\", #Activated neutrophils or myeloid-derived suppressor cells (MDSCs)\n",
    "                   \"24\":\"Mast cells\",#\n",
    "                   # \"25\":\"CD8- DC1\",#\n",
    "\n",
    "                   # The two DC1 clusters are more clearly separated by CyC than Cd8a (which has varied expression in CyC(hi))\n",
    "                   # Ref: The protease inhibitor cystatin C is differentially expressed among dendritic cell populations, but does not control antigen presentation\n",
    "                    #El-Sukkari et al. J Immunol. 2003 Nov 15;171(10):5003-11.  doi: 10.4049/jimmunol.171.10.5003. \n",
    "                   \"dc1r0\":\"CD8- CCR2+ cDC1\",#\n",
    "                   \"dc1r1\":\"CD8(low) cDC1\", #\n",
    "                   \"dc1r2\":\"CD8- CCR2- cDC1\",#\n",
    "                   \n",
    "                   \"26\":\"pDC\",#\n",
    "                   \"27\":\"Activated T cells\", #With B cell characteristics Cd79a, Pax5, and Ighd\n",
    "                   \"28\":\"Plasma cells\",#\n",
    "                   \"r0\":\"Mixed DC\",#\n",
    "                   \"r1\":\"Red pulp macrophages\", #\n",
    "                   \"r2\":\"Relb(int.) cDC2\",#\n",
    "                   \"r3\":\"Germinal center B cells\",\n",
    "                   \"r4\":\"Plasma cells\",#\n",
    "                   \"r5\":\"Treg\"\n",
    "                  }\n",
    "\n",
    "ann_colors = plt.colormaps['tab20'].colors\n",
    "ann_palette_all={\"Germinal center B cells\": ann_colors[18],\n",
    "                       \"Naive B cells\": ann_colors[0],\n",
    "                       \"CD8+\":ann_colors[4],\n",
    "                       \"Activated B cells\": ann_colors[13],\n",
    "                       \"Immature B cells\":plt.matplotlib.colors.to_rgb('dodgerblue'),\n",
    "                       \"Activated T cells\":ann_colors[2],\n",
    "                        \"Tcm\":plt.matplotlib.colors.to_rgb('lightseagreen'),\n",
    "                       \"CD4+\":plt.matplotlib.colors.to_rgb('greenyellow'),\n",
    "                       \"Treg\":ann_colors[16],\n",
    "                         \"Th\":plt.matplotlib.colors.to_rgb('lime'),\n",
    "                        \"Red pulp macrophages\":ann_colors[3],\n",
    "                       \"Heterogenous T cells\":ann_colors[15],\n",
    "                       \"NK cells\":ann_colors[17],\n",
    "                       \"Th cells\":plt.matplotlib.colors.to_rgb('mediumseagreen'),\n",
    "                       \"MZB and B-1 cells\":ann_colors[1],\n",
    "                       \"Mature follicular B cells\":plt.matplotlib.colors.to_rgb('navy'),\n",
    "                       \"Mixed DC\":ann_colors[5],\n",
    "                       \"CCR7+ DC1\":ann_colors[12],\n",
    "                       \"pDC\":ann_colors[6],\n",
    "                       \"Plasma cells\":ann_colors[19],\n",
    "                       \"Monocytes and macrophages\":ann_colors[7], #Probably Classical monocytes\n",
    "                       \"Lymphoid-resident cDC1\":ann_colors[8], #XCR1+ [Gurka et al]\n",
    "                        \"CD8- CCR2+ cDC1\":plt.matplotlib.colors.to_rgb('b'),#\n",
    "                       \"CD8(low) cDC1\":ann_colors[0], #\n",
    "                       \"CD8- CCR2- cDC1\":plt.matplotlib.colors.to_rgb('navy'),#\n",
    "                        \"Relb(int.) cDC2\":ann_colors[13],#\n",
    "                        \"WDFY4+ cDC2\":plt.matplotlib.colors.to_rgb('darkorchid'), #\n",
    "                        \"Relb(low) cDC2\":plt.matplotlib.colors.to_rgb('firebrick'),#\n",
    "                        \"Relb(int.) cDC2\":plt.matplotlib.colors.to_rgb('darkorchid'), \n",
    "                       \"Migratory cDC2\":plt.matplotlib.colors.to_rgb('cornflowerblue'), \n",
    "                       \"Proliferative\":plt.matplotlib.colors.to_rgb('b'),\n",
    "                       \"Mast cells\":plt.matplotlib.colors.to_rgb('cornflowerblue'),\n",
    "                       \"Neutrophils\":plt.matplotlib.colors.to_rgb('coral'),\n",
    "                        \"Proliferative B cells\":plt.matplotlib.colors.to_rgb('cyan'),\n",
    "                }"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7242a42d-7cc3-4a91-bc3a-f19ff5196524",
   "metadata": {},
   "outputs": [],
   "source": [
    "cell_types = list(adata.obs['cell_type_high_res'].unique())\n",
    "sc.pl.umap(adata, color=['cell_type_high_res'],add_outline=True, palette=ann_palette_all, alpha=0.7, s=10, title='Spleen',legend_loc='on data',outline_width = (0.2,0.8),legend_fontsize=8, legend_fontweight='heavy')\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "2f15943b-2a08-46e0-a765-d2914687ce8f",
   "metadata": {},
   "source": [
    "# Define matplotlib Axes\n",
    "# Number of Axes & plot size\n",
    "ncols = 2\n",
    "nrows = 2\n",
    "figsize = 8\n",
    "wspace = 0.1\n",
    "fig, axs = plt.subplots(\n",
    "    nrows=nrows,\n",
    "    ncols=ncols,\n",
    "    figsize=(ncols * figsize + figsize * wspace * (ncols - 1), nrows * figsize),\n",
    ")\n",
    "plt.subplots_adjust(wspace=wspace)\n",
    "\n",
    "sc.pl.umap(adata[adata.obs['sample']=='wt_naive'], ax=axs[0,0], show=False, color=['cell_type_high_res'], title=['Spleen node, Naive WT'], add_outline=True, outline_width = (0.2,0.8), palette=ann_palette_all, alpha=0.7, s=10,legend_loc='on data',legend_fontsize=12, legend_fontweight='medium')\n",
    "sc.pl.umap(adata[adata.obs['sample']=='wt_pathogenic'],  ax=axs[0,1], show=False, color=['cell_type_high_res'], title=['Spleen node, Pathogenic WT'],add_outline=True, outline_width = (0.2,0.8), palette=ann_palette_all, alpha=0.7, s=10,legend_loc='on data',legend_fontsize=12, legend_fontweight='medium')\n",
    "sc.pl.umap(adata[adata.obs['sample']=='ko_naive'],  ax=axs[1,0],show=False, color=['cell_type_high_res'], title=['Spleen node, Naive Cd74 KO'], add_outline=True, outline_width = (0.2,0.8), palette=ann_palette_all, alpha=0.7, s=10,legend_loc='on data',legend_fontsize=12, legend_fontweight='medium')\n",
    "sc.pl.umap(adata[adata.obs['sample']=='ko_pathogenic'], show=False, ax=axs[1,1],color=['cell_type_high_res'], title=['Spleen node, Pathogenic Cd74 KO'], add_outline=True, outline_width = (0.2,0.8), palette=ann_palette_all, alpha=0.7, s=10,legend_loc='on data',legend_fontsize=12, legend_fontweight='medium')\n",
    "\n",
    "plt.tight_layout()\n",
    "# plt.savefig('umap_spleen_annotations_samples.pdf',dpi=600)\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "cce1035c-bfe9-4974-ad19-1c8e7532db2c",
   "metadata": {},
   "source": [
    "## DC"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "76c38a2d-21af-48a0-90b4-238759bc7ebb",
   "metadata": {},
   "outputs": [],
   "source": [
    "dc_types = [\"Lymphoid-resident cDC1\",\"CD8(low) cDC1\",\"CD8- CCR2+ cDC1\",\"CD8- CCR2- cDC1\", \"CCR7+ DC1\", \n",
    "                        \"Relb(int.) cDC2\",\"Migratory cDC2\",\"WDFY4+ cDC2\",\"Relb(low) cDC2\",\"pDC\",\"Mixed DC\"]\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7ee39270-b598-42a2-96a9-f4697bba47f8",
   "metadata": {},
   "outputs": [],
   "source": [
    "adata_dc = adata[adata.obs['cell_type_high_res'].isin(dc_types)].copy()\n",
    "adata_dc.obs['cell_type_high_res']=pd.Categorical(adata_dc.obs['cell_type_high_res']).remove_unused_categories()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "794cdbd1-84fa-476e-8609-aebee8839de3",
   "metadata": {},
   "outputs": [],
   "source": [
    "print(\"X range:\", adata_dc.obsm['X_umap'][:, 0].min(), \"to\", adata_dc.obsm['X_umap'][:, 0].max())\n",
    "print(\"Y range:\", adata_dc.obsm['X_umap'][:, 1].min(), \"to\", adata_dc.obsm['X_umap'][:, 1].max())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7726b484-e8f1-45ac-8c83-c8a6734ecba6",
   "metadata": {},
   "outputs": [],
   "source": [
    "# adata_dc = adata[adata.obs['cell_type_high_res'].isin(dc_types)].copy()\n",
    "# adata_dc.obs['cell_type_high_res']=pd.Categorical(adata_dc.obs['cell_type_high_res']).remove_unused_categories()\n",
    "\n",
    "# Create the UMAP plot and capture the axis object that scanpy returns\n",
    "fig, axes = plt.subplots(1,1)\n",
    "sc.pl.umap(adata_dc,\n",
    "                color=['cell_type_high_res'],\n",
    "                add_outline=True,\n",
    "                outline_width=(0.2, 0.5),\n",
    "                palette=ann_palette_all,\n",
    "                alpha=0.7,\n",
    "                s=10,\n",
    "                title='Spleen DCs',\n",
    "                legend_loc='on data',\n",
    "                legend_fontsize=12,\n",
    "                legend_fontweight='medium',\n",
    "                show=False,\n",
    "                ax=axes,\n",
    "                )  \n",
    "\n",
    "# Set the axis limits on the returned axis object\n",
    "axes.set_xlim(-3.5, 8.)  # Adjust these values\n",
    "axes.set_ylim(-5.5, 5.0)  # Adjust these values\n",
    "\n",
    "plt.savefig('umap_spleen_dc_annotations.pdf', bbox_inches='tight')\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "7da7bfbd-f7b1-43ef-8be2-0bf741f62fd3",
   "metadata": {},
   "source": [
    "### 1. KO vs WT in pathogenic samples"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "fa0ec634-e239-453c-aba6-e1f546c1b07c",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Make a new column named 'condition' for DEA\n",
    "adata.obs['condition'] = adata.obs['sample']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0e48b853-6a48-4a5f-b6b2-058f038450bb",
   "metadata": {},
   "outputs": [],
   "source": [
    "# DEGs in aggregated DC\n",
    "\n",
    "dc_types = [\"Lymphoid-resident cDC1\",\"CD8(low) cDC1\",\"CD8- CCR2+ cDC1\",\"CD8- CCR2- cDC1\", \"CCR7+ DC1\", \n",
    "                        \"Relb(int.) cDC2\",\"Migratory cDC2\",\"WDFY4+ cDC2\",\"Relb(low) cDC2\",\"pDC\",\"Mixed DC\"]\n",
    "adata_subset = adata[adata.obs['cell_type_high_res'].isin(dc_types)].copy()\n",
    "\n",
    "for n_repeat in np.arange(1,2):\n",
    "\n",
    "    pseudodatas = []\n",
    "\n",
    "    for sample in adata_subset.obs['sample'].unique():\n",
    "    \n",
    "        adata_sample = adata_subset[adata_subset.obs['sample']==sample]\n",
    "        adata_sample.X = adata_sample.layers['counts']\n",
    "        \n",
    "        indices = list(adata_sample.obs_names)\n",
    "        random.shuffle(indices)\n",
    "    \n",
    "        num_pseudo_samples = 4\n",
    "        indices = np.array_split(np.array(indices), num_pseudo_samples)\n",
    "        print('Number of cells in each sample:',sample,len(indices[0]))\n",
    "        \n",
    "        for i, pseudo_rep in enumerate(indices) :\n",
    "            \n",
    "            rep_adata =  sc.AnnData(X = adata_sample[indices[i]].X.sum(axis = 0),\n",
    "                                       var = adata_sample[indices[i]].var[[]])\n",
    "            rep_adata.obs_names = [str(sample)+'_'+str(i)]\n",
    "            rep_adata.obs['condition'] = adata_sample.obs['condition'].iloc[0]\n",
    "            rep_adata.obs['wt/ko'] = adata_sample.obs['WT/KO'].iloc[0]\n",
    "            rep_adata.obs['pathogenicity'] = adata_sample.obs['pathogenicity'].iloc[0]\n",
    "            rep_adata.obs['replicate'] = i\n",
    "            \n",
    "            pseudodatas.append(rep_adata)\n",
    "    \n",
    "    pseudodata = sc.concat(pseudodatas)\n",
    "    \n",
    "    counts = pd.DataFrame(pseudodata.X, columns=pseudodata.var_names, index=pseudodata.obs.index)\n",
    "    \n",
    "    # DeSeq2 uses underscore internally to separate conditions. Therefore we need to remove them from the condition names. \n",
    "    for i in range(pseudodata.obs.shape[0]):\n",
    "        pseudodata.obs['condition'].values[i] = pseudodata.obs['condition'].values[i].replace('_', '')\n",
    "\n",
    "    # Save counts and metadata for plotiing heatmaps\n",
    "    counts.T.to_csv('/Users/oipulk/Documents/scRNASeq/data/Eleftheria_Maranou_Mar2024/analysis/figures/pseudobulk/spleen/counts/pseudobulk_counts_spleen_aggregatedDC_'+str(n_repeat)+'.csv')\n",
    "    metadata = pseudodata.obs\n",
    "    metadata.to_csv('/Users/oipulk/Documents/scRNASeq/data/Eleftheria_Maranou_Mar2024/analysis/figures/pseudobulk/spleen/counts/pseudobulk_metadata_spleen_aggregatedDC_'+str(n_repeat)+'.csv')\n",
    "    \n",
    "    dds = DeseqDataSet(counts = counts.astype(int),\n",
    "                       metadata = pseudodata.obs,\n",
    "                       design_factors=['condition'])\n",
    "    \n",
    "    # Filter out genes with zero counts\n",
    "    sc.pp.filter_genes(dds, min_cells = 1)\n",
    "    \n",
    "    dds.deseq2()\n",
    "\n",
    "    ## Statistics for pathogenic KO vs WT \n",
    "    stat_res_pat = DeseqStats(dds, contrast=['condition','kopathogenic','wtpathogenic'])\n",
    "    stat_res_pat.summary()\n",
    "\n",
    "    res_pat = stat_res_pat.results_df\n",
    "\n",
    "    res_pat = res_pat[res_pat.baseMean>0.25]\n",
    "    res_pat['log_pvalue'] = -np.log(eps+stat_res_pat.results_df['pvalue'])\n",
    "    res_pat['log_padj'] = -np.log(eps+stat_res_pat.results_df['padj'])\n",
    "    \n",
    "    \n",
    "    if n_repeat==1:\n",
    "    \n",
    "        de_kowt_pat = res_pat\n",
    "    \n",
    "    if n_repeat>1:\n",
    "\n",
    "        # Average log of adjusted p-value\n",
    "        de_kowt_pat.loc[np.intersect1d(de_kowt_pat.index, stat_res_pat.results_df.index)] = de_kowt_pat.loc[np.intersect1d(de_kowt_pat.index, stat_res_pat.results_df.index)]*(n_repeat-1)/n_repeat + res_pat/n_repeat\n",
    "        \n",
    "        #print(de_kowt_pat)\n",
    "\n",
    "    ## Statistics for naive KO vs WT \n",
    "    stat_res_naive = DeseqStats(dds, contrast=['condition','konaive', 'wtnaive'])\n",
    "    stat_res_naive.summary()\n",
    "\n",
    "    res_naive = stat_res_naive.results_df\n",
    "\n",
    "    res_naive = res_naive[res_naive.baseMean>0.25]\n",
    "    res_naive['log_pvalue'] = -np.log(eps+stat_res_naive.results_df['pvalue'])\n",
    "    res_naive['log_padj'] = -np.log(eps+stat_res_naive.results_df['padj'])\n",
    "    \n",
    "    if n_repeat==1:\n",
    "    \n",
    "        de_kowt_naive = res_naive\n",
    "    \n",
    "    if n_repeat>1:\n",
    "\n",
    "        # Average log of adjusted p-value\n",
    "        de_kowt_naive.loc[np.intersect1d(de_kowt_naive.index, stat_res_naive.results_df.index)] = de_kowt_naive.loc[np.intersect1d(de_kowt_naive.index, stat_res_naive.results_df.index)]*(n_repeat-1)/n_repeat + res_naive/n_repeat\n",
    "\n",
    "\n",
    "    ## Statistics for WT pathogenic vs naive \n",
    "    stat_res_wt = DeseqStats(dds, contrast=['condition', 'wtpathogenic', 'wtnaive'])\n",
    "    stat_res_wt.summary()\n",
    "\n",
    "    res_wt = stat_res_wt.results_df\n",
    "\n",
    "    print('results:')\n",
    "    print(res_wt)\n",
    "\n",
    "    res_wt = res_wt[res_wt.baseMean>0.25]\n",
    "    res_wt['log_pvalue'] = -np.log(eps+stat_res_wt.results_df['pvalue'])\n",
    "    res_wt['log_padj'] = -np.log(eps+stat_res_wt.results_df['padj'])\n",
    "    \n",
    "    if n_repeat==1:\n",
    "    \n",
    "        de_patnaive_wt = res_wt\n",
    "    \n",
    "    if n_repeat>1:\n",
    "\n",
    "        # Average log of adjusted p-value\n",
    "        de_patnaive_wt.loc[np.intersect1d(de_patnaive_wt.index, stat_res_wt.results_df.index)] = de_patnaive_wt.loc[np.intersect1d(de_patnaive_wt.index, stat_res_wt.results_df.index)]*(n_repeat-1)/n_repeat + res_wt/n_repeat\n",
    "\n",
    "\n",
    "    ## Statistics for KO pathogenic vs naive \n",
    "    stat_res_ko = DeseqStats(dds, contrast=['condition', 'kopathogenic', 'konaive'])\n",
    "    stat_res_ko.summary()\n",
    "\n",
    "    res_ko = stat_res_ko.results_df\n",
    "\n",
    "    print('results:')\n",
    "    print(res_ko)\n",
    "\n",
    "    res_ko = res_ko[res_ko.baseMean>0.25]\n",
    "    res_ko['log_pvalue'] = -np.log(eps+stat_res_ko.results_df['pvalue'])\n",
    "    res_ko['log_padj'] = -np.log(eps+stat_res_ko.results_df['padj'])\n",
    "    \n",
    "    if n_repeat==1:\n",
    "    \n",
    "        de_patnaive_ko = res_ko\n",
    "    \n",
    "    if n_repeat>1:\n",
    "\n",
    "        # Average log of adjusted p-value\n",
    "        de_patnaive_ko.loc[np.intersect1d(de_patnaive_ko.index, stat_res_ko.results_df.index)] = de_patnaive_ko.loc[np.intersect1d(de_patnaive_ko.index, stat_res_ko.results_df.index)]*(n_repeat-1)/n_repeat + res_ko/n_repeat\n",
    "\n",
    "    \n",
    "\n",
    "plt.figure(figsize=(6,4))\n",
    "plt.scatter(de_kowt_pat['log2FoldChange'], de_kowt_pat['log_padj'] , facecolor='dodgerblue', edgecolor='blue',linewidths=0.5, zorder=10,alpha=0.7, s=12)\n",
    "\n",
    "plt.axhline( -np.log(0.05) , linestyle='--', linewidth=0.8, alpha=0.7)\n",
    "\n",
    "degs_kowt_pat = de_kowt_pat[(de_kowt_pat['log_padj'] > -np.log(eps+0.05))&(np.abs(de_kowt_pat['log2FoldChange'])>0.5)]\n",
    "degs_kowt_naive = de_kowt_naive[(de_kowt_naive['log_padj'] > -np.log(eps+0.05))&(np.abs(de_kowt_naive['log2FoldChange'])>0.5)]\n",
    "degs_patnaive_wt = de_patnaive_wt[(de_patnaive_wt['log_padj'] > -np.log(eps+0.05))&(np.abs(de_patnaive_wt['log2FoldChange'])>0.5)]\n",
    "degs_patnaive_ko = de_patnaive_ko[(de_patnaive_ko['log_padj'] > -np.log(eps+0.05))&(np.abs(de_patnaive_ko['log2FoldChange'])>0.5)]\n",
    "\n",
    "sig_logfoldchanges_kowt_pat = degs_kowt_pat['log2FoldChange']\n",
    "sig_log_padj_kowt_pat = degs_kowt_pat['log_padj']\n",
    "sig_var_names_kowt_pat = list(degs_kowt_pat.index)\n",
    "\n",
    "for i, gene_name in enumerate(sig_var_names_kowt_pat):\n",
    "\n",
    "    if gene_name!='Cd74':\n",
    "    # print(i, gene_name)\n",
    "        plt.text(sig_logfoldchanges_kowt_pat[i]+0.1, sig_log_padj_kowt_pat[i]+0.05, str(gene_name), fontsize=8, zorder=15)\n",
    "\n",
    "\n",
    "if len(sig_var_names_kowt_pat)>0:\n",
    "    \n",
    "    plt.xlim([np.nanmin(de_kowt_pat.drop(['Cd74'])['log2FoldChange'])-0.5,np.nanmax(de_kowt_pat.drop(['Cd74'])['log2FoldChange'])+1.0])\n",
    "    plt.ylim([-0.05*np.nanmax(sig_log_padj_kowt_pat[np.asarray(sig_var_names_kowt_pat)!='Cd74']),np.nanmax(sig_log_padj_kowt_pat[np.asarray(sig_var_names_kowt_pat)!='Cd74'])*1.05+0.5])\n",
    "\n",
    "plt.title('KO vs. WT in pathogenic spleen samples, aggregated DCs', fontsize='small')\n",
    "\n",
    "plt.savefig('/Users/oipulk/Documents/scRNASeq/data/Eleftheria_Maranou_Mar2024/analysis/figures/pseudobulk/spleen/deseq2_pathogenic_results/volcano_ko_vs_wt_spleen_pahogenic_aggregatedDC.pdf', dpi=600,bbox_inches = \"tight\")\n",
    "\n",
    "plt.show()\n",
    "\n",
    "top_genes_pat=np.asarray(sig_var_names_kowt_pat)[np.argsort(np.abs(sig_logfoldchanges_kowt_pat)+sig_log_padj_kowt_pat)[::-1]]\n",
    "\n",
    "print('top_genes:',top_genes_pat )\n",
    "\n",
    "degs_kowt_pat.to_csv('/Users/oipulk/Documents/scRNASeq/data/Eleftheria_Maranou_Mar2024/analysis/figures/pseudobulk/spleen/deseq2_pathogenic_results/Spleen_pat_kowt_degs_aggregatedDC.csv', index=True)\n",
    "degs_kowt_naive.to_csv('/Users/oipulk/Documents/scRNASeq/data/Eleftheria_Maranou_Mar2024/analysis/figures/pseudobulk/spleen/deseq2_naive_results/Spleen_naive_kowt_degs_aggregatedDC.csv', index=True)\n",
    "degs_patnaive_wt.to_csv('/Users/oipulk/Documents/scRNASeq/data/Eleftheria_Maranou_Mar2024/analysis/figures/pseudobulk/spleen/deseq2_wt_results/Spleen_wt_patnaive_degs_aggregatedDC.csv', index=True)\n",
    "degs_patnaive_ko.to_csv('/Users/oipulk/Documents/scRNASeq/data/Eleftheria_Maranou_Mar2024/analysis/figures/pseudobulk/spleen/deseq2_ko_results/Spleen_ko_patnaive_degs_aggregatedDC.csv', index=True)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d47c52fe-e3e3-4e3c-91d9-a6c0c1b3ad8d",
   "metadata": {},
   "source": [
    "## Volcano plots and heatmaps"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "44fa3a68-71d0-4134-81bd-917b937a0c65",
   "metadata": {},
   "outputs": [],
   "source": [
    "de_kowt_naive = pd.read_csv('/Users/oipulk/Documents/scRNASeq/data/Eleftheria_Maranou_Mar2024/analysis/figures/pseudobulk/spleen/deseq2_naive_results/Spleen_naive_kowt_degs_aggregatedDC.csv')\n",
    "de_kowt_naive.index = de_kowt_naive[\"Unnamed: 0\"]\n",
    "de_kowt_pat = pd.read_csv('/Users/oipulk/Documents/scRNASeq/data/Eleftheria_Maranou_Mar2024/analysis/figures/pseudobulk/spleen/deseq2_pathogenic_results/Spleen_pat_kowt_degs_aggregatedDC.csv')\n",
    "de_kowt_pat.index = de_kowt_pat[\"Unnamed: 0\"]\n",
    "de_patnaive_wt = pd.read_csv('/Users/oipulk/Documents/scRNASeq/data/Eleftheria_Maranou_Mar2024/analysis/figures/pseudobulk/spleen/deseq2_wt_results/Spleen_wt_patnaive_degs_aggregatedDC.csv')\n",
    "de_patnaive_wt.index = de_patnaive_wt[\"Unnamed: 0\"]\n",
    "de_patnaive_ko = pd.read_csv('/Users/oipulk/Documents/scRNASeq/data/Eleftheria_Maranou_Mar2024/analysis/figures/pseudobulk/spleen/deseq2_ko_results/Spleen_ko_patnaive_degs_aggregatedDC.csv')\n",
    "de_patnaive_ko.index = de_patnaive_ko[\"Unnamed: 0\"]\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8d840a66-aa58-44ae-ba1e-a248a2b2f42f",
   "metadata": {},
   "outputs": [],
   "source": [
    "sns.set_style(\"white\")\n",
    "plt.figure(figsize=(9,6))\n",
    "plt.scatter(de_kowt_pat['log2FoldChange'], np.log(de_kowt_pat['log_padj']) , facecolor='dodgerblue', edgecolor='blue',linewidths=0.5, zorder=10,alpha=0.7, s=12)\n",
    "# plt.scatter(de_kowt['log2FoldChange'], de_kowt['log_pvalue'] , facecolor='dodgerblue', edgecolor='blue',linewidths=0.5, zorder=10,alpha=0.7, s=12)\n",
    "\n",
    "plt.axhline( np.log(-np.log(0.05)) , linestyle='--', linewidth=0.8, alpha=0.7)\n",
    "\n",
    "degs_kowt_pat = de_kowt_pat[(de_kowt_pat['log_padj'] > -np.log(eps+0.05))&(np.abs(de_kowt_pat['log2FoldChange'])>0.5)]\n",
    "degs_kowt_pat_up = degs_kowt_pat[degs_kowt_pat['log2FoldChange']>0.5]\n",
    "degs_kowt_pat_down = degs_kowt_pat[degs_kowt_pat['log2FoldChange']<-0.5]\n",
    "\n",
    "sig_logfoldchanges_kowt_pat = degs_kowt_pat['log2FoldChange']\n",
    "sig_log_padj_kowt_pat = degs_kowt_pat['log_padj']\n",
    "sig_var_names_kowt_pat = list(degs_kowt_pat.index)\n",
    "\n",
    "top_ind = np.argsort(np.abs(sig_logfoldchanges_kowt_pat)+np.log(sig_log_padj_kowt_pat))[::-1]\n",
    "top_names = np.asarray(sig_var_names_kowt_pat)[top_ind]\n",
    "\n",
    "for i, gene_name in enumerate(top_names[0:200]):\n",
    "\n",
    "    if gene_name!='Cd74':\n",
    "    # print(i, gene_name)\n",
    "        plt.text(sig_logfoldchanges_kowt_pat[top_ind[i]]+0.05, np.log(sig_log_padj_kowt_pat[top_ind[i]])+0.05, str(gene_name), fontsize=10, zorder=15)\n",
    "\n",
    "    # if len(sig_var_names_ckowt_pat)>0:\n",
    "        \n",
    "    #     plt.xlim([np.nanmin(de_kowt_pat.drop(['Cd74'])['log2FoldChange'])-0.5,np.nanmax(de_kowt_pat.drop(['Cd74'])['log2FoldChange'])+1.0])\n",
    "    #     plt.ylim([-0.05*np.nanmax(sig_log_padj_kowt_pat[np.asarray(sig_var_names_kowt_pat)!='Cd74']),np.nanmax(sig_log_padj_kowt_pat[np.asarray(sig_var_names_kowt_pat)!='Cd74'])*1.05+0.5])\n",
    "\n",
    "select_names = ['Jun','Junb']\n",
    "select_ind = []\n",
    "[select_ind.append(np.where(np.array(sig_var_names_kowt_pat, dtype='str')==select_names[k])[0][0]) for k in np.arange(0,len(select_names))]\n",
    "\n",
    "for i, gene_name in enumerate(select_names):\n",
    "    \n",
    "    plt.text(sig_logfoldchanges_kowt_pat[select_ind[i]]+0.05, np.log(sig_log_padj_kowt_pat[select_ind[i]])+0.05, str(gene_name), fontsize=10, zorder=15)\n",
    "\n",
    "#Only Cd74 will be outaide thia window\n",
    "plt.xlim(-5.5,7.5)\n",
    "plt.ylim(1,5.75)\n",
    "# plt.ylim([-0.2,np.max(sig_log_padj_kowt[np.asarray(sig_var_names_kowt)!='Cd74'])+0.5])\n",
    "plt.title('KO vs. WT in pathogenic spleen samples, aggregated DC',fontsize=14)\n",
    "\n",
    "plt.ylabel('log(-log p)', fontsize=13)\n",
    "plt.xlabel('LFC', fontsize=13)\n",
    "\n",
    "plt.savefig('/Users/oipulk/Documents/scRNASeq/data/Eleftheria_Maranou_Mar2024/analysis/figures/pseudobulk/spleen/deseq2_pathogenic_results/volcano_ko_vs_wt_spleen_pathogenic_aggregatedDC.pdf', dpi=600,bbox_inches = \"tight\")\n",
    "\n",
    "plt.show()\n",
    "\n",
    "top_genes_pat=np.asarray(sig_var_names_kowt_pat)[np.argsort(np.abs(sig_logfoldchanges_kowt_pat)+sig_log_padj_kowt_pat)[::-1]]\n",
    "\n",
    "sig_var_names_kowt_pat_up = list(degs_kowt_pat_up.index)\n",
    "sig_var_names_kowt_pat_down = list(degs_kowt_pat_down.index)\n",
    "\n",
    "sig_logfoldchanges_kowt_pat_up = degs_kowt_pat_up['log2FoldChange']\n",
    "sig_logfoldchanges_kowt_pat_down = degs_kowt_pat_down['log2FoldChange']\n",
    "\n",
    "sig_log_padj_kowt_pat_up = degs_kowt_pat_up['log_padj']\n",
    "sig_log_padj_kowt_pat_down = degs_kowt_pat_down['log_padj']\n",
    "\n",
    "top_genes_pat_up=np.asarray(sig_var_names_kowt_pat_up)[np.argsort(np.abs(sig_logfoldchanges_kowt_pat_up)+sig_log_padj_kowt_pat_up)[::-1]]\n",
    "top_genes_pat_down=np.asarray(sig_var_names_kowt_pat_down)[np.argsort(np.abs(sig_logfoldchanges_kowt_pat_down)+sig_log_padj_kowt_pat_down)[::-1]]\n",
    "\n",
    "print('top genes up:',top_genes_pat_up )\n",
    "print('top genes down:',top_genes_pat_down )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7d78e69e-aa3e-4e46-a9a5-f89ecf02e123",
   "metadata": {},
   "outputs": [],
   "source": [
    "sns.set_style(\"white\")\n",
    "plt.figure(figsize=(6,4))\n",
    "\n",
    "plt.scatter(de_kowt_naive['log2FoldChange'], np.log(de_kowt_naive['log_padj']) , facecolor='dodgerblue', edgecolor='blue',linewidths=0.5, zorder=10,alpha=0.7, s=12)\n",
    "\n",
    "plt.axhline(np.log( -np.log(0.05)) , linestyle='--', linewidth=0.8, alpha=0.7)\n",
    "\n",
    "degs_kowt_naive = de_kowt_naive[(de_kowt_naive['log_padj'] > -np.log(eps+0.05))&(np.abs(de_kowt_naive['log2FoldChange'])>0.5)]\n",
    "degs_kowt_naive_up = degs_kowt_naive[degs_kowt_naive['log2FoldChange']>0.5]\n",
    "degs_kowt_naive_down = degs_kowt_naive[degs_kowt_naive['log2FoldChange']<-0.5]\n",
    "\n",
    "sig_logfoldchanges_kowt_naive = degs_kowt_naive['log2FoldChange']\n",
    "sig_log_padj_kowt_naive = np.log(degs_kowt_naive['log_padj'])\n",
    "sig_var_names_kowt_naive = list(degs_kowt_naive.index)\n",
    "\n",
    "for i, gene_name in enumerate(sig_var_names_kowt_naive):\n",
    "\n",
    "    # Leave out the KO gene and 'Tpm3rs7' (LFC approx 6 in this set)\n",
    "    if (gene_name!='Cd74'):\n",
    "    # print(i, gene_name)\n",
    "        plt.text(sig_logfoldchanges_kowt_naive[i]+0.1, sig_log_padj_kowt_naive[i]+0.05, str(gene_name), fontsize=8, zorder=15)\n",
    "\n",
    "    if len(sig_var_names_kowt_naive)>0:\n",
    "        \n",
    "        plt.xlim([np.nanmin(de_kowt_naive.drop(['Cd74'])['log2FoldChange'])-0.5,np.nanmax(de_kowt_naive.drop(['Cd74'])['log2FoldChange'])+1.0])\n",
    "        plt.ylim([-0.05*np.nanmax(sig_log_padj_kowt_naive[np.asarray(sig_var_names_kowt_naive)!='Cd74']),np.nanmax(sig_log_padj_kowt_naive[np.asarray(sig_var_names_kowt_naive)!='Cd74'])*1.05+0.5])\n",
    "    \n",
    "plt.title('KO vs. WT in healthy mouse spleen', fontsize='small')\n",
    "\n",
    "plt.savefig('/Users/oipulk/Documents/scRNASeq/data/Eleftheria_Maranou_Mar2024/analysis/figures/pseudobulk/spleen/deseq2_pathogenic_results/volcano_ko_vs_wt_spleen_naive_aggregatedDC.pdf', dpi=600,bbox_inches = \"tight\")\n",
    "\n",
    "plt.ylim(1,4.25)\n",
    "# plt.xlim(-2.5,3.)\n",
    "\n",
    "plt.show()\n",
    "\n",
    "sig_var_names_kowt_naive_up = list(degs_kowt_naive_up.index)\n",
    "sig_var_names_kowt_naive_down = list(degs_kowt_naive_down.index)\n",
    "\n",
    "sig_logfoldchanges_kowt_naive_up = degs_kowt_naive_up['log2FoldChange']\n",
    "sig_logfoldchanges_kowt_naive_down = degs_kowt_naive_down['log2FoldChange']\n",
    "\n",
    "sig_log_padj_kowt_naive_up = degs_kowt_naive_up['log_padj']\n",
    "sig_log_padj_kowt_naive_down = degs_kowt_naive_down['log_padj']\n",
    "\n",
    "top_genes_naive_up=np.asarray(sig_var_names_kowt_naive_up)[np.argsort(np.abs(sig_logfoldchanges_kowt_naive_up)+sig_log_padj_kowt_naive_up)[::-1]]\n",
    "top_genes_naive_down=np.asarray(sig_var_names_kowt_naive_down)[np.argsort(np.abs(sig_logfoldchanges_kowt_naive_down)+sig_log_padj_kowt_naive_down)[::-1]]\n",
    "\n",
    "print('top genes up:',top_genes_naive_up )\n",
    "print('top genes down:',top_genes_naive_down )\n",
    "\n",
    "top_genes_naive=np.asarray(sig_var_names_kowt_naive)[np.argsort(np.abs(sig_logfoldchanges_kowt_naive)+sig_log_padj_kowt_naive)[::-1]]\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8aaec5fe-ca38-44a0-8da7-f011eca9eace",
   "metadata": {},
   "outputs": [],
   "source": [
    "sns.set_style(\"white\")\n",
    "plt.figure(figsize=(6,4))\n",
    "\n",
    "plt.scatter(de_patnaive_wt['log2FoldChange'], np.log(de_patnaive_wt['log_padj']) , facecolor='dodgerblue', edgecolor='blue',linewidths=0.5, zorder=10,alpha=0.7, s=12)\n",
    "\n",
    "plt.axhline(np.log( -np.log(0.05)) , linestyle='--', linewidth=0.8, alpha=0.7)\n",
    "\n",
    "degs_patnaive_wt = de_patnaive_wt[(de_patnaive_wt['log_padj'] > -np.log(eps+0.05))&(np.abs(de_patnaive_wt['log2FoldChange'])>0.5)]\n",
    "\n",
    "degs_patnaive_wt_up = de_patnaive_wt[(de_patnaive_wt['log_padj'] > -np.log(eps+0.05))&(de_patnaive_wt['log2FoldChange']>0.5)]\n",
    "degs_patnaive_wt_down = de_patnaive_wt[(de_patnaive_wt['log_padj'] > -np.log(eps+0.05))&(de_patnaive_wt['log2FoldChange']<-0.5)]\n",
    "\n",
    "sig_logfoldchanges_patnaive_wt = degs_patnaive_wt['log2FoldChange']\n",
    "sig_log_padj_patnaive_wt = degs_patnaive_wt['log_padj']\n",
    "sig_var_names_patnaive_wt = list(degs_patnaive_wt.index)\n",
    "\n",
    "for i, gene_name in enumerate(sig_var_names_patnaive_wt):\n",
    "\n",
    "    plt.text(sig_logfoldchanges_patnaive_wt[i]+0.1, np.log(sig_log_padj_patnaive_wt[i])+0.05, str(gene_name), fontsize=8, zorder=15)\n",
    "\n",
    "    # if len(sig_var_names_patnaive_wt)>0:\n",
    "        \n",
    "    #     plt.xlim([np.nanmin(de_patnaive_wt['log2FoldChange'])-0.5,np.nanmax(de_patnaive_wt['log2FoldChange'])+1.0])\n",
    "    #     plt.ylim([-0.05*np.nanmax(sig_log_padj_patnaive_wt[np.asarray(sig_var_names_patnaive_wt)!='Cd74']),np.nanmax(sig_log_padj_patnaive_wt[np.asarray(sig_var_names_patnaive_wt)!='Cd74'])*1.05+0.5])\n",
    "    \n",
    "plt.title('Pathogenic vs. Naive, WT mouse spleen', fontsize='small')\n",
    "\n",
    "plt.savefig('/Users/oipulk/Documents/scRNASeq/data/Eleftheria_Maranou_Mar2024/analysis/figures/pseudobulk/spleen/deseq2_wt_results/volcano_pat_vs_naive_spleen_wt_aggregatedDC.pdf', dpi=600,bbox_inches = \"tight\")\n",
    "\n",
    "plt.ylim(1,4.5)\n",
    "# plt.xlim(-2.5,3.)\n",
    "\n",
    "plt.show()\n",
    "\n",
    "sig_var_names_patnaive_wt_up = list(degs_patnaive_wt_up.index)\n",
    "sig_var_names_patnaive_wt_down = list(degs_patnaive_wt_down.index)\n",
    "\n",
    "sig_logfoldchanges_patnaive_wt_up = degs_patnaive_wt_up['log2FoldChange']\n",
    "sig_logfoldchanges_patnaive_wt_down = degs_patnaive_wt_down['log2FoldChange']\n",
    "\n",
    "sig_log_padj_patnaive_wt_up = degs_patnaive_wt_up['log_padj']\n",
    "sig_log_padj_patnaive_wt_down = degs_patnaive_wt_down['log_padj']\n",
    "\n",
    "top_genes_wt_up=np.asarray(sig_var_names_patnaive_wt_up)[np.argsort(np.abs(sig_logfoldchanges_patnaive_wt_up)+sig_log_padj_patnaive_wt_up)[::-1]]\n",
    "top_genes_wt_down=np.asarray(sig_var_names_patnaive_wt_down)[np.argsort(np.abs(sig_logfoldchanges_patnaive_wt_down)+sig_log_padj_patnaive_wt_down)[::-1]]\n",
    "\n",
    "\n",
    "print('top genes up:',top_genes_wt_up )\n",
    "print('top genes down:',top_genes_wt_down )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4c21b4e7-c831-485c-a0f2-9ee5ceaf449c",
   "metadata": {},
   "outputs": [],
   "source": [
    "sns.set_style(\"white\")\n",
    "plt.figure(figsize=(6,4))\n",
    "\n",
    "plt.scatter(de_patnaive_ko['log2FoldChange'], np.log(de_patnaive_ko['log_padj']) , facecolor='dodgerblue', edgecolor='blue',linewidths=0.5, zorder=10,alpha=0.7, s=12)\n",
    "\n",
    "plt.axhline(np.log( -np.log(0.05)) , linestyle='--', linewidth=0.8, alpha=0.7)\n",
    "\n",
    "degs_patnaive_ko = de_patnaive_ko[(de_patnaive_ko['log_padj'] > -np.log(eps+0.05))&(np.abs(de_patnaive_ko['log2FoldChange'])>0.5)]\n",
    "\n",
    "degs_patnaive_ko_up = de_patnaive_ko[(de_patnaive_ko['log_padj'] > -np.log(eps+0.05))&(de_patnaive_ko['log2FoldChange']>0.5)]\n",
    "degs_patnaive_ko_down = de_patnaive_ko[(de_patnaive_ko['log_padj'] > -np.log(eps+0.05))&(de_patnaive_ko['log2FoldChange']<-0.5)]\n",
    "\n",
    "sig_logfoldchanges_patnaive_ko = degs_patnaive_ko['log2FoldChange']\n",
    "sig_log_padj_patnaive_ko = degs_patnaive_ko['log_padj']\n",
    "sig_var_names_patnaive_ko = list(degs_patnaive_ko.index)\n",
    "\n",
    "for i, gene_name in enumerate(sig_var_names_patnaive_ko):\n",
    "\n",
    "    plt.text(sig_logfoldchanges_patnaive_ko[i]+0.1, np.log(sig_log_padj_patnaive_ko[i])+0.05, str(gene_name), fontsize=8, zorder=15)\n",
    "\n",
    "    # if len(sig_var_names_patnaive_ko)>0:\n",
    "        \n",
    "    #     plt.xlim([np.nanmin(de_patnaive_ko['log2FoldChange'])-0.5,np.nanmax(de_patnaive_ko['log2FoldChange'])+1.0])\n",
    "    #     plt.ylim([-0.05*np.nanmax(sig_log_padj_patnaive_ko[np.asarray(sig_var_names_patnaive_ko)!='Cd74']),np.nanmax(sig_log_padj_patnaive_ko[np.asarray(sig_var_names_patnaive_ko)!='Cd74'])*1.05+0.5])\n",
    "    \n",
    "plt.title('Pathogenic vs. Naive, WT mouse spleen', fontsize='small')\n",
    "\n",
    "plt.savefig('/Users/oipulk/Documents/scRNASeq/data/Eleftheria_Maranou_Mar2024/analysis/figures/pseudobulk/spleen/deseq2_ko_results/volcano_pat_vs_naive_spleen_ko_aggregatedDC.pdf', dpi=600,bbox_inches = \"tight\")\n",
    "\n",
    "plt.ylim(1,4.5)\n",
    "# plt.xlim(-2.5,3.)\n",
    "\n",
    "plt.show()\n",
    "\n",
    "sig_var_names_patnaive_ko_up = list(degs_patnaive_ko_up.index)\n",
    "sig_var_names_patnaive_ko_down = list(degs_patnaive_ko_down.index)\n",
    "\n",
    "sig_logfoldchanges_patnaive_ko_up = degs_patnaive_ko_up['log2FoldChange']\n",
    "sig_logfoldchanges_patnaive_ko_down = degs_patnaive_ko_down['log2FoldChange']\n",
    "\n",
    "sig_log_padj_patnaive_ko_up = degs_patnaive_ko_up['log_padj']\n",
    "sig_log_padj_patnaive_ko_down = degs_patnaive_ko_down['log_padj']\n",
    "\n",
    "top_genes_ko_up=np.asarray(sig_var_names_patnaive_ko_up)[np.argsort(np.abs(sig_logfoldchanges_patnaive_ko_up)+sig_log_padj_patnaive_ko_up)[::-1]]\n",
    "top_genes_ko_down=np.asarray(sig_var_names_patnaive_ko_down)[np.argsort(np.abs(sig_logfoldchanges_patnaive_ko_down)+sig_log_padj_patnaive_ko_down)[::-1]]\n",
    "\n",
    "\n",
    "print('top genes up:',top_genes_ko_up )\n",
    "print('top genes down:',top_genes_ko_down )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "098bddc4-e63a-42b9-a649-9690d59cae78",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Shared DEGs for wt_n -> ko_n and wt_pat -> ko_pat\n",
    "# These genes react similarly to Cd74 KO in naive and pathogenic conditions (provided the change is in the same direction!)\n",
    "\n",
    "print('Shared up:', len(list(set(top_genes_naive_up) & set(top_genes_pat_up))))\n",
    "print(list(set(top_genes_naive_up) & set(top_genes_pat_up)))\n",
    "print('Shared down:', len(list(set(top_genes_naive_down) & set(top_genes_pat_down))))\n",
    "print(list(set(top_genes_naive_down) & set(top_genes_pat_down)))\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "528a845a-35c5-4a33-95a1-55dc760f55ca",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Shared DEGs for wt_n -> wt_p and wt_n -> ko_n\n",
    "# These genes are markers that 'simulate' pathogenicity by Cd74 KO (provided the change is in the same direction!)\n",
    "print('Shared up:',len(list(set(top_genes_naive_up) & set(top_genes_wt_up))))\n",
    "print(list(set(top_genes_naive_up) & set(top_genes_wt_up)))\n",
    "print('Shared down:',len(list(set(top_genes_naive_down) & set(top_genes_wt_down))))\n",
    "print(list(set(top_genes_naive_down) & set(top_genes_wt_down)))\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d0a51fc3-67ac-43ea-892a-2ea20acbcad1",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Shared DEGs for wt_n -> wt_p and wt_pat -> ko_pat\n",
    "print('Shared up:',len(list(set(top_genes_wt_up) & set(top_genes_pat_up))))\n",
    "print(list(set(top_genes_wt_up) & set(top_genes_pat_up)))\n",
    "print('Shared down:',len(list(set(top_genes_wt_down) & set(top_genes_pat_down))))\n",
    "print(list(set(top_genes_wt_down) & set(top_genes_wt_down)))\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "33f88a70-cd61-4053-95f6-51bf656ce9e5",
   "metadata": {},
   "source": [
    "### Heatmap of top DE genes in pathogenic samples"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "21aafa06-d883-49ef-b271-5b494bcf3d70",
   "metadata": {},
   "outputs": [],
   "source": [
    "counts = pd.read_csv('/Users/oipulk/Documents/scRNASeq/data/Eleftheria_Maranou_Mar2024/analysis/figures/pseudobulk/spleen/counts/pseudobulk_counts_spleen_aggregatedDC_1.csv')\n",
    "counts = counts.rename(columns={\"Unnamed: 0\": \"Gene\"})\n",
    "counts.index = counts['Gene']\n",
    "counts = counts.drop('Gene', axis=1)\n",
    "\n",
    "metadata = pd.read_csv('/Users/oipulk/Documents/scRNASeq/data/Eleftheria_Maranou_Mar2024/analysis/figures/pseudobulk/spleen/counts/pseudobulk_metadata_spleen_aggregatedDC_1.csv')\n",
    "metadata = metadata.rename(columns={\"Unnamed: 0\": \"sample\"})\n",
    "metadata = metadata.rename(columns={\"wt/ko\": \"genotype\"})\n",
    "metadata.index = metadata['sample']\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7292ce80-b0c5-4563-a9fd-ff066c31f7fd",
   "metadata": {},
   "outputs": [],
   "source": [
    "dds = DeseqDataSet(counts = counts.astype(int).T,\n",
    "                           metadata = metadata,\n",
    "                           design='~genotype')\n",
    "        \n",
    "sc.pp.filter_genes(dds, min_cells = 1)\n",
    "        \n",
    "dds.deseq2()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b31a15d8-1ca4-4fef-964c-07d6d422f408",
   "metadata": {},
   "outputs": [],
   "source": [
    "#Apply a more strict p-value threshold for plotting \n",
    "degs = de_kowt_pat[(de_kowt_pat['log_padj'] > -np.log(eps+0.01))&(np.abs(de_kowt_pat['log2FoldChange'])>1.0)]\n",
    "\n",
    "sig_logfoldchanges = degs['log2FoldChange']\n",
    "sig_log_padj = degs['log_padj']\n",
    "sig_var_names = list(degs.index)\n",
    "\n",
    "dds.layers['log1p'] = np.log1p(dds.layers['normed_counts'])\n",
    "dds_sig_pat = dds[:, sig_var_names_kowt_pat]\n",
    "\n",
    "log1p_counts = pd.DataFrame(dds_sig_pat.layers['log1p'], index=dds_sig_pat.obs.index, columns=dds_sig_pat.var_names)\n",
    "\n",
    "top_genes_pat=np.asarray(sig_var_names)[np.argsort(np.abs(sig_logfoldchanges)+sig_log_padj)[::-1]]\n",
    "\n",
    "wt_pat_median = np.median(log1p_counts.loc[['wt_pathogenic_0','wt_pathogenic_1','wt_pathogenic_2','wt_pathogenic_3']],axis=0)\n",
    "ko_pat_median = np.median(log1p_counts.loc[['ko_pathogenic_0','ko_pathogenic_1','ko_pathogenic_2','ko_pathogenic_3']],axis=0)\n",
    "\n",
    "median_log1p_counts_pathogenic = pd.DataFrame([wt_pat_median,ko_pat_median], index=['WT', 'KO'], columns=dds_sig_pat.var_names)\n",
    "\n",
    "sns.set_style(\"white\")\n",
    "plt.figure()\n",
    "\n",
    "\n",
    "# Exclude Cd74 (KO) because it's on a different scale\n",
    "g=sns.clustermap(median_log1p_counts_pathogenic[top_genes_pat[0:len(top_genes_pat)]].T,\n",
    "               cmap='RdYlBu_r',\n",
    "               yticklabels=True,\n",
    "               xticklabels=True,\n",
    "               col_cluster=False,\n",
    "               row_cluster=True,\n",
    "               cbar_pos=None,\n",
    "               dendrogram_ratio=(.5, 0.),\n",
    "               # cbar_pos=(0.02, .32, .03, .2)\n",
    "               figsize = (4,18)\n",
    "              )\n",
    "\n",
    "g.ax_heatmap.set_yticklabels(g.ax_heatmap.get_ymajorticklabels(), fontsize = 9)\n",
    "g.ax_heatmap.set_xticklabels(g.ax_heatmap.get_xmajorticklabels(), fontsize = 12)\n",
    "\n",
    "\n",
    "# Rotate labels for for landcape format\n",
    "# g.ax_heatmap.set_yticklabels(g.ax_heatmap.get_ymajorticklabels(), fontsize = 9, rotation=180)\n",
    "# g.ax_heatmap.set_xticklabels(g.ax_heatmap.get_xmajorticklabels(), fontsize = 12, rotation=90)\n",
    "\n",
    "\n",
    "diff_ko_wt = median_log1p_counts_pathogenic.loc['KO'] - median_log1p_counts_pathogenic.loc['WT']\n",
    "\n",
    "# Create a list of '+' and '-' signs based on the difference\n",
    "signs_ko_wt = {gene: '+' if d > 0 else '-' for gene, d in diff_ko_wt[top_genes_pat].items()}\n",
    "\n",
    "reordered_genes = g.dendrogram_row.reordered_ind\n",
    "\n",
    "ax = g.ax_heatmap\n",
    "\n",
    "# Add the signs on top of the heatmap\n",
    "for idx, gene_idx in enumerate(reordered_genes):\n",
    "# for idx, sign in enumerate(signs_ko):\n",
    "\n",
    "    gene = top_genes_pat[gene_idx]\n",
    "    sign_ko_wt = signs_ko_wt[gene]\n",
    "\n",
    "    #Sign of ko-wt response difference\n",
    "    rect = patches.Rectangle((0.9, idx), 0.2, 1, fill=True, facecolor='white', edgecolor='none', alpha=0.9)\n",
    "    ax.add_patch(rect)\n",
    "    ax.text(1, idx+0.5, sign_ko_wt, ha='center', va='center', fontweight='bold', fontsize=10)\n",
    "\n",
    "# plt.title('Heatmap of Top Differentially Expressed Genes in Pathogenic Samples')\n",
    "plt.tight_layout()\n",
    "\n",
    "plt.savefig('/Users/oipulk/Documents/scRNASeq/data/Eleftheria_Maranou_Mar2024/analysis/figures/pseudobulk/spleen/deseq2_pathogenic_results/heatmap_ko_vs_wt_in_pathogenic_spleen_aggregatedDC.pdf', dpi=600,bbox_inches = \"tight\")\n",
    "\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "16201a62-b38b-4687-ae7e-9d1c38b45544",
   "metadata": {},
   "source": [
    "## Interaction analysis: How WT and KO respond to introduction of a pathogenic condition? "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "06c6e010-912a-4fb8-bf60-90f656875e1c",
   "metadata": {},
   "outputs": [],
   "source": [
    "#Import interaction DEA results generated in R\n",
    "\n",
    "def read_results(filename):\n",
    "    res = pd.read_csv(filename)\n",
    "    res.set_index('gene', inplace=True)\n",
    "    return res\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d44eb060-4345-4736-a2da-45726901c46e",
   "metadata": {},
   "outputs": [],
   "source": [
    "from statsmodels.stats.multitest import multipletests"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "044e6511-4830-406e-a4d4-06348af45fa8",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Initialize dictionary to store results across repeats\n",
    "all_results = {}\n",
    "eps = 1e-300  # small constant to avoid log(0)\n",
    "\n",
    "# Average stats over repeats\n",
    "for n_repeat in np.arange(1, 6):\n",
    "    # Read results for this repeat\n",
    "    results_df = read_results('/Users/oipulk/Documents/scRNASeq/data/Eleftheria_Maranou_Mar2024/analysis/figures/pseudobulk/spleen/deseq2_interaction_results/spleen_aggregatedDC_deseq2_interaction_results_'+str(n_repeat)+'.csv')\n",
    "\n",
    "    # Filter by base mean\n",
    "    res = results_df[results_df.baseMean > 0.25]\n",
    "    \n",
    "    # Store log of raw p-values (and other statistics)\n",
    "    res['log_pvalue'] = -np.log(eps + res['pvalue'])\n",
    "    \n",
    "    if n_repeat == 1:\n",
    "        # Initialize storage for all genes we'll see\n",
    "        de_res = res.copy()\n",
    "        # Initialize storage for running average of log p-values\n",
    "        all_genes = set(res.index)\n",
    "    else:\n",
    "        # Update set of all genes we've seen\n",
    "        all_genes.update(res.index)\n",
    "    \n",
    "    # Store results for each repeat\n",
    "    for idx in res.index:\n",
    "        if idx not in all_results:\n",
    "            all_results[idx] = {\n",
    "                'log_pvalue': [],\n",
    "                'baseMean': [],\n",
    "                'log2FoldChange': [],\n",
    "                'lfcSE': [],\n",
    "                'stat': []\n",
    "            }\n",
    "        \n",
    "        all_results[idx]['log_pvalue'].append(res.loc[idx, 'log_pvalue'])\n",
    "        all_results[idx]['baseMean'].append(res.loc[idx, 'baseMean'])\n",
    "        all_results[idx]['log2FoldChange'].append(res.loc[idx, 'log2FoldChange'])\n",
    "        all_results[idx]['lfcSE'].append(res.loc[idx, 'lfcSE'])\n",
    "        all_results[idx]['stat'].append(res.loc[idx, 'stat'])\n",
    "\n",
    "# Create final results dataframe\n",
    "final_results = []\n",
    "for gene in all_genes:\n",
    "    if gene in all_results:\n",
    "        # Calculate mean of log p-values\n",
    "        mean_log_pvalue = np.mean(all_results[gene]['log_pvalue'])\n",
    "        # Convert back to p-value\n",
    "        pvalue = np.exp(-mean_log_pvalue)\n",
    "        \n",
    "        # Calculate means of other statistics\n",
    "        row = {\n",
    "            'gene': gene,\n",
    "            'pvalue': pvalue,\n",
    "            'baseMean': np.mean(all_results[gene]['baseMean']),\n",
    "            'log2FoldChange': np.mean(all_results[gene]['log2FoldChange']),\n",
    "            'lfcSE': np.mean(all_results[gene]['lfcSE']),\n",
    "            'stat': np.mean(all_results[gene]['stat'])\n",
    "        }\n",
    "        final_results.append(row)\n",
    "\n",
    "# Convert to dataframe\n",
    "final_df = pd.DataFrame(final_results)\n",
    "final_df.set_index('gene', inplace=True)\n",
    "\n",
    "# Remove NaN values before running multipletests\n",
    "valid_mask = ~np.isnan(final_df['pvalue'])\n",
    "valid_pvals = final_df.loc[valid_mask, 'pvalue']\n",
    "\n",
    "# Run multipletests only on valid p-values\n",
    "adj_pvals = np.full(len(final_df), np.nan)\n",
    "adj_pvals[valid_mask] = multipletests(valid_pvals.values, method='fdr_bh')[1]\n",
    "final_df['padj'] = adj_pvals\n",
    "\n",
    "# # Store log-transformed versions\n",
    "final_df['log_pvalue'] = -np.log(eps + final_df['pvalue'])\n",
    "final_df['log_padj'] = -np.log(eps + final_df['padj'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "76f5e06a-75ac-436a-8007-b3b1e05179bf",
   "metadata": {},
   "outputs": [],
   "source": [
    "final_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a120d0a3-d7c7-4540-9bed-9c0072799af6",
   "metadata": {},
   "outputs": [],
   "source": [
    "de_res = final_df\n",
    "\n",
    "sns.set_style(\"white\")\n",
    "plt.figure(figsize=(9,6))\n",
    "\n",
    "# pat_intersection= de_kowt_pat.loc[de_kowt_pat.index.intersection(top_names)]\n",
    "\n",
    "# colors = pat_intersection['log2FoldChange'][dDeg_inds]\n",
    "# ringcolors = np.sign(colors-np.mean(colors))\n",
    "\n",
    "# plt.scatter(sig_dLFC,np.log(-np.log(sig_dpadj)),c=ringcolors, cmap = 'coolwarm', zorder=5,alpha=1, s=15)\n",
    "# plt.scatter(sig_dLFC,np.log(-np.log(sig_dpadj)),c=colors, cmap = 'coolwarm', zorder=10,alpha=1, s=12)\n",
    "\n",
    "plt.scatter(de_res['log2FoldChange'], np.log(de_res['log_padj']) , facecolor='dodgerblue', edgecolor='blue',linewidths=0.5, zorder=10,alpha=0.7, s=14)\n",
    "\n",
    "plt.axhline( np.log(-np.log(0.05)) , linestyle='--', linewidth=0.8, alpha=0.7)\n",
    "\n",
    "degs = de_res[(de_res['log_padj'] > -np.log(eps+0.05))&(np.abs(de_res['log2FoldChange'])>0.5)]\n",
    "\n",
    "sig_logfoldchanges = degs['log2FoldChange']\n",
    "sig_log_padj = degs['log_padj']\n",
    "sig_var_names = list(degs.index)\n",
    "\n",
    "top_ind = np.argsort(np.abs(sig_logfoldchanges)+3*np.log(sig_log_padj))[::-1]\n",
    "top_names = np.asarray(sig_var_names)[top_ind]\n",
    "\n",
    "for i, gene_name in enumerate(top_names[0:100]):\n",
    "\n",
    "    # if gene_name=='Cd86':\n",
    "    # print(i, gene_name)\n",
    "    plt.text(sig_logfoldchanges[top_ind[i]]+0.03, np.log(sig_log_padj[top_ind[i]])+0.05, str(gene_name), fontsize=10, zorder=15)\n",
    "\n",
    "select_names = ['Fos','Jun','Junb']\n",
    "select_ind = []\n",
    "[select_ind.append(np.where(np.array(sig_var_names, dtype='str')==select_names[k])[0][0]) for k in np.arange(0,len(select_names))]\n",
    "\n",
    "for i, gene_name in enumerate(select_names):\n",
    "    \n",
    "    plt.text(sig_logfoldchanges[select_ind[i]]+0.03, np.log(sig_log_padj[select_ind[i]])+0.05, str(gene_name), fontsize=10, zorder=15)\n",
    "\n",
    "\n",
    "plt.xlim([-8,9])\n",
    "plt.ylim([0.75,5])\n",
    "\n",
    "plt.ylabel('log(-log p)', fontsize=13)\n",
    "plt.xlabel('LFC', fontsize=13)\n",
    "\n",
    "plt.title('Diff. response to pathogenicity in spleen aggregated DCs: KO vs WT', fontsize='small')\n",
    "\n",
    "plt.savefig('/Users/oipulk/Documents/scRNASeq/data/Eleftheria_Maranou_Mar2024/analysis/figures/pseudobulk/spleen/deseq2_interaction_results/volcano_kopat_vs_rest_pathogenic_spleen_aggregatedDC_interaction.pdf', dpi=600,bbox_inches = \"tight\")\n",
    "\n",
    "plt.show()\n",
    "\n",
    "top_genes=np.asarray(sig_var_names)[np.argsort(np.abs(sig_logfoldchanges)+sig_log_padj)[::-1]]\n",
    "\n",
    "print('top_genes:',top_genes )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "24dd753e-f3ce-4ab6-9c6c-b4d390ad040f",
   "metadata": {},
   "outputs": [],
   "source": [
    "degs.iloc[np.where(degs.index=='Cd74')[0],:]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "abb99de0-eea8-49c6-8ab8-aefa6b2174db",
   "metadata": {},
   "outputs": [],
   "source": [
    "degs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0a1f76e9-1bdc-4603-9d98-137d628c7d65",
   "metadata": {},
   "outputs": [],
   "source": [
    "# degs.to_csv('/Users/oipulk/Documents/scRNASeq/data/Eleftheria_Maranou_Mar2024/analysis/figures/pseudobulk/spleen/deseq2_interaction_results/Spleen_kopat_vs_rest_interaction_degs_CD8minusDC1.csv', index=True)\n",
    "# degs.to_csv('/Users/oipulk/Documents/scRNASeq/data/Eleftheria_Maranou_Mar2024/analysis/figures/pseudobulk/spleen/deseq2_interaction_results/Spleen_kopat_vs_rest_interaction_degs_DC2.csv', index=True)\n",
    "degs.to_csv('/Users/oipulk/Documents/scRNASeq/data/Eleftheria_Maranou_Mar2024/analysis/figures/pseudobulk/spleen/deseq2_interaction_results/Spleen_kopat_vs_rest_interaction_degs_aggregatedDC.csv', index=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2ef494e3-fefb-4240-808b-7f5ad551b113",
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "# Read the files made in R\n",
    "raw_counts = pd.read_csv(\"/Users/oipulk/Documents/scRNASeq/data/Eleftheria_Maranou_Mar2024/analysis/figures/pseudobulk/spleen/deseq2_interaction_results/res_raw_counts.csv\", index_col=0)\n",
    "norm_counts = pd.read_csv(\"/Users/oipulk/Documents/scRNASeq/data/Eleftheria_Maranou_Mar2024/analysis/figures/pseudobulk/spleen/deseq2_interaction_results/res_normalized_counts.csv\", index_col=0)\n",
    "gene_info = pd.read_csv(\"/Users/oipulk/Documents/scRNASeq/data/Eleftheria_Maranou_Mar2024/analysis/figures/pseudobulk/spleen/deseq2_interaction_results/res_gene_info.csv\", index_col=0)\n",
    "sample_info = pd.read_csv(\"/Users/oipulk/Documents/scRNASeq/data/Eleftheria_Maranou_Mar2024/analysis/figures/pseudobulk/spleen/deseq2_interaction_results/res_sample_info.csv\", index_col=0)\n",
    "\n",
    "# Create AnnData object with raw counts as main matrix\n",
    "dds_adata = ad.AnnData(\n",
    "    X=raw_counts.T.values,\n",
    "    var=gene_info,\n",
    "    obs=sample_info\n",
    ")\n",
    "\n",
    "# Set the index names\n",
    "dds_adata.var_names = raw_counts.index\n",
    "dds_adata.obs_names = raw_counts.columns\n",
    "\n",
    "# Add normalized counts in layers\n",
    "dds_adata.layers['normalized'] = norm_counts.T.values\n",
    "\n",
    "# Add size factors to obs\n",
    "dds_adata.obs['size_factors'] = sample_info['sizeFactor']\n",
    "\n",
    "# Store information about what's in the layers\n",
    "dds_adata.uns['layer_description'] = {\n",
    "    'X': 'raw_counts',\n",
    "    'normalized': 'deseq2_normalized_counts'\n",
    "}\n",
    "\n",
    "# Verify the structure\n",
    "print(dds_adata)\n",
    "print(\"\\nAvailable layers:\", list(dds_adata.layers.keys()))\n",
    "print(\"\\nShape of raw counts:\", dds_adata.X.shape)\n",
    "print(\"Shape of normalized counts:\", dds_adata.layers['normalized'].shape)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "edd4cb64-9b62-445b-95a5-1501045244de",
   "metadata": {},
   "source": [
    "# Not run. We import dds from R with fully implemented interaction analysis\n",
    "dds = DeseqDataSet(counts = counts.astype(int).T,\n",
    "                           metadata = metadata,\n",
    "                           design='~genotype + pathogenicity + genotype:pathogenicity')\n",
    "        \n",
    "sc.pp.filter_genes(dds, min_cells = 1)\n",
    "        \n",
    "dds.deseq2()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e988d6f3-86ea-4e3c-bbcf-967b7c98eec7",
   "metadata": {},
   "outputs": [],
   "source": [
    "#Apply a more strict p-value threshold for plotting \n",
    "dds =dds_adata.copy()\n",
    "\n",
    "# The same thresholds will be used in GOA\n",
    "degs = de_res[(de_res['log_padj'] > -np.log(eps+0.05))&(np.abs(de_res['log2FoldChange'])>1.0)]\n",
    "\n",
    "sig_logfoldchanges = degs['log2FoldChange']\n",
    "sig_log_padj = degs['log_padj']\n",
    "sig_var_names = list(degs.index)\n",
    "\n",
    "top_genes=np.asarray(sig_var_names)[np.argsort(np.abs(sig_logfoldchanges)+sig_log_padj)[::-1]]\n",
    "\n",
    "dds.layers['log1p'] = np.log1p(dds.layers['normalized'])\n",
    "dds_sig = dds[:, sig_var_names]\n",
    "\n",
    "log1p_counts = pd.DataFrame(dds_sig.layers['log1p'], index=dds_sig.obs.index, columns=dds_sig.var_names)\n",
    "\n",
    "# conditions = ['wt_naive','wt_pathogenic','ko_naive','ko_pathogenic']  \n",
    "conditions = ['ko_pathogenic','ko_naive','wt_pathogenic','wt_naive']  \n",
    "\n",
    "\n",
    "#Construct a dataframe that will hold the mean expression levels of the filtered genes in various conditions\n",
    "mean_log1p_counts = pd.DataFrame(index=conditions, columns=dds_sig.var_names, dtype='float')\n",
    "\n",
    "for condition in conditions:\n",
    "    # Get all rows that start with the current condition\n",
    "    rows = [row for row in log1p_counts.index if row.startswith(condition)]\n",
    "    \n",
    "    # Calculate the mean of these rows\n",
    "    condition_mean = np.nanmean(log1p_counts.loc[rows],axis=0)\n",
    "    \n",
    "    mean_log1p_counts.loc[condition] =  condition_mean.astype(float)\n",
    "\n",
    "\n",
    "diff_ko = mean_log1p_counts.loc['ko_pathogenic'] - mean_log1p_counts.loc['ko_naive']\n",
    "diff_wt = mean_log1p_counts.loc['wt_pathogenic'] - mean_log1p_counts.loc['wt_naive']\n",
    "ddiff_ko_wt = diff_ko - diff_wt\n",
    "\n",
    "plot_genes = top_genes\n",
    "\n",
    "# Create a list of '+' and '-' signs based on the difference\n",
    "signs_ko = {gene: '+' if d > 0 else '-' for gene, d in diff_ko[plot_genes].items()}\n",
    "signs_wt = {gene: '+' if d > 0 else '-' for gene, d in diff_wt[plot_genes].items()}\n",
    "signs_ko_wt = {gene: '+' if d > 0 else '-' for gene, d in ddiff_ko_wt[plot_genes].items()}\n",
    "\n",
    "\n",
    "sns.set_style(\"white\")\n",
    "plt.figure(dpi=1200)\n",
    "\n",
    "g=sns.clustermap(mean_log1p_counts[plot_genes].T,\n",
    "               z_score=0,\n",
    "               # standard_scale=0,  \n",
    "               cmap='RdYlBu_r',\n",
    "               yticklabels=True,\n",
    "               xticklabels=['KO Pathogenic', 'KO Naive' ,'WT Pathogenic','WT Naive'],\n",
    "               col_cluster=False,\n",
    "               row_cluster=True,\n",
    "               cbar_pos=None,\n",
    "               dendrogram_ratio=(.3, 0.),\n",
    "               tree_kws=dict(linewidths=1.0, colors=(0.2, 0.2, 0.4)),\n",
    "               figsize = (8,20)\n",
    "               # cbar_pos=(0.85, 0.8, 0.1, 0.05),\n",
    "              )\n",
    "\n",
    "g.ax_heatmap.set_yticklabels(g.ax_heatmap.get_ymajorticklabels(), fontsize = 9, rotation=180)\n",
    "g.ax_heatmap.set_xticklabels(g.ax_heatmap.get_xmajorticklabels(), fontsize = 9, rotation=180)\n",
    "\n",
    "labels = g.ax_heatmap.yaxis.get_majorticklabels()\n",
    "\n",
    "# plt.title('Top DEGs in tumor response difference: Spleen data')\n",
    "\n",
    "# Get the reordered index of genes\n",
    "reordered_genes = g.dendrogram_row.reordered_ind\n",
    "\n",
    "ax = g.ax_heatmap\n",
    "\n",
    "# Add the signs on top of the heatmap\n",
    "for idx, gene_idx in enumerate(reordered_genes):\n",
    "# for idx, sign in enumerate(signs_ko):\n",
    "\n",
    "    gene = plot_genes[gene_idx]\n",
    "    sign_ko = signs_ko[gene]\n",
    "    sign_wt = signs_wt[gene]\n",
    "    sign_ko_wt = signs_ko_wt[gene]\n",
    "\n",
    "    #Sign of ko response\n",
    "    rect = patches.Rectangle((0.925, idx), 0.15, 1, fill=True, facecolor='white', edgecolor='none', alpha=0.8)\n",
    "    ax.add_patch(rect)\n",
    "    if sign_ko=='+':\n",
    "        ax.text(1.01, idx+0.5, sign_ko, ha='center', va='center', fontweight='medium', fontsize=10, rotation=90)\n",
    "    else:\n",
    "        ax.text(1.0, idx+0.5, sign_ko, ha='center', va='center', fontweight='medium', fontsize=10, rotation=90)\n",
    "\n",
    "    \n",
    "    #Sign of wt response\n",
    "    rect = patches.Rectangle((2.925, idx), 0.15, 1, fill=True, facecolor='white', edgecolor='none', alpha=0.8)\n",
    "    ax.add_patch(rect)\n",
    "    if sign_wt=='+':\n",
    "        ax.text(3.01, idx+0.5, sign_wt, ha='center', va='center', fontweight='medium', fontsize=10, rotation=90)\n",
    "    else:\n",
    "        ax.text(3.0, idx+0.5, sign_wt, ha='center', va='center', fontweight='medium', fontsize=10, rotation=90)\n",
    "\n",
    "    #Sign of ko-wt response difference\n",
    "    rect = patches.Rectangle((1.9, idx), 0.2, 1, fill=True, facecolor='white', edgecolor='none', alpha=0.9)\n",
    "    ax.add_patch(rect)\n",
    "\n",
    "    if sign_ko_wt=='+':\n",
    "        ax.text(2.01, idx+0.5, sign_ko_wt, ha='center', va='center', fontweight='bold', fontsize=14, rotation=90)\n",
    "    else:\n",
    "        ax.text(2.0, idx+0.5, sign_ko_wt, ha='center', va='center', fontweight='bold', fontsize=14, rotation=90)\n",
    "\n",
    "\n",
    "plt.tight_layout()\n",
    "plt.savefig('/Users/oipulk/Documents/scRNASeq/data/Eleftheria_Maranou_Mar2024/analysis/figures/pseudobulk/spleen/deseq2_interaction_results/Spleen_DC_interactionDEA_heatmap.pdf', dpi=600,bbox_inches = \"tight\")\n",
    "\n",
    "fig = plt.gcf()\n",
    "\n",
    "# Add a colorbar\n",
    "# cax = fig.add_axes([0.8, 0.8, 0.1, 0.1])  # [left, bottom, width, height]\n",
    "# fig.colorbar(g.ax_heatmap.collections[0], cax=cax)\n",
    "\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e0b2e09f-1419-4a34-894a-f7720beaa0b9",
   "metadata": {},
   "outputs": [],
   "source": [
    "labels = [t.get_text() for t in g.ax_heatmap.yaxis.get_majorticklabels()]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4692ac03-4fd0-45c7-b85a-d6573f7da8ba",
   "metadata": {},
   "outputs": [],
   "source": [
    "[print(deg) for deg in labels]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "60179546-4bd4-47c1-89f3-34fcbda5aa6e",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.12"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
