{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "efb33280-17bd-45c2-9b7d-7353eafe4bb8",
   "metadata": {},
   "source": [
    "## Annotation of cells from tumor samples"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "40506a85-8ec8-4660-be16-2fcafb64d0f3",
   "metadata": {},
   "outputs": [],
   "source": [
    "import numpy as np\n",
    "import seaborn as sns\n",
    "\n",
    "import matplotlib\n",
    "import matplotlib.pyplot as plt\n",
    "from matplotlib.legend import Legend\n",
    "import matplotlib.colors as colors\n",
    "from matplotlib import cm\n",
    "from mpl_toolkits.mplot3d import axes3d\n",
    "import pandas as pd\n",
    "import scipy\n",
    "import scanpy as sc\n",
    "import anndata as ad\n",
    "    \n",
    "from sklearn import datasets\n",
    "from sklearn.decomposition import PCA\n",
    "\n",
    "from numba import jit\n",
    "\n",
    "import celltypist\n",
    "from celltypist import models\n",
    "\n",
    "from matplotlib.cm import ScalarMappable"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "858b2b14-4ab3-4dff-b882-212090c7ce45",
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "#Custom colormap\n",
    "\n",
    "from matplotlib.cm import register_cmap\n",
    "from matplotlib.colors import ListedColormap\n",
    "\n",
    "tab20b = matplotlib.colormaps['tab20b']\n",
    "tab20c = matplotlib.colormaps['tab20c']\n",
    "colors1 = tab20b(np.linspace(3.001/5., 1, 9))\n",
    "colors2 = tab20c(np.linspace(0, 3.999/5., 16))\n",
    "\n",
    "colors = np.concatenate([colors1, colors2])\n",
    "\n",
    "map_name = 'op_tab25'\n",
    "op_cmap = ListedColormap(colors, name=map_name )\n",
    "matplotlib.colormaps.register(name=map_name, cmap=op_cmap)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0ac4ce9c-eba7-4316-8f76-b10a09495101",
   "metadata": {},
   "outputs": [],
   "source": [
    "sc.set_figure_params(scanpy=True, dpi=300, dpi_save=1200, frameon=True, vector_friendly=True, fontsize=14,\n",
    "                         figsize=(9,8),  format='pdf', facecolor=None, transparent=False, ipython_format='png2x')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a8198d1a-4772-4f3d-ac54-2cbf76c78b45",
   "metadata": {},
   "outputs": [],
   "source": [
    "#Load tumor data\n",
    "adata = sc.read(\"maranou_032024_tumor_integrated.h5ad\")\n",
    "tissue = 'tumor'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d6c02f47-fa6a-497b-a1f6-471176ab1d0c",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Create a new index by combining 'sample' column with the current index\n",
    "adata.obs['unique_cell_name'] = adata.obs['sample'].astype(str) + '_' + adata.obs.index.astype(str)\n",
    "\n",
    "# Set this new column as the index\n",
    "adata.obs.index = adata.obs['unique_cell_name']\n",
    "\n",
    "# Optionally, you might want to drop the 'unique_cell_name' column afterward if it's no longer needed\n",
    "adata.obs.drop('unique_cell_name', axis=1, inplace=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "cbcf716b-ee80-4f82-b1cf-7e22d4cdeabf",
   "metadata": {},
   "outputs": [],
   "source": [
    "adata.obs['WT/KO']=='wt'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3a8bc916-3e3b-4bec-8da0-6e928bfdd2e8",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Define matplotlib Axes\n",
    "# Number of Axes & plot size\n",
    "ncols = 2\n",
    "nrows = 1\n",
    "figsize = 8\n",
    "wspace = 0.1\n",
    "fig, axs = plt.subplots(\n",
    "    nrows=nrows,\n",
    "    ncols=ncols,\n",
    "    figsize=(ncols * figsize + figsize * wspace * (ncols - 1), nrows * figsize),\n",
    ")\n",
    "plt.subplots_adjust(wspace=wspace)\n",
    "\n",
    "sc.pl.umap(adata[adata.obs['WT/KO']=='wt'], color=['Calr'],palette='tab20',cmap='coolwarm',ax=axs[0], show=False) \n",
    "sc.pl.umap(adata[adata.obs['WT/KO']=='ko'], color=['Calr'],palette='tab20',cmap='coolwarm',ax=axs[1]) \n",
    "plt.tight_layout()\n",
    "# plt.savefig('Calr_tumor.pdf',dpi=600)\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "139a0fe6-35cb-4d83-ba29-2a1a0b4c484e",
   "metadata": {},
   "source": [
    "### Cancer cell and Cancer stem cell markers"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c15d642f-2581-440f-a1b9-6dc4df42e64c",
   "metadata": {},
   "outputs": [],
   "source": [
    "cancer_genes = ['Mlana','Tyr', 'Pmel']\n",
    "adata.obs['cancer_cell_score'] = 0.333*(np.asarray(adata[:,'Mlana'].X.sum(1)/np.max(adata[:,'Mlana'].X.sum(1)))+np.asarray(adata[:,'Tyr'].X.sum(1)/np.max(adata[:,'Tyr'].X.sum(1)))+np.asarray(adata[:,'Pmel'].X.sum(1)/np.max(adata[:,'Pmel'].X.sum(1))))*np.asarray(1-adata[:,'Ptprc'].X.sum(1)/np.max(adata[:,'Ptprc'].X.sum(1)))\n",
    "\n",
    "cancer_stem_cell_genes = ['Cd34','Pdgfra']\n",
    "adata.obs['cancer_stem_cell_score'] = adata[:,cancer_stem_cell_genes].X.sum(1)\n",
    "\n",
    "sc.pl.umap(adata, color=['cancer_cell_score','cancer_stem_cell_score'],palette='tab20',cmap='coolwarm', vmax=1) \n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "65d749f7-d7fa-4d23-96a7-a2c5151189dc",
   "metadata": {},
   "outputs": [],
   "source": [
    "sc.pl.umap(adata, color=['WT/KO'], title=[str(tissue) +' Leiden'],  legend_loc='right margin')\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d301b7e4-eb3d-434c-b472-875721f1e914",
   "metadata": {},
   "outputs": [],
   "source": [
    "sns.set_style('ticks')\n",
    "plt.figure(figsize=(5,4))\n",
    "plt.hist(adata.obsm['X_umap'][:,0], bins=200)\n",
    "plt.xticks(np.arange(-5,20))\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c4819a04-3b86-43cc-af24-1aa927a03416",
   "metadata": {},
   "source": [
    "## Let's split the data to immune cells and other"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c15012c5-4ea9-417f-badc-ac87f5b30095",
   "metadata": {},
   "outputs": [],
   "source": [
    "adata_immune = adata[adata.obsm['X_umap'][:,0]<=1.5].copy()\n",
    "adata_other = adata[adata.obsm['X_umap'][:,0]>1.5].copy()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "0fa4bcaf-943c-4371-846a-25a5cd475c3c",
   "metadata": {},
   "source": [
    "### Annotate immune cells"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "993e2d01-db42-4a6f-8bdb-8b35f0ca358e",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Leiden clustering\n",
    "sc.tl.leiden(adata_immune, resolution=1.0,key_added='leiden_initial')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ad0d8dcc-4cb1-48ef-b16e-d1e6fe509621",
   "metadata": {},
   "outputs": [],
   "source": [
    "# # Create a new index by combining 'sample' column with the current index\n",
    "# adata.obs['unique_cell_name'] = adata.obs['sample'].astype(str) + '_' + adata.obs.index.astype(str)\n",
    "\n",
    "# # Set this new column as the index\n",
    "# adata.obs.index = adata.obs['unique_cell_name']\n",
    "\n",
    "# # Optionally, you might want to drop the 'unique_cell_name' column afterward if it's no longer needed\n",
    "# adata.obs.drop('unique_cell_name', axis=1, inplace=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "067ef764-d890-4f3e-baae-32d3d66ef8ac",
   "metadata": {},
   "outputs": [],
   "source": [
    "sc.set_figure_params(scanpy=True, dpi=300, dpi_save=1200, frameon=True, vector_friendly=True, fontsize=14,\n",
    "                         figsize=(9,9),  format='pdf', facecolor=None, transparent=False, ipython_format='png2x')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5d796d48-38c9-43b6-ad7f-9f56609bd3ca",
   "metadata": {},
   "outputs": [],
   "source": [
    "sc.pl.umap(adata_immune, color=['leiden_initial'], title=[str(tissue) +' Leiden'], palette='tab20', legend_loc='on data')\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7f739b12-eba4-4945-b51a-52a16fac9cd8",
   "metadata": {},
   "outputs": [],
   "source": [
    "##Refine clustering of 16 and 23\n",
    "# 2. Identify the cluster(s) containing DC2, monocytes, and macrophages\n",
    "target_clusters = ['9','10']  # Replace with your actual cluster IDs\n",
    "\n",
    "# 3. Subset the data\n",
    "adata_subset = adata_immune[adata_immune.obs['leiden_initial'].isin(target_clusters)].copy()\n",
    "\n",
    "# 4. Recompute the neighborhood graph on the subset\n",
    "sc.pp.neighbors(adata_subset)\n",
    "\n",
    "# 5. Perform Leiden clustering at higher resolution on the subset\n",
    "sc.tl.leiden(adata_subset, resolution=0.3, key_added='leiden_refined')\n",
    "\n",
    "# 6. Prepare categories for the combined clustering\n",
    "initial_categories = list(adata_immune.obs['leiden_initial'].cat.categories)\n",
    "refined_categories = list(adata_subset.obs['leiden_refined'].cat.categories)\n",
    "\n",
    "# Remove target clusters from initial categories\n",
    "initial_categories_filtered = [cat for cat in initial_categories if cat not in target_clusters]\n",
    "\n",
    "# Create new category names for refined clusters\n",
    "refined_categories_renamed = [f'r{cat}' for cat in refined_categories]\n",
    "\n",
    "# Combine filtered initial categories with renamed refined categories\n",
    "combined_categories = initial_categories_filtered + refined_categories_renamed\n",
    "\n",
    "# 7. Create new column for combined clustering\n",
    "adata_immune.obs['leiden'] = pd.Categorical(\n",
    "    adata_immune.obs['leiden_initial'],\n",
    "    categories=combined_categories\n",
    ")\n",
    "\n",
    "# 8. Update the combined clustering for the refined subset\n",
    "for idx in adata_subset.obs.index:\n",
    "    refined_value = adata_subset.obs.loc[idx, 'leiden_refined']\n",
    "    adata_immune.obs.at[idx, 'leiden'] = f'r{refined_value}'\n",
    "\n",
    "# 9. Optionally, sort the categories for better readability\n",
    "adata_immune.obs['leiden'] = adata_immune.obs['leiden'].cat.reorder_categories(sorted(adata_immune.obs['leiden'].cat.categories))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "64e4a68b-ce89-4903-8949-85578d45f6cf",
   "metadata": {},
   "outputs": [],
   "source": [
    "sc.pl.umap(adata_immune, color=['leiden'], title=[str(tissue) +' Leiden'], palette='tab20', legend_loc='on data')\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "5731c792-9635-4a48-bb2e-febe060e9472",
   "metadata": {},
   "source": [
    "## Use celltypist for immune cell annotation"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2c3b7466-4815-4813-9f6b-51349b525d9e",
   "metadata": {},
   "outputs": [],
   "source": [
    "import celltypist\n",
    "from celltypist import models\n",
    "\n",
    "#Take only immune cells\n",
    "adata_human = adata_immune.copy()\n",
    "\n",
    "#Mouse-human gene conversion\n",
    "mouse_human_genes = pd.read_csv(\"http://www.informatics.jax.org/downloads/reports/HOM_MouseHumanSequence.rpt\",sep=\"\\t\", index_col=False)\n",
    "mouse = mouse_human_genes[mouse_human_genes['Common Organism Name']=='mouse, laboratory']\n",
    "human = mouse_human_genes[mouse_human_genes['Common Organism Name']=='human']\n",
    "\n",
    "# Collect dataframes for conversion\n",
    "mouse = mouse[['DB Class Key', 'Symbol']]\n",
    "mouse.index = np.arange(mouse.shape[0])\n",
    "\n",
    "human = human[['DB Class Key', 'Symbol']]\n",
    "human.index = np.arange(human.shape[0])\n",
    "\n",
    "gene_list = np.array([])\n",
    "\n",
    "# Convert from human to mouse\n",
    "for sym_idx in np.arange(adata.var_names.shape[0]):\n",
    "\n",
    "    row_idx  = np.where(mouse['Symbol']== adata_human.var_names[sym_idx])[0]\n",
    "\n",
    "    if len(row_idx)>0:\n",
    "        \n",
    "        dbck = mouse['DB Class Key'][row_idx].values[0]\n",
    "        human_gene = human[human['DB Class Key']==dbck]['Symbol'].values\n",
    "\n",
    "        if len(human_gene)>0:\n",
    "            gene_list =  np.append(gene_list,human_gene[0])\n",
    "\n",
    "        else:\n",
    "\n",
    "            gene_list =  np.append(gene_list,adata_human.var_names[sym_idx])\n",
    "\n",
    "    else:\n",
    "\n",
    "        gene_list =  np.append(gene_list,adata_human.var_names[sym_idx])\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4bf86403-e92f-41df-84da-d5162b339904",
   "metadata": {},
   "outputs": [],
   "source": [
    "sc.pp.normalize_per_cell(adata_human, counts_per_cell_after=1e4)\n",
    "sc.pp.log1p(adata_human)\n",
    "np.expm1(adata_human.X).sum(1)\n",
    "\n",
    "adata_human.var_names = gene_list\n",
    "predictions = celltypist.annotate(adata_human, model = 'Immune_All_Low.pkl', majority_voting = True)\n",
    "print(predictions.predicted_labels)\n",
    "adata_human = predictions.to_adata()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1a817b5d-66b1-4c53-a3d7-5da5bfbd7982",
   "metadata": {},
   "outputs": [],
   "source": [
    "sc.pl.umap(adata_human, color = ['majority_voting','leiden'], palette='tab20',  legend_loc = 'on data', legend_fontsize=12, legend_fontweight='heavy')\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "462da0da-30a3-4693-ad26-5d0f739c1154",
   "metadata": {},
   "outputs": [],
   "source": [
    "sc.tl.rank_genes_groups(adata_immune,'leiden', method='wilcoxon')\n",
    "sc.pl.rank_genes_groups(adata_immune, n_genes=20, sharey=False)\n",
    "\n",
    "immune_markers = sc.get.rank_genes_groups_df(adata_immune, None)\n",
    "immune_markers = immune_markers[(immune_markers.pvals_adj < 0.03) & (immune_markers.logfoldchanges > 1.0)]\n",
    "immune_markers"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8760a0bd-00ab-48eb-b2f5-f2e72054cbc9",
   "metadata": {},
   "outputs": [],
   "source": [
    "cluster ='0'\n",
    "immune_markers[immune_markers['group']==cluster].iloc[np.argsort(immune_markers[immune_markers['group']==cluster]['logfoldchanges'])[::-1],:][0:20]\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9074e5d0-6639-4cc1-9cd9-159854aac6ac",
   "metadata": {},
   "outputs": [],
   "source": [
    "sc.pl.umap(adata_immune, color=['leiden'], title=[str(tissue) +' Leiden'],add_outline=True, outline_width = (0.2,0.2), palette='op_tab25', alpha=0.7, s=36,legend_loc='on data', legend_fontsize=12, legend_fontweight='heavy')\n",
    "# sc.pl.umap(adata_immune, color=['leiden'], title=[str(tissue) +' Leiden'],add_outline=True, outline_width = (0.2,0.2), palette='tab20', alpha=0.7, s=36,legend_loc='right margin', legend_fontsize=12, legend_fontweight='heavy')\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f11a7b86-3e09-4092-a5ff-609276973cda",
   "metadata": {},
   "outputs": [],
   "source": [
    "#for i in np.arange(26):\n",
    "#    print(f'\"{i}\":\"\",')\n",
    "\n",
    "annotation_dict = {\"0\":\"Tissue-resident macrophages\",\n",
    "\"1\":\"M2 tumor-associated macrophages\",\n",
    "\"2\":\"Activated/TRM CD8+\", #effector/memory T cells with potential tissue-resident characteristics\n",
    "\"3\":\"NK cells\",#\n",
    "\"4\":\"Immunosuppressive, M2-like macrophages\", #\n",
    "\"5\":\"Monocytes and macrophages\",\n",
    "\"6\":\"M-MDSCs\", #Monocytic Myeloid-Derived Suppressor Cells: monocytic origin, immature phenotype, lack of antigen presentation capability, and likely role in the immunosuppressive tumor microenvironment\n",
    "\"7\":\"CD8+\",\n",
    "\"8\":\"Tex\",\n",
    "\"9\":\"DC\",\n",
    "\n",
    "\"11\":\"T cells under regulated activation\" ,#T cell-DC interactions: co-stimulatory (CD28-CD80/CD86) and inhibitory (PD-1 and CTLA-4 pathways) signals. \n",
    "\"r0\":\"DC2\",\n",
    "\"r1\":\"DC1\",\n",
    "\"r2\":\"Migratory DC\",\n",
    "\"r3\":\"Mature B cells\",#activated, tumor-infiltrating B cells that have not yet differentiated into plasma cells\n",
    "\"r4\":\"pDC\",\n",
    "\"r5\":\"Migratory DC\",\n",
    "           }\n",
    "\n",
    "ann_colors = plt.colormaps['tab20'].colors\n",
    "ann_palette={\"Germinal center B cells\": ann_colors[18],\n",
    "                       \"B cells\": ann_colors[0],\n",
    "                       \"CD8+\":ann_colors[4],\n",
    "                       \"Activated B cells\": ann_colors[13],\n",
    "                       \"Tissue-resident macrophages\":plt.matplotlib.colors.to_rgb('dodgerblue'),\n",
    "                       \"Activated/TRM CD8+\":ann_colors[2],\n",
    "                        \"Tcm\":plt.matplotlib.colors.to_rgb('lightseagreen'),\n",
    "                       \"CD4+\":plt.matplotlib.colors.to_rgb('greenyellow'),\n",
    "                       \"T cells under regulated activation\":ann_colors[16],\n",
    "                         \"Th\":plt.matplotlib.colors.to_rgb('lime'),\n",
    "                        \"M2 tumor-associated macrophages\":ann_colors[3],\n",
    "                       \"CD8-CD4-CD3E- T cell-like\":ann_colors[15],\n",
    "                       \"NK cells\":ann_colors[17],\n",
    "                       \"Tex\":plt.matplotlib.colors.to_rgb('lime'),\n",
    "                       \"MZB and B-1 cells\":ann_colors[1],\n",
    "                       \"Activated B cells\":ann_colors[9],\n",
    "                       \"Mature follicular B cells\":plt.matplotlib.colors.to_rgb('navy'),\n",
    "                       \"Immunosuppressive, M2-like macrophages\":ann_colors[5],\n",
    "                       \"Activated DC1-like\":ann_colors[12],\n",
    "                       \"pDC\":ann_colors[6],\n",
    "                       \"M-MDSCs\":ann_colors[19],\n",
    "                       \"Monocytes and macrophages\":ann_colors[7], #Probably Classical monocytes\n",
    "                       \"DC1\":ann_colors[8], #XCR1+ [Gurka et al]\n",
    "                       \"Migratory DC\":ann_colors[9], #[Gurka et al]\n",
    "                       \"DC2\":plt.matplotlib.colors.to_rgb('darkorchid'), \n",
    "                       \"Proliferative\":plt.matplotlib.colors.to_rgb('b'),\n",
    "                       \"Mast cells\":plt.matplotlib.colors.to_rgb('cornflowerblue'),\n",
    "                       \"Neutrophils\":plt.matplotlib.colors.to_rgb('coral'),\n",
    "                        \"Mature B cells\":plt.matplotlib.colors.to_rgb('cyan'),\n",
    "                }"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "bf583760-4717-403c-8c81-41ddacd07049",
   "metadata": {},
   "outputs": [],
   "source": [
    "adata_immune.obs['cell_type'] = adata_immune.obs.leiden.map(annotation_dict)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3ebf1b3d-d0a9-41f6-927c-f4dd1451c9b7",
   "metadata": {},
   "outputs": [],
   "source": [
    "sc.pl.umap(adata_immune, color=['cell_type'],palette=ann_palette, alpha=0.75, add_outline=True, outline_width = (0.2,0.2), title='Tumor immune cell cluster',legend_loc='on data',legend_fontsize=12, legend_fontweight='heavy', save='_tumor_immune_annotations.pdf')\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2fe1cb2c-d6c6-4b09-ba65-00203c49f0bc",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Save data\n",
    "adata_immune.write_h5ad(\"maranou_032024_tumor_immune_cells_annotated.h5ad\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "beca76ae-df2a-4051-b3a8-e0a6f844a1bf",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Define matplotlib Axes\n",
    "# Number of Axes & plot size\n",
    "ncols = 2\n",
    "nrows = 1\n",
    "figsize = 8\n",
    "wspace = 0.1\n",
    "fig, axs = plt.subplots(\n",
    "    nrows=nrows,\n",
    "    ncols=ncols,\n",
    "    figsize=(ncols * figsize + figsize * wspace * (ncols - 1), nrows * figsize),\n",
    ")\n",
    "plt.subplots_adjust(wspace=wspace)\n",
    "\n",
    "sc.pl.umap(adata_immune[adata_immune.obs['WT/KO']=='wt'], ax=axs[0], show=False, color=['cell_type'], title=['Tumor, WT'], add_outline=True, outline_width = (0.2,0.5), palette=ann_palette, alpha=0.7, s=10,legend_loc='on data',legend_fontsize=12, legend_fontweight='medium')\n",
    "sc.pl.umap(adata_immune[adata_immune.obs['WT/KO']=='ko'],  ax=axs[1], show=False, color=['cell_type'], title=['Tumor, KO'],add_outline=True, outline_width = (0.2,0.5), palette=ann_palette, alpha=0.7, s=10,legend_loc='on data',legend_fontsize=12, legend_fontweight='medium')\n",
    "\n",
    "plt.tight_layout()\n",
    "plt.savefig('umap_tumor_immune_annotations_samples.pdf',dpi=600)\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "87d9bebc-8cd1-4e0f-8b07-a2f2fb9e7b19",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Get cell type counts\n",
    "cell_counts = adata_immune.obs['cell_type'].value_counts()\n",
    "\n",
    "# Calculate proportions\n",
    "total_cells = cell_counts.sum()\n",
    "cell_proportions = cell_counts / total_cells\n",
    "\n",
    "# Sort cell types by proportion (descending)\n",
    "cell_proportions_sorted = cell_proportions.sort_values(ascending=True)\n",
    "\n",
    "# Create the plot\n",
    "fig, ax = plt.subplots(figsize=(6, 8))\n",
    "\n",
    "# Plot the stacked bar\n",
    "bottom = 0\n",
    "top = 1\n",
    "for cell_type, proportion in cell_proportions_sorted.items():\n",
    "    ax.bar(0, proportion, bottom=bottom, width=0.5, \n",
    "           color=ann_palette[cell_type], label=cell_type)\n",
    "    bottom += proportion\n",
    "    top -= proportion\n",
    "\n",
    "# Customize the plot\n",
    "ax.set_ylabel('Proportion of Cells')\n",
    "ax.set_title('Cell Type Proportions', fontsize=16)\n",
    "ax.set_xlim(-0.3, 0.3)\n",
    "ax.set_xticks([])  # Remove x-axis ticks\n",
    "\n",
    "bottom = 0\n",
    "top = 1\n",
    "# Add cell count and proportion labels\n",
    "for i, (cell_type, proportion) in enumerate(cell_proportions_sorted.items()):\n",
    "    count = cell_counts[cell_type]\n",
    "    y_position = bottom + proportion/2\n",
    "\n",
    "    if proportion>0.015:\n",
    "        # ax.text(0, y_position, f'{cell_type} {count} ({proportion:.1%})', \n",
    "                # ha='center', va='center', fontsize=14)\n",
    "        ax.text(0, y_position, f'{cell_type} ({proportion:.1%})', \n",
    "                ha='center', va='center', fontsize=14)\n",
    "    bottom += proportion\n",
    "    top -= proportion\n",
    "\n",
    "# Add a legend\n",
    "# plt.legend(bbox_to_anchor=(1.05, 1), loc='upper left', borderaxespad=0.)\n",
    "\n",
    "# Optionally, save the figure\n",
    "plt.savefig('tumor_cell_type_proportions_stacked.png', dpi=300, bbox_inches='tight')\n",
    "\n",
    "plt.tight_layout()\n",
    "plt.show()\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ae2190c2-38f5-40a5-9822-7a47c42c5f10",
   "metadata": {},
   "outputs": [],
   "source": [
    "sns.set_style(\"ticks\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a9d78b23-b22d-46a5-9143-89e300a1b507",
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_cell_proportions(adata):\n",
    "    \n",
    "    cell_counts = adata.obs['cell_type'].value_counts()\n",
    "    total_cells = cell_counts.sum()\n",
    "    return cell_counts / total_cells\n",
    "\n",
    "# Get cell proportions for both AnnData objects\n",
    "adata1 = adata_immune[adata_immune.obs['WT/KO']=='wt']\n",
    "adata2 = adata_immune[adata_immune.obs['WT/KO']=='ko']\n",
    "\n",
    "cell_proportions1 = get_cell_proportions(adata1)\n",
    "cell_proportions2 = get_cell_proportions(adata2)\n",
    "\n",
    "# Combine all cell types from both datasets\n",
    "all_cell_types = sorted(set(cell_proportions1.index) | set(cell_proportions2.index))\n",
    "# Create a DataFrame with proportions from all datasets\n",
    "df = pd.DataFrame({\n",
    "    'WT': cell_proportions1.reindex(all_cell_types).fillna(0),\n",
    "    'KO': cell_proportions2.reindex(all_cell_types).fillna(0),\n",
    " \n",
    "})\n",
    "\n",
    "# Sort by the maximum proportion across all datasets\n",
    "df = df.sort_values(by=df.columns.tolist(), ascending=False)\n",
    "\n",
    "# Create the plot\n",
    "fig, ax = plt.subplots(figsize=(12, 10))\n",
    "\n",
    "ax.set_xlim(0., 1.0)  # Adjust x-axis limits\n",
    "\n",
    "# Plot the stacked bars\n",
    "x = [0.1, 0.9]  # x-coordinates for the four bars\n",
    "width = 0.5  # width of the bars\n",
    "\n",
    "for cell_type in df.index:\n",
    "    bottoms = [df.loc[:cell_type, col].sum() - df.loc[cell_type, col] for col in df.columns]\n",
    "    \n",
    "    ax.bar(x, df.loc[cell_type], bottom=bottoms, width=width, alpha=0.6,\n",
    "           color=ann_palette[cell_type], label=cell_type if x[0] == 0 else \"\")\n",
    "\n",
    "# Customize the plot\n",
    "ax.set_ylabel('Proportion of Cells', fontsize=14)\n",
    "# ax.set_title('Cell Type Proportions Comparison', fontsize=18)\n",
    "ax.set_xticks(x)\n",
    "ax.set_xticklabels(['WT', 'Cd74 KO'], fontsize=12, rotation=0, ha='center')\n",
    "# ax.set_xlim(-0.5, 2.5)\n",
    "\n",
    "# Add cell type labels\n",
    "for i, dataset in enumerate(df.columns):\n",
    "    bottom = 0\n",
    "    for cell_type in df.index:\n",
    "        if df.loc[cell_type, dataset] > 0.015:  # Only label if proportion > 1%\n",
    "            ax.text(x[i], bottom + df.loc[cell_type, dataset]/2, \n",
    "                    f'{cell_type} {df.loc[cell_type, dataset]:.1%}', \n",
    "                    ha='center', va='center', fontsize=12)\n",
    "        bottom += df.loc[cell_type, dataset]\n",
    "\n",
    "# Add a legend\n",
    "# plt.legend(bbox_to_anchor=(1.05, 1), loc='upper left', borderaxespad=0., fontsize=10)\n",
    "\n",
    "# Remove top and right spines\n",
    "ax.spines['top'].set_visible(False)\n",
    "ax.spines['right'].set_visible(False)\n",
    "\n",
    "ax.set_xlim(-0.4, 1.6)  # Adjust x-axis limits\n",
    "# ax.text(0.25, 1.02, 'Naive', transform=ax.transAxes, ha='center', va='bottom', fontsize=16, fontweight='bold')\n",
    "# ax.text(0.77, 1.02, 'Pathogenic', transform=ax.transAxes, ha='center', va='bottom', fontsize=16, fontweight='bold')\n",
    "\n",
    "# Optionally, save the figure\n",
    "plt.savefig('tumor_cell_type_proportions_WT_vs_KO.pdf', dpi=600, bbox_inches='tight')\n",
    "\n",
    "plt.show()\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "bddf38f4-3442-4d89-82ef-fc2dbee80cca",
   "metadata": {},
   "source": [
    "## Check expression of Pdcd1, Tigit, Ctla4, Havcr2, and Eomes in T cells"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "005570b9-0e91-4c4a-9a80-0270c1b7eccf",
   "metadata": {},
   "outputs": [],
   "source": [
    "def create_split_violin_plot(df_wt, df_ko, genes_to_plot, figsize=(10, 6)):\n",
    "    # First, let's reshape the data to long format\n",
    "    # Add a condition column to each dataframe\n",
    "    df_wt_long = df_wt[genes_to_plot].copy()\n",
    "    df_wt_long['condition'] = 'WT'\n",
    "    df_wt_long = df_wt_long.melt(id_vars=['condition'], \n",
    "                                value_vars=genes_to_plot,\n",
    "                                var_name='gene', \n",
    "                                value_name='expression')\n",
    "\n",
    "    df_ko_long = df_ko[genes_to_plot].copy()\n",
    "    df_ko_long['condition'] = 'KO'\n",
    "    df_ko_long = df_ko_long.melt(id_vars=['condition'], \n",
    "                                value_vars=genes_to_plot,\n",
    "                                var_name='gene', \n",
    "                                value_name='expression')\n",
    "\n",
    "    # Combine the dataframes\n",
    "    df_combined = pd.concat([df_wt_long, df_ko_long], axis=0)\n",
    "    df_combined = df_combined[(df_combined['gene']!='cell_type')&(df_combined['gene']!='WT/KO')]\n",
    "\n",
    "    # df_combined['expression'] = np.exp(df_combined['expression'])\n",
    "    \n",
    "    # return df_combined\n",
    "\n",
    "    # Create the plot\n",
    "    plt.figure(figsize=figsize)\n",
    "    ax = sns.violinplot(data=df_combined, \n",
    "                        x='gene', \n",
    "                        y='expression', \n",
    "                        hue='condition',\n",
    "                        split=True,\n",
    "                        gap=0.1,\n",
    "                        # split=False,\n",
    "                        density_norm = 'width',\n",
    "                        inner='stick',\n",
    "                        cut=0,\n",
    "                        palette=['dodgerblue','orchid'],\n",
    "                        alpha=0.7)\n",
    "\n",
    "    plt.xticks(rotation=45, ha='right')\n",
    "    plt.tight_layout()\n",
    "    plt.xlim(-0.5, len(genes_to_plot)+0.25)\n",
    "    \n",
    "    return ax, df_combined\n",
    "    \n",
    "\n",
    "# Statistical testing\n",
    "from scipy import stats\n",
    "from statsmodels.stats.multitest import multipletests\n",
    "\n",
    "def add_statistical_annotation(ax, df_long, genes):\n",
    "    # Perform statistical tests\n",
    "    p_values = []\n",
    "    directions = []\n",
    "    y_max = []\n",
    "    \n",
    "    for gene in genes:\n",
    "        gene_data = df_long[df_long['gene'] == gene]\n",
    "        # Convert to numeric and handle any non-numeric values\n",
    "        wt_expr = pd.to_numeric(gene_data[gene_data['condition'] == 'WT']['expression'], errors='coerce')\n",
    "        ko_expr = pd.to_numeric(gene_data[gene_data['condition'] == 'KO']['expression'], errors='coerce')\n",
    "\n",
    "    \n",
    "        # Remove NaN values\n",
    "        wt_expr = wt_expr.dropna()\n",
    "        ko_expr = ko_expr.dropna()\n",
    "        \n",
    "        # Convert to numpy arrays\n",
    "        wt_expr = wt_expr.to_numpy()\n",
    "        ko_expr = ko_expr.to_numpy()\n",
    "        \n",
    "        # Mann-Whitney U test\n",
    "        try:\n",
    "            if len(wt_expr) > 0 and len(ko_expr) > 0:\n",
    "                _, p_value = stats.mannwhitneyu(wt_expr, ko_expr, alternative='two-sided')\n",
    "                if np.mean(wt_expr) > np.mean(ko_expr):\n",
    "                    direction = '>'\n",
    "    \n",
    "                if np.mean(wt_expr) < np.mean(ko_expr):\n",
    "                    direction = '<'\n",
    "                    \n",
    "            else:\n",
    "                p_value = 1.0\n",
    "                direction = ' '\n",
    "                \n",
    "        except ValueError:\n",
    "            p_value = 1.0\n",
    "\n",
    "        \n",
    "        p_values.append(p_value)\n",
    "\n",
    "        directions.append(direction)\n",
    "        \n",
    "        # Get max y value for positioning asterisks\n",
    "        # Handle empty arrays\n",
    "        wt_max = np.max(wt_expr) if len(wt_expr) > 0 else -np.inf\n",
    "        ko_max = np.max(ko_expr) if len(ko_expr) > 0 else -np.inf\n",
    "        current_max = max(wt_max, ko_max)\n",
    "        \n",
    "        # If both arrays were empty, use 0 or another suitable default\n",
    "        if current_max == -np.inf:\n",
    "            current_max = 0\n",
    "            \n",
    "        y_max.append(current_max)\n",
    "    \n",
    "    # Only proceed with multiple testing correction if we have valid p-values\n",
    "    if len(p_values) > 0:\n",
    "        _, p_values_corrected, _, _ = multipletests(p_values, method='fdr_bh')\n",
    "    else:\n",
    "        p_values_corrected = []\n",
    "\n",
    "    p_values_corrected_short = np.ones(len(genes))\n",
    "    # Add asterisks\n",
    "    if len(y_max) > 0:  # Only proceed if we have valid y_max values\n",
    "        y_range = max(y_max) - min(y_max) if len(y_max) > 1 else 1\n",
    "        \n",
    "        for idx, (p, dir, ymax) in enumerate(zip(p_values_corrected, directions, y_max)):\n",
    "            asterisk = ('****' if p < 0.0001 else \n",
    "                       '***' if p < 0.001 else \n",
    "                       '**' if p < 0.01 else \n",
    "                       '*' if p < 0.05 else \n",
    "                       'ns')\n",
    "\n",
    "            p_values_corrected_short= \"{:.2E}\".format(p_values_corrected[idx])\n",
    "            \n",
    "            # Position the asterisk above the violin plot\n",
    "            y_position = ymax +  y_range\n",
    "            ax.text(idx, y_position, dir+asterisk, ha='center', va='bottom')\n",
    "            ax.text(idx, y_position-0.2, 'p='+ str(p_values_corrected_short) , ha='center', va='bottom')\n",
    "        \n",
    "        # Adjust the plot limits to show asterisks\n",
    "        plt.ylim(0, max(y_max) * 1.5)\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "38ab96c3-ef4d-4c0d-a6d1-038a37aabd1b",
   "metadata": {},
   "outputs": [],
   "source": [
    "list(adata_immune.obs['cell_type'].unique())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8d042230-1964-49f5-be59-8b5858576e34",
   "metadata": {},
   "outputs": [],
   "source": [
    "gene_list = ['Pdcd1', 'Tigit', 'Ctla4', 'Havcr2', 'Eomes']\n",
    "\n",
    "Tcells = ['Tex','CD8+','Activated/TRM CD8+','T cells under regulated activation']\n",
    "\n",
    "adata_Tcells = adata_immune[adata_immune.obs['cell_type'].isin(Tcells)].copy()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "cb2ac5b6-1881-4a21-9e50-e597b2e30f72",
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "vars = gene_list.copy()\n",
    "vars.append('WT/KO')\n",
    "vars.append('cell_type')\n",
    "\n",
    "df = sc.get.obs_df(adata_immune, vars)\n",
    "\n",
    "# Consider only DC clusters\n",
    "df_Tcells = df[df['cell_type'].isin(Tcells)]\n",
    "# df_Tcells = df_Tcells[df_Tcells['Tigit']>0]\n",
    "\n",
    "df_Tcells_ko = df_Tcells[df_Tcells['WT/KO']=='ko'][gene_list]\n",
    "df_Tcells_wt = df_Tcells[df_Tcells['WT/KO']=='wt'][gene_list]\n",
    "\n",
    "ax, df_long = create_split_violin_plot(df_Tcells_wt, df_Tcells_ko, gene_list)\n",
    "add_statistical_annotation(ax, df_long, gene_list)\n",
    "\n",
    "\n",
    "plt.savefig('tumor_dysfunction_markers_in_T_cells.pdf', dpi=600, bbox_inches='tight')\n",
    "plt.show()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "19e7099b-8e29-42d6-9aa3-4dfe5d6e4772",
   "metadata": {},
   "outputs": [],
   "source": [
    "def compare_gene_expression_single_celltype(df, features, condition_key, condition1, condition2):\n",
    "    \"\"\"\n",
    "    Compare gene expression between conditions for multiple genes in a single cell type.\n",
    "    \n",
    "    Parameters:\n",
    "    -----------\n",
    "    df : pandas DataFrame\n",
    "        Input data frame containing expression data\n",
    "    features : list\n",
    "        List of genes to analyze\n",
    "    condition_key : str\n",
    "        Column name for condition (e.g., 'WT/KO')\n",
    "    condition1: str\n",
    "        Name of condition 1 (e.g., 'wt')\n",
    "    condition2: str\n",
    "        Name of condition 2 (e.g., 'ko')\n",
    "    \n",
    "    Returns:\n",
    "    --------\n",
    "    pandas DataFrame with results including statistics and significance\n",
    "    \"\"\"\n",
    "    \n",
    "    all_results = []\n",
    "    all_p_values = []  # Store all p-values for global correction\n",
    "    \n",
    "    # Calculate p-values for each gene\n",
    "    for feature in features:\n",
    "        c1_expr = df[df[condition_key] == condition1][feature]\n",
    "        c2_expr = df[df[condition_key] == condition2][feature]\n",
    "        \n",
    "        # Calculate p-value if enough samples\n",
    "        if (len(c1_expr) > 1) & (len(c2_expr) > 1):\n",
    "            statistic, p_value = stats.mannwhitneyu(c1_expr, c2_expr, alternative='two-sided')\n",
    "        else:\n",
    "            p_value = 1.0\n",
    "            \n",
    "        direction = '_'\n",
    "        if (len(c1_expr) > 0) & (len(c2_expr) > 0):\n",
    "            if np.mean(c1_expr) > np.mean(c2_expr):\n",
    "                direction = '>'\n",
    "            if np.mean(c1_expr) < np.mean(c2_expr):\n",
    "                direction = '<'\n",
    "                \n",
    "        # Calculate mean expression for each condition\n",
    "        mean_c1 = np.mean(c1_expr)\n",
    "        mean_c2 = np.mean(c2_expr)\n",
    "                \n",
    "        all_results.append({\n",
    "            'feature': feature,\n",
    "            'direction': direction,\n",
    "            'p_value': p_value,\n",
    "            f'{condition1}_mean': mean_c1,\n",
    "            f'{condition2}_mean': mean_c2\n",
    "        })\n",
    "        all_p_values.append(p_value)\n",
    "    \n",
    "    # Convert to DataFrame\n",
    "    results_df = pd.DataFrame(all_results)\n",
    "    \n",
    "    # Correct p-values\n",
    "    _, p_values_corrected, _, _ = multipletests(all_p_values, method='fdr_bh')\n",
    "    results_df['p_value_corrected'] = p_values_corrected\n",
    "    \n",
    "    # Add significance asterisks based on corrected p-values\n",
    "    results_df['significance'] = results_df['p_value_corrected'].apply(\n",
    "        lambda p: '****' if p < 0.0001 else\n",
    "                 ('***' if p < 0.001 else\n",
    "                  ('**' if p < 0.01 else\n",
    "                   ('*' if p < 0.05 else '')))\n",
    "    )\n",
    "    \n",
    "    return results_df\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c17a3035-a886-491b-aad9-719d873f1a91",
   "metadata": {},
   "outputs": [],
   "source": [
    "T_cell_types =['Tex','CD8+','Activated/TRM CD8+','T cells under regulated activation']\n",
    "adata_T_cells = adata_immune[adata_immune.obs['cell_type'].isin(T_cell_types)].copy()\n",
    "adata_T_cells.obs['cell_type']=pd.Categorical(adata_T_cells.obs['cell_type']).remove_unused_categories()\n",
    "\n",
    "dysfunction_markers = ['Pdcd1', 'Tigit', 'Ctla4', 'Havcr2', 'Eomes']\n",
    "activation_markers = ['Mki67', 'Ifng', 'Gzmb', 'Prf1']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2e170d10-482c-4b38-96d3-8718c13d5344",
   "metadata": {},
   "outputs": [],
   "source": [
    "features = dysfunction_markers+activation_markers\n",
    "cell_key = 'cell_type'\n",
    "condition_key = 'WT/KO'\n",
    "\n",
    "df = sc.get.obs_df(adata_immune[adata_immune.obs['pathogenicity']=='pathogenic'], features+ [condition_key, cell_key])\n",
    "\n",
    "df = df[df['cell_type'].isin(T_cell_types)]\n",
    "\n",
    "df[cell_key] = pd.Categorical(df[cell_key]).remove_unused_categories()\n",
    "df_all_cells = df.copy()\n",
    "df_all_cells[cell_key] = 'Combined T Cells'\n",
    "\n",
    "# Concatenate original and combined data\n",
    "df = pd.concat([df, df_all_cells])\n",
    "og_cell_list=list(df[cell_key].unique())\n",
    "\n",
    "cell_ids = []\n",
    "for cell_id in np.arange(len(og_cell_list)):\n",
    "\n",
    "    cell_type = og_cell_list[cell_id]\n",
    "    \n",
    "    if ((len(df[(df[features[0]]>0)&(df[cell_key]==cell_type)&(df[condition_key]=='wt')])<1)&\n",
    "        (len(df[(df[features[0]]>0)&(df[cell_key]==cell_type)&(df[condition_key]=='ko')])<1)):\n",
    "   \n",
    "        df = df[df[cell_key]!=cell_type]\n",
    "        \n",
    "\n",
    "    else:\n",
    "\n",
    "        cell_ids.append(cell_id)\n",
    "\n",
    "cell_list=list(df[cell_key].unique())\n",
    "\n",
    "df_T_cells = df_all_cells[df_all_cells[cell_key] == 'Combined T Cells'].copy()\n",
    "\n",
    "\n",
    "plt.figure(figsize=(12,4))\n",
    "\n",
    "df_melted = pd.melt(df_T_cells, \n",
    "                    id_vars=['WT/KO'], \n",
    "                    value_vars=features,\n",
    "                    var_name='Gene',\n",
    "                    value_name='Expression') \n",
    "\n",
    "# For analyzing level of expression in expressing cells)\n",
    "df_T_cells_expressing = df_T_cells.copy()\n",
    "df_melted_expressing = df_melted.copy()\n",
    "df_melted_expressing = df_melted_expressing[df_melted_expressing['Expression']>0]\n",
    "\n",
    "# ax = sns.violinplot(data=df_melted, x='Gene', y='Expression' , hue=condition_key, split=True, cut=0, inner='box', gap=.2, \n",
    "#                   density_norm='count', width=0.8, palette=['darkgreen','lightgreen'], legend=True, linewidth=1,\n",
    "#                   inner_kws=dict(box_width=3, whis_width=0, color='k', marker='s', zorder=10), alpha=0.5)\n",
    "\n",
    "ax = sns.violinplot(data=df_melted, x='Gene', y='Expression' , hue=condition_key, split=True, cut=0, inner='stick', gap=.2, \n",
    "                  density_norm='count', width=0.8, palette=['darkgreen','lightgreen'], legend=True, linewidth=1,\n",
    "                   alpha=0.5)\n",
    "\n",
    "\n",
    "# Customize the plot\n",
    "plt.xticks(rotation=30,rotation_mode=\"anchor\", fontsize=14, ha='right')\n",
    "plt.ylabel('Expression Level')\n",
    "plt.title('Gene Expression in Combined T Cells in Tumor Samples (WT vs KO)')\n",
    "\n",
    "\n",
    "# # Get the maximum y value for positioning asterisks\n",
    "y_max = df_melted.groupby('Gene')['Expression'].max()\n",
    "\n",
    "# # Add asterisks above each violin plot\n",
    "stat_res = compare_gene_expression_single_celltype(df_all_cells[df_all_cells[cell_key] == 'Combined T Cells'], features, condition_key, 'wt', 'ko')\n",
    "\n",
    "for idx, (gene, direction, asterisk) in enumerate(zip(stat_res['feature'], stat_res['direction'], stat_res['significance'])):\n",
    "    \n",
    "    if asterisk:  # Only add text if there is a significance marker\n",
    "        # Add some padding above the maximum value\n",
    "        y_position = y_max[gene] + 0.2 * (y_max.max() - y_max.min())\n",
    "        ax.text(idx, y_position,  direction+asterisk, ha='center', va='bottom', fontsize=12)\n",
    "\n",
    "plt.ylim([-0.25,np.max(y_max)+1])\n",
    "plt.xlim([-1.0,len(features)+0.5])\n",
    "plt.savefig('tumor_combined_Tcells_wt_vs_ko_split.pdf', dpi=600, bbox_inches = \"tight\")\n",
    "\n",
    "plt.show()\n",
    "stat_res"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "404b64b0-6f1f-40fe-a26c-74e9cb5a31c9",
   "metadata": {},
   "outputs": [],
   "source": [
    "def compare_gene_expression_expressing_cells(df, features, condition_key, condition1, condition2):\n",
    "    \"\"\"\n",
    "    Compare gene expression between conditions for multiple genes in expressing cells only.\n",
    "    \n",
    "    Parameters:\n",
    "    -----------\n",
    "    df : pandas DataFrame\n",
    "        Input data frame containing expression data\n",
    "    features : list\n",
    "        List of genes to analyze\n",
    "    condition_key : str\n",
    "        Column name for condition\n",
    "    condition1: str\n",
    "        Name of condition 1 (e.g., 'wt')\n",
    "    condition2: str\n",
    "        Name of condition 2 (e.g., 'ko')\n",
    "    \n",
    "    Returns:\n",
    "    --------\n",
    "    pandas DataFrame with results including statistics and significance\n",
    "    \"\"\"\n",
    "    \n",
    "    all_results = []\n",
    "    all_p_values = []\n",
    "    \n",
    "    # Calculate p-values for each gene\n",
    "    for feature in features:\n",
    "        # Filter for expressing cells only for this specific gene\n",
    "        expressing_cells = df[df[feature] > 0]\n",
    "        \n",
    "        c1_expr = expressing_cells[expressing_cells[condition_key] == condition1][feature]\n",
    "        c2_expr = expressing_cells[expressing_cells[condition_key] == condition2][feature]\n",
    "        \n",
    "        # Calculate p-value if enough samples\n",
    "        if (len(c1_expr) > 1) & (len(c2_expr) > 1):\n",
    "            statistic, p_value = stats.mannwhitneyu(c1_expr, c2_expr, alternative='two-sided')\n",
    "        else:\n",
    "            p_value = 1.0\n",
    "            \n",
    "        direction = '_'\n",
    "        if (len(c1_expr) > 0) & (len(c2_expr) > 0):\n",
    "            if np.mean(c1_expr) > np.mean(c2_expr):\n",
    "                direction = '>'\n",
    "            if np.mean(c1_expr) < np.mean(c2_expr):\n",
    "                direction = '<'\n",
    "                \n",
    "        # Calculate mean expression and percent expressing for each condition\n",
    "        mean_c1 = np.mean(c1_expr) if len(c1_expr) > 0 else 0\n",
    "        mean_c2 = np.mean(c2_expr) if len(c2_expr) > 0 else 0\n",
    "        \n",
    "        # Calculate percentage of expressing cells\n",
    "        total_c1 = len(df[df[condition_key] == condition1])\n",
    "        total_c2 = len(df[df[condition_key] == condition2])\n",
    "        pct_c1 = (len(c1_expr) / total_c1) * 100 if total_c1 > 0 else 0\n",
    "        pct_c2 = (len(c2_expr) / total_c2) * 100 if total_c2 > 0 else 0\n",
    "                \n",
    "        all_results.append({\n",
    "            'feature': feature,\n",
    "            'direction': direction,\n",
    "            'p_value': p_value,\n",
    "            f'{condition1}_mean': mean_c1,\n",
    "            f'{condition2}_mean': mean_c2,\n",
    "            f'{condition1}_pct': pct_c1,\n",
    "            f'{condition2}_pct': pct_c2\n",
    "        })\n",
    "        all_p_values.append(p_value)\n",
    "    \n",
    "    # Convert to DataFrame\n",
    "    results_df = pd.DataFrame(all_results)\n",
    "    \n",
    "    # Correct p-values\n",
    "    _, p_values_corrected, _, _ = multipletests(all_p_values, method='fdr_bh')\n",
    "    results_df['p_value_corrected'] = p_values_corrected\n",
    "    \n",
    "    # Add significance asterisks\n",
    "    results_df['significance'] = results_df['p_value_corrected'].apply(\n",
    "        lambda p: '****' if p < 0.0001 else\n",
    "                 ('***' if p < 0.001 else\n",
    "                  ('**' if p < 0.01 else\n",
    "                   ('*' if p < 0.05 else '')))\n",
    "    )\n",
    "    \n",
    "    return results_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5609266e-0e96-45e6-9b35-43306fad0985",
   "metadata": {},
   "outputs": [],
   "source": [
    "features = dysfunction_markers+activation_markers\n",
    "cell_key = 'cell_type'\n",
    "condition_key = 'WT/KO'\n",
    "\n",
    "df = sc.get.obs_df(adata_immune[adata_immune.obs['pathogenicity']=='pathogenic'], features+ [condition_key, cell_key])\n",
    "\n",
    "df = df[df['cell_type'].isin(T_cell_types)]\n",
    "\n",
    "df[cell_key] = pd.Categorical(df[cell_key]).remove_unused_categories()\n",
    "df_all_cells = df.copy()\n",
    "df_all_cells[cell_key] = 'Combined T Cells'\n",
    "\n",
    "# Concatenate original and combined data\n",
    "df = pd.concat([df, df_all_cells])\n",
    "og_cell_list=list(df[cell_key].unique())\n",
    "\n",
    "cell_ids = []\n",
    "for cell_id in np.arange(len(og_cell_list)):\n",
    "\n",
    "    cell_type = og_cell_list[cell_id]\n",
    "    \n",
    "    if ((len(df[(df[features[0]]>0)&(df[cell_key]==cell_type)&(df[condition_key]=='wt')])<1)&\n",
    "        (len(df[(df[features[0]]>0)&(df[cell_key]==cell_type)&(df[condition_key]=='ko')])<1)):\n",
    "   \n",
    "        df = df[df[cell_key]!=cell_type]\n",
    "        \n",
    "\n",
    "    else:\n",
    "\n",
    "        cell_ids.append(cell_id)\n",
    "\n",
    "cell_list=list(df[cell_key].unique())\n",
    "\n",
    "df_T_cells = df_all_cells[df_all_cells[cell_key] == 'Combined T Cells'].copy()\n",
    "\n",
    "\n",
    "plt.figure(figsize=(12,4))\n",
    "\n",
    "df_melted = pd.melt(df_T_cells, \n",
    "                    id_vars=['WT/KO'], \n",
    "                    value_vars=features,\n",
    "                    var_name='Gene',\n",
    "                    value_name='Expression') \n",
    "\n",
    "# For analyzing level of expression in expressing cells)\n",
    "df_T_cells_expressing = df_T_cells.copy()\n",
    "df_melted_expressing = df_melted.copy()\n",
    "df_melted_expressing = df_melted_expressing[df_melted_expressing['Expression']>0]\n",
    "\n",
    "# ax = sns.violinplot(data=df_melted, x='Gene', y='Expression' , hue=condition_key, split=True, cut=0, inner='box', gap=.2, \n",
    "#                   density_norm='count', width=0.8, palette=['darkgreen','lightgreen'], legend=True, linewidth=1,\n",
    "#                   inner_kws=dict(box_width=3, whis_width=0, color='k', marker='s', zorder=10), alpha=0.5)\n",
    "\n",
    "ax = sns.violinplot(data=df_melted_expressing, x='Gene', y='Expression' , hue=condition_key, split=True, cut=0, inner='stick', gap=.2, \n",
    "                  density_norm='count', width=0.8, palette=['darkgreen','lightgreen'], legend=True, linewidth=1,\n",
    "                   alpha=0.5)\n",
    "\n",
    "\n",
    "# Customize the plot\n",
    "plt.xticks(rotation=30,rotation_mode=\"anchor\", fontsize=14, ha='right')\n",
    "plt.ylabel('Expression Level')\n",
    "plt.title('Gene Expression in Combined T Cells in tdLN (WT vs KO)')\n",
    "\n",
    "\n",
    "# # Get the maximum y value for positioning asterisks\n",
    "y_max = df_melted.groupby('Gene')['Expression'].max()\n",
    "\n",
    "# # Add asterisks above each violin plot\n",
    "stat_res = compare_gene_expression_expressing_cells(df_all_cells[df_all_cells[cell_key] == 'Combined T Cells'], features, condition_key, 'wt', 'ko')\n",
    "\n",
    "for idx, (gene, direction, asterisk) in enumerate(zip(stat_res['feature'], stat_res['direction'], stat_res['significance'])):\n",
    "    \n",
    "    if asterisk:  # Only add text if there is a significance marker\n",
    "        # Add some padding above the maximum value\n",
    "        y_position = y_max[gene] + 0.2 * (y_max.max() - y_max.min())\n",
    "        ax.text(idx, y_position,  direction+asterisk, ha='center', va='bottom', fontsize=12)\n",
    "\n",
    "plt.ylim([-0.25,np.max(y_max)+1])\n",
    "plt.xlim([-1.0,len(features)+0.5])\n",
    "plt.savefig('tumor_combined_Tcells_expressing_wt_vs_ko_split.pdf', dpi=600, bbox_inches = \"tight\")\n",
    "\n",
    "plt.show()\n",
    "stat_res"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f231af87-2a3d-41ad-9d7d-0336980eaa55",
   "metadata": {},
   "source": [
    "## Markers used in annotation"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "cf94970d-b03f-4396-aa2b-f9fe6b92b915",
   "metadata": {},
   "source": [
    "### T cell markers"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a3aea5b2-7db1-4686-8099-cbfa254de71b",
   "metadata": {},
   "outputs": [],
   "source": [
    "cd4t_act = ['Cd3e','Cd4','Gzmb']\n",
    "cd8t_act = ['Cd3e','Cd8a','Gzmk','Clec4e','Gzmb','Pdcd1']\n",
    "activated_cd8t_act = ['Cd3e','Cd8a','Cd28','Cd69']\n",
    "effector_cd8t_act = ['Cd3e','Cd8a','Cd44','Cd69']\n",
    "\n",
    "sc.tl.score_genes(adata_immune, cd8t_act, score_name='cd8t_act_score')\n",
    "sc.tl.score_genes(adata_immune, cd4t_act, score_name='cd4t_act_score')\n",
    "sc.tl.score_genes(adata_immune, activated_cd8t_act, score_name='activated_cd8t_act_score')\n",
    "sc.tl.score_genes(adata_immune, effector_cd8t_act, score_name='effector_cd8t_act_score')\n",
    "\n",
    "\n",
    "sc.pl.umap(adata_immune, color=['cd8t_act_score','cd4t_act_score','Cd4'],palette='tab20',cmap='coolwarm',vmax=2)\n",
    "sc.pl.umap(adata_immune, color=['activated_cd8t_act_score','effector_cd8t_act_score','Pdcd1'],palette='tab20',cmap='coolwarm',vmax=2)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "749426e5-9d7d-4dc7-b1b5-6b037c31867f",
   "metadata": {},
   "outputs": [],
   "source": [
    "#Itga (CD49A) is a tissue resident cytotoxic CD8+ T marker (doi: 10.1016/j.immuni.2017.01.009)\n",
    "sc.pl.umap(adata_immune, color=['Itga1','Ctla4','Cd8a'],palette='tab20',cmap='coolwarm')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "206385c3-8ae7-414c-8df8-772b816d84ae",
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "exhausted_cd8t_act = ['Gzmb','Havcr2','Lag3','Pdcd1','Prf1','Tigit']\n",
    "dneg_thymocyte_act = ['Cd3e','Cd3g','Klrd1','Nkg7']\n",
    "th_act = ['Cd4','Cxcl5']\n",
    "treg_act = ['Cd4','Foxp3','Cd3e','Odc1']\n",
    "t_naive_act = ['Ccr7','Lef1','Sell','Tcf7']\n",
    "\n",
    "sc.tl.score_genes(adata_immune, exhausted_cd8t_act, score_name='exhausted_cd8t_act_score')\n",
    "sc.tl.score_genes(adata_immune, dneg_thymocyte_act, score_name='dneg_thymocyte_act_score')\n",
    "sc.tl.score_genes(adata_immune, th_act, score_name='th_act_score')\n",
    "sc.tl.score_genes(adata_immune, treg_act, score_name='treg_act_score')\n",
    "sc.tl.score_genes(adata_immune, t_naive_act, score_name='t_naive_act_score')\n",
    "\n",
    "\n",
    "sc.pl.umap(adata_immune, color=['exhausted_cd8t_act_score','dneg_thymocyte_act_score'],palette='tab20',cmap='coolwarm')\n",
    "sc.pl.umap(adata_immune, color=['th_act_score','treg_act_score','t_naive_act_score'],palette='tab20',cmap='coolwarm')\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f992fd7c-c164-4e6d-9135-de9956a21f80",
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "Treg_genes =['Ikzf2','Foxp3', 'Ctla4','Il2ra']\n",
    "adata_immune.obs['Treg_score'] = adata_immune[:,Treg_genes].X.sum(1)\n",
    "\n",
    "Th_genes = ['Cxcr6','Bcl6','Prdm1','Tbx21']\n",
    "adata_immune.obs['Th_score'] = adata_immune[:,Th_genes].X.sum(1)\n",
    "\n",
    "cd8effector_genes = ['Cd8a','Cx3cr1','Gzmb','Lgals1','S1pr5']\n",
    "adata_immune.obs['cd8effector_score'] = adata_immune[:,cd8effector_genes].X.sum(1)\n",
    "TRM_genes = ['Itgae','Cd69']\n",
    "adata_immune.obs['TRM_score'] = adata_immune[:,TRM_genes].X.sum(1)\n",
    "\n",
    "sc.pl.umap(adata_immune, color=['Treg_score','Th_score'],palette='tab20',cmap='coolwarm', vmax=5)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "433c1eb3-c086-40f5-8e27-3c6519c8eafd",
   "metadata": {},
   "outputs": [],
   "source": [
    "sc.pl.umap(adata_immune, color=['cd8effector_score','TRM_score'],palette='tab20',cmap='coolwarm')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "51187185-70fc-4770-b91a-f0041135ef14",
   "metadata": {},
   "outputs": [],
   "source": [
    "# This shows CTLA4+ CD8+ T cells if present\n",
    "adata_immune.obs['ctla4_cd8_score'] = np.asarray(adata_immune[:,'Ctla4'].X.sum(1))*np.asarray(adata_immune[:,'Cd8a'].X.sum(1))/(np.max(np.asarray(adata_immune[:,'Ctla4'].X.sum(1)))*np.max(np.asarray(adata_immune[:,'Cd8a'].X.sum(1))))\n",
    "\n",
    "sc.pl.umap(adata_immune, color=['ctla4_cd8_score'],palette='tab20',cmap='coolwarm', vmax=0.8)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0f998aec-f112-40e5-a5f1-6a3dd4b04888",
   "metadata": {},
   "outputs": [],
   "source": [
    "#CD3E+ CD8A- CD4-\n",
    "adata_immune.obs['DNTscore'] = (np.asarray(1-adata_immune[:,'Cd8a'].X.sum(1)/np.max(adata_immune[:,'Cd8a'].X.sum(1)))*np.asarray(1-adata_immune[:,'Cd4'].X.sum(1)/np.max(adata_immune[:,'Cd4'].X.sum(1))))*np.asarray(adata_immune[:,'Cd3e'].X.sum(1)/np.max(adata_immune[:,'Cd3e'].X.sum(1)))\n",
    "\n",
    "#CD3E- CD8A- CD4-\n",
    "adata_immune.obs['TNTscore'] = (np.asarray(1 - adata_immune[:,'Cd8a'].X.sum(1)/np.max(adata_immune[:,'Cd8a'].X.sum(1)))*np.asarray(1 - adata_immune[:,'Cd4'].X.sum(1)/np.max(adata_immune[:,'Cd4'].X.sum(1))))*np.asarray(1- adata_immune[:,'Cd3e'].X.sum(1)/np.max(adata_immune[:,'Cd3e'].X.sum(1)))\n",
    "ncols = 2\n",
    "nrows = 1\n",
    "figsize = 8\n",
    "wspace = 0.1\n",
    "fig, axs = plt.subplots(\n",
    "    nrows=nrows,\n",
    "    ncols=ncols,\n",
    "    figsize=(ncols * figsize + figsize * wspace * (ncols - 1), nrows * figsize),\n",
    ")\n",
    "sc.pl.umap(adata_immune, color=['DNTscore'],palette='tab20',cmap='coolwarm', ax=axs[0], show=False)\n",
    "sc.pl.umap(adata_immune, color=['TNTscore'],palette='tab20',cmap='coolwarm', ax=axs[1],vmax=1)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9240b933-cc73-4e59-bddc-835022b1917b",
   "metadata": {},
   "source": [
    "### NK cells"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "46888644-88f2-4a19-8571-cc1172330acb",
   "metadata": {},
   "outputs": [],
   "source": [
    "nk_genes = ['Klrb1c', 'Ncr1','Itgal']\n",
    "adata_immune.obs['nk_score'] = adata_immune[:,nk_genes].X.sum(1)\n",
    "sc.pl.umap(adata_immune, color=['nk_score','Fcgr3'],palette='tab20',cmap='coolwarm',vmax=4)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "409b973f-16ea-47c3-9b3c-1efe298f831a",
   "metadata": {},
   "source": [
    "### Naive B"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0927670b-d34d-4af7-bf24-a1a41e6ee52c",
   "metadata": {},
   "outputs": [],
   "source": [
    "naiveB_genes = ['Ighd', 'Ighm']\n",
    "memoryB_genes = ['Itgam','Cd80','Cxcr3','Nt5e','Pdcd1lg2']\n",
    "adata_immune.obs['naiveB_score'] = adata_immune[:,naiveB_genes].X.sum(1)\n",
    "adata_immune.obs['memoryB_score'] = adata_immune[:,memoryB_genes].X.sum(1)\n",
    "sc.pl.umap(adata_immune, color=['naiveB_score','memoryB_score'],palette='tab20',cmap='coolwarm', vmax=4)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "10beb764-7df8-43df-bc3b-3c8f87e46211",
   "metadata": {},
   "outputs": [],
   "source": [
    "## memoryB_genes2 = ['Aicda']\n",
    "adata_immune.obs['memoryB_score2'] = adata_immune[:,memoryB_genes2].X.sum(1)\n",
    "\n",
    "memoryB_genes3 = ['Cd27', 'Cd80', 'Cd86']\n",
    "adata_immune.obs['memoryB_score3'] = adata_immune[:,memoryB_genes3].X.sum(1)\n",
    "sc.pl.umap(adata_immune, color=['memoryB_score2','memoryB_score3'],palette='tab20',cmap='coolwarm', vmax=3)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d1d92615-f222-408b-ab9c-70f849c0833f",
   "metadata": {},
   "source": [
    "### Age-associated B"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "252bbccc-6acc-4be7-87ac-6a6a00f33187",
   "metadata": {},
   "outputs": [],
   "source": [
    "ageB_genes = ['Ighm','Il10', 'Ifng', 'Itgax', 'Tbx2', 'Itgam', 'Fas']\n",
    "\n",
    "adata_immune.obs['ageB_score'] = adata_immune[:,ageB_genes].X.sum(1)\n",
    "sc.pl.umap(adata_immune, color=['ageB_score'],palette='tab20',cmap='coolwarm',vmax=5)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "17298963-3b59-4d7e-a16f-b911c723410a",
   "metadata": {},
   "outputs": [],
   "source": [
    "AAB_markers = ['Ighm', 'Il10', 'Ifng', 'Itgax', 'Tbx2', 'Itgam', 'Fas','Cd19', 'Cd38', 'Bcl6', 'Tbx21', 'Irf4', 'Cxcr5']\n",
    "\n",
    "adata_immune.obs['AAB_score'] = adata_immune[:,AAB_markers].X.sum(1)\n",
    "sc.pl.umap(adata_immune, color=['AAB_score'],palette='tab20',cmap='coolwarm',vmax=10)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "bddf7447-d222-4278-b5c0-650e01ef0dc4",
   "metadata": {},
   "outputs": [],
   "source": [
    "#B cell activation markers\n",
    "sc.pl.umap(adata_immune, color=['Cd86','Cd69'],palette='tab20',cmap='coolwarm',vmax=3)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f88938cf-a9d0-4540-b9ff-a71ba8d723b9",
   "metadata": {},
   "outputs": [],
   "source": [
    "sc.pl.umap(adata_immune, color=['Prdm1','Xbp1'],palette='tab20',cmap='coolwarm')\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "fcf80936-95e7-448a-9df3-549f8d6e5b60",
   "metadata": {},
   "outputs": [],
   "source": [
    "# MZB and B-1 cells\n",
    "mzB_genes = ['Cd9','Cr2','Spib']\n",
    "\n",
    "adata_immune.obs['mzB_score'] = adata_immune[:,mzB_genes].X.sum(1)\n",
    "sc.pl.umap(adata_immune, color=mzB_genes,palette='tab20',cmap='coolwarm',vmax=2)\n",
    "\n",
    "B1_genes = ['Cd5','Spn','Ptpn22']\n",
    "sc.pl.umap(adata_immune, color=B1_genes,palette='tab20',cmap='coolwarm',vmax=1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4cc5606b-f64d-4df3-8ca0-ba608c90c34b",
   "metadata": {},
   "outputs": [],
   "source": [
    "#Follicular B cells\n",
    "folB_genes = ['Cd19', 'Cd79a', 'Ms4a1']\n",
    "folB_low_genes = ['Prdm1', 'Xbp1']\n",
    "sc.pl.umap(adata_immune, color=folB_genes,palette='tab20',cmap='coolwarm', vmax=2)\n",
    "sc.pl.umap(adata_immune, color=folB_low_genes,palette='tab20',cmap='coolwarm', vmax=2)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "3ac36284-f37a-4e8b-b299-9ccbe27b7a96",
   "metadata": {},
   "source": [
    "### DC"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0465fc5e-8fcf-45d0-bddc-942733d6381c",
   "metadata": {},
   "outputs": [],
   "source": [
    "# dc1_genes = ['Cd8a','Irf8']\n",
    "dc1_genes = ['Cd8a','Irf8', 'Batf3', 'Nfil3','Id2', 'Bcl6', 'Xcr1','Rab43', 'Itgax', 'Itgae', 'Cd24a']\n",
    "adata_immune.obs['dc1_score'] = adata_immune[:,dc1_genes].X.sum(1)\n",
    "# dc2_genes = ['Relb', 'Esam', 'Itgam','Irf4', 'Sirpa']\n",
    "dc2_genes = ['H2-Ab1', 'Sirpa', 'Cd4', 'Notch2', 'Clec4a2','Esam', 'Irf4', 'Relb', 'Zeb2', 'Klf4']\n",
    "adata_immune.obs['dc2_score'] = np.asarray(adata_immune[:,dc2_genes].X.sum(1)/np.max(adata_immune[:,dc2_genes].X.sum(1)))*np.asarray(1-adata_immune[:,'Cd8a'].X.sum(1)/np.max(adata_immune[:,'Cd8a'].X.sum(1)))\n",
    "pdc_genes =['Cd209a','Lag3','Lifr','Tcf4', 'Zeb2', 'Pacsin1', 'Spib']\n",
    "adata_immune.obs['pdc_score'] = adata_immune[:,pdc_genes].X.sum(1)\n",
    "\n",
    "migratorydc_genes = ['Ccr7','Itgae', 'Cd207', 'Cx3cr1', 'Itgax', 'H2-Ab1', 'Cd40', 'Cd80', 'Cd86', 'Relb', 'Zbtb46']\n",
    "adata_immune.obs['migratorydc_score'] = adata_immune[:,migratorydc_genes].X.sum(1)\n",
    "\n",
    "sc.pl.umap(adata_immune, color=['dc1_score'],palette='tab20',cmap='coolwarm',vmax=15)\n",
    "sc.pl.umap(adata_immune, color=['dc2_score'],palette='tab20',cmap='coolwarm')\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d94ee848-1778-484d-b721-46760294fc7b",
   "metadata": {},
   "outputs": [],
   "source": [
    "sc.pl.umap(adata_immune, color=dc1_genes,palette='tab20',cmap='coolwarm')\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ddcd7ca0-d0b2-49b1-b56e-dd814f46e17d",
   "metadata": {},
   "outputs": [],
   "source": [
    "# For testing DC2 hypothesis\n",
    "dc2_genes2= ['Itgam', 'Sirpa', 'Zbtb46','Flt3']\n",
    "sc.pl.umap(adata_immune, color=dc2_genes2,palette='tab20',cmap='coolwarm', vmax=0.9)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "673993ed-d141-4642-9ccb-0c811e3d024f",
   "metadata": {},
   "outputs": [],
   "source": [
    "## Division to Cross-presenting and Sirpa+ DC\n",
    "## See Gurka et al, Front. Immunol., 04 February 2015 Sec. Antigen Presenting Cell Biology\n",
    "## Volume 6 - 2015 | https://doi.org/10.3389/fimmu.2015.00035\n",
    "\n",
    "#Sirpa and Xcr1 ar enot expressed in the tumor cells\n",
    "\n",
    "cdc1_genes1 = ['Cd8a','Clec9a','Batf3','Xcr1']\n",
    "\n",
    "adata_immune.obs['dc_xcr_score'] = (np.asarray(adata_immune[:,'Cd8a'].X.sum(1)/np.max(adata_immune[:,'Cd8a'].X.sum(1)))*np.asarray(adata_immune[:,'Clec9a'].X.sum(1)/np.max(adata_immune[:,'Clec9a'].X.sum(1))))*np.asarray(adata_immune[:,'Batf3'].X.sum(1)/np.max(adata_immune[:,'Batf3'].X.sum(1)))*np.asarray(adata_immune[:,'Xcr1'].X.sum(1)/np.max(adata_immune[:,'Xcr1'].X.sum(1)))*np.asarray(1.-adata_immune[:,'Sirpa'].X.sum(1)/np.max(adata_immune[:,'Sirpa'].X.sum(1)))\n",
    "\n",
    "cdc1_genes2 = ['Cd8a','Clec9a','Batf3','Sirpa']\n",
    "adata_immune.obs['dc_sirpa_score'] = (np.asarray(adata_immune[:,'Cd8a'].X.sum(1)/np.max(adata_immune[:,'Cd8a'].X.sum(1)))*np.asarray(adata_immune[:,'Clec9a'].X.sum(1)/np.max(adata_immune[:,'Clec9a'].X.sum(1))))*np.asarray(adata_immune[:,'Batf3'].X.sum(1)/np.max(adata_immune[:,'Batf3'].X.sum(1)))*np.asarray(1.-adata_immune[:,'Xcr1'].X.sum(1)/np.max(adata_immune[:,'Xcr1'].X.sum(1)))*np.asarray(adata_immune[:,'Sirpa'].X.sum(1)/np.max(adata_immune[:,'Sirpa'].X.sum(1)))\n",
    "\n",
    "sc.pl.umap(adata_immune, color=['dc_xcr_score'],palette='tab20',cmap='coolwarm')\n",
    "sc.pl.umap(adata_immune, color=['dc_sirpa_score'],palette='tab20',cmap='coolwarm')\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ccf414d9-0068-4a22-affc-156382cfb7be",
   "metadata": {},
   "outputs": [],
   "source": [
    "sc.pl.umap(adata_immune, color=['pdc_score'],palette='tab20',cmap='coolwarm',vmax=10)\n",
    "sc.pl.umap(adata_immune, color=['migratorydc_score'],palette='tab20',cmap='coolwarm',vmax=15)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a3ab1a8a-e51b-4564-9d0c-5908e301151f",
   "metadata": {},
   "outputs": [],
   "source": [
    "#pDCs express Siglech\n",
    "sc.pl.umap(adata_immune, color=['Siglech'],palette='tab20',cmap='coolwarm')\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e43649d7-fadd-464c-8ea3-a1319c67bd8e",
   "metadata": {},
   "outputs": [],
   "source": [
    "### Activated cDCs\n",
    "acdc_genes = ['Itgax','Cd83' ,'H2-Aa', 'H2-Ab1', 'H2-Eb1']\n",
    "\n",
    "adata_immune.obs['acdc_score'] = adata_immune[:,acdc_genes].X.sum(1)\n",
    "\n",
    "sc.pl.umap(adata_immune, color=acdc_genes,palette='tab20',cmap='coolwarm')\n",
    "sc.pl.umap(adata_immune, color=['acdc_score'],palette='tab20',cmap='coolwarm')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b231e6d5-e0e7-411f-a10e-192e710c7f30",
   "metadata": {},
   "source": [
    "### Monocytes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "937d0f49-afeb-4851-a1b8-5885ae483d4c",
   "metadata": {},
   "outputs": [],
   "source": [
    "clmono_genes=['Ccl9', 'Ccr2','Cd68','Ly6c2']\n",
    "nclmono_genes = ['Csf1r','Cx3cr1','Fabp4']\n",
    "adata_immune.obs['clmono_score'] = adata_immune[:,clmono_genes].X.sum(1)\n",
    "adata_immune.obs['nclmono_score'] = adata_immune[:,nclmono_genes].X.sum(1)\n",
    "adata_immune.obs['mono_score'] = (adata_immune.obs['clmono_score']+adata_immune.obs['nclmono_score'])/2\n",
    "sc.pl.umap(adata_immune, color=['clmono_score', 'nclmono_score', 'mono_score'],palette='tab20',cmap='coolwarm',vmax=10)\n",
    "sc.pl.umap(adata_immune, color=clmono_genes,palette='tab20',cmap='coolwarm')\n",
    "sc.pl.umap(adata_immune, color=nclmono_genes,palette='tab20',cmap='coolwarm')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "18a01cc4-0f42-4bfc-b024-e7dfa27f3261",
   "metadata": {},
   "source": [
    "### Plasma cells"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c95a17ec-7e96-4767-aef2-27a863fcd199",
   "metadata": {},
   "outputs": [],
   "source": [
    "plasma_genes = ['Jchain','Sdc1','Prdm1','Xbp1']\n",
    "adata_immune.obs['plasma_score'] = adata_immune[:,plasma_genes].X.sum(1)\n",
    "sc.pl.umap(adata_immune, color=['plasma_score'], palette='tab20',cmap='coolwarm', vmax=4.)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9ecb9c3d-3344-40cc-b38b-7581142d2a1d",
   "metadata": {},
   "source": [
    "### Macrophages"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b4a6cf09-de56-4eca-b1be-b93aa17808d4",
   "metadata": {},
   "outputs": [],
   "source": [
    "adata.var_names[adata.var_names.str.startswith('Sep')]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "69c82d2f-f1c4-4388-a04d-cf685a99cbe1",
   "metadata": {},
   "outputs": [],
   "source": [
    "# macroph_genes = ['Adgre1','Itgam','Ly6g']\n",
    "\n",
    "#PanglaoDB\n",
    "macroph_genes = ['Cd68','Fcgr1','Naaa','Lyz2','Ccl12']\n",
    "#['Adgre1','Itgam','Apoe','C1qa','Cx3cr1']\n",
    "adata_immune.obs['macroph_score'] = adata_immune[:,macroph_genes].X.sum(1)\n",
    "\n",
    "sc.pl.umap(adata_immune, color=['macroph_score'],palette='tab20',cmap='coolwarm', vmax=15)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "0f81c4e2-2f89-472a-a61d-080e7401dd5c",
   "metadata": {},
   "source": [
    "### Erythrophagocytic macrophages"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1e09e171-7ce7-44c0-a8aa-3e69ab7021f3",
   "metadata": {},
   "outputs": [],
   "source": [
    "ephM_genes = ['Cd68', 'Adgre1', 'Mrc1']\n",
    "adata_immune.obs['ephM_score'] = adata_immune[:,ephM_genes].X.sum(1)\n",
    "sc.pl.umap(adata_immune, color=ephM_genes, palette='tab20',cmap='coolwarm') "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7988692c-0800-4c6b-a2c9-3b2ba479680a",
   "metadata": {},
   "outputs": [],
   "source": [
    "### Activated macrophages\n",
    "actM_genes = [\n",
    "    'Cd80', 'Cd86', 'H2-Ab1', 'H2-DMb1', 'Itgam',\n",
    "    'Tnf', 'Il1b', 'Il6', 'Ccl2', \n",
    "    'F4/80', 'Cd11b', 'Cd68', \n",
    "    'Arg1', 'Nos2', 'Irf5', 'Cd274'\n",
    "]\n",
    "\n",
    "# Check if the genes are in the dataset\n",
    "actM_genes = [gene for gene in actM_genes if gene in adata.var_names]\n",
    "\n",
    "adata_immune.obs['actM_score'] = adata_immune[:,actM_genes].X.sum(1)\n",
    "\n",
    "# Plot UMAP with marker gene expression\n",
    "sc.pl.umap(adata_immune, color='actM_score',cmap='coolwarm', vmax=30)\n",
    "\n",
    "# Show the plot\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7121dc95-8cc3-42b1-8736-92e4d7fc8ab2",
   "metadata": {},
   "outputs": [],
   "source": [
    "sc.pl.umap(adata_immune, color=['Cd68', 'Cd14', 'Adgre1'],palette='tab20',cmap='coolwarm',vmax=2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "89f3fb39-8390-4416-94a0-57c3e56ded34",
   "metadata": {},
   "outputs": [],
   "source": [
    "# M1 markers\n",
    "sc.pl.umap(adata_immune, color=['Nos2', 'Il1b', 'Tnf'],palette='tab20',cmap='coolwarm')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9acbe2c6-636a-4af5-94e1-d31fbac727fc",
   "metadata": {},
   "outputs": [],
   "source": [
    "# M2 markers\n",
    "sc.pl.umap(adata_immune, color=['Zeb2', 'Gab2','Mitf'],palette='tab20',cmap='coolwarm')\n",
    "sc.pl.umap(adata_immune, color=['Arg1', 'Mrc1', 'Il10'],palette='tab20',cmap='coolwarm',vmax=1)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "181563f4-4e53-437c-9e7f-8d1e4f424325",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Test antigen-presenting properties\n",
    "sc.pl.umap(adata_immune, color=['H2-Aa', 'H2-Ab1','H2-Eb1'],palette='tab20',cmap='coolwarm')\n",
    "sc.pl.umap(adata_immune, color=['Ciita', 'Cd80','Cd86'],palette='tab20',cmap='coolwarm')\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "df8d8853-d538-442a-a210-4e905a5a3ac8",
   "metadata": {},
   "source": [
    "### Neutrophils"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b851ed93-840d-4b6c-beb9-46ee11cfcf71",
   "metadata": {},
   "outputs": [],
   "source": [
    "neutrophil_genes1 = ['Csf3r','S100a8','Il1r2']\n",
    "neutrophil_genes2 = ['S100a9','Elane']\n",
    "sc.pl.umap(adata_immune, color=neutrophil_genes1, palette='tab20',cmap='coolwarm', vmax=2)\n",
    "sc.pl.umap(adata_immune, color=neutrophil_genes2, palette='tab20',cmap='coolwarm', vmax=1)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "70e355c1-f503-4d6e-af65-11c30bdb1786",
   "metadata": {},
   "source": [
    "### Progenitor cells"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f45aead6-d997-4e3a-848b-fd7ba240b25c",
   "metadata": {},
   "outputs": [],
   "source": [
    "progenitor_genes = ['Cxcr5', 'Id3','Slamf6','Tcf7']\n",
    "adata_immune.obs['progenitor_score'] = adata_immune[:,progenitor_genes].X.sum(1)\n",
    "sc.pl.umap(adata_immune, color=['progenitor_score'], palette='tab20',cmap='coolwarm')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "947fedf2-4a28-41f6-a270-d98ded845cd6",
   "metadata": {},
   "source": [
    "### NKT cells"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2850993d-519c-4ef7-8a37-7aa0e9693dc6",
   "metadata": {},
   "outputs": [],
   "source": [
    "NKT_genes = ['Ncam1', 'Gata3', 'Il2rb'] #From PanglaoDB\n",
    "adata_immune.obs['NKT_score'] = adata_immune[:,NKT_genes].X.sum(1)\n",
    "sc.pl.umap(adata_immune, color=['NKT_score'], palette='tab20',cmap='coolwarm', vmax=5)\n",
    "#NK and NKT markers (necessary)\n",
    "sc.pl.umap(adata_immune, color=['Klrb1c','Itga2'], palette='tab20',cmap='coolwarm')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "75388834-10cc-46bc-bb89-910c7a3c6040",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Further markers for annotating\n",
    "# the high-Tox cluster,  cluster 8, https://doi.org/10.1038/s41586-019-1325-x\n",
    "# https://doi.org/10.3389/fimmu.2023.990419.)\n",
    "\n",
    "exhaustion_markers = [\n",
    "    \"Pdcd1\",    # PD-1\n",
    "    \"Havcr2\",   # Tim-3\n",
    "    \"Lag3\",\n",
    "    \"Tigit\",\n",
    "    \"Tox\",\n",
    "    \"Eomes\",\n",
    "    \"Gzmb\",     # Granzyme B\n",
    "    \"Prf1\",     # Perforin\n",
    "]\n",
    "\n",
    "general_T_markers = [\n",
    "    \"Cd2\",\n",
    "    \"Cd5\",\n",
    "    \"Cd28\",\n",
    "    \"Cd3e\",     # Already mentioned as low, but good to confirm\n",
    "    \"Cd8a\",     # Already mentioned as low, but good to confirm\n",
    "]\n",
    "\n",
    "sc.pl.umap(adata_immune, color=exhaustion_markers, palette='tab20',cmap='coolwarm')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c5219370-589b-4eae-9dab-9edf3ead95ec",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Final stage of exhaustion: TCF1neg T-betlo TOXhi Eomeshi\n",
    "\n",
    "sc.pl.umap(adata_immune, color=['Tcf7','Tbx21','Tox','Eomes'], palette='tab20',cmap='coolwarm')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6ac3de71-edce-44e1-b821-aff57b72e335",
   "metadata": {},
   "outputs": [],
   "source": [
    "#additional exhaustion markers\n",
    "#progenitor\n",
    "sc.pl.umap(adata_immune, color=['Tcf7','Il7r'], palette='tab20',cmap='coolwarm')\n",
    "\n",
    "#Tissue residency markers\n",
    "sc.pl.umap(adata_immune, color=['Cd69', 'Itgae'], palette='tab20',cmap='coolwarm', vmax=2.5)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "81ed7582-98b9-414b-96b4-d72f2dab741b",
   "metadata": {},
   "outputs": [],
   "source": [
    "#Migratory properties\n",
    "chemokine_receptors = ['Ccr7', 'Cxcr3', 'Cxcr4', 'Ccr5']\n",
    "sc.pl.umap(adata_immune, color=chemokine_receptors, palette='tab20',cmap='coolwarm', vmax=4)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a4bb9646-5fd9-4721-acae-185e7693e7ee",
   "metadata": {},
   "outputs": [],
   "source": [
    "#Investigate T cell-DC interactions (cluster 11)\n",
    "sc.pl.umap(adata_immune, color=['Cd28','Cd80'], palette='tab20',cmap='coolwarm',vmax=2)\n",
    "sc.pl.umap(adata_immune, color=['Cd40lg','Cd40'], palette='tab20',cmap='coolwarm',vmax=2)\n",
    "sc.pl.umap(adata_immune, color=['Pdcd1','Cd247', 'Pdcd1lg2'], palette='tab20',cmap='coolwarm',vmax=2)\n",
    "sc.pl.umap(adata_immune, color=['Ctla4','Cd86'], palette='tab20',cmap='coolwarm',vmax=2)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "2478a187-3863-4098-a928-6fa855ee816a",
   "metadata": {},
   "source": [
    "### gamma delta T"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e03742d6-fe8d-42fb-83df-b94b63cdfce5",
   "metadata": {},
   "outputs": [],
   "source": [
    "gamma_delta_genes = ['Tcrg-V4', 'Tcrg-V6', 'Tcrg-V1','Tcrg', 'Tcrd', 'Trgv1', 'Trgv2', 'Trgv3', 'Trgv4', 'Trgv5', 'Trgv6', 'Trgv7','Trdv1', 'Trdv3', 'Trdv4', 'Trdv5','Trg', 'Trd', 'Sox13', 'Id3', 'Blk', 'Il17a', 'Ifng']\n",
    "present_gamma_delta_genes = [gene for gene in gamma_delta_genes if gene in adata.var_names]\n",
    "print(\"Present  T cell genes:\", present_gamma_delta_genes)\n",
    "adata_immune.obs['gamma_delta_score'] = adata_immune[:,present_gamma_delta_genes].X.sum(1)\n",
    "sc.pl.umap(adata_immune, color=['gamma_delta_score'], palette='tab20',cmap='coolwarm', vmax=4)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d2990f6d-7718-4206-b025-ad65205c7c93",
   "metadata": {},
   "source": [
    "## Tumor-associated suppressor cells"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b90c7768-bdae-415a-bf2f-08fbac67d6f6",
   "metadata": {},
   "outputs": [],
   "source": [
    "suppressor_markers =[\"Itgam\",    # CD11b, myeloid marker\n",
    "    \"Cd33\",     # myeloid marker\n",
    "    \"Arg1\",     # often expressed in MDSCs\n",
    "    \"Nos2\",     # iNOS, often in MDSCs\n",
    "    \"S100a8\",   # often in MDSCs\n",
    "    \"S100a9\",\n",
    "]\n",
    "\n",
    "sc.pl.umap(adata_immune, color=suppressor_markers, palette='tab20',cmap='coolwarm')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ce4578a5-a297-4819-9179-a91bc8b88b5e",
   "metadata": {},
   "source": [
    "### DecoupleR"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c5b8227d-d360-4f81-b63b-205225fbf448",
   "metadata": {},
   "outputs": [],
   "source": [
    "import decoupler as dc"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "34949bdb-db88-450f-9410-2740efac6d7e",
   "metadata": {},
   "outputs": [],
   "source": [
    "markers = dc.get_resource('PanglaoDB')\n",
    "markers"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "91c2103d-7ffb-449b-b5c8-2b4958104ea9",
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "markers = dc.get_resource('PanglaoDB')\n",
    "\n",
    "# Convert datatypes from string to float and bool\n",
    "markers['mouse_sensitivity'] = markers['mouse_sensitivity'].astype(float)\n",
    "markers['mouse'] = markers['mouse'].astype(bool)\n",
    "markers['canonical_marker'] = markers['canonical_marker'].astype(bool)\n",
    "\n",
    "# Filter by canonical_marker and mouse\n",
    "markers = markers[(markers['mouse']==True) & (markers['canonical_marker']==True) & (markers['mouse_sensitivity'] > 0.5)]\n",
    "\n",
    "# Remove duplicated entries\n",
    "markers = markers[~markers.duplicated(['cell_type', 'genesymbol'])]\n",
    "markers.index = np.arange(markers.shape[0])\n",
    "\n",
    "# PanglaoDB genes are in human nomenclature. Download conversion list\n",
    "mouse_human_genes = pd.read_csv(\"http://www.informatics.jax.org/downloads/reports/HOM_MouseHumanSequence.rpt\",sep=\"\\t\", index_col=False)\n",
    "mouse = mouse_human_genes[mouse_human_genes['Common Organism Name']=='mouse, laboratory']\n",
    "human = mouse_human_genes[mouse_human_genes['Common Organism Name']=='human']\n",
    "\n",
    "# Collect dataframes for conversion\n",
    "mouse = mouse[['DB Class Key', 'Symbol']]\n",
    "mouse.index = np.arange(mouse.shape[0])\n",
    "\n",
    "human = human[['DB Class Key', 'Symbol']]\n",
    "human.index = np.arange(human.shape[0])\n",
    "\n",
    "# Convert from human to mouse\n",
    "for sym_idx in np.arange(markers.shape[0]):\n",
    "\n",
    "    row_idx  = np.where(human['Symbol']== markers['genesymbol'].iloc[sym_idx])[0]\n",
    "\n",
    "    if len(row_idx)>0:\n",
    "        \n",
    "        dbck = human['DB Class Key'][row_idx].values[0]\n",
    "        mouse_gene = mouse[mouse['DB Class Key']==dbck]['Symbol'].values\n",
    "        markers['genesymbol'].iloc[sym_idx] =  mouse_gene[0]\n",
    "\n",
    "# Remove duplicated entries\n",
    "markers = markers[~markers.duplicated(['cell_type', 'genesymbol'])]\n",
    "\n",
    "markers"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "da9dec38-0cc2-4f21-9516-04ad715755c3",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Run Overrepresentation analysis\n",
    "tissue = 'tumor'\n",
    "\n",
    "dc.run_ora(\n",
    "    mat=adata_immune,\n",
    "    net=markers,\n",
    "    source='cell_type',\n",
    "    target='genesymbol',\n",
    "    min_n=3,\n",
    "    verbose=True,\n",
    "    use_raw=False\n",
    ")\n",
    "\n",
    "# New anndata object with ora estimates\n",
    "acts = dc.get_acts(adata_immune, obsm_key='ora_estimate')\n",
    "\n",
    "# We need to remove inf and set them to the maximum value observed for pvals=0\n",
    "acts_v = acts.X.ravel()\n",
    "max_e = np.nanmax(acts_v[np.isfinite(acts_v)])\n",
    "acts.X[~np.isfinite(acts.X)] = max_e\n",
    "\n",
    "acts\n",
    "\n",
    "df = dc.rank_sources_groups(acts, groupby='leiden', reference='rest', method='t-test_overestim_var')\n",
    "\n",
    "# List most prominent predicted cell types in each Leiden cluster\n",
    "n_ctypes = 3\n",
    "ctypes_dict = df.groupby('group').head(n_ctypes).groupby('group')['names'].apply(lambda x: list(x)).to_dict()\n",
    "ctypes_dict\n",
    "\n",
    "# Dictionary for annotation\n",
    "annotation_dict_PanglaoDB = df.groupby('group').head(1).set_index('group')['names'].to_dict()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a6e4589a-1ef5-425f-b05c-095024517360",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Add cell type column based on annotation\n",
    "adata_immune.obs['cell_type_PanglaoDB'] = [annotation_dict_PanglaoDB[clust] for clust in (adata_immune.obs['leiden'])]\n",
    "\n",
    "# Plot results\n",
    "sc.pl.umap(adata_immune, color=['leiden','cell_type_PanglaoDB'], title=['Leiden','PanglaoDB'], palette='tab20')\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "886cc9f3-bfad-41e4-b7f2-90bce91effab",
   "metadata": {},
   "source": [
    "### Non-immune cluster"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7d4ff367-7761-492d-8c51-7720891f5793",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Fine-resolution clustering\n",
    "sc.tl.leiden(adata_other, resolution=0.5)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9d5ed464-d668-443a-b5da-4ff28d8a26c3",
   "metadata": {},
   "outputs": [],
   "source": [
    "sc.pl.umap(adata_other, color=['leiden','WT/KO'], title=[str(tissue) +' Leiden',str(tissue) +' WT/KO'], palette='tab20', legend_loc='on data')\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "990b30e3-0542-4481-bdb3-39781cddbed4",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Run Overrepresentation analysis\n",
    "tissue = 'tumor'\n",
    "\n",
    "dc.run_ora(\n",
    "    mat=adata_other,\n",
    "    net=markers,\n",
    "    source='cell_type',\n",
    "    target='genesymbol',\n",
    "    min_n=3,\n",
    "    verbose=True,\n",
    "    use_raw=False\n",
    ")\n",
    "\n",
    "# New anndata object with ora estimates\n",
    "acts = dc.get_acts(adata_other, obsm_key='ora_estimate')\n",
    "\n",
    "# We need to remove inf and set them to the maximum value observed for pvals=0\n",
    "acts_v = acts.X.ravel()\n",
    "max_e = np.nanmax(acts_v[np.isfinite(acts_v)])\n",
    "acts.X[~np.isfinite(acts.X)] = max_e\n",
    "\n",
    "acts\n",
    "\n",
    "df = dc.rank_sources_groups(acts, groupby='leiden', reference='rest', method='t-test_overestim_var')\n",
    "\n",
    "# List most prominent predicted cell types in each Leiden cluster\n",
    "n_ctypes = 3\n",
    "ctypes_dict = df.groupby('group').head(n_ctypes).groupby('group')['names'].apply(lambda x: list(x)).to_dict()\n",
    "ctypes_dict\n",
    "\n",
    "# Dictionary for annotation\n",
    "annotation_dict_PanglaoDB = df.groupby('group').head(1).set_index('group')['names'].to_dict()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "933d51eb-e9e1-4bfc-9c56-1b270f8091a5",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Add cell type column based on annotation\n",
    "adata_other.obs['cell_type_PanglaoDB'] = [annotation_dict_PanglaoDB[clust] for clust in (adata_other.obs['leiden'])]\n",
    "\n",
    "# Plot results\n",
    "sc.pl.umap(adata_other, color=['leiden','cell_type_PanglaoDB'], title=['Leiden','PanglaoDB'], palette='tab20')\n"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.12"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
